---
title: 'GBS Impatiens capensis Toronto populations bioinformatic pipeline'
author: "Ruth Rivkin"
date: "2020-2021"
output:
  html_document: 
    toc: yes
---

# Downloaded reads from KCCG Sequencing Center server (Oct 19-2020)

- Download multiplexed reads (forward and reverse), keyfile, and qc report to personal computer

``` {bash eval = FALSE}
curl -o 200813_FD07777933_FastQC_collated_report.html 'https://filesender.public.garvan.org.au/download.php?token=4ae684a4-7c85-4f90-8a90-424b17d5eabd&files_ids=8411' # Expires 07/01/2021
curl -o DeliverySummary.csv 'https://filesender.public.garvan.org.au/download.php?token=47e9e393-da96-4ade-8850-6d2a972b646e&files_ids=8409' # Expires 07/01/2021
curl -o HF2NMCCX2_1_200813_FD07777933_Other__R_200605_ROBELS1_LIBX10_M006_R1.fastq.gz 'https://filesender.public.garvan.org.au/download.php?token=93daf319-a092-427d-ab64-448707a8b5a9&files_ids=8392' # Expires 07/01/2021
curl -o HF2NMCCX2_1_200813_FD07777933_Other__R_200605_ROBELS1_LIBX10_M006_R1.fastq.gz.md5 'https://filesender.public.garvan.org.au/download.php?token=e34f7636-5abf-4b20-ab9d-570c5af331df&files_ids=8412' # Expires 07/01/2021
curl -o HF2NMCCX2_1_200813_FD07777933_Other__R_200605_ROBELS1_LIBX10_M006_R2.fastq.gz 'https://filesender.public.garvan.org.au/download.php?token=9ffc1d68-e10f-4c5e-aecc-a05e6cf0340f&files_ids=8407' # Expires 07/01/2021
curl -o HF2NMCCX2_1_200813_FD07777933_Other__R_200605_ROBELS1_LIBX10_M006_R2.fastq.gz.md5 'https://filesender.public.garvan.org.au/download.php?token=3841a159-7f4e-434b-a656-674a327d0009&files_ids=8408' # Expires 07/01/2021
curl -o R_200605_ROBELS1_LIBX10_M006.csv 'https://filesender.public.garvan.org.au/download.php?token=752ce0bc-3831-42d5-90e0-b1e20b94513d&files_ids=8415' # Expires 07/01/2021
curl -o README.txt 'https://filesender.public.garvan.org.au/download.php?token=9542dca9-5d77-434e-82d8-5f6e6bfab5c0&files_ids=8400' # Expires 07/01/2021
curl -o SequencingDeliverablesReport.html 'https://filesender.public.garvan.org.au/download.php?token=0769a076-cf0e-45ad-a03c-1a43cbe8a43a&files_ids=8396' # Expires 07/01/2021
curl -o SequencingProjectReport.html 'https://filesender.public.garvan.org.au/download.php?token=dc73b285-f6b5-4c27-9dd8-0b4048bed0ed&files_ids=8381' # Expires 07/01/2021'
```

- Transferred .fastq and keyfile to calculon server using Cyberduck

# Demultiplexed reads (Nov 9-2020)

-	Demultiplexed reads using axe_demux (https://gitlab.com/relshire/gbs-pipeline-scripts/-/tree/master/axe-demultiplexer) with -m 0 flag

``` {bash eval = FALSE}
module load axe/0.3.3
mkdir demult-output
axe-demux -m0 -z 3 -c -2 -v \ 
-t 40473_demult_summary.txt \ 
-b ApeK1-Plate2Key.txt \ 
-f HF2NMCCX2_1_200813_FD07777933_Other__R_200605_ROBELS1_LIBX10_M006_R1.fastq.gz \ 
-r HF2NMCCX2_1_200813_FD07777933_Other__R_200605_ROBELS1_LIBX10_M006_R2.fastq.gz \ 
-F demult-output/40473_R1 \ 
-R demult-output/40473_R2
```

-	Moved into new file: Demultiplex

# Trim Reads
## Trim adaptors (Nov 25-2020- Mar 1-2021)

-	Renamed .fastq.gz to make it easier to read/transcribe
- Created keyfile of forward and reverse reads and barcodes 
- Copied files into new director: TrimGalore
- Removed blank wells (G3, E5, D9, H11, A-H12) from keyfile and directory
-	Trim reads with TrimGalore wrapper batch_trim.pl (https://github.com/Lanilen/GBS-PreProcess) with Phred cutoff: 0. Note: perl, not bash (Mar 1-2021)
- Note: all files (including trim_keyfile and batch_trim) must be in same folder as samples

``` {bash eval = FALSE}
Trim Galore version: 0.6.6
Cutadapt version: 1.14

perl batch_trim.pl trim_keyfile.txt
```


# Trim Reads with Process Radtags (Stacks)
- Trim reads with standard qc and 110 bp
- Generate fastqc reports
- Rename files to standard notation for processing in stacks
- Record sequence length
``` {bash eval = FALSE}
module load stacks/2.55
module load gcc/8.3.0
module load fastqc/0.11.8

# specify the input samples file, where the third                                                                                               
# column is the path to each sample FASTQ file                                                                                                        
sample_info=./Scripts/ProcessRadtags/procrad_samples.txt

sample_names=($(cut -f 1 "$sample_info" | uniq))

for sample in ${sample_names[@]}
do
process_radtags -1 ./Data/TrimGalore/${sample}_R1_val_1.fq -2 ./Data/TrimGalore/${sample}_R2_val_2.fq -o ./Data/ProcessRadtags -e apeKI -c -r -q -t 110
done

#fastQc reports
fastqc ./Data/ProcessRadtags/*_R{1,2}_val_{1,2}.{1,2}.fq -o ./Data/ProcessRadtags/FastQC/


#Rename files                                                                                                                                                       
rename -v _R2_val_2.2.fq .2.fq ./Data/ProcessRadtags/*.fq
rename -v _R1_val_1.1.fq .1.fq ./Data/ProcessRadtags/*.fq
rename -v _R2_val_2.fq.rem.2.fq .rem.2.fq ./Data/ProcessRadtags/*.fq
rename -v _R1_val_1.fq.rem.1.fq .rem.1.fq ./Data/ProcessRadtags/*.fq

#Record sequence lengths                                                                                                                                            
for file in ./Data/ProcessRadtags/*.fq
do
test -f "$file" || continue
awk 'NR%4 == 2 {lengths[length($0)]++} END {for (l in lengths) {print l, lengths[l]}}' "$file" >> ./Data/ProcessRadtags/sequencelength.txt
done
```

# Align Reads with denovo_map (Stacks)
``` {bash eval = FALSE}
#Align
module load stacks/2.55
module load gcc/8.3.0

denovo_map.pl -T 8 -m 5 -M 4 -n 4 -o ./Data/DenovoMap/ --popmap ./Scripts/DenovoMap/popmap.txt --samples ./Data/ProcessRadtags --paired -r 75 -X "ustacks:--deleverage"


#Populations
##Do not restrict SNPs or pops

populations -P ./Data/DenovoMap -O ./Data/DenovoMap/PopulationsAllSNPS --popmap ./Scripts/DenovoMap/popmap.txt -t 8 -r 75 --min-maf 0.05 --hwe --fstats --plink --structure --vcf --verbose	

##Restrict snps

populations -P ./Data/DenovoMap -O ./Data/DenovoMap/Populations --popmap ./Scripts/DenovoMap/popmap.txt -t 8 -p 4 -r 75 --min-maf 0.05 --hwe --fstats --plink --structure --vcf --verbose	--write-single-snp

```

```{bash eval = FALSE}
module load python/2.7.14

#Make genotype file 

python Make.Borice.gfile.stacks.vcf.py slim.populations.snps.vcf data.key.txt slim.populations

#Find bad loci (mutliple calls/mismatches between parents/offsrping)

python locus.police.py slim.populations

##Create bad.loci file
awk '(NR>1) && ($2 > 0 ) ' slim.populations.impossibles.bySNP.txt > bad.txt
awk '{print $1}' bad.txt > bad.loci.txt

#Clean genotypes

python clean.genotypes.py slim.populations

#Run BORICE

python Borice.Genomic.v3.py 12.4.21.v3

```


#Generate PLINK files 
```{bash eval = FALSE}
module load plink/1.90

#plink --file ./SinglePop/populations.plink --geno 0.5  --double-id --allow-extra-chr  --recode A --out all.trim.plink

#plink --file ./SingleParents/populations.plink --geno 0.5  --double-id --allow-extra-chr  --recode A --out parents.trim.plink

#plink --file ./SingleOffspring/populations.plink --geno 0.5  --double-id --allow-extra-chr  --recode A --out offspring.trim.plink

#plink --file ./SinglePop/populations.plink --geno 0.5  --double-id --allow-extra-chr  --recode A --het --out 11.5.21

plink --file ./SingleOffspring/populations.plink --geno 0.5  --double-id --allow-extra-chr  --recode A --het --out 15.5.21.off

plink --file ./SingleParents/populations.plink --geno 0.5  --double-id --allow-extra-chr  --recode A --het --out 15.5.21.par
```


#Run ADMIXTURE 
```{bash eval = FALSE}
# Create BED files
#module load plink/1.90
#plink --file ./SingleParents/populations.plink -geno 0.5 --make-bed --double-id --allow-extra-chr --recode A --out ad.parents
#plink --file ./SingleOffspring/populations.plink -geno 0.5 --make-bed --double-id --allow-extra-chr --recode A --out ad.offspring    

#plink --file ../SinglePop/populations.plink -geno 0.5 --make-bed --double-id --allow-extra-chr --recode A --out ad.all  

module load admixture/1.3.0

FILEP=ad.parents
FILEO=ad.offspring
FILEA=ad.all

#Remove chrom name
#awk '{$1=0;print $0}' $FILEP.bim > $FILEP.bim.tmp
#mv $FILEP.bim.tmp $FILEP.bim

#awk '{$1=0;print $0}' $FILEO.bim > $FILEO.bim.tmp
#mv $FILEO.bim.tmp $FILEO.bim

awk '{$1=0;print $0}' $FILEA.bim > $FILEA.bim.tmp                                                                                                                     m
mv $FILEA.bim.tmp $FILEA.bim 


#Run k=1-6 clusters
#for i in {1..6}
#do
# admixture --cv $FILEP.bed $i > $FILEP.log${i}.out
#done

#for i in {1..6}
#do
# admixture --cv $FILEO.bed $i > $FILEO.log${i}.out
#done
for i in {1..6}
do
 admixture --cv $FILEA.bed $i > $FILEA.log${i}.out
done

#Identify the best value of k clusters which is the value with lowest cross-validation error, CV error for each corresponding K

#grep "CV" $FILEP.*out | awk '{print $3,$4}' | cut -c 4,7-20 > $FILEP.cv.error
#grep "CV" $FILEO.*out | awk '{print $3,$4}' | cut -c 4,7-20 > $FILEO.cv.error
grep "CV" $FILEA.*out | awk '{print $3,$4}' | cut -c 4,7-20 > $FILEA.cv.error
```