# Set up notebook
## Load libraries & functions
```{r warning=F}
source(here::here("libraries.R"))
source(here::here("functions.R"))
```

## Import files
### Genetic data
#### All individuals
```{r}
# vcf
vcf <- read.vcfR(
  here::here("./clean_data/stacks_output/output_populations_no-IBD/mmaf0.05_R0.75/populations.snps.vcf"),
  verbose = FALSE)

## Convert vcf to genind object
my_genind <- vcfR2genind(vcf)

```

#### Urban individuals
##### Urb = dist to CC
```{r}
# vcf
vcf_urb <- read.vcfR(
  here::here("./clean_data/filtered_vcfs/urban_pops_dist_mmaf0.05_variant_sites.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_urb <- vcfR2genind(vcf_urb)

```

##### Urb = urb score
```{r}
# vcf
vcf_urb2 <- read.vcfR(
  here::here("./clean_data/filtered_vcfs/urban_pops_usc_mmaf0.05_variant_sites.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_urb2 <- vcfR2genind(vcf_urb2)

```

#### Rural individuals
##### Urb = dist to CC
```{r}
# vcf
vcf_rur <- read.vcfR(
  here::here("./clean_data/filtered_vcfs/rural_pops_dist_mmaf0.05_variant_sites.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_rur <- vcfR2genind(vcf_rur)

```

##### Urb = urb score
```{r}
# vcf
vcf_rur2 <- read.vcfR(
  here::here("./clean_data/filtered_vcfs/rural_pops_usc_mmaf0.05_variant_sites.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_rur2 <- vcfR2genind(vcf_rur2)

```

#### 1 individual/population
```{r}
# vcf
vcf_subsamp <- read.vcfR(
  here::here("./clean_data/filtered_vcfs/1indiv_per_mmaf0.05_variant_sites.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_subsamp <- vcfR2genind(vcf_subsamp)
```


#### 1 individual/population- URBAN ONLY
##### Urb = dist to CC
```{r}
# vcf
vcf_urb_subsamp <- read.vcfR(
  here::here("./clean_data/filtered_vcfs/1indiv_per_mmaf0.05_variant_sites_URBAN.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_urb_subsamp <- vcfR2genind(vcf_urb_subsamp)
```

##### Urb = urb score
```{r}
# vcf
vcf_urb_subsamp2 <- read.vcfR(
  here::here("./clean_data/filtered_vcfs/1indiv_per_pop_mmaf0.05_variant_sites_URBAN_usc.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_urb_subsamp2 <- vcfR2genind(vcf_urb_subsamp2)
```

#### 1 individual/population- RURAL ONLY
##### Urb = dist to CC
```{r}
# vcf
vcf_rur_subsamp <- read.vcfR(
  here::here("./clean_data/filtered_vcfs/1indiv_per_mmaf0.05_variant_sites_RURAL.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_rur_subsamp <- vcfR2genind(vcf_rur_subsamp)
```

##### Urb = urb score
```{r}
# vcf
vcf_rur_subsamp2 <- read.vcfR(
  here::here("./clean_data/filtered_vcfs/1indiv_per_pop_mmaf0.05_variant_sites_RURAL_usc.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_rur_subsamp2 <- vcfR2genind(vcf_rur_subsamp2)
```

### Coordinates
#### All individuals
```{r}
coords_utm <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./genomic_resources/pop_map3.csv")
```

#### Urban
To make these pop map csvs, I took the txt, made it into a csv, and removed the "MW0" and "MW00" from the population IDs to fit this function
##### Urb = dist to CC
```{r}
coords_utm_urb <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_maps/pop_map3_U_dist.csv")
```

##### Urb = urb score
```{r}
coords_utm_urb2 <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_maps/pop_map3_U_usc.csv")
```

#### Rural
##### Urb = dist to CC
```{r}
coords_utm_rur <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_maps/pop_map3_R_dist.csv")
```

##### Urb = urb score
```{r}
coords_utm_rur2 <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_maps/pop_map3_R_usc.csv")
```

#### 1 individual/population
```{r}
coords_utm_subsamp <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_maps/pop_map3_1indiv_per_pop.csv")
```


#### 1 individual/population- URBAN ONLY
##### Urb = dist to CC
```{r}
coords_utm_urb_subsamp <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_maps/pop_map3_1indiv_per_pop_URBAN.csv")
```

##### Urb = urb score
```{r}
coords_utm_urb_subsamp2 <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv", 
  "./clean_data/pop_maps/pop_map3_1indiv_per_pop_URBAN_usc.csv")
```

#### 1 individual/population- RURAL ONLY
##### Urb = dist to CC
```{r}
coords_utm_rur_subsamp <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_maps/pop_map3_1indiv_per_pop_RURAL.csv")
```

##### Urb = urb score
```{r}
coords_utm_rur_subsamp2 <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_maps/pop_map3_1indiv_per_pop_RURAL_usc.csv") 
```

# Create genetic distance matrices
```{r}
# all individuals-----
genetic_distance_matrix <- dist(my_genind, method = "euclidean")

# urban individuals-----
## urb = dist to CC
genetic_distance_matrix_urb <- dist(my_genind_urb, method = "euclidean")

## urb = urb score
genetic_distance_matrix_urb2 <- dist(my_genind_urb2, method = "euclidean")


# rural individuals-----
## urb = dist to CC
genetic_distance_matrix_rur <- dist(my_genind_rur, method = "euclidean")

## urb = urb score
genetic_distance_matrix_rur2 <- dist(my_genind_rur2, method = "euclidean")


# 1 individual/site-----
genetic_distance_matrix_subsamp <- dist(my_genind_subsamp, method = "euclidean")

# 1 individual/site: urban-----
## urb = dist to CC
genetic_distance_matrix_urb_subsamp <- dist(my_genind_urb_subsamp, method = "euclidean")

## urb = urb score
genetic_distance_matrix_urb_subsamp2 <- dist(my_genind_urb_subsamp2, method = "euclidean")


# 1 individual/site: rural-----
## urb = dist to CC
genetic_distance_matrix_rur_subsamp <- dist(my_genind_rur_subsamp, method = "euclidean")

## urb = urb score
genetic_distance_matrix_rur_subsamp2 <- dist(my_genind_rur_subsamp2, method = "euclidean")
```

# Mantel correlogram (UTM)
## Create breakpoints vectors
### Urban: 2.5km spacing btwn sampling sites
```{r}
break_points_2500m <- c(0, 2500, 5000, 7500, 10000, 12500, 15000, 17500, 20000, 22500, 25000, 27500, 30000, 32500, 35000, 37500, 40000, 42500, 45000, 47500, 50000, 52500, 55000, 57500, 60000, 62500, 65000, 67500, 70000 )
```

### Urban: 3km spacing btwn sampling sites
```{r}
break_points_3000m <- c(0, 3000, 6000, 9000, 12000, 15000, 18000, 21000, 24000, 27000, 30000, 33000, 36000, 39000, 42000, 45000, 48000, 51000, 54000, 57000, 60000, 63000, 66000, 69000)
```

### Rural: 10km spacing btwn sampling sites
```{r}
break_points_10km <- c(0, 10000, 20000, 30000, 40000, 50000, 60000, 70000)
```

### Rural: 12km spacing btwn sampling sites
```{r}
break_points_12km <- c(0, 12000, 24000, 36000, 48000, 60000, 72000)
```

## All individuals
### 2.5km spacing
```{r}
correlogram_utm <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix,
   D.geo = coords_utm,
   nperm = 1000,
  cutoff = T,
  break.pts = break_points_2500m
   )

plot(correlogram_utm)
```

### 10km spacing
```{r}
correlogram_utm_10km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix,
  D.geo = coords_utm,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_10km
   )

plot(correlogram_utm_10km)

```

## Urban
### Urb = dist to CC
#### 2.5km spacing
```{r}
correlogram_utm_urb <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb,
D.geo = coords_utm_urb,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_2500m
   )

plot(correlogram_utm_urb)

```

#### 3km spacing
```{r}
correlogram_utm_urb_3km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb,
  D.geo = coords_utm_urb,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_3000m
   )

plot(correlogram_utm_urb_3km)

```

#### 10km spacing
```{r}
correlogram_utm_urb_10km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb,
 D.geo = coords_utm_urb,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_10km
   )

plot(correlogram_utm_urb_10km)

```

#### 12km spacing
```{r}
correlogram_utm_urb_12km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb,
   D.geo = coords_utm_urb,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_12km
   )

plot(correlogram_utm_urb_12km)

```

### Urb = urb score
#### 2.5km spacing
```{r}
correlogram_utm_urb2 <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb2,
  D.geo = coords_utm_urb2,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_2500m
   )

plot(correlogram_utm_urb2)

```

#### 3km spacing
```{r}
correlogram_utm_urb2_3km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb2,
    D.geo = coords_utm_urb2,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_3000m
   )

plot(correlogram_utm_urb2_3km)

```

#### 10km spacing
```{r}
correlogram_utm_urb2_10km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb2,
   D.geo = coords_utm_urb2,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_10km
   )

plot(correlogram_utm_urb2_10km)

```

#### 12km spacing
```{r}
correlogram_utm_urb2_12km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb2,
   D.geo = coords_utm_urb2,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_12km
   )

plot(correlogram_utm_urb2_12km)

```

## Rural
### Urb = dist to CC
#### 10km spacing
```{r}
correlogram_utm_rur <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur,
  D.geo = coords_utm_rur,
  nperm = 1000,
  cutoff = T,
   break.pts = break_points_10km
   )

plot(correlogram_utm_rur)

```

#### 12km spacing
```{r}
correlogram_utm_rur_12km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur,
   D.geo = coords_utm_rur,
  nperm = 1000,
  cutoff = T,
   break.pts = break_points_12km
   )

plot(correlogram_utm_rur_12km)

```

### Urb = urb score
#### 10km spacing
```{r}
correlogram_utm_rur2 <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur2,
 D.geo = coords_utm_rur2,
  nperm = 1000,
  cutoff = T,
   break.pts = break_points_10km
   )

plot(correlogram_utm_rur2)

```

#### 12km spacing
```{r}
correlogram_utm_rur2_12km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur2,
  D.geo = coords_utm_rur2,
  nperm = 1000,
  cutoff = T,
   break.pts = break_points_12km
   )

plot(correlogram_utm_rur2_12km)

```

## 1 individual/site
### 2.5km spacing
```{r}
correlogram_utm_subsamp <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_subsamp,
   D.geo = coords_utm_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_2500m
   )

plot(correlogram_utm_subsamp)

```

### 10km spacing
```{r}
correlogram_utm_subsamp_10km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_subsamp,
  D.geo = coords_utm_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_10km
   )

plot(correlogram_utm_subsamp_10km)

```

## 1 individual/site- Urban
### Urb = dist to CC
#### 2.5km spacing
```{r}
correlogram_utm_urb_subsamp <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb_subsamp,
  D.geo = coords_utm_urb_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_2500m
   )

plot(correlogram_utm_urb_subsamp)

```

#### 3km spacing
```{r}
correlogram_utm_urb_subsamp_3km <- vegan::mantel.correlog(
D.eco = genetic_distance_matrix_urb_subsamp,
 D.geo = coords_utm_urb_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_3000m
   )

plot(correlogram_utm_urb_subsamp_3km)

```

#### 10km spacing
```{r}
correlogram_utm_urb_subsamp_10km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb_subsamp,
  D.geo = coords_utm_urb_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_10km
   )

plot(correlogram_utm_urb_subsamp_10km)

```

#### 12km spacing
```{r}
correlogram_utm_urb_subsamp_12km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb_subsamp,
  D.geo = coords_utm_urb_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_12km
   )

plot(correlogram_utm_urb_subsamp_12km)

```

### Urb = urb score
#### 2.5km spacing
```{r}
correlogram_utm_urb_subsamp2 <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb_subsamp2,
  D.geo = coords_utm_urb_subsamp2,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_2500m
   )

plot(correlogram_utm_urb_subsamp2)

```

#### 3km spacing
```{r}
correlogram_utm_urb_subsamp2_3km <- vegan::mantel.correlog(
D.eco = genetic_distance_matrix_urb_subsamp2,
 D.geo = coords_utm_urb_subsamp2,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_3000m
   )

plot(correlogram_utm_urb_subsamp2_3km)

```

#### 10km spacing
```{r}
correlogram_utm_urb_subsamp2_10km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb_subsamp2,
  D.geo = coords_utm_urb_subsamp2,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_10km
   )

plot(correlogram_utm_urb_subsamp2_10km)

```

#### 12km spacing
```{r}
correlogram_utm_urb_subsamp2_12km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb_subsamp2,
  D.geo = coords_utm_urb_subsamp2,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_12km
   )

plot(correlogram_utm_urb_subsamp2_12km)

```

## 1 individual/site- Rural
### Urb = dist to CC
#### 10km spacing
```{r}
correlogram_utm_rur_subsamp <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur_subsamp,
  D.geo = coords_utm_rur_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_10km
   )

plot(correlogram_utm_rur_subsamp)

```

#### 12km spacing
```{r}
correlogram_utm_rur_subsamp_12km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur_subsamp,
 D.geo = coords_utm_rur_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_12km
   )

plot(correlogram_utm_rur_subsamp_12km)

```

### Urb = urb score
#### 10km spacing
```{r}
correlogram_utm_rur_subsamp2 <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur_subsamp2,
  D.geo = coords_utm_rur_subsamp2,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_10km
   )

plot(correlogram_utm_rur_subsamp2)

```

#### 12km spacing
```{r}
correlogram_utm_rur_subsamp2_12km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur_subsamp2,
 D.geo = coords_utm_rur_subsamp2,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_12km
   )

plot(correlogram_utm_rur_subsamp2_12km)

```

# Export
## All individuals
```{r}
# 2.5 km
export_corr_table_fig(correlogram_utm, all, all_indiv_2500m)

# 10 km
export_corr_table_fig(correlogram_utm_10km, all, all_indiv_10km)
```

## Urban
### Urb = dist to CC
```{r}
# 2.5 km
export_corr_table_fig(correlogram_utm_urb, distance, urban_2500m)

# 3 km
export_corr_table_fig(correlogram_utm_urb_3km, distance, urban_3km)

# 10 km
export_corr_table_fig(correlogram_utm_urb_10km, distance, urban_10km)

# 12 km
export_corr_table_fig(correlogram_utm_urb_12km, distance, urban_12km)

```

### Urb = urb score
```{r}
# 2.5 km
export_corr_table_fig(correlogram_utm_urb2, urb_score, urban_2500m)

# 3 km
export_corr_table_fig(correlogram_utm_urb2_3km, urb_score, urban_3km)

# 10 km
export_corr_table_fig(correlogram_utm_urb2_10km, urb_score, urban_10km)

# 12 km
export_corr_table_fig(correlogram_utm_urb2_12km, urb_score, urban_12km)
```

## Rural
### Urb = dist to CC
```{r}
# 10 km
export_corr_table_fig(correlogram_utm_rur, distance, rural_10km)

# 12 km
export_corr_table_fig(correlogram_utm_rur_12km, distance, rural_12km)
```

### Urb = urb score
```{r}
# 10 km
export_corr_table_fig(correlogram_utm_rur2, urb_score, rural_10km)

# 12 km
export_corr_table_fig(correlogram_utm_rur2_12km, urb_score, rural_12km)
```

## 1 individual/site
```{r}
# 2.5 km
export_corr_table_fig(correlogram_utm_subsamp, all, one_indiv_2500m)

# 10 km
export_corr_table_fig(correlogram_utm_subsamp_10km, all, one_indiv_10km)
```

## 1 individual/site: Urban
### Urb = dist to CC
```{r}
# 2.5 km
export_corr_table_fig(correlogram_utm_urb_subsamp, distance, one_indiv_urban_2500m)

# 3 km
export_corr_table_fig(correlogram_utm_urb_subsamp_3km, distance, one_indiv_urban_3km)

# 10 km
export_corr_table_fig(correlogram_utm_urb_subsamp_10km, distance, one_indiv_urban_10km)

# 12 km
export_corr_table_fig(correlogram_utm_urb_subsamp_12km, distance, one_indiv_urban_12km)
```

### Urb = urb score
```{r}
# 2.5 km
export_corr_table_fig(correlogram_utm_urb_subsamp2, urb_score, one_indiv_urban_2500m_urbscore)

# 3 km
export_corr_table_fig(correlogram_utm_urb_subsamp2_3km, urb_score, one_indiv_urban_3km_urbscore)

# 10 km
export_corr_table_fig(correlogram_utm_urb_subsamp2_10km, urb_score, one_indiv_urban_10km_urbscore)

# 12 km
export_corr_table_fig(correlogram_utm_urb_subsamp2_12km, urb_score, one_indiv_urban_12km_urbscore)
```

## 1 individual/site: Rural
### Urb = dist to CC
```{r}
# 10 km
export_corr_table_fig(correlogram_utm_rur_subsamp, distance, one_indiv_rural_10km)

# 12 km
export_corr_table_fig(correlogram_utm_rur_subsamp_12km, distance, one_indiv_rural_12km)
```

### Urb = urb score
```{r}
# 10 km
export_corr_table_fig(correlogram_utm_rur_subsamp2, urb_score, one_indiv_rural_10km_urbscore)

# 12 km
export_corr_table_fig(correlogram_utm_rur_subsamp2_12km, urb_score, one_indiv_rural_12km_urbscore)
```

# Re-import images, then create composite figures
## Distance
### All individuals
```{r}
urb_2500m_dist <- rasterGrob(readPNG(here::here("./Figures_Tables/correlogram/distance/figures/urban_2500m.png")))

urb_10km_dist <- rasterGrob(readPNG(here::here("./Figures_Tables/correlogram/distance/figures/urban_10km.png")))

rur_10km_dist <- rasterGrob(readPNG(here::here("./Figures_Tables/correlogram/distance/figures/rural_10km.png")))



corr_dist_plots <- gridExtra::grid.arrange(urb_2500m_dist, 
                        urb_10km_dist,
                        rur_10km_dist,
                        ncol = 3,
             left = textGrob("A", 
                            x = unit(1, "npc"), 
                            y = unit(0.98, "npc"),
                            gp = gpar(fontsize = 12)),
             center = textGrob("B", 
                            x = unit(1, "npc"), 
                            y = unit(1.96, "npc"),
                            gp = gpar(fontsize = 12)),
             right = textGrob("C", 
                            x = unit(-9.75, "npc"), 
                            y = unit(0.98, "npc"),
                            gp = gpar(fontsize = 12)))

ggsave("./Figures_Tables/correlogram/distance/figures/main_figs.png", 
       plot = corr_dist_plots,
       width = 7, height = 4,
       units = "in",
       dpi = 400)
```

### 1 individual/pop
```{r}
urb_2500m_dist2 <- rasterGrob(readPNG(here::here("./Figures_Tables/correlogram/distance/figures/one_indiv_urban_2500m.png")))

urb_10km_dist2 <- rasterGrob(readPNG(here::here("./Figures_Tables/correlogram/distance/figures/one_indiv_urban_10km.png")))

rur_10km_dist2 <- rasterGrob(readPNG(here::here("./Figures_Tables/correlogram/distance/figures/one_indiv_rural_10km.png")))



corr_dist_plots2 <- gridExtra::grid.arrange(urb_2500m_dist2, 
                        urb_10km_dist2,
                        rur_10km_dist2,
                        ncol = 3,
             left = textGrob("A", 
                            x = unit(1, "npc"), 
                            y = unit(0.98, "npc"),
                            gp = gpar(fontsize = 12)),
             center = textGrob("B", 
                            x = unit(1, "npc"), 
                            y = unit(1.96, "npc"),
                            gp = gpar(fontsize = 12)),
             right = textGrob("C", 
                            x = unit(-9.75, "npc"), 
                            y = unit(0.98, "npc"),
                            gp = gpar(fontsize = 12)))

ggsave("./Figures_Tables/correlogram/distance/figures/main_figs_1indiv_per_pop.png", 
       plot = corr_dist_plots2,
       width = 7, height = 4,
       units = "in",
       dpi = 400)
```

## Urb score
### All individuals
```{r}
urb_2500m_usc <- rasterGrob(readPNG(here::here("./Figures_Tables/correlogram/urb_score/figures/urban_2500m.png")))

urb_10km_usc <- rasterGrob(readPNG(here::here("./Figures_Tables/correlogram/urb_score/figures/urban_10km.png")))

rur_10km_usc <- rasterGrob(readPNG(here::here("./Figures_Tables/correlogram/urb_score/figures/rural_10km.png")))



corr_usc_plots <- gridExtra::grid.arrange(urb_2500m_usc, 
                        urb_10km_usc,
                        rur_10km_usc,
                        ncol = 3,
             left = textGrob("A", 
                            x = unit(1, "npc"), 
                            y = unit(0.98, "npc"),
                            gp = gpar(fontsize = 12)),
             center = textGrob("B", 
                            x = unit(1, "npc"), 
                            y = unit(1.96, "npc"),
                            gp = gpar(fontsize = 12)),
             right = textGrob("C", 
                            x = unit(-9.75, "npc"), 
                            y = unit(0.98, "npc"),
                            gp = gpar(fontsize = 12)))

ggsave("./Figures_Tables/correlogram/urb_score/figures/main_figs.png", 
       plot = corr_usc_plots,
       width = 7, height = 4,
       units = "in",
       dpi = 400)
```

### 1 individual/pop
```{r}
urb_2500m_usc2 <- rasterGrob(readPNG(here::here("./Figures_Tables/correlogram/urb_score/figures/one_indiv_urban_2500m_urbscore.png")))

urb_10km_usc2 <- rasterGrob(readPNG(here::here("./Figures_Tables/correlogram/urb_score/figures/one_indiv_urban_10km_urbscore.png")))

rur_10km_usc2 <- rasterGrob(readPNG(here::here("./Figures_Tables/correlogram/urb_score/figures/one_indiv_rural_10km_urbscore.png")))



corr_usc_plots2 <- gridExtra::grid.arrange(urb_2500m_usc2, 
                        urb_10km_usc2,
                        rur_10km_usc2,
                        ncol = 3,
             left = textGrob("A", 
                            x = unit(1, "npc"), 
                            y = unit(0.98, "npc"),
                            gp = gpar(fontsize = 12)),
             center = textGrob("B", 
                            x = unit(1, "npc"), 
                            y = unit(1.96, "npc"),
                            gp = gpar(fontsize = 12)),
             right = textGrob("C", 
                            x = unit(-9.75, "npc"), 
                            y = unit(0.98, "npc"),
                            gp = gpar(fontsize = 12)))

ggsave("./Figures_Tables/correlogram/urb_score/figures/main_figs_1indiv_per_pop.png", 
       plot = corr_usc_plots2,
       width = 7, height = 4,
       units = "in",
       dpi = 400)
```