# Set up notebook
## Load libraries & functions
```{r warning=F}
source(here::here("libraries.R"))
source(here::here("functions.R"))
```

# Milkweed data
## Import files
### Genetic data
#### Function
```{r}
process_vcf_data <- function(vcf_filepath) {
  
  # Read vcf file
  vcf <- read.vcfR(vcf_filepath, verbose = FALSE)
  
  # Convert vcf to genind object
  my_genind <- vcfR2genind(vcf)
  
  # Tidy up colnames (remove periods)
  inputdata1 <- my_genind$tab
  colnames(inputdata1) <- gsub("\\.", "_", colnames(inputdata1))
  
  ## Produce a proportion of shared alleles genetic distance matrix using the convenience wrapper function provided with the package
  gen_data <- codomToPropShared(as.data.frame(inputdata1), missingData = NA)
  
  return(gen_data)
}
```

#### All individuals
```{r}
gen_data <- process_vcf_data(here::here("./clean_data/stacks_output/output_populations_no-IBD/mmaf0.05_R0.75/populations.snps.vcf"))

```

#### JUST URBAN individuals
##### urb = dist to CC
```{r}
gen_data_urban <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/urban_pops_dist_mmaf0.05_variant_sites.recode.vcf")
  )
```

##### urb = urb score
```{r}
gen_data_urban_usc <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/urban_pops_usc_mmaf0.05_variant_sites.recode.vcf")
  )
```

#### JUST RURAL individuals
##### urb = dist to CC
```{r}
gen_data_rural <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/rural_pops_dist_mmaf0.05_variant_sites.recode.vcf")
  )
```

##### urb = urb score
```{r}
gen_data_rural_usc <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/rural_pops_usc_mmaf0.05_variant_sites.recode.vcf")
  )
```

#### 1 individual/population
```{r}
gen_data_subsamp <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/1indiv_per_mmaf0.05_variant_sites.recode.vcf")
  )
```

#### 1 individual/population- URBAN ONLY
##### urb = dist to CC
```{r}
gen_data_subsamp_urban <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/1indiv_per_mmaf0.05_variant_sites_URBAN.recode.vcf")
  )
```

##### urb = urb score
```{r}
# gen_data_subsamp_urban_usc <- process_vcf_data(
#   here::here("./clean_data/filtered_vcfs/1indiv_per_pop_mmaf0.05_variant_sites_URBAN_usc.recode.vcf")
#   )

# locus "26386:21:+.0" doesn't have a paired "26386:21:+.1" locus, so I'll remove it from the vcf manually because, otherwise, this function won't work (Error: memgene:  Alleles matrix must have paired columns (i.e. diploid genotypes)

 # Read vcf file, convert to genind
gen_data_subsamp_urban_usc <- vcfR2genind(read.vcfR(here::here("./clean_data/filtered_vcfs/1indiv_per_pop_mmaf0.05_variant_sites_URBAN_usc.recode.vcf"), verbose = FALSE))
  

# ID locus to remove
locus_to_remove <- "26386:21:+"  

# remove locus from tab
gen_data_subsamp_urban_usc@tab <- gen_data_subsamp_urban_usc$tab[, -1019]  

# remove locus from all.names
gen_data_subsamp_urban_usc@all.names <- lapply(gen_data_subsamp_urban_usc@all.names, function(x) x[x != locus_to_remove])

# find index of locus to remove
allele_index <- which(unlist(gen_data_subsamp_urban_usc@all.names) == locus_to_remove)
    
# remove locus from loc.fac
gen_data_subsamp_urban_usc@loc.fac <- gen_data_subsamp_urban_usc@loc.fac[-allele_index]

  
# Tidy up colnames (remove periods)
inputdata1 <- gen_data_subsamp_urban_usc$tab
colnames(inputdata1) <- gsub("\\.", "_", colnames(inputdata1))
  
  
## Produce a proportion of shared alleles genetic distance matrix using the convenience wrapper function provided with the package
gen_data_subsamp_urban_usc <- codomToPropShared(as.data.frame(inputdata1),
                                missingData = NA)
```

#### 1 individual/population- RURAL ONLY
##### urb = dist to CC
```{r}
gen_data_subsamp_rural <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/1indiv_per_mmaf0.05_variant_sites_RURAL.recode.vcf")
  )
```

##### urb = urb score
```{r}
gen_data_subsamp_rural_usc <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/1indiv_per_pop_mmaf0.05_variant_sites_RURAL_usc.recode.vcf")
  )
```

### Coordinates
#### Function
```{r}
process_location_data <- function(urb_filepath, pop_map_filepath) {
  
  # lat/longs of sampling sites
  urb <- read.csv(urb_filepath)
  
  # sampling site IDs and genetic sample names
  coords <- read.csv(pop_map_filepath) %>%
    
    # if population doesn't start with "MWI", it should start with "MW"
  # rename col 1
    dplyr::rename(pop_id = 2) %>%
    dplyr::mutate(pop_id = as.factor(pop_id)) %>%
    
    # take entries with numeric populations and add "MW" as prefix
    add_MW_IDs() %>%

    # left vs. full join because some populations in urb_clean weren't genotyped due to lack of material
    left_join(., urb, by = "patch_id") %>%
    dplyr::mutate(patch_id = as.factor(patch_id),
                  pop_id = as.factor(pop_id)) %>%
    dplyr::select(c(1,2,6,7)) %>%
    dplyr::rename("y" = 3,
                  "x" = 4) %>%
    dplyr::arrange(sample) %>%
    dplyr::select("x", "y")
  
  return(coords)
  }
```

#### All individuals
```{r}
coords <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./genomic_resources/pop_map3.csv"))

```

#### All URBAN individuals
##### urb = dist to CC
```{r}
coords_urban <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_U_dist.csv"))

# I converted the pop_map3_U_dist.txt into csv, then removed the "MW" prefixes so that the function would work
```

##### urb = urb score
```{r}
coords_urban_usc <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_U_usc.csv"))

# I converted the pop_map3_U_dist.txt into csv, then removed the "MW" prefixes so that the function would work
```

#### All RURAL individuals
##### urb = dist to CC
```{r}
coords_rural <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_R_dist.csv"))

```

##### urb = urb score
```{r}
coords_rural_usc <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_R_usc.csv"))

```

#### 1 individual/population
```{r}
coords_subsamp <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_1indiv_per_pop.csv"))

```

#### 1 individual/population- RURAL ONLY
##### urb = dist to CC
```{r}
coords_subsamp_rural <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_1indiv_per_pop_RURAL.csv"))

```

##### urb = urb score
```{r}
coords_subsamp_rural_usc <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_1indiv_per_pop_RURAL_usc.csv"))

```

#### 1 individual/population- URBAN ONLY
##### urb = dist to CC
```{r}
coords_subsamp_urban <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_1indiv_per_pop_URBAN.csv"))

```

##### urb = urb score
```{r}
coords_subsamp_urban_usc <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_1indiv_per_pop_URBAN_usc.csv"))

```


## Run analysis
### All individuals
```{r}
## Run the MEMGENE analysis
set.seed(1)
if (!exists("milkweed_analysis"))
milkweed_analysis <- mgQuick(gen_data, coords, forwardPerm = 1000)
```

### Just URBAN individuals
#### Urb = dist to CC
```{r}
set.seed(1)
if (!exists("milkweed_analysis_urban"))
milkweed_analysis_urban <- mgQuick(gen_data_urban, coords_urban, forwardPerm = 1000)

# No significant MEMs detected
```

#### Urb = urb score
```{r}
set.seed(1)
if (!exists("milkweed_analysis_urban_usc"))
milkweed_analysis_urban_usc <- mgQuick(gen_data_urban_usc, coords_urban_usc, forwardPerm = 1000)

# No significant MEMs detected
```

### Just RURAL individuals
#### Urb = dist to CC
```{r}
set.seed(1)
if (!exists("milkweed_analysis_rural"))
milkweed_analysis_rural <- mgQuick(gen_data_rural, coords_rural,  forwardPerm = 1000)

# No significant MEMs detected
```

#### Urb = urb score
```{r}
set.seed(1)
if (!exists("milkweed_analysis_rural_usc"))
milkweed_analysis_rural_usc <- mgQuick(gen_data_rural_usc, coords_rural_usc,  forwardPerm = 1000)

# No significant MEMs detected
```

### 1 individual per population
```{r}
set.seed(1)
if (!exists("milkweed_analysis_subsamp"))
milkweed_analysis_subsamp <- mgQuick(gen_data_subsamp, coords_subsamp,  forwardPerm = 1000)
# No significant MEMs detected
```

### 1 individual per population- URBAN
#### Urb = dist to CC
```{r}
set.seed(1)
if (!exists("milkweed_analysis_subsamp_urban"))
milkweed_analysis_subsamp_urban <- mgQuick(gen_data_subsamp_urban, coords_subsamp_urban,  forwardPerm = 1000)
# 0 significant positive MEMs detected; 
# Forward selections of positive Moran's eigenvectors
# ----Selected: None 
# Forward selections of negative Moran's eigenvectors
# ----Selected: 7, 8, 17, 23, 25, 37, 41 

```

#### Urb = urb score
```{r}
set.seed(1)
if (!exists("milkweed_analysis_subsamp_urban_usc"))
milkweed_analysis_subsamp_urban_usc <- mgQuick(gen_data_subsamp_urban_usc, coords_subsamp_urban_usc,  forwardPerm = 1000)
# 0 significant MEMs detected

```

### 1 individual per population- RURAL
#### Urb = dist to CC
```{r}
set.seed(1)
if (!exists("milkweed_analysis_subsamp_rural"))
milkweed_analysis_subsamp_rural <- mgQuick(gen_data_subsamp_rural, coords_subsamp_rural,  forwardPerm = 1000)
# 2 significant MEMs detected (2,4)

```

#### Urb = urb score
```{r}
set.seed(1)
if (!exists("milkweed_analysis_subsamp_rural_usc"))
milkweed_analysis_subsamp_rural_usc <- mgQuick(gen_data_subsamp_rural_usc, coords_subsamp_rural_usc,  forwardPerm = 1000)
# 4 significant MEMs detected (3, 4, 8, 14)

```

## Visualize results- NOTE: the terrain basemap chunks use stamen basemaps that I can't obtain anymore. These maps aren't necessary to convey my results anymore so I'm not going to fix the code. But if I didn, I'd go to sample_site_map.rmd and use the tmap code to create the maps.
### MOST basic plot (built-in functions)
#### 1 individual per population- RURAL
```{r}
plot(coords_subsamp_rural, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords_subsamp_rural, milkweed_analysis_subsamp_rural$memgene[, 1], add.plot=TRUE, legend=TRUE)
box()
```

### Basic plots
#### Function
```{r}
basic_plot_mems <- function(group, # e.g., all_indiv
                            coords_df, 
                            milkweed_analysis_list
                            ){
  
  # MEM1-----
  
  # with legend
png(#here::here("./Figures_Tables/memgene/memgene_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_legend_var1_",
           today(),
           ".png"),
    width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
#base_map
mgMap(coords_df,
      milkweed_analysis_list$memgene[, 1], 
      add.plot=TRUE,
      legend=TRUE)
box()

# Close device
dev.off()


# without legend
png(#here::here("./Figures_Tables/memgene/memgene_no_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_nolegend_var1_",
           today(),
               ".png"),
        width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords_df,
      milkweed_analysis_list$memgene[, 1], 
      add.plot=TRUE,
      legend=FALSE)
box()

# Close device
dev.off()

  # MEM2-----
  
  # with legend
png(#here::here("./Figures_Tables/memgene/memgene_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_legend_var2_",
           today(),
           ".png"),
    width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
#base_map
mgMap(coords_df,
      milkweed_analysis_list$memgene[, 2], 
      add.plot=TRUE,
      legend=TRUE)
box()

# Close device
dev.off()


# without legend
png(#here::here("./Figures_Tables/memgene/memgene_no_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_nolegend_var2_",
           today(),
               ".png"),
        width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords_df,
      milkweed_analysis_list$memgene[, 2], 
      add.plot=TRUE,
      legend=FALSE)
box()

# Close device
dev.off()
  # MEM3-----
  
  # with legend
png(#here::here("./Figures_Tables/memgene/memgene_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_legend_var3_",
           today(),
           ".png"),
    width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
#base_map
mgMap(coords_df,
      milkweed_analysis_list$memgene[, 3], 
      add.plot=TRUE,
      legend=TRUE)
box()

# Close device
dev.off()


# without legend
png(#here::here("./Figures_Tables/memgene/memgene_no_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_nolegend_var3_",
           today(),
               ".png"),
        width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords_df,
      milkweed_analysis_list$memgene[, 3], 
      add.plot=TRUE,
      legend=FALSE)
box()

# Close device
dev.off()
}
```

#### All individuals
```{r}
# Moran's Eigenvectors 12,20,5
basic_plot_mems("all_indiv",
                coords,
                milkweed_analysis)
```

#### 1 individual/population- RURAL
```{r}
# Moran's Eigenvectors 2,4
basic_plot_mems("subsamp_rural",
                coords_subsamp_rural,
                milkweed_analysis_subsamp_rural)
```

### Plots with terrain basemap
#### Code for each var
```{r}
basemap_terrain <- ggmap(get_map(location = c(-80.5, 43.2, -78.5, 44.25), 
              zoom = 11,
                          maptype = 'terrain',
                          source = 'stamen'),
        darken = c(0.3, "white"))
```

#### Functions
```{r}
# create mem1, mem2, and mem3 objects-----
create_mem <- function(milkweed_analysis_list,
                       coords_df,
                       mem_number #e.g., 1, 2, or 3
                       ){
  
  mem <- milkweed_analysis_list$memgene[, mem_number] %>%
  as.data.frame() %>%
  dplyr::rename("var" = 1) %>%
  cbind(coords_df) %>%
  dplyr::mutate("negative" = ifelse(var < 0, "negative", "positive")) %>%
  dplyr::mutate("abs_val" = abs(var))
  
  return(mem)

}


# create terrain plots-----
terrain_plot_mems <- function(group, # e.g., all_indiv
                              coords_df, 
                              milkweed_analysis_list,
                              mem1_df,
                              mem2_df,
                              mem3_df
){

  for (i in 1:3) {
    
    mem_df <- get(paste0("mem", i, "_df"))  # Access mem1_df, mem2_df, mem3_df dynamically
    
    
    # with legend
    # black/white discrete-----
    basemap_terrain +
      geom_point(data = mem_df,
                 mapping = aes(x = x,
                               y = y, 
                               fill = negative,
                               #    size = abs(var)),
                               size = abs_val),
                 shape = 21) +
      labs(x = 'Longitude', y = 'Latitude',
           fill = paste0("MEMgene\nvariable ", i),
           size = paste0("MEMgene\nvariable ", i)) +
      scale_fill_manual(values = c("#333333", "white")) +
      scale_size_continuous(limits = c(0, 0.15))
    
    ggsave(filename =
             paste0("./Figures_Tables/memgene/",
                    group,
                    "/memgene_terrain_bw_var",
               i,
               "_",
               today(),
               ".png"),
           plot = last_plot(),
           width = 15,
        height = 10,
        units = "cm")

    
    # black/white gradient, instead of discrete-----
    basemap_terrain +
      geom_point(data = mem_df,
                 mapping = aes(x = x,
                               y = y,
                               fill = var,
                               size = abs_val),
                 shape = 21) +
      labs(x = 'Longitude', y = 'Latitude',
           fill = paste0("MEMgene\nvariable ", i),
           size = paste0("MEMgene\nvariable ", i)) +
      scale_fill_gradient(high = "white", low = "black",
                          limits = c(-0.15, 0.1),
                          breaks = c(-0.15, 0, 0.1)) +
      scale_size_continuous(limits = c(0, 0.14)) +
      guides(size = FALSE)

      ggsave(filename =
               paste0("./Figures_Tables/memgene/",
                    group,
               "/memgene_terrain_greyscale_var",
               i,
               "_",
               today(),
               ".png"),
           plot = last_plot(),
           width = 15,
        height = 10,
        units = "cm")

    # colors: plasma, unidirectional size scale-----
    basemap_terrain +
      geom_point(data = mem_df,
                 mapping = aes(x = x,
                               y = y,
                               fill = var),
                 shape = 21,
                 size = 3) +
      labs(x = 'Longitude', y = 'Latitude',
           fill = paste0("MEMgene\nvariable ", i)) +  scale_fill_viridis(option = "plasma")

    ggsave(filename =
             paste0("./Figures_Tables/memgene/",
                    group,
                    "/memgene_terrain_color_var",
               i,
               "_",
               group,
               today(),
               ".png"),
           plot = last_plot(),
           width = 15,
        height = 10,
        units = "cm")

  }
}
```

#### All individuals
```{r}
# create MEM dfs
mem1 <- create_mem(milkweed_analysis, coords, 1)
mem2 <- create_mem(milkweed_analysis, coords, 2)
mem3 <- create_mem(milkweed_analysis, coords, 3)

# create maps
terrain_plot_mems("all_indiv",
                            coords, 
                            milkweed_analysis,
                            mem1,
                            mem2,
                            mem3)
```

#### 1 individual/population- RURAL
```{r}
# create MEM dfs
mem1_subsamp_rural <- create_mem(milkweed_analysis_subsamp_rural,
                          coords_subsamp_rural, 1)
mem2_subsamp_rural <- create_mem(milkweed_analysis_subsamp_rural,
                          coords_subsamp_rural, 2)
mem3_subsamp_rural <- create_mem(milkweed_analysis_subsamp_rural,
                          coords_subsamp_rural, 3)

# create maps
terrain_plot_mems("subsamp_rural",
                            coords_subsamp_rural, 
                            milkweed_analysis_subsamp_rural,
                            mem1_subsamp_rural,
                            mem2_subsamp_rural,
                            mem3_subsamp_rural)
```


### map ALL combined significant memgene variables
#### Function
```{r}
sig_combined_mem_plot <- function(group,
                                  coords_df,
                                  milkweed_analysis_list
                                  ){
  
  total_var <- apply(milkweed_analysis_list$memSelected, 1, sum) %>%
  as.data.frame() %>%
  dplyr::rename("var" = 1)

# sum all SGS explained by all significant memgene variables
# Combine 'var' from 'total_var' with 'coords'
spatial_data <- data.frame(var = total_var$var,
                           coords_df)

# map with colors-----
basemap_terrain +
  geom_point(data = spatial_data,
             mapping = aes(x = x,
                           y = y, 
                           fill = var,
                           size = var),
             shape = 21,
             color = "black",
             ) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = "Variation") +
  scale_fill_viridis(option="plasma") + 
  guides(size = FALSE) 

ggsave(last_plot(),
       file =  paste0("./Figures_Tables/memgene/",
                      group,
                      "/memgene_allMEMs_", today(), ".png"),
       width = 19,
       height = 19,
       units = "cm")

# colors: black/white, big to little (0) to big circles-----

basemap_terrain +
  geom_point(data = spatial_data,
             mapping = aes(x = x,
                           y = y, 
                           fill = ifelse(var < 0, "negative", "positive"),
                           size = abs(var)),
             shape = 21#,
            # color = "black"
            ) +
  labs(x = 'Longitude', y = 'Latitude',
       size = "Variation",
       fill = "Variation") +
  scale_fill_manual(values = c("negative" = "darkgrey",
                               "positive" = "white")) +
  scale_size_continuous(
    range = c(1.5, 12 * max(abs(spatial_data$var))), 
    limits = c(0, max(abs(spatial_data$var)))) 

ggsave(last_plot(),
       file =  paste0("./Figures_Tables/memgene/",
                      group,
                      "/memgene_allMEMs_BW_", today(), ".png"),
       width = 19,
       height = 19,
       units = "cm")

}
```

#### All individuals
```{r}
sig_combined_mem_plot("all_indiv",
                            coords, 
                            milkweed_analysis)
```

#### 1 individual/population- RURAL
```{r}
sig_combined_mem_plot("subsamp_rural",
                            coords_subsamp_rural, 
                            milkweed_analysis_subsamp_rural)
```


### Rescaling, then combine MEMs from 0-255
#### Rescale
```{r}
mems <- list(mem1, mem2, mem3,
            # mem1_urban, mem2_urban, mem3_urban,
           #  mem1_subsamp, mem2_subsamp, mem3_subsamp
             mem1_subsamp_rural, mem2_subsamp_rural, mem3_subsamp_rural
             )

# rescale mems from 0-255
rescale_var_to_255 <- function(df) {
  df %>%
    mutate(mem_255 = scales::rescale(var, to = c(0, 255)))
}

# map the function to each element of the list
mems <- purrr::map(mems, rescale_var_to_255)

names(mems) <- list("mem1", "mem2", "mem3",
            # "mem1_urban", "mem2_urban", "mem3_urban",
            # "mem1_subsamp", "mem2_subsamp", "mem3_subsamp")
             "mem1_subsamp_rural", "mem2_subsamp_rural", "mem3_subsamp_rural")

list2env(mems, globalenv())
```

#### Create individual maps for top 3 MEMs
##### Function
```{r}
rescale_mems <- function(group,
                         milkweed_analysis_list,
                         mem1_df,
                         mem2_df,
                         mem3_df){
  
  # plot with colors: plasma, unidirectional size scale
  
      # MEM1-----
    basemap_terrain +
    geom_point(data = mem1_df,
             mapping = aes(x = x,
                           y = y,
                           fill = mem_255),
             shape = 21,
             size = 3) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = paste0("MEMGENE\nvariable 1"),
       caption = paste0("Moran's Eigenvector: ",
         milkweed_analysis_list$whichSelectedPos[1])) +
  scale_fill_gradient(low = "white",
                      high = "red") 
    

    ggsave(filename = paste0("./Figures_Tables/memgene/",
                             group,
"/memgene_terrain_color_255_var1_",
           today(),
           ".png"),
           plot = last_plot(),
    width = 15,
    height = 10,
    units = "cm")
    
      # MEM2-----
    basemap_terrain +
    geom_point(data = mem2_df,
             mapping = aes(x = x,
                           y = y,
                           fill = mem_255),
             shape = 21,
             size = 3) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = paste0("MEMGENE\nvariable 2"),
       caption = paste0("Moran's Eigenvector: ",
         milkweed_analysis_list$whichSelectedPos[2])) +
  scale_fill_gradient(low = "white",
                      high = "green") 
    

    ggsave(filename = paste0("./Figures_Tables/memgene/",
                             group,
"/memgene_terrain_color_255_var2_",
           today(),
           ".png"),
           plot = last_plot(),
    width = 15,
    height = 10,
    units = "cm")
    
      # MEM3-----
    basemap_terrain +
    geom_point(data = mem3_df,
             mapping = aes(x = x,
                           y = y,
                           fill = mem_255),
             shape = 21,
             size = 3) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = paste0("MEMGENE\nvariable 3"),
       caption = paste0("Moran's Eigenvector: ",
         milkweed_analysis_list$whichSelectedPos[3])) +
  scale_fill_gradient(low = "white",
                      high = "blue") 
    

    ggsave(filename = paste0("./Figures_Tables/memgene/",
                             group,
"/memgene_terrain_color_255_var3_",
           today(),
           ".png"),
           plot = last_plot(),
    width = 15,
    height = 10,
    units = "cm")

}
```

##### All individuals
```{r}
rescale_mems("all_indiv",
             milkweed_analysis,
             mem1,
             mem2,
             mem3)
```

##### 1 individual/population- RURAL
```{r}
rescale_mems("subsamp_rural",
             milkweed_analysis_subsamp_rural,
             mem1_subsamp_rural,
             mem2_subsamp_rural,
             mem3_subsamp_rural)

```

#### Combine top 3 MEMs into one figure
##### Function
```{r}
combine_mems <- function(group,
                         milkweed_analysis_list,
                         mem1_df,
                         mem2_df,
                         mem3_df){
  
  make_color_df <- function(df, color){
    df %>%
    dplyr::select(x,y,mem_255) %>%
    dplyr::rename_with(~color, mem_255) %>%
    dplyr::mutate(ID = row_number())
  }

  red_data <- make_color_df(mem1_df, "red")
  green_data <- make_color_df(mem2_df, "green")
  blue_data <- make_color_df(mem3_df, "blue")

  color_data <- inner_join(red_data,
                        green_data,
                        by = c("x","y","ID")) %>%
  inner_join(.,
             blue_data,
             by = c("x","y","ID")) %>%
  dplyr::select(ID, x, y, red, green, blue)

  color_data$color <- rgb(color_data$red,
                        color_data$green, 
                        color_data$blue,
                        maxColorValue = 255)


  basemap_terrain +
    geom_point(data = color_data,
             mapping = aes(x = x,
                           y = y,
                           fill = color),
             shape = 21,
             size = 3) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = "MEMGENE\nvariables 1-3",
       caption = paste0("Moran's Eigenvectors: ",
         milkweed_analysis_list$whichSelectedPos[1],
         ", ",
         milkweed_analysis_list$whichSelectedPos[2],
         ", ",
         milkweed_analysis_list$whichSelectedPos[3]
         ))  +
    scale_fill_identity()

    ggsave(filename = paste0("./Figures_Tables/memgene/",
                             group,
"/memgene_terrain_color_255_vars1_to_3",
           today(),
           ".png"),
           plot = last_plot(),
    width = 15,
    height = 10,
    units = "cm")
    
}
```

##### All individuals
```{r}
combine_mems("all_indiv",
             milkweed_analysis,
                         mem1,
                         mem2,
                         mem3)

# good visualization of color space in 3D: https://www.whydomath.org/node/wavlets/images/rgbcube.jpg
```

##### 1 individual/population- RURAL
```{r}
combine_mems("subsamp_rural",
             milkweed_analysis_subsamp_rural,
                         mem1_subsamp_rural,
                         mem2_subsamp_rural,
                         mem3_subsamp_rural)

# good visualization of color space in 3D: https://www.whydomath.org/node/wavlets/images/rgbcube.jpg
```


## Export stats
### Function
```{r}
export_stats <- function(group,
                         milkweed_analysis_list){
  
  # Adj R2
  R2_adj <- paste0("Adjusted RÂ² for model: ", 
                   round(milkweed_analysis_list$RsqAdj,
                         4))
  
  
  # Extracting relevant data from the milkweed_analysis object
  whichSelectedPos <- milkweed_analysis_list$whichSelectedPos
  memgene_values <- milkweed_analysis_list$mem$valuesMEM
  sdev_values <- milkweed_analysis_list$sdev
  
  # Calculating R-squared values per MEMGENE variable
  r_squared_list <- milkweed_analysis_list$sdev/sum(milkweed_analysis_list$sdev)
  
  r2_df <- as.data.frame(r_squared_list) %>%
    rownames_to_column("MEMGENE_var") %>%
    slice(1:length(whichSelectedPos)) %>%
    dplyr::rename(R_squared = r_squared_list)
  
  
  # Creating the dataframe
  mg_table <- data.frame(
    MEMGENE_variable = 1:length(whichSelectedPos),
    Moran_Eigenvector = whichSelectedPos,
    R_squared = r2_df$R_squared) %>%
    dplyr::mutate(R_squared = round(R_squared, 3)) %>%
    dplyr::mutate(R_squared = ifelse(
      R_squared < 0.001, 
      "<0.001",
      R_squared)) %>%
    rownames_to_column(var = "Row") %>%
    dplyr::select(-Row) %>%
    dplyr::rename("MEMGENE Variable" = 1,
                  "Moran's Eigenvector" = 2,
                  "R2" = 3)
  

  mg_table %>%
    flextable() %>%
    flextable::align(., align = "center", part = "all") %>%
    flextable::compose(i = 1,
                       j = 3,
                       part = "header",
                       value = as_paragraph("R", as_sup("2"))) %>%
    footnote(.,
             i = 1, j = 1,
            value = as_paragraph(
              R2_adj),
            ref_symbols = c(" "),
            part = "header") %>%
    flextable::autofit() %>%
    flextable::save_as_docx(.,
                            path = paste0("./Figures_Tables/memgene/",
                                          group,
                                          "/memgene_stats_",
                                          today(),
                                          ".docx"))
  
}
```

### All individuals
```{r}
export_stats("all_indiv",
             milkweed_analysis)
```

### 1 individual/population- RURAL
#### Urb = dist to CC
```{r}
export_stats("rural_subsamp",
             milkweed_analysis_subsamp_rural)
```

#### Urb = urb score
```{r}
export_stats("rural_subsamp_usc",
             milkweed_analysis_subsamp_rural_usc)
```

