# Set up notebook
## Load libraries & functions
```{r warning=F}
source(here::here("libraries.R"))
source(here::here("functions.R"))
```

# Import files
## Genetic data
### All individuals
```{r}
gen_data <- process_vcf_data(here::here("./clean_data/stacks_output/output_populations_no-IBD/mmaf0.05_R0.75/populations.snps.vcf"))

```

### JUST URBAN individuals
#### urb = dist to CC
```{r}
gen_data_urban <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/urban_pops_dist_mmaf0.05_variant_sites.recode.vcf")
  )
```

#### urb = urb score
```{r}
gen_data_urban_usc <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/urban_pops_usc_mmaf0.05_variant_sites.recode.vcf")
  )
```

### JUST RURAL individuals
#### urb = dist to CC
```{r}
gen_data_rural <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/rural_pops_dist_mmaf0.05_variant_sites.recode.vcf")
  )
```

#### urb = urb score
```{r}
gen_data_rural_usc <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/rural_pops_usc_mmaf0.05_variant_sites.recode.vcf")
  )
```

### 1 individual/population
```{r}
gen_data_subsamp <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/1indiv_per_mmaf0.05_variant_sites.recode.vcf")
  )
```

### 1 individual/population- URBAN ONLY
#### urb = dist to CC
```{r}
gen_data_subsamp_urban <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/1indiv_per_mmaf0.05_variant_sites_URBAN.recode.vcf")
  )
```

#### urb = urb score
```{r}
# gen_data_subsamp_urban_usc <- process_vcf_data(
#   here::here("./clean_data/filtered_vcfs/1indiv_per_pop_mmaf0.05_variant_sites_URBAN_usc.recode.vcf")
#   )

# locus "26386:21:+.0" doesn't have a paired "26386:21:+.1" locus, so I'll remove it from the vcf manually because, otherwise, this function won't work (Error: memgene:  Alleles matrix must have paired columns (i.e. diploid genotypes)

 # Read vcf file, convert to genind
gen_data_subsamp_urban_usc <- vcfR2genind(read.vcfR(here::here("./clean_data/filtered_vcfs/1indiv_per_pop_mmaf0.05_variant_sites_URBAN_usc.recode.vcf"), verbose = FALSE))
  

# ID locus to remove
locus_to_remove <- "26386:21:+"  

# remove locus from tab
gen_data_subsamp_urban_usc@tab <- gen_data_subsamp_urban_usc$tab[, -1019]  

# remove locus from all.names
gen_data_subsamp_urban_usc@all.names <- lapply(gen_data_subsamp_urban_usc@all.names, function(x) x[x != locus_to_remove])

# find index of locus to remove
allele_index <- which(unlist(gen_data_subsamp_urban_usc@all.names) == locus_to_remove)
    
# remove locus from loc.fac
gen_data_subsamp_urban_usc@loc.fac <- gen_data_subsamp_urban_usc@loc.fac[-allele_index]

  
# Tidy up colnames (remove periods)
inputdata1 <- gen_data_subsamp_urban_usc$tab
colnames(inputdata1) <- gsub("\\.", "_", colnames(inputdata1))
  
  
## Produce a proportion of shared alleles genetic distance matrix using the convenience wrapper function provided with the package
gen_data_subsamp_urban_usc <- codomToPropShared(as.data.frame(inputdata1),
                                missingData = NA)
```

### 1 individual/population- RURAL ONLY
#### urb = dist to CC
```{r}
gen_data_subsamp_rural <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/1indiv_per_mmaf0.05_variant_sites_RURAL.recode.vcf")
  )
```

#### urb = urb score
```{r}
gen_data_subsamp_rural_usc <- process_vcf_data(
  here::here("./clean_data/filtered_vcfs/1indiv_per_pop_mmaf0.05_variant_sites_RURAL_usc.recode.vcf")
  )
```

## Coordinates
### All individuals
```{r}
coords <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./genomic_resources/pop_map3.csv"))

```

### All URBAN individuals
#### urb = dist to CC
```{r}
coords_urban <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_U_dist.csv"))

# I converted the pop_map3_U_dist.txt into csv, then removed the "MW" prefixes so that the function would work
```

#### urb = urb score
```{r}
coords_urban_usc <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_U_usc.csv"))

# I converted the pop_map3_U_dist.txt into csv, then removed the "MW" prefixes so that the function would work
```

### All RURAL individuals
#### urb = dist to CC
```{r}
coords_rural <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_R_dist.csv"))

```

#### urb = urb score
```{r}
coords_rural_usc <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_R_usc.csv"))

```

### 1 individual/population
```{r}
coords_subsamp <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_1indiv_per_pop.csv"))

```

### 1 individual/population- URBAN ONLY
#### urb = dist to CC
```{r}
coords_subsamp_urban <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_1indiv_per_pop_URBAN.csv"))

```

#### urb = urb score
```{r}
coords_subsamp_urban_usc <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_1indiv_per_pop_URBAN_usc.csv"))

```


### 1 individual/population- RURAL ONLY
#### urb = dist to CC
```{r}
coords_subsamp_rural <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_1indiv_per_pop_RURAL.csv"))

```

#### urb = urb score
```{r}
coords_subsamp_rural_usc <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_maps/pop_map3_1indiv_per_pop_RURAL_usc.csv"))

```

# Run memgene analysis
## All individuals
```{r}
## Run the MEMGENE analysis
set.seed(1)
if (!exists("memgene_all"))
memgene_all <- mgQuick(gen_data, coords, forwardPerm = 1000)

# Forward selections of positive Moran's eigenvectors
# ----Selected: 2, 3, 4, 5, 6, 12, 14, 16, 17, 18, 19, 20, 21, 28, 29, 34, 39, 40, 42, 45 
# Forward selections of negative Moran's eigenvectors
# ----Selected: None 
```

## Just URBAN individuals
### Urb = dist to CC
```{r}
set.seed(1)
if (!exists("memgene_urban"))
memgene_urban <- mgQuick(gen_data_urban, coords_urban, forwardPerm = 1000)

# No significant MEMs detected
```

### Urb = urb score
```{r}
set.seed(1)
if (!exists("memgene_urban_usc"))
memgene_urban_usc <- mgQuick(gen_data_urban_usc, coords_urban_usc, forwardPerm = 1000)

# No significant MEMs detected
```

## Just RURAL individuals
### Urb = dist to CC
```{r}
set.seed(1)
if (!exists("memgene_rural"))
memgene_rural <- mgQuick(gen_data_rural, coords_rural,  forwardPerm = 1000)

# No significant MEMs detected
```

### Urb = urb score
```{r}
set.seed(1)
if (!exists("memgene_rural_usc"))
memgene_rural_usc <- mgQuick(gen_data_rural_usc, coords_rural_usc,  forwardPerm = 1000)

# No significant MEMs detected
```

## 1 individual per population
```{r}
set.seed(1)
if (!exists("memgene_subsamp"))
memgene_subsamp <- mgQuick(gen_data_subsamp, coords_subsamp,  forwardPerm = 1000)
# No significant MEMs detected
```

## 1 individual per population- URBAN
### Urb = dist to CC
```{r}
set.seed(1)
if (!exists("memgene_subsamp_urban"))
memgene_subsamp_urban <- mgQuick(gen_data_subsamp_urban, coords_subsamp_urban,  forwardPerm = 1000)
# 0 significant positive MEMs detected; 
# Forward selections of positive Moran's eigenvectors
# ----Selected: None 
# Forward selections of negative Moran's eigenvectors
# ----Selected: 7, 8, 17, 23, 25, 37, 41 

```

### Urb = urb score
```{r}
set.seed(1)
if (!exists("memgene_subsamp_urban_usc"))
memgene_subsamp_urban_usc <- mgQuick(gen_data_subsamp_urban_usc, coords_subsamp_urban_usc,  forwardPerm = 1000)
# 0 significant MEMs detected

```

## 1 individual per population- RURAL
### Urb = dist to CC
```{r}
set.seed(1)
if (!exists("memgene_subsamp_rural"))
memgene_subsamp_rural <- mgQuick(gen_data_subsamp_rural, coords_subsamp_rural,  forwardPerm = 1000)
# 2 significant MEMs detected (2,4)

```

### Urb = urb score
```{r}
set.seed(1)
if (!exists("memgene_subsamp_rural_usc"))
memgene_subsamp_rural_usc <- mgQuick(gen_data_subsamp_rural_usc, coords_subsamp_rural_usc,  forwardPerm = 1000)
# 4 significant MEMs detected (3, 4, 8, 14)

```

# Visualize, export maps
## MOST basic plot (built-in functions)
### All individuals
```{r}
plot(coords, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords, memgene_all$memgene[, 1], add.plot=TRUE, legend=TRUE)
box()
```

### 1 individual per population- RURAL
#### Urb = dist to CC
```{r}
plot(coords_subsamp_rural, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords_subsamp_rural, memgene_subsamp_rural$memgene[, 1], add.plot=TRUE, legend=TRUE)
box()
```

#### Urb = urb score
```{r}
plot(coords_subsamp_rural_usc, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords_subsamp_rural_usc, memgene_subsamp_rural_usc$memgene[, 1], add.plot=TRUE, legend=TRUE)
box()
```

## Basic plots: 1 plot per memgene variable (for 3 vars)
### Function
```{r}
basic_plot_mems <- function(group, # e.g., all_indiv
                            coords_df, 
                            memgene_list
                            ){
  
  # MEM1-----
  
  # with legend
png(#here::here("./Figures_Tables/memgene/memgene_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_legend_var1_",
           today(),
           ".png"),
    width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
#base_map
mgMap(coords_df,
      memgene_list$memgene[, 1], 
      add.plot=TRUE,
      legend=TRUE)
box()

# Close device
dev.off()


# without legend
png(#here::here("./Figures_Tables/memgene/memgene_no_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_nolegend_var1_",
           today(),
               ".png"),
        width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords_df,
      memgene_list$memgene[, 1], 
      add.plot=TRUE,
      legend=FALSE)
box()

# Close device
dev.off()

  # MEM2-----
  
  # with legend
png(#here::here("./Figures_Tables/memgene/memgene_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_legend_var2_",
           today(),
           ".png"),
    width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
#base_map
mgMap(coords_df,
      memgene_list$memgene[, 2], 
      add.plot=TRUE,
      legend=TRUE)
box()

# Close device
dev.off()


# without legend
png(#here::here("./Figures_Tables/memgene/memgene_no_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_nolegend_var2_",
           today(),
               ".png"),
        width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords_df,
      memgene_list$memgene[, 2], 
      add.plot=TRUE,
      legend=FALSE)
box()

# Close device
dev.off()
  # MEM3-----
  
  # with legend
png(#here::here("./Figures_Tables/memgene/memgene_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_legend_var3_",
           today(),
           ".png"),
    width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
#base_map
mgMap(coords_df,
      memgene_list$memgene[, 3], 
      add.plot=TRUE,
      legend=TRUE)
box()

# Close device
dev.off()


# without legend
png(#here::here("./Figures_Tables/memgene/memgene_no_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_nolegend_var3_",
           today(),
               ".png"),
        width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords_df,
      memgene_list$memgene[, 3], 
      add.plot=TRUE,
      legend=FALSE)
box()

# Close device
dev.off()
}
```

### All individuals
```{r}
# Moran's Eigenvectors 12,20,5
basic_plot_mems("all_indiv",
                coords,
                memgene_all)
```

### 1 individual/population- RURAL
#### Urb = dist to CC
```{r}
# Moran's Eigenvectors 2,4
basic_plot_mems("rural_subsamp",
                coords_subsamp_rural,
                memgene_subsamp_rural)
```

#### Urb = urb score
```{r}
# Moran's Eigenvectors 3, 4, 8, 14
basic_plot_mems("rural_subsamp_usc",
                coords_subsamp_rural_usc,
                memgene_subsamp_rural_usc)
```

## Plots with terrain basemap
### All individuals
```{r}
# will do this for first 3 memgene vars (i.e., those which explain the most variation in spatial patterns)... there are 20 total, though

# create MEM dfs-----
mem1_all <- create_mem(
  memgene_all,
  coords,
  1) %>%
  dplyr::mutate("mem" = "Memgene Variable 1")

mem2_all <- create_mem(
  memgene_all,
  coords,
  2) %>%
  dplyr::mutate("mem" = "Memgene Variable 2")

mem3_all <- create_mem(
  memgene_all,
  coords,
  3) %>%
  dplyr::mutate("mem" = "Memgene Variable 3")

mems_all <- full_join(mem1_all,
                                mem2_all) %>%
  dplyr::full_join(., mem3_all) 


# make, export maps (4 memgene variables)-----
memgene_terrain_map(mems_all,
                    "1",
                    "all_indiv")

memgene_terrain_map(mems_all,
                    "2",
                    "all_indiv")

memgene_terrain_map(mems_all,
                    "3",
                    "all_indiv")

```

### 1 individual/population- RURAL
#### Urb = dist to CC
```{r}
# create MEM dfs-----
mem1_subsamp_rural <- create_mem(memgene_subsamp_rural,
                                 coords_subsamp_rural,
                                 1) %>%
  dplyr::mutate("mem" = "Memgene Variable 1")

mem2_subsamp_rural <- create_mem(memgene_subsamp_rural,
                                 coords_subsamp_rural,
                                 2) %>%
  dplyr::mutate("mem" = "Memgene Variable 2")

mems_subsamp_rural <- full_join(mem1_subsamp_rural,
                                mem2_subsamp_rural)


# make, export maps (2 memgene variables)-----
memgene_terrain_map(mems_subsamp_rural,
                    "1",
                    "rural_subsamp")

memgene_terrain_map(mems_subsamp_rural,
                    "2",
                    "rural_subsamp")
```

#### Urb = urb score
```{r}
# create MEM dfs-----
mem1_subsamp_rural_usc <- create_mem(
  memgene_subsamp_rural_usc,
  coords_subsamp_rural_usc,
  1) %>%
  dplyr::mutate("mem" = "Memgene Variable 1")

mem2_subsamp_rural_usc <- create_mem(
  memgene_subsamp_rural_usc,
  coords_subsamp_rural_usc,
  2) %>%
  dplyr::mutate("mem" = "Memgene Variable 2")

mem3_subsamp_rural_usc <- create_mem(
  memgene_subsamp_rural_usc,
  coords_subsamp_rural_usc,
  3) %>%
  dplyr::mutate("mem" = "Memgene Variable 3")

mem4_subsamp_rural_usc <- create_mem(
  memgene_subsamp_rural_usc,
  coords_subsamp_rural_usc,
  4) %>%
  dplyr::mutate("mem" = "Memgene Variable 4")

mems_subsamp_rural_usc <- full_join(mem1_subsamp_rural_usc,
                                mem2_subsamp_rural_usc) %>%
  dplyr::full_join(., mem3_subsamp_rural_usc) %>%
  dplyr::full_join(., mem4_subsamp_rural_usc)


# make, export maps (4 memgene variables)-----
memgene_terrain_map(mems_subsamp_rural_usc,
                    "1",
                    "rural_subsamp_usc")

memgene_terrain_map(mems_subsamp_rural_usc,
                    "2",
                    "rural_subsamp_usc")

memgene_terrain_map(mems_subsamp_rural_usc,
                    "3",
                    "rural_subsamp_usc")

memgene_terrain_map(mems_subsamp_rural_usc,
                    "4",
                    "rural_subsamp_usc")
```

## Re-import terrain maps, then create composite figures
### All individuals
```{r}
var1_all_indiv <- rasterGrob(readPNG(here::here("./Figures_Tables/memgene/all_indiv/memgene_terrain_var1_all_indiv.png")))

var2_all_indiv <- rasterGrob(readPNG(here::here("./Figures_Tables/memgene/all_indiv/memgene_terrain_var2_all_indiv.png")))

var3_all_indiv <- rasterGrob(readPNG(here::here("./Figures_Tables/memgene/all_indiv/memgene_terrain_var3_all_indiv.png")))


memgene_all_indiv_plots <- gridExtra::grid.arrange(var1_all_indiv, 
                        var2_all_indiv,
                        var3_all_indiv,
                        ncol = 3,
             left = textGrob("A", 
                            x = unit(1.75, "npc"), 
                            y = unit(0.98, "npc"),
                            gp = gpar(fontsize = 12)),
             center = textGrob("B", 
                            x = unit(1.1, "npc"), 
                            y = unit(1.96, "npc"),
                            gp = gpar(fontsize = 12)),
             right = textGrob("C", 
                            x = unit(-8.35, "npc"), 
                            y = unit(0.98, "npc"),
                            gp = gpar(fontsize = 12)))

ggsave("./Figures_Tables/memgene/all_indiv/main_figs.png", 
       plot = memgene_all_indiv_plots,
       width = 6.5, height = 3.5,
       units = "in",
       dpi = 400)
```

### 1 indiv/pop: RURAL
#### Urb = dist to CC
```{r}
var1_subsamp_rural <- rasterGrob(readPNG(here::here("./Figures_Tables/memgene/rural_subsamp/memgene_terrain_var1_rural_subsamp.png")))

var2_subsamp_rural <- rasterGrob(readPNG(here::here("./Figures_Tables/memgene/rural_subsamp/memgene_terrain_var2_rural_subsamp.png")))


memgene_subsamp_rural_plots <- gridExtra::grid.arrange(var1_subsamp_rural, 
                        var2_subsamp_rural,
                        ncol = 3,
             left = textGrob("A", 
                            x = unit(1.95, "npc"), 
                            y = unit(0.75, "npc"),
                            gp = gpar(fontsize = 12)),
             center = textGrob("B", 
                            x = unit(-0.89, "npc"), 
                            y = unit(0.75, "npc"),
                            gp = gpar(fontsize = 12)))

ggsave("./Figures_Tables/memgene/rural_subsamp/main_figs.png", 
       plot = memgene_subsamp_rural_plots,
       width = 6.5, height = 3.5,
       units = "in",
       dpi = 400)
```

#### Urb = urb score
```{r}
var1_subsamp_rural_usc <- rasterGrob(readPNG(here::here("./Figures_Tables/memgene/rural_subsamp_usc/memgene_terrain_var1_rural_subsamp_usc.png")))

var2_subsamp_rural_usc <- rasterGrob(readPNG(here::here("./Figures_Tables/memgene/rural_subsamp_usc/memgene_terrain_var2_rural_subsamp_usc.png")))

var3_subsamp_rural_usc <- rasterGrob(readPNG(here::here("./Figures_Tables/memgene/rural_subsamp_usc/memgene_terrain_var3_rural_subsamp_usc.png")))


memgene_subsamp_rural_usc_plots <- gridExtra::grid.arrange(var1_subsamp_rural_usc, 
                        var2_subsamp_rural_usc,
                        var3_subsamp_rural_usc,
                        ncol = 3,
             left = textGrob("A", 
                            x = unit(1.75, "npc"), 
                            y = unit(0.98, "npc"),
                            gp = gpar(fontsize = 12)),
             center = textGrob("B", 
                            x = unit(1.1, "npc"), 
                            y = unit(1.96, "npc"),
                            gp = gpar(fontsize = 12)),
             right = textGrob("C", 
                            x = unit(-8.35, "npc"), 
                            y = unit(0.98, "npc"),
                            gp = gpar(fontsize = 12)))

ggsave("./Figures_Tables/memgene/rural_subsamp_usc/main_figs.png", 
       plot = memgene_subsamp_rural_usc_plots,
       width = 6.5, height = 3.5,
       units = "in",
       dpi = 400)
```

# Export stats
## All individuals
```{r}
export_stats("all_indiv",
             memgene_all)
```

## 1 individual/population- RURAL
### Urb = dist to CC
```{r}
export_stats("rural_subsamp",
             memgene_subsamp_rural)
```

### Urb = urb score
```{r}
export_stats("rural_subsamp_usc",
             memgene_subsamp_rural_usc)
```

