# Set up notebook
## load libraries
```{r}
source("libraries.R")
```

# Create sfs
## Load vcf2sfs code
```{r}
# can't install this package so copied the R code into my own file, load it

source(here::here("./scripts/vcf2sfs_code.R"))
```

## get path to vcf
```{r}
vcf_address <- here::here("./results/pi_fst_TajD_estimates/pi/global_pi/global_pi_vcf-all/populations.all.vcf")
```

## Notes for understanding vcfR::maf()'s output

*Each row is a locus.
*nAllele = the number of different alleles sequenced at this
 locus. At most, it's 2n = 261*2=522 because it's a diploid-
 so there's at most 2 alleles per locus for each individual.
 It can drop to numbers like 50 if only, for example,
 50/2n=50/2*261=50/522=~10% of the loci were sequenced.
*Count = the number of times a specific allele was seen at
 this locus.
*Frequency = Count/nAlleles to give the total frequency of
 that specific allele.

There can be bars in the 1 bins on the x axis because of
 missing values. Some loci were only sequenced in a few of
 the individuals (depending on R), so if an allele was seen
 only once but its locus was sequenced in only ~25% of
 individuals, it could still have a frequency >0.05 and
 wind up in the "1 individual" bin.

This holds up because the mmaf = 0.05, R = 1 plot should
 have no missing data and a min minor allele frequency of
 5%, so the first individual should show up at
 0.05*2n=0.05*2*261=26 individuals. It does.
 
## Read the VCF file and the popmap file and create a gt object
### SFS for all individuals in 1 population
```{r}
mygt <- vcf2gt(vcf_address,
               here::here("./genomic_resources/pop_map3_all1pop_for_sfs.txt"))

# look at all unique populations
unique(mygt$popmap)
```

### SFS for URBAN or RURAL individuals in 1 population
#### Urbanization = dist to CC
```{r}
mygt_urb <- vcf2gt(vcf_address,
               here::here("./clean_data/pop_maps/pop_map3_UR_dist.txt"))

# look at all unique populations
unique(mygt_urb$popmap)
```

#### Urbanization = urbanization score
```{r}
mygt_urb2 <- vcf2gt(vcf_address,
               here::here("./clean_data/pop_maps/pop_map3_UR_usc.txt"))

# look at all unique populations
unique(mygt_urb2$popmap)
```

## Create sfs
### SFS for all individuals in 1 population
```{r}
sfs <- gt2sfs.raw(mygt,
           "1") %>%
        as.data.frame() %>%
  dplyr::filter(X1 != 0) %T>%
  write_delim(here::here("./clean_data/sfs/sfs.txt"))

```

### SFS for URBAN individuals in 1 population
#### Urbanization = dist to CC
```{r}
sfs_urb <- gt2sfs.raw(mygt_urb,
           "Urban") %>%
        as.data.frame() %>%
  dplyr::filter(Urban != 0) %T>%
  write_delim(here::here("./clean_data/sfs/sfs_urban.txt"))

```

#### Urbanization = urbanization score
```{r}
sfs_urb2 <- gt2sfs.raw(mygt_urb2,
           "Urban") %>%
        as.data.frame() %>%
  dplyr::filter(Urban != 0) %T>%
  write_delim(here::here("./clean_data/sfs/sfs_urban_urbscore.txt"))

```

### SFS for RURAL individuals in 1 population
#### Urbanization = dist to CC
```{r}
sfs_rur <- gt2sfs.raw(mygt_urb,
           "Rural") %>%
        as.data.frame() %>%
  dplyr::filter(Rural != 0) %T>%
  write_delim(here::here("./clean_data/sfs/sfs_rural.txt"))

```

#### Urbanization = urbanization score
```{r}
sfs_rur2 <- gt2sfs.raw(mygt_urb2,
           "Rural") %>%
        as.data.frame() %>%
  dplyr::filter(Rural != 0) %T>%
  write_delim(here::here("./clean_data/sfs/sfs_rural_urbscore.txt"))

```

## Prepare to export sfs
### SFS for all individuals in 1 population
```{r}
# get freqs for SFS
freqs_for_stairwayplot <- sfs %>%
                   dplyr::select(Freq) %>%
  pull()


freqs_string <- paste(freqs_for_stairwayplot,
                      collapse = " ")

# Write the string to a txt file
writeLines(freqs_string,
           here::here("./clean_data/sfs/sfs_freqs.txt"))

# length of SFS
length(freqs_for_stairwayplot)
```

### SFS for URBAN individuals in 1 population
#### Urbanization = dist to CC
```{r}
# get freqs for SFS
sfs_urb %>%
  dplyr::select(Freq) %>%
  pull() %>%
  paste(collapse = " ") %T>%
# Write the string to a txt file
  writeLines(here::here("./clean_data/sfs/sfs_freqs_urb.txt"))

# length of SFS
length(sfs_urb %>%
  dplyr::select(Freq) %>%
  pull())
```

#### Urbanization = urbanization score
```{r}
# get freqs for SFS
sfs_urb2 %>%
  dplyr::select(Freq) %>%
  pull() %>%
  paste(collapse = " ") %T>%
# Write the string to a txt file
  writeLines(here::here("./clean_data/sfs/sfs_freqs_urb_urbscore.txt"))

# length of SFS
length(sfs_urb2 %>%
  dplyr::select(Freq) %>%
  pull())
```

### SFS for RURAL individuals in 1 population
#### Urbanization = dist to CC
```{r}
# get freqs for SFS
sfs_rur %>%
  dplyr::select(Freq) %>%
  pull() %>%
  paste(collapse = " ") %T>%
# Write the string to a txt file
  writeLines(here::here("./clean_data/sfs/sfs_freqs_rur.txt"))

# length of SFS
length(sfs_rur %>%
  dplyr::select(Freq) %>%
  pull())
```

#### Urbanization = urbanization score
```{r}
# get freqs for SFS
sfs_rur2 %>%
  dplyr::select(Freq) %>%
  pull() %>%
  paste(collapse = " ") %T>%
# Write the string to a txt file
  writeLines(here::here("./clean_data/sfs/sfs_freqs_rur_urbscore.txt"))

# length of SFS
length(sfs_rur2 %>%
  dplyr::select(Freq) %>%
  pull())
```

# Stairway Plot figures
## Import stairway plot data
### mu = 1.0x10^-8 & generation time = 1 year
#### All sites in 1 population
```{r}
full_fig_1e8 <- read.delim(here::here("./results/StairwayPlot/mu_1e-8/gen_time_1yr/full_mmaf0.007813_mut_rate_1e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

### mu = 1.0x10^-8 & generation time = 1.5 years
#### All sites in 1 population
```{r}
full_fig_1e8_1.5yrs <- read.delim(here::here("./results/StairwayPlot/mu_1e-8/gen_time_1.5yr/full_mmaf0.007813_mut_rate_1e-8_gt1.5.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

### mu = 1.2x10^-8 & generation time = 1 year
#### All sites in 1 population
```{r}
full_fig <- read.delim(here::here("./results/StairwayPlot/mu_1.2e-8/gen_time_1yr/full_mmaf0.007813.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

#### Urban
##### Dist to CC
```{r}
urb_fig <- read.delim(here::here("./results/StairwayPlot/mu_1.2e-8/gen_time_1yr/urban_dist_mmaf0.007813.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

##### Urb Score
```{r}
urb_fig2 <- read.delim(here::here("./results/StairwayPlot/mu_1.2e-8/gen_time_1yr/urban_usc_mmaf0.007813.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

#### Rural
##### Dist to CC
```{r}
rur_fig <- read.delim(here::here("./results/StairwayPlot/mu_1.2e-8/gen_time_1yr/rural_dist_mmaf0.007813.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

##### Urb score
```{r}
rur_fig2 <- read.delim(here::here("./results/StairwayPlot/mu_1.2e-8/gen_time_1yr/rural_usc_mmaf0.007813.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

#### Merge dfs
##### Urb = dist to CC
```{r}
all_dist <- dplyr::full_join(urb_fig %>%
                    dplyr::mutate(Group = "Urban"),
                  rur_fig %>%
                    dplyr::mutate(Group = "Rural")) %>%
  dplyr::full_join(.,
                   full_fig %>%
                    dplyr::mutate(Group = "Entire population"))

# remove duplicate rows
all_dist %<>%
  distinct(.keep_all = TRUE)
```


##### Urb = urb score
```{r}
all_usc <- dplyr::full_join(urb_fig2 %>%
                    dplyr::mutate(Group = "Urban"),
                  rur_fig2 %>%
                    dplyr::mutate(Group = "Rural")) %>%
  dplyr::full_join(.,
                   full_fig %>%
                    dplyr::mutate(Group = "Entire population"))

# remove duplicate rows
all_usc %<>%
  distinct(.keep_all = TRUE)
```

### mu = 1.2x10^-8 & generation time = 1.5 years
#### All sites in 1 population
```{r}
full_fig_1.5yrs <- read.delim(here::here("./results/StairwayPlot/mu_1.2e-8/gen_time_1.5yr/full_mmaf0.007813_mut_rate_1.2e-8_gt1.5.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

### mu = 1.8x10^-8 & generation time = 1 year
#### All sites in 1 population
```{r}
full_fig_1.8e8 <- read.delim(here::here("./results/StairwayPlot/mu_1.8e-8/gen_time_1yr/full_mmaf0.007813_mut_rate_1.8e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

#### Urban
##### Dist to CC
```{r}
urb_fig_1.8e8 <- read.delim(here::here("./results/StairwayPlot/mu_1.8e-8/gen_time_1yr/urban_dist_mmaf0.007813_mut_rate_1.8e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

##### Urb Score
```{r}
urb_fig2_1.8e8 <- read.delim(here::here("./results/StairwayPlot/mu_1.8e-8/gen_time_1yr/urban_usc_mmaf0.007813_mut_rate_1.8e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

#### Rural
##### Dist to CC
```{r}
rur_fig_1.8e8 <- read.delim(here::here("./results/StairwayPlot/mu_1.8e-8/gen_time_1yr/rural_dist_mmaf0.007813_mut_rate_1.8e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

##### Urb score
```{r}
rur_fig2_1.8e8 <- read.delim(here::here("./results/StairwayPlot/mu_1.8e-8/gen_time_1yr/rural_usc_mmaf0.007813_mut_rate_1.8e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

#### Merge dfs
##### Urb = dist to CC
```{r}
all_dist_1.8e8 <- dplyr::full_join(urb_fig_1.8e8 %>%
                    dplyr::mutate(Group = "Urban"),
                  rur_fig_1.8e8 %>%
                    dplyr::mutate(Group = "Rural")) %>%
  dplyr::full_join(.,
                   full_fig_1.8e8 %>%
                    dplyr::mutate(Group = "Entire population"))

# remove duplicate rows
all_dist_1.8e8 %<>%
  distinct(.keep_all = TRUE)
```


##### Urb = urb score
```{r}
all_usc_1.8e8 <- dplyr::full_join(urb_fig2_1.8e8 %>%
                    dplyr::mutate(Group = "Urban"),
                  rur_fig2_1.8e8 %>%
                    dplyr::mutate(Group = "Rural")) %>%
  dplyr::full_join(.,
                   full_fig_1.8e8 %>%
                    dplyr::mutate(Group = "Entire population"))

# remove duplicate rows
all_usc_1.8e8 %<>%
  distinct(.keep_all = TRUE)
```

### mu = 1.8x10^-8 & generation time = 1.5 years
#### All sites in 1 population
```{r}
full_fig_1.8e8_1.5yr <- read.delim(here::here("./results/StairwayPlot/mu_1.8e-8/gen_time_1.5yr/full_mmaf0.007813_mut_rate_1.8e-8_gt1.5.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

### mu = 2.6x10^-8 & generation time = 1 year
#### All sites in 1 population
```{r}
full_fig_2.6e8 <- read.delim(here::here("./results/StairwayPlot/mu_2.6e-8/gen_time_1yr/full_mmaf0.007813_mut_rate_2.6e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

### mu = 2.6x10^-8 & generation time = 1.5 years
#### All sites in 1 population
```{r}
full_fig_2.6e8_1.5yr <- read.delim(here::here("./results/StairwayPlot/mu_2.6e-8/gen_time_1.5yr/full_mmaf0.007813_mut_rate_2.6e-8_gt1.5.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

## Create functions
### Export plots
#### All, urban, and rural separately
```{r}
# regular y axis
make_ne_plot_reg_y <- function(df, exported_filepath){
  
  max_y <- df %>%
    dplyr::select(Ne_median) %>%
    dplyr::filter(Ne_median == max(Ne_median)) %>%
    dplyr::slice(1) %>%
    dplyr::mutate(max = Ne_median*1.1) %>%
    dplyr::select(max) %>%
    as.numeric()
  
  # full plot with red horizontal line at 500 years ago
  full_plot <- 
    ggplot(
    df,
    aes(
      x = year,
      y = Ne_median))  +

    geom_point(size = 0.75) +

    # add vertical line at 12,000 yrs (glaciers retreat)
    geom_segment( 
      aes(
        x = 12000,
        xend = 12000,
        y = 0,
        yend = max_y),
      color = "blue",
      linetype = 5 ) +

    # add vertical line at 1500 yrs ago (corn cultivation begins around Grand River in southern Ontario)
      geom_segment( 
      aes(
        x = 1500,
        xend = 1500,
        y = 0,
        yend = max_y),
      color = "cyan",
      linetype = "longdash") +
    
    scale_x_continuous(labels = scales::comma,
                       limits = c(NA, 40000)) +
    scale_y_continuous(breaks=c(0, 25000, 50000, 75000, 100000, 125000, 150000),
                       labels = scales::comma,
                       limits = c(0, 
                                  max_y
                                  )
                       ) +
    xlab("Years ago") +
    ylab("Median Nₑ (1k individuals)") +
    theme_bw() +
    labs_pubr() +
    theme(plot.margin = unit(c(0.5,1,0.5,0.5), "cm"))
  
  ### Plot inset (past 2300 years)
  full_inset <- 
    full_plot +
    scale_x_continuous(limits = c(NA, 2300),
                       labels = scales::comma) +
    theme(axis.title = element_blank(),
          plot.background = element_rect(
            colour = "black", fill = "white", size = 1)) +

       # add vertical line around 700 yrs ago, when Iroquois start cultivating crops in Halton, Ontario
      geom_segment(
      aes(
        x = 723,
        xend = 723,
        y = 0,
        yend = max_y),
      color = "green",
      linetype = "longdash")  +

      # add vertical line ~400 yrs ago, when European diseases spread to Eastern North America
      geom_segment( 
      aes(
        x = 423,
        xend = 423,
        y = 0,
        yend = max_y),
      color = "orange",
      linetype = "twodash")   +
  
      # add vertical line ~100 yrs ago, when reforestation efforts begin in earnest
      geom_segment( 
      aes(
        x = 123,
        xend = 123,
        y = 0,
        yend = max_y),
      color = "tomato",
      linetype = "dotdash")
    
  
  ### Overlay inset in full figure
  cowplot::ggdraw() +
    draw_plot(full_plot) +
    draw_plot(full_inset,
              x = 0.5,
              y = 0.4, 
              width = 0.4,
              height = 0.5)
  
  
  #### Export
  ggsave(filename = here::here(exported_filepath),
         height = 5,
         width = 8,
         dpi = 200)
}
```

#### All plots together
```{r}
export_all_plots <- function(df_distCC, df_urbscore, mu, gen_time){
  
  # All plots will have regular y axis
  
  # first, find maximum y axis values
    max_y_distCC <- df_distCC %>%
    dplyr::select(Ne_median) %>%
    dplyr::filter(Ne_median == max(Ne_median)) %>%
    dplyr::slice(1) %>%
    dplyr::mutate(max = Ne_median*1.1) %>%
    dplyr::select(max) %>%
    as.numeric()
    
    max_y_urbscore <- df_urbscore %>%
    dplyr::select(Ne_median) %>%
    dplyr::filter(Ne_median == max(Ne_median)) %>%
    dplyr::slice(1) %>%
    dplyr::mutate(max = Ne_median*1.1) %>%
    dplyr::select(max) %>%
    as.numeric()
  
  # DISTANCE TO CC-----
  exported_filename_distCC <- paste0("./Figures_Tables/Stairway_plot/mu_",
                              mu,
                              "/gen_time_",
                              gen_time,
                              "/distance",
                              "/all_with_inset.png")
  
  all_plot <- ggplot(df_distCC)  +

    geom_point(aes(x = year,
                   y = Ne_median,
                   color = Group),
               size = 0.25) +
  
        
    # add vertical line at 12,000 yrs (glaciers retreat)
    geom_segment( 
      aes(
        x = 12000,
        xend = 12000,
        y = 0,
        yend = max_y_distCC),
      color = "blue",
      linetype = 5 ) +

    # add vertical line at 1500 yrs ago (corn cultivation begins around Grand River in southern Ontario)
      geom_segment( 
      aes(
        x = 1500,
        xend = 1500,
        y = 0,
        yend = max_y_distCC),
      color = "cyan",
      linetype = "longdash") +
    
    scale_x_continuous(labels = scales::comma,
                       limits = c(NA, 40000)) +
    scale_y_continuous(breaks=c(0, 25000, 50000, 75000, 100000, 125000, 150000),
                       labels = scales::comma,
                       limits = c(0, 
                                  max_y_distCC
                                  )) +
    xlab("Years ago") +
    ylab("Median Nₑ (1k individuals)") +
    theme_bw() +
    labs_pubr() +
  scale_color_grey() +
    theme(plot.margin = unit(c(0.5,1,0.5,0.5), "cm"),
          legend.position = "top") +
  # make legend point size larger
  guides(colour = guide_legend(
    override.aes = list(size=3)))
  

  ### Plot inset (past 2300 years)
all_plot_inset <- all_plot +
    scale_x_continuous(limits = c(NA, 2300),
                       labels = scales::comma) +
  
         # add vertical line around 700 yrs ago, when Iroquois start cultivating crops in Halton, Ontario
      geom_segment(
      aes(
        x = 723,
        xend = 723,
        y = 0,
        yend = max_y_distCC),
      color = "green",
      linetype = "longdash")  +

      # add vertical line ~400 yrs ago, when European diseases spread to Eastern North America
      geom_segment( 
      aes(
        x = 423,
        xend = 423,
        y = 0,
        yend = max_y_distCC),
      color = "orange",
      linetype = "twodash")   +
  
      # add vertical line ~100 yrs ago, when reforestation efforts begin in earnest
      geom_segment( 
      aes(
        x = 123,
        xend = 123,
        y = 0,
        yend = max_y_distCC),
      color = "tomato",
      linetype = "dotdash") +
  
    theme(axis.title = element_blank(),
          plot.background = element_rect(
            colour = "black",
            fill = "white",
            size = 1),
          legend.position = "none")

  
  
#### Export
ggsave(plot =
### Overlay inset in full figure
cowplot::ggdraw() +
    draw_plot(all_plot) +
    draw_plot(all_plot_inset,
              x = 0.55,
              y = 0.3, 
              width = 0.4,
              height = 0.5),     
           filename = here::here(exported_filename_distCC),
         height = 5,
         width = 9,
         dpi = 200)


  # URBANIZATION SCORE-----

 exported_filename_urbscore <- paste0("./Figures_Tables/Stairway_plot/mu_",
                              mu,
                              "/gen_time_",
                              gen_time,
                              "/urb_score",
                              "/all_with_inset.png")

all_plot_usc <- ggplot(df_urbscore)  +

    geom_point(aes(x = year,
                   y = Ne_median,
                   color = Group),
               size = 0.25) +
  
      # add vertical line at 12,000 yrs (glaciers retreat)
    geom_segment( 
      aes(
        x = 12000,
        xend = 12000,
        y = 0,
        yend = max_y_urbscore),
      color = "blue",
      linetype = 5 ) +

    # add vertical line at 1500 yrs ago (corn cultivation begins around Grand River in southern Ontario)
      geom_segment( 
      aes(
        x = 1500,
        xend = 1500,
        y = 0,
        yend = max_y_urbscore),
      color = "cyan",
      linetype = "longdash") +
  
    scale_x_continuous(labels = scales::comma,
                       limits = c(NA, 40000)) +
    scale_y_continuous(breaks=c(0, 25000, 50000, 75000, 100000, 125000, 150000),
                       labels = scales::comma,
                       limits = c(0, 
                                  max_y_urbscore)) +
    xlab("Years ago") +
    ylab("Median Nₑ (1k individuals)") +
  theme_bw() +
    labs_pubr() +
  scale_color_grey() +
    theme(plot.margin = unit(c(0.5,1,0.5,0.5), "cm"),
          legend.position = "top") +
  # make legend point size larger
  guides(colour = guide_legend(
    override.aes = list(size=3)))
  

  ### Plot inset (past 2300 years)
all_plot_inset_usc <- all_plot_usc +
    scale_x_continuous(limits = c(NA, 2300),
                       labels = scales::comma) +
  
         # add vertical line around 700 yrs ago, when Iroquois start cultivating crops in Halton, Ontario
      geom_segment(
      aes(
        x = 723,
        xend = 723,
        y = 0,
        yend = max_y_urbscore),
      color = "green",
      linetype = "longdash")  +

      # add vertical line ~400 yrs ago, when European diseases spread to Eastern North America
      geom_segment( 
      aes(
        x = 423,
        xend = 423,
        y = 0,
        yend = max_y_urbscore),
      color = "orange",
      linetype = "twodash")   +
  
      # add vertical line ~100 yrs ago, when reforestation efforts begin in earnest
      geom_segment( 
      aes(
        x = 123,
        xend = 123,
        y = 0,
        yend = max_y_urbscore),
      color = "tomato",
      linetype = "dotdash") +
  
    theme(axis.title = element_blank(),
          plot.background = element_rect(
            colour = "black",
            fill = "white",
            size = 1),
          legend.position = "none")

  
  
#### Export
ggsave(plot =
### Overlay inset in full figure
cowplot::ggdraw() +
    draw_plot(all_plot_usc) +
    draw_plot(all_plot_inset_usc,
              x = 0.55,
              y = 0.3, 
              width = 0.4,
              height = 0.5),     
  filename = here::here(exported_filename_urbscore),
         height = 5,
         width = 9,
         dpi = 200)
}
```

### Find Ne highs, lows
#### Median Ne
```{r}
Ne_stats <- function(Ne_df){

max_ne <- max(Ne_df$Ne_median)  

# find year of max Ne
max_ne_years <- Ne_df %>%
  dplyr::filter(Ne_median == max_ne) %>%
   dplyr::select(year) %>%
   dplyr::mutate(year = round(year, 0))

print(paste0("Years of max Ne: ", max_ne_years))


# find median Ne around 750 years ago
median_ne_750YA <- Ne_df %>%
  dplyr::filter(year > 750 & year < 751) %>%
  dplyr::select(Ne_median) %>%
  dplyr::mutate(Ne_median = round(Ne_median, 0)) %>%
  slice(1) %>%
  as.numeric()

print(paste0("Median Ne ~750 years ago: ", median_ne_750YA*1000))

min_year <- min(Ne_df$year)  

# find most recent median Ne
most_recent_ne <- Ne_df %>%
  dplyr::filter(year == min_year) %>%
  dplyr::select(Ne_median) %>%
  dplyr::mutate(Ne_median = round(Ne_median, 0)) %>%
  slice(1) %>%
  as.numeric()

print(paste0("Most recent median Ne: ", most_recent_ne*1000))


# find % decrease from max to min
perc_decrease <- round(((most_recent_ne-median_ne_750YA)/median_ne_750YA ) * 100, 3)

print(paste0("% decrease in Ne: ", perc_decrease))
}

```

#### 97.5% conf int Ne
```{r}
Ne_stats_CI <- function(Ne_df){

min_year <- min(Ne_df$year)  


# find most recent 2.5% CI Ne
most_recent_ne_2 <- Ne_df %>%
  dplyr::filter(year == min_year) %>%
  dplyr::select(Ne_2.5.) %>%
  dplyr::mutate(Ne_2.5. = round(Ne_2.5., 0)) %>%
  slice(1) %>%
  as.numeric()

# find most recent 97.5% CI Ne
most_recent_ne_97 <- Ne_df %>%
  dplyr::filter(year == min_year) %>%
  dplyr::select(Ne_97.5.) %>%
  dplyr::mutate(Ne_97.5. = round(Ne_97.5., 0)) %>%
  slice(1) %>%
  as.numeric()


print(paste0("Most recent Ne CI: [",
             most_recent_ne_2*1000,
             ",",
             most_recent_ne_97*1000,
             "]"))

}

```

## Export plots
### Mu = 1.0e-8 & gen time = 1 yr
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_1e8,
                   "./Figures_Tables/Stairway_plot/mu_1e-8/gen_time_1yr/full_with_inset.png")
```

### Mu = 1.0e-8 & gen time = 1.5 yrs
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_1e8_1.5yrs,
                   "./Figures_Tables/Stairway_plot/mu_1e-8/gen_time_1.5yr/full_with_inset.png")
```

### Mu = 1.2e-8 & gen time = 1 yr
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig,
                   "./Figures_Tables/Stairway_plot/mu_1.2e-8/gen_time_1yr/full_with_inset.png")
```

#### Export all plots
```{r}
export_all_plots(df_distCC = all_dist,
                 df_urbscore = all_usc,
                 mu = "1.2e-8",
                 gen_time = "1yr")
```

### Mu = 1.2e-8 & gen time = 1.5 yrs
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_1.5yrs,
                   "./Figures_Tables/Stairway_plot/mu_1.2e-8/gen_time_1.5yr/full_with_inset.png")
```

### Mu = 1.8e-8 & gen time = 1 yr
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_1.8e8,
                   "./Figures_Tables/Stairway_plot/mu_1.8e-8/gen_time_1yr/full_with_inset.png")
```

#### Export main plot WITH conf intervals
```{r}
#max_y <- max(urb_fig_1.8e8$Ne_97.5./1000)
max_y <- 300

ggplot(
    full_fig_1.8e8,
    aes(x = year))  +

    geom_point(aes(y = Ne_median/1000),
               size = 0.75,
               color = "black") +
  
    geom_point(aes(y = Ne_97.5./1000),
               size = 0.75,
               color = "lightgrey") +
  
      geom_point(aes(y = Ne_2.5./1000),
               size = 0.75,
               color = "lightgrey") +
  
     geom_line(aes(y = Ne_median/1000),
               size = 0.75,
               color = "black") +
  
    geom_line(aes(y = Ne_97.5./1000),
               size = 0.75,
               color = "lightgrey") +
  
      geom_line(aes(y = Ne_2.5./1000),
               size = 0.75,
               color = "lightgrey") +

    # add vertical line at 12,000 yrs (glaciers retreat)
    geom_segment( 
      aes(
        x = 12000,
        xend = 12000,
        y = 0,
        yend = max_y),
      color = "blue",
      linetype = 5 ) +

    # add vertical line at 1500 yrs ago (corn cultivation begins around Grand River in southern Ontario)
      geom_segment( 
      aes(
        x = 1500,
        xend = 1500,
        y = 0,
        yend = max_y),
      color = "cyan",
      linetype = "longdash") +
    
    scale_x_continuous(labels = scales::comma,
                       limits = c(NA, 30000)) +
    scale_y_continuous(
      #breaks=c(0, 25000, 50000, 75000, 100000, 125000, 150000),
                       labels = scales::comma,
                       limits = c(0, 
                                  max_y
                                  )
                       ) +
    xlab("Years ago") +
    ylab("Median Nₑ (million individuals)") +
    theme_bw() +
    labs_pubr() +
    theme(plot.margin = unit(c(0.5,1,0.5,0.5), "cm"))
  
ggsave("./Figures_Tables/Stairway_plot/mu_1.8e-8/gen_time_1yr/distance/entirepop_conf_ints.png",
       width = 17,
       height = 9,
       units = "cm")
```


#### Export urb plot WITH conf intervals
```{r}
# max_y_urb <- max(urb_fig_1.8e8$Ne_97.5./1000)
max_y_urb <- 300

ggplot(
    urb_fig_1.8e8,
    aes(x = year))  +

    geom_point(aes(y = Ne_median/1000),
               size = 0.75,
               color = "black") +
  
    geom_point(aes(y = Ne_97.5./1000),
               size = 0.75,
               color = "lightgrey") +
  
      geom_point(aes(y = Ne_2.5./1000),
               size = 0.75,
               color = "lightgrey") +
  
     geom_line(aes(y = Ne_median/1000),
               size = 0.75,
               color = "black") +
  
    geom_line(aes(y = Ne_97.5./1000),
               size = 0.75,
               color = "lightgrey") +
  
      geom_line(aes(y = Ne_2.5./1000),
               size = 0.75,
               color = "lightgrey") +

    # add vertical line at 12,000 yrs (glaciers retreat)
    geom_segment( 
      aes(
        x = 12000,
        xend = 12000,
        y = 0,
        yend = max_y_urb),
      color = "blue",
      linetype = 5 ) +

    # add vertical line at 1500 yrs ago (corn cultivation begins around Grand River in southern Ontario)
      geom_segment( 
      aes(
        x = 1500,
        xend = 1500,
        y = 0,
        yend = max_y_urb),
      color = "cyan",
      linetype = "longdash") +
    
    scale_x_continuous(labels = scales::comma,
                       limits = c(NA, 30000)) +
    scale_y_continuous(
      #breaks=c(0, 25000, 50000, 75000, 100000, 125000, 150000),
                       labels = scales::comma,
                       limits = c(0, 
                                  max_y_urb
                                  )
                       ) +
    xlab("Years ago") +
    ylab("Median Nₑ (million individuals)") +
    theme_bw() +
    labs_pubr() +
    theme(plot.margin = unit(c(0.5,1,0.5,0.5), "cm"))
  
ggsave("./Figures_Tables/Stairway_plot/mu_1.8e-8/gen_time_1yr/distance/urb_conf_ints.png",
       width = 17,
       height = 9,
       units = "cm")
```

#### Export rur plot WITH conf intervals
```{r}
# max_y_rur <- max(rur_fig_1.8e8$Ne_97.5./1000)
max_y_rur <- 300

ggplot(
    rur_fig_1.8e8,
    aes(x = year))  +

    geom_point(aes(y = Ne_median/1000),
               size = 0.75,
               color = "black") +
  
    geom_point(aes(y = Ne_97.5./1000),
               size = 0.75,
               color = "lightgrey") +
  
      geom_point(aes(y = Ne_2.5./1000),
               size = 0.75,
               color = "lightgrey") +
  
     geom_line(aes(y = Ne_median/1000),
               size = 0.75,
               color = "black") +
  
    geom_line(aes(y = Ne_97.5./1000),
               size = 0.75,
               color = "lightgrey") +
  
      geom_line(aes(y = Ne_2.5./1000),
               size = 0.75,
               color = "lightgrey") +

    # add vertical line at 12,000 yrs (glaciers retreat)
    geom_segment( 
      aes(
        x = 12000,
        xend = 12000,
        y = 0,
        yend = max_y_rur),
      color = "blue",
      linetype = 5 ) +

    # add vertical line at 1500 yrs ago (corn cultivation begins around Grand River in southern Ontario)
      geom_segment( 
      aes(
        x = 1500,
        xend = 1500,
        y = 0,
        yend = max_y_rur),
      color = "cyan",
      linetype = "longdash") +
    
    scale_x_continuous(labels = scales::comma,
                       limits = c(NA, 30000)) +
    scale_y_continuous(
      #breaks=c(0, 25000, 50000, 75000, 100000, 125000, 150000),
                       labels = scales::comma,
                       limits = c(0, 
                                  max_y_rur
                                  )
                       ) +
    xlab("Years ago") +
    ylab("Median Nₑ (million individuals)") +
    theme_bw() +
    labs_pubr() +
    theme(plot.margin = unit(c(0.5,1,0.5,0.5), "cm"))
  
ggsave("./Figures_Tables/Stairway_plot/mu_1.8e-8/gen_time_1yr/distance/rur_conf_ints.png",
       width = 17,
       height = 9,
       units = "cm")
```

#### Export all plots
```{r}
export_all_plots(df_distCC = all_dist_1.8e8,
                 df_urbscore = all_usc_1.8e8,
                 mu = "1.8e-8",
                 gen_time = "1yr")
```


### Mu = 1.8e-8 & gen time = 1.5 yrs
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_1.8e8_1.5yr,
                   "./Figures_Tables/Stairway_plot/mu_1.8e-8/gen_time_1.5yr/full_with_inset.png")
```

### Mu = 2.6e-8 & gen time = 1 yr
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_2.6e8,
                   "./Figures_Tables/Stairway_plot/mu_2.6e-8/gen_time_1yr/full_with_inset.png")
```

### Mu = 2.6e-8 & gen time = 1.5 yrs
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_2.6e8_1.5yr,
                   "./Figures_Tables/Stairway_plot/mu_2.6e-8/gen_time_1.5yr/full_with_inset.png")
```

## Calculate Ne stats
### mu = 1.8 x 10^-8, gen time = 1 yr
```{r}
Ne_stats(full_fig_1.8e8)
Ne_stats(urb_fig_1.8e8)
#Ne_stats(urb_fig2_1.8e8)
Ne_stats(rur_fig_1.8e8)
#Ne_stats(rur_fig2_1.8e8)

Ne_stats_CI(full_fig_1.8e8)
Ne_stats_CI(urb_fig_1.8e8)
Ne_stats_CI(rur_fig_1.8e8)

```
