# Set up notebook
## Load libraries & functions
```{r}
source(here::here("libraries.R"))
```

# Visualize Tajima's D across the genome
## Import files
### Global (all sites in one large population)
#### Window size: 1 million bp
```{r}
global_taj <- read.delim(
  here::here("./results/pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/global_TajD_window_1mil/global_TajD_window_1mil.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(global_taj$TajimaD)
```

#### Window size: 100,000 bp
```{r}
global_taj_100k <- read.delim(
  here::here("./results/pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/global_TajD_window_100k/global_TajD_window_100k.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(global_taj_100k$TajimaD)
```

### Urban (based on distance to CC)
#### Window size: 1 million bp
```{r}
urb_taj <- read.delim(
  here::here("./results/pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Urban_dist_TajD_window_1mil/Urban_dist_TajD_window_1mil.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(urb_taj$TajimaD)
```

#### Window size: 100,000 bp
```{r}
urb_taj_100k <- read.delim(
  here::here("./results/pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Urban_dist_TajD_window_100k/Urban_dist_TajD_window_100k.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(urb_taj_100k$TajimaD)
```

### Rural (based on distance to CC)
#### Window size: 1 million bp
```{r}
rur_taj <- read.delim(
  here::here("./results/pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Rural_dist_TajD_window_1mil/Rural_dist_TajD_window_1mil.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(rur_taj$TajimaD)
```

#### Window size: 100,000 bp
```{r}
rur_taj_100k <- read.delim(
  here::here("./results/pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Rural_dist_TajD_window_100k/Rural_dist_TajD_window_100k.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(rur_taj_100k$TajimaD)
```

### Urban (based on urb score)
#### Window size: 1 million bp
```{r}
urb_taj_usc <- read.delim(
  here::here("./results/pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Urban_usc_TajD_window_1mil/Urban_usc_TajD_window_1mil.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(urb_taj_usc$TajimaD)
```

#### Window size: 100,000 bp
```{r}
urb_taj_100k_usc <- read.delim(
  here::here("./results/pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Urban_usc_TajD_window_100k/Urban_usc_TajD_window_100k.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(urb_taj_100k_usc$TajimaD)
```

### Rural (based on urb score)
#### Window size: 1 million bp
```{r}
rur_taj_usc <- read.delim(
  here::here("./results/pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Rural_usc_TajD_window_1mil/Rural_usc_TajD_window_1mil.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(rur_taj_usc$TajimaD)
```

#### Window size: 100,000 bp
```{r}
rur_taj_100k_usc <- read.delim(
  here::here("./results/pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Rural_usc_TajD_window_100k/Rural_usc_TajD_window_100k.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(rur_taj_100k_usc$TajimaD)
```

## Visualize
### Global (all sites in one large population)
#### Window size: 1 million bp
```{r}
global_taj_plot <- ggplot(global_taj,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(global_taj$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
    scale_y_continuous(limits = c(-2,4)) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=150, 
           y=3,
           size = 4.5,
           label = paste("Mean:",
                         round(mean(global_taj$TajimaD),3)))

global_taj_plot

# add "0" x axis mark
global_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(global_taj_plot, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
global_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(global_taj_plot, 
                   x = 0.87,  y = 3.2, as.numeric(sum(global_taj$BIN_START)),
                   size = 12))


global_taj_plot



mean(global_taj$TajimaD)
sum(global_taj$BIN_START) 
```

#### Window size: 100,000 bp
```{r}
global_taj_plot2 <- ggplot(global_taj_100k,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(global_taj_100k$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
  scale_y_continuous(limits = c(-2,4)) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=400, 
           y=3.5,
           size = 4.5,
           label = paste("Mean:",
                         round(mean(global_taj_100k$TajimaD),3)))

global_taj_plot2

# add "0" x axis mark
global_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(global_taj_plot2, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
global_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(global_taj_plot2, 
                   x = 0.88,  y = 3.2, as.numeric(sum(global_taj_100k$BIN_START)),
                   size = 12))


global_taj_plot2



mean(global_taj_100k$TajimaD)
sum(global_taj_100k$BIN_START) 
```

### Urban (based on distance to CC)
#### Window size: 1 million bp
```{r}
urb_taj_plot <- ggplot(urb_taj,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(urb_taj$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
    scale_y_continuous(limits = c(-2,4)) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=150, 
           y=3,
           size = 4.5,
           label = paste("Mean:",
                         round(mean(urb_taj$TajimaD),3)))

urb_taj_plot

# add "0" x axis mark
urb_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
urb_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot, 
                    x = 0.87,  y = 3.2, as.numeric(sum(urb_taj$BIN_START)),
                   size = 12))


urb_taj_plot



mean(urb_taj$TajimaD)
sum(urb_taj$BIN_START) 
```


#### Window size: 100,000 bp
```{r}
urb_taj_plot2 <- ggplot(urb_taj_100k,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(urb_taj_100k$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
    scale_y_continuous(limits = c(-2,4)) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=400, 
           y=3.5,
           size = 4.5,
           label = paste("Mean:",
                         round(mean(urb_taj_100k$TajimaD),3)))

urb_taj_plot2

# add "0" x axis mark
urb_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot2, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
urb_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot2, 
                   x = 0.88,  y = 3.2, format(as.numeric(sum(urb_taj_100k$BIN_START)), scientific = TRUE),
                   size = 12))


urb_taj_plot2



mean(urb_taj_100k$TajimaD)
sum(urb_taj_100k$BIN_START) 
```


### Rural (based on distance to CC)
#### Window size: 1 million bp
```{r}
rur_taj_plot <- ggplot(rur_taj,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(rur_taj$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
    scale_y_continuous(limits = c(-2,4)) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=150, 
           y=3,
           size = 4.5,
           label = paste("Mean:",
                         round(mean(rur_taj$TajimaD),3)))

rur_taj_plot

# add "0" x axis mark
rur_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
rur_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot, 
                    x = 0.87,  y = 3.2, as.numeric(sum(rur_taj$BIN_START)),
                   size = 12))


rur_taj_plot


mean(rur_taj$TajimaD)
sum(rur_taj$BIN_START) 
```


#### Window size: 100,000 bp
```{r}
rur_taj_plot2 <- ggplot(rur_taj_100k,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(rur_taj_100k$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
    scale_y_continuous(limits = c(-2,4)) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=400, 
           y=3.5,
           size = 4.5,
           label = paste("Mean:",
                         round(mean(rur_taj_100k$TajimaD),3)))

rur_taj_plot2

# add "0" x axis mark
rur_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot2, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
rur_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot2, 
                   x = 0.88,  y = 3.2, format(as.numeric(sum(rur_taj_100k$BIN_START)), scientific = TRUE),
                   size = 12))


rur_taj_plot2



mean(rur_taj_100k$TajimaD)
sum(rur_taj_100k$BIN_START) 
```

### Urban (based on urb score)
#### Window size: 1 million bp
```{r}
urb_taj_plot_usc <- ggplot(urb_taj_usc,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(urb_taj_usc$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
    scale_y_continuous(limits = c(-2,4)) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=150, 
           y=3,
           size = 4.5,
           label = paste("Mean:",
                         round(mean(urb_taj_usc$TajimaD),3)))

urb_taj_plot_usc

# add "0" x axis mark
urb_taj_plot_usc <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot_usc, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
urb_taj_plot_usc <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot_usc, 
                    x = 0.87,  y = 3.2, as.numeric(sum(urb_taj_usc$BIN_START)),
                   size = 12))


urb_taj_plot_usc



mean(urb_taj_usc$TajimaD)
sum(urb_taj_usc$BIN_START) 
```


#### Window size: 100,000 bp
```{r}
urb_taj_plot2_usc <- ggplot(urb_taj_100k_usc,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(urb_taj_100k_usc$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
    scale_y_continuous(limits = c(-2,4)) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=400, 
           y=3.5,
           size = 4.5,
           label = paste("Mean:",
                         round(mean(urb_taj_100k_usc$TajimaD),3)))

urb_taj_plot2_usc

# add "0" x axis mark
urb_taj_plot2_usc <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot2_usc, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
urb_taj_plot2_usc <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot2_usc, 
                   x = 0.88,  y = 3.2, as.numeric(sum(urb_taj_100k_usc$BIN_START)),
                   size = 12))


urb_taj_plot2_usc



mean(urb_taj_100k_usc$TajimaD)
sum(urb_taj_100k_usc$BIN_START) 
```


### Rural (based on urb score)
#### Window size: 1 million bp
```{r}
rur_taj_plot_usc <- ggplot(rur_taj_usc,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(rur_taj_usc$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
    scale_y_continuous(limits = c(-2.1,4)) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=150, 
           y=3,
           size = 4.5,
           label = paste("Mean:",
                         round(mean(rur_taj_usc$TajimaD),3)))

rur_taj_plot_usc

# add "0" x axis mark
rur_taj_plot_usc <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot_usc, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
rur_taj_plot_usc <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot_usc, 
                    x = 0.87,  y = 3.2, as.numeric(sum(rur_taj_usc$BIN_START)),
                   size = 12))


rur_taj_plot_usc



mean(rur_taj_usc$TajimaD)
sum(rur_taj_usc$BIN_START) 
```

#### Window size: 100,000 bp
```{r}
rur_taj_plot2_usc <- ggplot(rur_taj_100k_usc,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(rur_taj_100k_usc$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
    scale_y_continuous(limits = c(-2,4)) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=400, 
           y=3.5,
           size = 4.5,
           label = paste("Mean:",
                         round(mean(rur_taj_100k_usc$TajimaD),3)))

rur_taj_plot2_usc

# add "0" x axis mark
rur_taj_plot2_usc <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot2_usc, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
rur_taj_plot2_usc <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot2_usc, 
                   x = 0.88,  y = 3.2, format(as.numeric(sum(rur_taj_100k_usc$BIN_START)), scientific = TRUE),
                   size = 12))


rur_taj_plot2_usc



mean(rur_taj_100k_usc$TajimaD)
sum(rur_taj_100k_usc$BIN_START) 
```

### All 1 mil bp
```{r}
global_taj_plot + 
  urb_taj_plot + 
  urb_taj_plot_usc + 
  rur_taj_plot + 
  rur_taj_plot_usc +
  plot_annotation(tag_levels = 'A') +
  plot_layout(ncol = 3) +
  patchwork::plot_layout(axes = "collect_y",
                         axis_titles = "collect_y")


ggsave(here::here("./Figures_Tables/tajimas_d/tajd_1mil.png"),
       width = 11,
       height = 7,
       units = "in")
```

### All 100k bp
```{r}
global_taj_plot2 + 
  urb_taj_plot2 + 
  urb_taj_plot2_usc + 
  rur_taj_plot2 + 
  rur_taj_plot2_usc +
  plot_annotation(tag_levels = 'A') +
  plot_layout(ncol = 3) +
  patchwork::plot_layout(axes = "collect_y",
                         axis_titles = "collect_y")


ggsave(here::here("./Figures_Tables/tajimas_d/tajd_100k.png"),
       width = 11,
       height = 7,
       units = "in")
```


# Significance testing: Are the Tajima's D values significantly different from zero?
## Create function
```{r}
t_test_tajd <- function(tajd_df){
 
   # create random normal distribution with a mean of zero and the same SD as dataset
  set.seed(1)
  
  norm_distr <- rnorm(n = nobs(tajd_df),
                    mean = 0,
                    sd = sd(tajd_df[,"TajimaD"])) %>%
  as.data.frame() %>%
  dplyr::rename("simulated_value" = 1)

  # create df with both sets of observations
  combined_data <- cbind(tajd_df %>%
                           dplyr::select(TajimaD),
                         norm_distr)
  
  # t-test
  ttest <- t.test(combined_data %>%
                    dplyr::select(TajimaD),
                  combined_data %>%
                    dplyr::select(simulated_value))
  
  return(ttest)

}
```

## Global
```{r}
global_ttest <- t_test_tajd(global_taj)
```

## Urban
### City_dist
#### 1 mil bp
```{r}
urban_1mil_ttest <- t_test_tajd(urb_taj)
```

#### 100k bp
```{r}
urban_100k_ttest <- t_test_tajd(urb_taj_100k)

# data:  combined_data %>% dplyr::select(TajimaD) and combined_data %>% dplyr::select(simulated_value)
# t = -7.6843, df = 1733.6, p-value = 2.558e-14
# alternative hypothesis: true difference in means is not equal to 0
# 95 percent confidence interval:
#  -0.4308573 -0.2556369
# sample estimates:
#   mean of x   mean of y 
# -0.35951650 -0.01626939
```

### Urb_score
#### 1 mil bp
```{r}
urban_1mil_ttest_usc <- t_test_tajd(urb_taj_usc)
```

#### 100k bp
```{r}
urban_100k_ttest_usc <- t_test_tajd(urb_taj_100k_usc)

# data:  combined_data %>% dplyr::select(TajimaD) and combined_data %>% dplyr::select(simulated_value)
# t = -7.8198, df = 1729.7, p-value = 9.127e-15
# alternative hypothesis: true difference in means is not equal to 0
# 95 percent confidence interval:
#  -0.4362091 -0.2612697
# sample estimates:
#  mean of x  mean of y 
# -0.3634625 -0.0147231 
```

## Rural
### City_dist
#### 1 mil bp
```{r}
rural_1mil_ttest <- t_test_tajd(rur_taj)
```

#### 100k bp
```{r}
rural_100k_ttest <- t_test_tajd(rur_taj_100k)

# data:  combined_data %>% dplyr::select(TajimaD) and combined_data %>% dplyr::select(simulated_value)
# t = -7.6723, df = 1701.6, p-value = 2.827e-14
# alternative hypothesis: true difference in means is not equal to 0
# 95 percent confidence interval:
#  -0.4467366 -0.2648296
# sample estimates:
#   mean of x   mean of y 
# -0.36714839 -0.01136528 
```

### Urb_score
#### 1 mil bp
```{r}
rural_1mil_ttest_usc <- t_test_tajd(rur_taj_usc)
```

#### 100k bp
```{r}
rural_100k_ttest_usc <- t_test_tajd(rur_taj_100k_usc)

# data:  combined_data %>% dplyr::select(TajimaD) and combined_data %>% dplyr::select(simulated_value)
# t = -7.8536, df = 1709.4, p-value = 7.082e-15
# alternative hypothesis: true difference in means is not equal to 0
# 95 percent confidence interval:
#  -0.4481927 -0.2690658
# sample estimates:
#   mean of x   mean of y 
# -0.37454920 -0.01591998 
```

# Export means with p-values
```{r}
tajd_table <- data.frame(
  Group = c("Global",
            "Urban",
            "Urban",
            "Rural",
            "Rural"),
  Urbanization = c("-",
                   "Distance to City Center",
                   "Urbanization Score",
                   "Distance to City Center",
                   "Urbanization Score"),
  TajD_window_1mil = c(
    round(mean(global_taj$TajimaD),3),
    round(mean(urb_taj$TajimaD),3),
    round(mean(rur_taj$TajimaD),3),
    round(mean(urb_taj_usc$TajimaD),3),
    round(mean(rur_taj_usc$TajimaD),3)
    ),
  TajD_window_100k = c(
    round(mean(global_taj_100k$TajimaD),3),
    round(mean(urb_taj_100k$TajimaD),3),
    round(mean(rur_taj_100k$TajimaD),3),
    round(mean(urb_taj_100k_usc$TajimaD),3),
    round(mean(rur_taj_100k_usc$TajimaD),3)
  ),
  p_1mil_bp = c(
    round(global_ttest$p.value, 3),
    round(urban_1mil_ttest$p.value, 3),
    round(urban_1mil_ttest_usc$p.value, 3),
    round(rural_1mil_ttest$p.value, 3),
    round(rural_1mil_ttest_usc$p.value, 3)
  ),
    p_100k_bp = c(
    round(global_ttest$p.value, 3),
    round(urban_100k_ttest$p.value, 3),
    round(urban_100k_ttest_usc$p.value, 3),
    round(rural_100k_ttest$p.value, 3),
    round(rural_100k_ttest_usc$p.value, 3)
  )) %>%
  dplyr::mutate(p_1mil_bp = ifelse(
      p_1mil_bp < 0.001, 
      "<0.001",
      p_1mil_bp)) %>%
  dplyr::mutate(p_100k_bp = ifelse(
      p_100k_bp < 0.001, 
      "<0.001",
      p_100k_bp))

tajd_table %>%
  # dplyr::rename("Tajima's D" = 3
  #   #"Window = 1,000,000 bp" = 3,
  #              # "Window = 100,000 bp" = 4
  #              ) %>%
  flextable::flextable() %>%
  # flextable::italic(i = 1, j = 4, part = "header") %>%
  flextable::align(align = "center", i = c(1:5), part = "body") %>%
  flextable::align(align = "left", j = 1:2, part = "all") %>%
  flextable::align(align = "center", i = 1, j = c(3:6), part = "header") %>%
  flextable::autofit() %>%
  flextable::save_as_docx(path = here::here(
    "./Figures_Tables/tajimas_d/taj_d_means.docx"))
```


# Significance testing: Are the urban and rural values different?
## Create function
```{r}
perform_ttest_UvsR <- function(urb_taj_df,
                               rur_taj_df){
  
  # join urb, rur Tajimas D estimates
  combined_tajd <- dplyr::full_join(urb_taj_df,
                                rur_taj_df,
                                by = c("CHROM",
                                       "BIN_START"),
                                suffix = c(".urb", ".rur")) %>%
    dplyr::select(starts_with("Taj"))
  
  # t-test
  set.seed(1)
  
  ttest_result <- t.test(combined_tajd[,"TajimaD.urb"], 
         combined_tajd["TajimaD.rur"],
         alternative = "two.sided")
  
  return(ttest_result)
}
```

## City_dist
### 1 mil bp
```{r}
dist_1mil_ttest_UvsR <- perform_ttest_UvsR(urb_taj,
                              rur_taj)
```

### 100k bp
```{r}
dist_100k_ttest_UvsR <- perform_ttest_UvsR(urb_taj_100k,
                              rur_taj_100k)

# data:  combined_tajd[, "TajimaD.urb"] and combined_tajd["TajimaD.rur"]
# t = 0.17085, df = 1716.3, p-value = 0.8644
# alternative hypothesis: true difference in means is not equal to 0
# 95 percent confidence interval:
#  -0.07998265  0.09524643
# sample estimates:
#  mean of x  mean of y 
# -0.3595165 -0.3671484 
```

## Urb_score
### 1 mil bp
```{r}
usc_1mil_ttest_UvsR <- perform_ttest_UvsR(urb_taj_usc,
                              rur_taj_usc)
```

### 100k bp
```{r}
usc_100k_ttest_UvsR <- perform_ttest_UvsR(urb_taj_100k_usc,
                              rur_taj_100k_usc)

# data:  combined_tajd[, "TajimaD.urb"] and combined_tajd["TajimaD.rur"]
# t = 0.25043, df = 1720.7, p-value = 0.8023
# alternative hypothesis: true difference in means is not equal to 0
# 95 percent confidence interval:
#  -0.07574213  0.09791554
# sample estimates:
#  mean of x  mean of y 
# -0.3634625 -0.3745492 
```

## Export
### Significance testing: Are the Tajima's D values significantly different from zero?
#### Function
```{r}
export_ttests <- function(ttest_obj){
  
  filepath <- paste0("./Figures_Tables/tajimas_d/sig_testing/",
                     deparse(substitute(ttest_obj)),
                     ".json")
  
  list() %>%
  add_stats(ttest_obj) %>%
  write_stats(here::here(filepath))
}

```

#### Export all
```{r}
export_ttests(global_ttest)

export_ttests(urban_1mil_ttest)
export_ttests(urban_100k_ttest)
export_ttests(urban_1mil_ttest_usc)
export_ttests(urban_100k_ttest_usc)

export_ttests(rural_1mil_ttest)
export_ttests(rural_100k_ttest)
export_ttests(rural_1mil_ttest_usc)
export_ttests(rural_100k_ttest_usc)
```

### Significance testing: Are the urban and rural values different?
#### Export test results
```{r}
export_ttests(dist_1mil_ttest_UvsR)
export_ttests(dist_100k_ttest_UvsR)

export_ttests(usc_1mil_ttest_UvsR)
export_ttests(usc_100k_ttest_UvsR)

```


#### Export results in one table
##### 1 million bp windows
```{r}
tajd_table2 <- data.frame(
  Urbanization = c("Distance to City Center",
                   "Urbanization Score"),
  t = c(
    round((dist_1mil_ttest_UvsR$statistic),3),
    round((usc_1mil_ttest_UvsR$statistic),3)
    ),
  df = c(
    round(dist_1mil_ttest_UvsR$parameter, 3),
    round(usc_1mil_ttest_UvsR$parameter, 3)
    ),
  p = c(
    round(dist_1mil_ttest_UvsR$p.value, 3),
    round(usc_1mil_ttest_UvsR$p.value, 3)
  )) 


tajd_table2 %>%
  flextable::flextable() %>%
  flextable::italic(i = 1, j = 4, part = "header") %>%
  flextable::align(align = "center", j = c(2:4), part = "all") %>%
  flextable::align(align = "left", j = 1, part = "all") %>%
  flextable::autofit() %>%
  flextable::save_as_docx(path = here::here(
    "./Figures_Tables/tajimas_d/taj_d_UvsR_1mil_bp.docx"))
```


##### 100k bp windows
```{r}
tajd_table3 <- data.frame(
  Urbanization = c("Distance to City Center",
                   "Urbanization Score"),
  t = c(
    round((dist_100k_ttest_UvsR$statistic),3),
    round((usc_100k_ttest_UvsR$statistic),3)
    ),
  df = c(
    round(dist_100k_ttest_UvsR$parameter, 3),
    round(usc_100k_ttest_UvsR$parameter, 3)
    ),
  p = c(
    round(dist_100k_ttest_UvsR$p.value, 3),
    round(usc_100k_ttest_UvsR$p.value, 3)
  )) 


tajd_table3 %>%
  flextable::flextable() %>%
  flextable::italic(i = 1, j = 4, part = "header") %>%
  flextable::align(align = "center", j = c(2:4), part = "all") %>%
  flextable::align(align = "left", j = 1, part = "all") %>%
  flextable::autofit() %>%
  flextable::save_as_docx(path = here::here(
    "./Figures_Tables/tajimas_d/taj_d_UvsR_100k_bp.docx"))
```