To do:
- export anova
- export figures

- looks like there are differences btwn the ch2 and ch3 plants re: pop sizes, which is associated with many effects. What if I removed the ch2 plants? what would I see then?


- updated "urb is categorical" section since I changed the "continous" section a lot. Copy/paste it, then rename vars/change out response vars?
-Compare data from all_data csv w/ variants csv

# Set up notebook
## Load libraries
```{r warning=F}
source(here::here("libraries.R"))
```

## Import data
### Urbanization data
```{r}
urb <- read.csv(here::here("./clean_data/urb_metrics.csv")) %>%
  dplyr::mutate(site_id = as.factor(site_id),
                patch_id = as.factor(patch_id),
                urb_rur = as.factor(urb_rur),
                quadrant = as.factor(quadrant),
                river_valley = as.factor(river_valley)
                )
```

### Genetic data
#### Functions that read populations.sumstats_summary.tsv files, extracts data re: positions, then puts it into a dataframe
```{r}

## Extract variant positions
extract_variant_position_summary <- function(file) {
  
    text <- readLines(file)
    begin_line <- grep("# Variant positions", text)
    end_line <- grep("# All positions", text)
    data_text <- text[(begin_line + 1):(end_line - 1)]
    data_df <- read.table(text = data_text, header = FALSE, sep = "\t")
    header <- text[begin_line + 1]
    colnames(data_df) <- unlist(strsplit(header, "\t"))

  return(data_df)
}


# Extract ALL positions 
extract_all_position_summary <- function(file) {
  
    text <- readLines(file)
    begin_line <- grep("# All positions", text)
    end_line <- length(text)
    data_text <- text[(begin_line + 1):(end_line - 1)]
    data_df <- read.table(text = data_text, header = FALSE, sep = "\t")
    header <- text[begin_line + 1]
    colnames(data_df) <- unlist(strsplit(header, "\t"))

  return(data_df)
}

```

#### Read in genetic data
```{r}

input_file <- here::here("./summary_stats/trial_02/output_refmap_run2/populations.sumstats_summary.tsv")


div <- extract_variant_position_summary(input_file) %>%
# if population doesn't start with "MWI", it should start with "MW"
  # clean names
  janitor::clean_names() %>%
  # rename col 1
  dplyr::rename(pop_id = 1) %>%
  dplyr::mutate(pop_id = as.factor(pop_id))


div_all <- extract_all_position_summary(input_file) %>%
# if population doesn't start with "MWI", it should start with "MW"
  # clean names
  janitor::clean_names() %>%
  # rename col 1
  dplyr::rename(pop_id = 1) %>%
  dplyr::mutate(pop_id = as.factor(pop_id))

```

#### Add "MW" to numeric populations
```{r}
# take entries with numeric populations and add "MW" as prefix
add_MW_IDs <- function(div_df) {
  # Filter out populations that start with "MWI" or "UTSC"
  MW_pops <- div_df %>%
    filter(!str_detect(pop_id, "MWI")) %>%
    filter(!str_detect(pop_id, "UTSC"))

  # Get populations that start with "MWI" or "UTSC"
  MWI_UTSC_pops <- anti_join(div_df, MW_pops)

  # Split up populations MW001->MW009 vs MW010->MW079
  MW_pops_singledigits <- MW_pops[nchar(as.character(MW_pops$pop_id)) == 1, ]
  MW_pops_doubledigits <- MW_pops[nchar(as.character(MW_pops$pop_id)) != 1, ]

  # Add "MW00" to single-digit populations
  MW_pops_singledigits %<>%
    dplyr::mutate(patch_id = paste0("MW00", pop_id))

  # Add "MW0" to double-digit populations
  MW_pops_doubledigits %<>%
    dplyr::mutate(patch_id = paste0("MW0", pop_id))

  # Bring all entries together again
  all_pops <- full_join(MW_pops_singledigits,
                         MW_pops_doubledigits) %>%
    dplyr::relocate(patch_id, .before = 1) %>%
    full_join(.,
              MWI_UTSC_pops) %>%
    # If patch_id column is NA (for MWI and UTSC values), replace it with value from pop_id
    dplyr::mutate(patch_id = coalesce(patch_id, pop_id))

  return(all_pops)
}

diversity_all_sites <- add_MW_IDs(div_all)
diversity_variant_sites <- add_MW_IDs(div)

```

## Join genetic data with urbanization data
```{r}
# left vs. full join because some populations in urb_clean weren't genotyped due to lack of material
diversity_all_sites %<>%
  left_join(., urb, by = "patch_id") %>%
  dplyr::mutate(patch_id = as.factor(patch_id),
                pop_id = as.factor(pop_id))

diversity_variant_sites %<>%
  left_join(., urb, by = "patch_id") %>%
  dplyr::mutate(patch_id = as.factor(patch_id),
                pop_id = as.factor(pop_id))


# make new columns: u_r_dist & u_r_usc (urban or rural)
# this column is different from urb_rur, which I made when scouting sites for the 100 populations. I don't have those types of designations for the sites I imported from my previous observational study (note the NAs). So I'll calculate two new columns that assign a population an "urban" or "rural" status based on its distance from city center (u_r_dist) or urbanization score (u_r_usc)
diversity_variant_sites %<>%
  dplyr::mutate(u_r_dist = case_when(
    City_dist <= 40 ~ "Urban",
    TRUE ~ "Rural")) %>%
  dplyr::mutate(u_r_usc = case_when(
    urb_score > 0 ~ "Urban",
    TRUE ~ "Rural"))


diversity_all_sites %<>%
  dplyr::mutate(u_r_dist = case_when(
    City_dist <= 40 ~ "Urban",
    TRUE ~ "Rural")) %>%
  dplyr::mutate(u_r_usc = case_when(
    urb_score > 0 ~ "Urban",
    TRUE ~ "Rural"))

```

## Make plotting theme
```{r}
plot_aesthetics <- c(geom_point(size = 1.5,
                                shape = 1),
                     
                     geom_smooth(method = lm,
                                 color = "black",
                                 size = 0.5,
                                 se = TRUE))

# for ggpredict objects
rep_geoms <- c(geom_smooth(aes(x = x,
                               y = predicted),
                           color = "#66a182",
                           size = 1,
                           se = F),
               geom_ribbon(aes(x = x,
                               ymin = conf.low,
                               ymax = conf.high),
                           fill = "#66a182",
                           alpha = 0.3))

```

# Does urbanization predict genetic diversity within pops?
## Explore data
### All sites
```{r}
str(diversity_all_sites)

head(diversity_all_sites)

diversity_all_sites %>%
  dplyr::select(1, 3, "exp_het", "obs_het", "exp_hom", "obs_hom", "pi", "num_indv", "City_dist", "urb_score") %>%
  view

# quick boxplots
boxplot(diversity_all_sites$private)
boxplot(diversity_all_sites$private / diversity_all_sites$sites)
boxplot(diversity_all_sites$fis)
boxplot(diversity_all_sites$obs_het)
boxplot(diversity_all_sites$exp_het)
boxplot(diversity_all_sites$obs_hom)
boxplot(diversity_all_sites$exp_hom)
boxplot(diversity_all_sites$pi)

# quick regressions of genetic vars against distance from cc
qplot(data = diversity_all_sites,
      City_dist, private)

qplot(data = diversity_all_sites,
      City_dist, private/sites)

qplot(data = diversity_all_sites,
      City_dist, fis)

qplot(data = diversity_all_sites,
      City_dist, obs_het)

qplot(data = diversity_all_sites,
      City_dist, exp_het)

qplot(data = diversity_all_sites,
      City_dist, obs_hom)

qplot(data = diversity_all_sites,
      City_dist, exp_hom)

qplot(data = diversity_all_sites,
      exp_het, obs_het)

qplot(data = diversity_all_sites,
      City_dist, pi)

qplot(data = diversity_all_sites,
      City_dist, num_indv)

# regress all selected variables against one another
diversity_all_sites %>%
  dplyr::select(private, pi, exp_het, obs_het, exp_hom, obs_hom, num_indv, fis, urb_score, City_dist) %>%
  pairs()

# look at distributions of urb vs rural pops
qplot(data = diversity_all_sites,
      u_r_dist, pi)

qplot(data = diversity_all_sites,
      u_r_dist, private)

qplot(data = diversity_all_sites,
      u_r_usc, pi)

qplot(data = diversity_all_sites,
      u_r_usc, private)


# does urb correlated with other population descriptors?

## number of individuals/population
test1 <- glmmTMB(City_dist ~ num_indv, diversity_all_sites)
performance::check_model(test1)
car::Anova(test1) # yes
r.squaredGLMM(test1) # ~4% variance explained

## number of sites/population
test1 <- glmmTMB(City_dist ~ sites, diversity_all_sites)
performance::check_model(test1)
car::Anova(test1) # yes
r.squaredGLMM(test1) # ~4% variance explained

## number of polymorphic sites/population
test1 <- glmmTMB(City_dist ~ polymorphic_sites, diversity_all_sites)
performance::check_model(test1)
car::Anova(test1) # yes
r.squaredGLMM(test1) # ~5% variance explained



## number of sites vs. individuals/pop
test1 <- glmmTMB(sites ~ num_indv, diversity_all_sites, family = "nbinom2")
performance::check_model(test1)
car::Anova(test1) # yes
r.squaredGLMM(test1) # >74% variance explained!!

# num indivs vs fis
qplot(diversity_all_sites$num_indv,
      diversity_all_sites$fis)
```

### Variant sites
```{r}
str(diversity_variant_sites)

head(diversity_variant_sites)

diversity_variant_sites %>%
  dplyr::select(1, 3, 4, "exp_het", "obs_het", "exp_hom", "obs_hom", "pi", "City_dist", "urb_score") %>%
  view

# quick boxplots
boxplot(diversity_variant_sites$private)
boxplot(diversity_variant_sites$private / diversity_variant_sites$num_indv)
boxplot(diversity_variant_sites$fis)
boxplot(diversity_variant_sites$obs_het)
boxplot(diversity_variant_sites$exp_het)
boxplot(diversity_variant_sites$obs_hom)
boxplot(diversity_variant_sites$exp_hom)
boxplot(diversity_variant_sites$pi)

# quick regressions of genetic vars against distance from cc
qplot(data = diversity_variant_sites,
      City_dist, private)

qplot(data = diversity_variant_sites,
      City_dist, private/num_indv)

qplot(data = diversity_variant_sites,
      City_dist, fis)

qplot(data = diversity_variant_sites,
      City_dist, obs_het)

qplot(data = diversity_variant_sites,
      City_dist, exp_het)

qplot(data = diversity_variant_sites,
      City_dist, obs_hom)

qplot(data = diversity_variant_sites,
      City_dist, exp_hom)

qplot(data = diversity_variant_sites,
      exp_het, obs_het)

qplot(data = diversity_variant_sites,
      City_dist, pi)

qplot(data = diversity_variant_sites,
      City_dist, num_indv)

# regress all selected variables against one another
diversity_variant_sites %>%
  dplyr::select(private, pi, exp_het, obs_het, exp_hom, obs_hom, num_indv, fis, urb_score, City_dist) %>%
  pairs()

# look at distributions of urb vs rural pops
qplot(data = diversity_variant_sites,
      u_r_dist, pi)

qplot(data = diversity_variant_sites,
      u_r_dist, private)

qplot(data = diversity_variant_sites,
      u_r_usc, pi)

qplot(data = diversity_variant_sites,
      u_r_usc, private)


# does urb correlated with other population descriptors?

## number of individuals/population
test1 <- glmmTMB(City_dist ~ num_indv, diversity_variant_sites)
performance::check_model(test1)
car::Anova(test1) # yes
r.squaredGLMM(test1) # ~4% variance explained
```

## try to automate diagnostics/model fitting
```{r}
create_glm <- function(df, model_family, response_var, predictor_var) {
  # create the formula string
  formula_str <- paste(response_var, "~", predictor_var)
  
  # create the formula object
  formula_obj <- as.formula(formula_str)
  
  # fit the model
  model <- glmmTMB(formula_obj, data = df, family = model_family)
  
  # plot the residuals
  plot(DHARMa::simulateResiduals(model))
  
  # create the output data frame
  output_df <- data.frame(Response_Variable = rep(response_var, length(predictor_var)),
                           Predictor_Variable = predictor_var,
                           Model_Family = rep(model_family, length(predictor_var)),
                           Model_Formula = rep(formula_str, length(predictor_var)))
  
  print(output_df)
  
  return(model)
}

test <- create_glm(diversity, "gaussian", "private", "City_dist")

```

## Urbanization is continuous
### Fit linear models/diagnostics
#### Private alleles
##### City_dist
```{r}
priv_dist_rar <- glmmTMB(private ~ City_dist +
                           (1|num_indv) 
                    #   +    (1|sites)
                         ,
           diversity_all_sites,
           family = "nbinom2")

performance::check_model(priv_dist_rar)
car::Anova(priv_dist_rar) # becomes insignificant after accounting for sites and indivs/pop
```

##### Urb_score
```{r}
priv_usc_rar <- glmmTMB(private ~ urb_score  +
                           (1|num_indv) 
                     #    +  (1|sites)
                        ,
           diversity_all_sites,
           family = "nbinom2")

performance::check_model(priv_usc_rar)
```

#### Observed heterozygosity (Het_obs)
##### City_dist
```{r}
Het_obs_dist <- glmmTMB(obs_het ~ City_dist,
           diversity_all_sites)

performance::check_model(Het_obs_dist)
```

##### Urb_score
```{r}
Het_obs_usc <- glmmTMB(obs_het ~ urb_score,
           diversity_all_sites)

performance::check_model(Het_obs_usc)
```

#### Expected heterozygosity (Het_exp)
##### City_dist
```{r}
Het_exp_dist <- glmmTMB(exp_het ~ City_dist,
           diversity_all_sites)

performance::check_model(Het_exp_dist)
```

##### Urb_score
```{r}
Het_exp_usc <- glmmTMB(exp_het ~ urb_score,
           diversity_all_sites)

performance::check_model(Het_exp_usc)
```

#### Observed homozygosity (Hom_obs)
##### City_dist
```{r}
Hom_obs_dist <- glmmTMB(obs_hom ~ City_dist,
           diversity_all_sites)

performance::check_model(Hom_obs_dist)
```

##### Urb_score
```{r}
Hom_obs_usc <- glmmTMB(obs_hom ~ urb_score,
           diversity_all_sites)

performance::check_model(Hom_obs_usc)
```

#### Expected homozygosity (Hom_exp)
##### City_dist
```{r}
Hom_exp_dist <- glmmTMB(exp_hom ~ City_dist,
           diversity_all_sites)

performance::check_model(Hom_exp_dist)
```

##### Urb_score
```{r}
Hom_exp_usc <- glmmTMB(exp_hom ~ urb_score,
           diversity_all_sites)

performance::check_model(Hom_exp_usc)
```

#### Inbreeding coefficient (FIS)
This is extremely zero-inflated. I've tried fitting models with zero inflation parameters w/glmmTMB but they don't converge and/or diagnostics don't pass tests. I'll fit manual hurdle models instead.
##### City_dist
```{r}
diversity_all_sites %<>%
  mutate(fis_bin = case_when(
    fis == 0 ~ 0,
    fis > 0 ~ 1)
  )
  
# binary
fis_dist_bin <- glmmTMB(fis_bin ~ City_dist 
                        + (1|num_indv)
                        ,
           diversity_all_sites,
           family = "binomial")

performance::check_model(fis_dist_bin)


# quantitative
# wont' converge unless I multiply it
fis_dist_quant <- glmmTMB(fis * 10 ~  City_dist 
                        + (1|num_indv)
                        ,
           diversity_all_sites %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_dist_quant)
```

##### Urb_score
```{r}
# binary
fis_usc_bin <- glmmTMB(fis_bin ~ urb_score 
                        + (1|num_indv)
                        ,
           diversity_all_sites,
           family = "binomial")

performance::check_model(fis_usc_bin)


# quantitative
fis_usc_quant <- glmmTMB(fis ~ urb_score 
                        + (1|num_indv)
                        ,
           diversity_all_sites %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_usc_quant)
```

#### Pi
##### City_dist
```{r}
pi_dist_rar <- glmmTMB(pi ~ City_dist +
                         (1|num_indv),
           diversity_all_sites)

performance::check_model(pi_dist_rar)
```

##### Urb_score
```{r}
pi_usc_rar <- glmmTMB(pi ~ urb_score +
                        (1|num_indv),
           diversity_all_sites)

performance::check_model(pi_usc_rar)
```

### ANOVA & effect sizes
```{r}

mod_list <- list(priv_dist_rar,
                 priv_usc_rar,
                 Het_obs_dist,
                 Het_obs_usc,
                 Hom_obs_dist,
                 Hom_obs_usc,
                 fis_dist_bin,
                 fis_dist_quant,
                 fis_usc_bin,
                 fis_usc_quant,
                 pi_dist_rar,
                 pi_usc_rar)

create_anova_df <- lapply(mod_list, function(model) {
  anova_result <- car::Anova(model)
  
  formula <- formula(model)[2] %>%
    unlist()
  
  Rsq <- MuMIn::r.squaredGLMM(model)[1] %>%
  round(digits = 3) %>%
  format(nsmall = 3)
  
  # ADD EFFECT SIZES
  
  anova_tidy <- broom::tidy(anova_result) %>%
    dplyr::mutate("Variable" = as.character(formula),
                  "Rsq" = as.numeric(Rsq)) %>%
    dplyr::rename("Chi_sq" = statistic,
                  "p" = p.value,
                  "Urbanization" = term) %>%
    dplyr::mutate("Sig" = case_when(
      p <= 0.05 & p > 0.01 ~ "*",
      p <= 0.01 & p > 0.001 ~ "**",
      p <= 0.001 ~ "***",
      TRUE ~ "")) %>%
    dplyr::mutate(Chi_sq = format(round(Chi_sq, digits = 3), nsmall = 3),
                  p = format(round(p, digits = 3), nsmall = 3)) %>%
  dplyr::arrange(Urbanization, Variable) %>%
  dplyr::mutate("Variable" = case_when(
      Variable == "exp_het" ~ "Expected heterozygosity",
      Variable == "obs_het" ~ "Observed heterozygosity",
      Variable == "exp_hom" ~ "Expected homozygosity",
      Variable == "obs_hom" ~ "Observed homozygosity",
      Variable == "fis" ~ "FIS (quantitative)",
      Variable == "fis * 10" ~ "FIS (quantitative)",
      Variable == "fis_bin" ~ "FIS (binary)",
      Variable == "pi" ~ "Pi",
      Variable == "private" ~ "Private alleles",
      TRUE ~ ""))
  })

all_anovas <- do.call(rbind, create_anova_df) %>%
  dplyr::select(1,5,2,3,4,7,6) %T>%
  view

# create flextable
all_anovas %>%
  dplyr::select(-1, -Sig) %>%
  flextable::flextable() %>%
  flextable::compose(i = 1, j = 2, part = "header",
                     value = as_paragraph("Χ", as_sup("2"))) %>%
  flextable::compose(i = 1, j = 5, part = "header",
                     value = as_paragraph("R", as_sup("2"))) %>%
  bold(~ p <= 0.05) %>%
  flextable::autofit()
```

### Figures
#### Private alleles
##### City_dist
```{r}
priv_dist_rar_pred <- ggeffects::ggpredict(priv_dist_rar,
                                   terms = c("City_dist"),
                                   type = "fe")

plot(priv_dist_rar_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Private alleles") +
  ylim(0, 500) +
  labs(title = "") +
  theme_pubr()

ggsave(here::here("./Figures_Tables/regressions/private_alleles.png"))
```

##### Urb_score
```{r}
ggplot(diversity_all_sites,
       aes(x = urb_score,
           y = private/sites)) +
  xlab("Urbanization Score") +
  ylab("Private Alleles per Site") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### Expected heterozygosity
##### City_dist
```{r}
ggplot(diversity_all_sites,
       aes(x = City_dist,
           y = exp_het)) +
  xlab("Distance to City Center (km)") +
  ylab("Expected Heterozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
ggplot(diversity_all_sites,
       aes(x = urb_score,
           y = exp_het)) +
  xlab("Urbanization Score") +
  ylab("Expected Heterozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)
```

#### Obs heterozygosity
##### City_dist
```{r}
ggplot(diversity_all_sites,
       aes(x = City_dist,
           y = obs_het)) +
  xlab("Distance to City Center (km)") +
  ylab("Observed Heterozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
ggplot(diversity_all_sites,
       aes(x = urb_score,
           y = obs_het)) +
  xlab("Urbanization Score") +
  ylab("Observed Heterozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### Expected homozygosity
##### City_dist
```{r}
ggplot(diversity_all_sites,
       aes(x = City_dist,
           y = exp_hom)) +
  xlab("Distance to City Center (km)") +
  ylab("Expected Homozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
ggplot(diversity_all_sites,
       aes(x = urb_score,
           y = exp_hom)) +
  xlab("Urbanization Score") +
  ylab("Expected Homozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)
```

#### Obs homozygosity
##### City_dist
```{r}
ggplot(diversity_all_sites,
       aes(x = City_dist,
           y = obs_hom)) +
  xlab("Distance to City Center (km)") +
  ylab("Observed Homozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
ggplot(diversity_all_sites,
       aes(x = urb_score,
           y = obs_hom)) +
  xlab("Urbanization Score") +
  ylab("Observed Homozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### FIS
##### City_dist
```{r}
# binary
ggplot(diversity_all_sites,
       aes(x = City_dist,
           y = fis_bin)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr() 


# quantitative
ggplot(diversity_all_sites %>%
         dplyr::filter(fis != 0),
       aes(x = City_dist,
           y = fis)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 
```

##### Urb_score
```{r}
# binary
ggplot(diversity_all_sites,
       aes(x = urb_score,
           y = fis_bin)) +
  xlab("Urbanization Score") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr()  +
  xlim(4, -4)


# quantitative
ggplot(diversity_all_sites %>%
         dplyr::filter(fis != 0),
       aes(x = urb_score,
           y = fis)) +
  xlab("Urbanization Score") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### Pi
##### City_dist
```{r}
# ggplot(diversity_all_sites,
#        aes(x = City_dist,
#            y = pi/sites)) +
#   xlab("Distance to City Center (km)") +
#   ylab("Pi/sites") +
#   plot_aesthetics +
#   theme_pubr() +
#   labs_pubr() 

pi_dist_rar_pred <- ggeffects::ggpredict(pi_dist_rar,
                                   terms = c("City_dist"),
                                   type = "fe")

plot(pi_dist_rar_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Pi * 10") +
  ylim(NA, 0.04) +
  labs(title = "") +
  theme_pubr()

ggsave(here::here("./Figures_Tables/regressions/pi.png"))
```

###### Urb_score
```{r}
ggplot(diversity_all_sites,
       aes(x = urb_score,
           y = pi/sites)) +
  xlab("Urbanization Score") +
  ylab("Pi/sites") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### Sites
##### City_dist
```{r}
ggplot(diversity_all_sites,
       aes(x = City_dist,
           y = sites)) +
  xlab("Distance to City Center (km)") +
  ylab("Sites") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
ggplot(diversity_all_sites,
       aes(x = urb_score,
           y = sites)) +
  xlab("Urbanization Score") +
  ylab("Sites") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### Number of individuals per pop
##### City_dist
```{r}
ggplot(diversity_all_sites,
       aes(x = City_dist,
           y = num_indv)) +
  xlab("Distance to City Center (km)") +
  ylab("Sites") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
ggplot(diversity_all_sites,
       aes(x = urb_score,
           y = num_indv)) +
  xlab("Urbanization Score") +
  ylab("Sites") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

## Urbanization is categorical
### Fit linear models/diagnostics
#### Private alleles
###### u_r_dist
```{r}
priv_dist_rar2 <- glmmTMB(private ~ u_r_dist + (1|sites),
           diversity,
           family = "nbinom2")

performance::check_model(priv_dist_rar2)
```

###### u_r_usc
```{r}
priv_usc_rar2 <- glmmTMB(private ~ u_r_usc + (1|sites),
           diversity,
           family = "nbinom2")

performance::check_model(priv_usc_rar2)
```

#### Observed heterozygosity (Het_obs)
##### u_r_dist
```{r}
Het_obs_dist2 <- glmmTMB(obs_het ~ u_r_dist,
           diversity)

performance::check_model(Het_obs_dist2)
```

##### u_r_usc
```{r}
Het_obs_usc2 <- glmmTMB(obs_het ~ u_r_usc,
           diversity)

performance::check_model(Het_obs_usc2)
```

#### Expected heterozygosity (Het_exp)
##### u_r_dist
```{r}
Het_exp_dist2 <- glmmTMB(exp_het ~ u_r_dist,
           diversity)

performance::check_model(Het_exp_dist2)
```

##### u_r_usc
```{r}
Het_exp_usc2 <- glmmTMB(exp_het ~ u_r_usc,
           diversity)

performance::check_model(Het_exp_usc2)
```

#### Observed homozygosity (Hom_obs)
##### u_r_dist
```{r}
Hom_obs_dist2 <- glmmTMB(obs_hom ~ u_r_dist,
           diversity)

performance::check_model(Hom_obs_dist2)
```

##### u_r_usc
```{r}
Hom_obs_usc2 <- glmmTMB(obs_hom ~ u_r_usc,
           diversity)

performance::check_model(Hom_obs_usc2)
```

#### Expected homozygosity (Hom_exp)
##### u_r_dist
```{r}
Hom_exp_dist2 <- glmmTMB(exp_hom ~ u_r_dist,
           diversity)

performance::check_model(Hom_exp_dist2)
```

##### u_r_usc
```{r}
Hom_exp_usc2 <- glmmTMB(exp_hom ~ u_r_usc,
           diversity)

performance::check_model(Hom_exp_usc2)
```

#### Inbreeding coefficient (FIS)
This is extremely zero-inflated. I've tried fitting models with zero inflation parameters w/glmmTMB but they don't converge and/or diagnostics don't pass tests. I'll fit manual hurdle models instead.
##### u_r_dist
```{r}
diversity %<>%
  mutate(fis_bin = case_when(
    fis == 0 ~ 0,
    fis > 0 ~ 1)
  )
  
# binary
fis_dist_bin2 <- glmmTMB(fis_bin ~ u_r_dist,
           diversity,
           family = "binomial")

performance::check_model(fis_dist_bin2)


# quantitative
fis_dist_quant2 <- glmmTMB(fis ~ u_r_dist,
           diversity %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_dist_quant2)
```

##### u_r_usc
```{r}
# binary
fis_usc_bin2 <- glmmTMB(fis_bin ~ u_r_usc,
           diversity,
           family = "binomial")

performance::check_model(fis_usc_bin2)


# quantitative
fis_usc_quant2 <- glmmTMB(fis ~ u_r_usc,
           diversity %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_usc_quant2)
```

#### Pi
##### u_r_dist
```{r}
# have to mult by 10 or else doesn't converge
pi_dist_rar2 <- glmmTMB(pi*10 ~ u_r_dist + (1|sites),
           diversity)

performance::check_model(pi_dist_rar2)
```

##### u_r_usc
```{r}
pi_usc_rar2 <- glmmTMB(pi*10 ~ u_r_usc + (1|sites),
           diversity)

performance::check_model(pi_usc_rar2)
```


### ANOVA & effect sizes
#### Private alleles
##### City_dist
```{r}
# ANOVA
car::Anova(priv_dist_rar2)
summary(priv_dist_rar2)

test <- ggeffects::ggpredict(priv_dist_rar2)
# Pi(Urban) = 269.6539
# Pi(Rural) = 181.9130

(181.9130-269.6539)/181.9130
# 48% mean increase from rural to urban!
```

##### Urb_score
```{r}
car::Anova(priv_usc_rar2)
```

#### Expected heterozygosity
##### City_dist
```{r}
# ANOVA
car::Anova(Het_exp_dist2)
summary(Het_exp_dist2)
```

##### Urb_score
```{r}
# ANOVA
car::Anova(Het_exp_usc2)
```

#### Observed heterozygosity
##### City_dist
```{r}
# ANOVA
car::Anova(Het_obs_dist2)

```

##### Urb_score
```{r}
# ANOVA
car::Anova(Het_obs_usc2)
```

#### Expected homozygosity
##### City_dist
```{r}
# ANOVA
car::Anova(Hom_exp_dist2)
summary(Hom_exp_dist2)

```

##### Urb_score
```{r}
car::Anova(Hom_exp_usc2)
```

#### Observed homozygosity
##### City_dist
```{r}
car::Anova(Hom_obs_dist2)
```

##### Urb_score
```{r}
car::Anova(Hom_obs_usc2)
```

#### FIS
##### City_dist
```{r}
# ANOVA
# bin
car::Anova(fis_dist_bin2)

# quant
car::Anova(fis_dist_quant2)
```

##### Urb_score
```{r}
# ANOVA
# bin
car::Anova(fis_usc_bin2)

# quant
car::Anova(fis_usc_quant2)
```

#### Pi
##### City_dist
```{r}
car::Anova(pi_dist_rar2)

test <- ggeffects::ggpredict(pi_dist_rar2)
# Pi(Urban) = 0.03276026
# Pi(Rural) = 0.02818478

(0.02818478-0.03276026)/0.02818478
# 16% increase from rural to urban!
```

###### Urb_score
```{r}
car::Anova(pi_usc_rar2)
```

#### All
```{r}

mod_list <- list(#priv_dist2,
                 priv_dist_rar2,
                 #priv_usc2,
                 priv_usc_rar2,
                 Het_obs_dist2,
                 Het_obs_usc2,
                 Hom_obs_dist2,
                 Hom_obs_usc2,
                 fis_dist_bin2,
                 fis_dist_quant2,
                 fis_usc_bin2,
                 fis_usc_quant2,
                 #pi_dist2,
                 #pi_usc2,
                 pi_dist_rar2,
                 pi_usc_rar2)

anova_l2 <- lapply(mod_list, function(model) {
  anova_result <- car::Anova(model)
  
  formula <- formula(model)[2] %>%
    unlist()
  
  Rsq <- MuMIn::r.squaredGLMM(model)[1] %>%
  round(digits = 3) %>%
  format(nsmall = 3)
  
  # ADD EFFECT SIZES
  
  anova_tidy <- broom::tidy(anova_result) %>%
    dplyr::mutate("Variable" = as.character(formula),
                  "Rsq" = as.numeric(Rsq)) %>%
    dplyr::rename("Chi_sq" = statistic,
                  "p" = p.value,
                  "Urbanization" = term) %>%
    dplyr::mutate("Sig" = case_when(
      p <= 0.05 & p > 0.01 ~ "*",
      p <= 0.01 & p > 0.001 ~ "**",
      p <= 0.001 ~ "***",
      TRUE ~ "")) %>%
    dplyr::mutate(Chi_sq = format(round(Chi_sq, digits = 3), nsmall = 3),
                  p = format(round(p, digits = 3), nsmall = 3))
  })

all_anovas2 <- do.call(rbind, anova_l2) %>%
  dplyr::arrange(Urbanization, Variable) %>%
  dplyr::select(1,5,2,3,4,7,6) %>%
  dplyr::mutate("Variable" = case_when(
      Variable == "exp_het" ~ "Expected heterozygosity",
      Variable == "obs_het" ~ "Observed heterozygosity",
      Variable == "exp_hom" ~ "Expected homozygosity",
      Variable == "obs_hom" ~ "Observed homozygosity",
      Variable == "fis" ~ "FIS (quantitative)",
      Variable == "fis_bin" ~ "FIS (binary)",
      Variable == "pi * 10" ~ "Pi (standardized)",
      Variable == "private" ~ "Private alleles (standardized)",
      TRUE ~ "")) %>%
  dplyr::mutate("Urbanization" = case_when(
      Urbanization == "u_r_dist" ~ "Distance to City Center",
      Urbanization == "u_r_usc" ~ "Urbanization Score",
      TRUE ~ "")) %>%
  dplyr::arrange(p) %T>%
  view


all_anovas2 %>%
  dplyr::filter(Sig != "") %>%
  dplyr::select(-Sig) %>%
  flextable::flextable() %>%
  flextable::compose(i = 1, j = 3, part = "header",
                     value = as_paragraph("Χ", as_sup("2"))) %>%
  flextable::compose(i = 1, j = 6, part = "header",
                     value = as_paragraph("R", as_sup("2"))) %>%
  flextable::autofit()
```


### Figures- HAVEN'T UPDATED YET
#### Private alleles
###### Without rarefaction
####### City_dist
```{r}
ggplot(diversity,
       aes(x = u_r_dist,
           y = private)) +
  geom_violin(aes(fill = u_r_dist)) +
  geom_dotplot(binaxis = "y",
               stackdir = "center",
               binwidth = 20) +
  xlab("Urbanization based on Distance to City Center") +
  ylab("Private Alleles") +
  #plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  theme(legend.position = "none" )

```

####### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = private)) +
  xlab("Urbanization Score") +
  ylab("Private Alleles") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

###### With rarefaction
####### City_dist
```{r}
# ggplot(diversity,
#        aes(x = City_dist,
#            y = private/sites)) +
#   xlab("Distance to City Center (km)") +
#   ylab("Private Alleles per Site") +
#   plot_aesthetics +
#   theme_pubr() +
#   labs_pubr() 



priv_dist_rar_pred2 <- ggeffects::ggpredict(priv_dist_rar2,
                     terms = c("u_r_dist"),
                     type = "fe")

plot(priv_dist_rar_pred2, 
     add.data = F,
     colors = "bw") +
  xlab("Urbanization based on\n Distance to City Center") +
  ylab("Private alleles (standardized)") +
  ylim(0, 400) +
  labs(title = "") +
  theme_pubr()

# ggsave(here::here("./Figures_Tables/regressions/private_alleles.png"))
```

####### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = private/sites)) +
  xlab("Urbanization Score") +
  ylab("Private Alleles per Site") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### Expected heterozygosity
###### City_dist
```{r}
ggplot(diversity,
       aes(x = City_dist,
           y = exp_het)) +
  xlab("Distance to City Center (km)") +
  ylab("Expected Heterozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

###### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = exp_het)) +
  xlab("Urbanization Score") +
  ylab("Expected Heterozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)
```

#### Obs heterozygosity
###### City_dist
```{r}
ggplot(diversity,
       aes(x = City_dist,
           y = obs_het)) +
  xlab("Distance to City Center (km)") +
  ylab("Observed Heterozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

###### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = obs_het)) +
  xlab("Urbanization Score") +
  ylab("Observed Heterozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### Expected homozygosity
###### City_dist
```{r}
ggplot(diversity,
       aes(x = City_dist,
           y = exp_hom)) +
  xlab("Distance to City Center (km)") +
  ylab("Expected Homozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

###### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = exp_hom)) +
  xlab("Urbanization Score") +
  ylab("Expected Homozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)
```

#### Obs homozygosity
###### City_dist
```{r}
ggplot(diversity,
       aes(x = City_dist,
           y = obs_hom)) +
  xlab("Distance to City Center (km)") +
  ylab("Observed Homozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

###### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = obs_hom)) +
  xlab("Urbanization Score") +
  ylab("Observed Homozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### FIS
###### City_dist
```{r}
# binary
ggplot(diversity,
       aes(x = City_dist,
           y = fis_bin)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr() 


# quantitative
ggplot(diversity %>%
         dplyr::filter(fis != 0),
       aes(x = City_dist,
           y = fis)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 
```

###### Urb_score
```{r}
# binary
ggplot(diversity,
       aes(x = urb_score,
           y = fis_bin)) +
  xlab("Urbanization Score") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr()  +
  xlim(4, -4)


# quantitative
ggplot(diversity %>%
         dplyr::filter(fis != 0),
       aes(x = urb_score,
           y = fis)) +
  xlab("Urbanization Score") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### Pi
###### With rarefaction
####### City_dist
```{r}
# ggplot(diversity,
#        aes(x = City_dist,
#            y = pi/sites)) +
#   xlab("Distance to City Center (km)") +
#   ylab("Pi/sites") +
#   plot_aesthetics +
#   theme_pubr() +
#   labs_pubr() 

pi_dist_rar_pred2 <- ggeffects::ggpredict(pi_dist_rar2,
                                   terms = c("u_r_dist"),
                                   type = "fe")

plot(pi_dist_rar_pred2, 
     add.data = F,
     colors = "bw") +
  xlab("Distance to City Center (km)") +
  ylab("Pi * 10") +
  #ylim(NA, 0.04) +
  labs(title = "") +
  theme_pubr()

# ggsave(here::here("./Figures_Tables/regressions/pi.png"))
```

####### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = pi/sites)) +
  xlab("Urbanization Score") +
  ylab("Pi/sites") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

## Compare expected w/observed heterozygosity, homozygosity
```{r}
# All populations
chisq.test(diversity_all_sites$obs_hom, diversity_all_sites$exp_hom)
chisq.test(diversity_all_sites$obs_het, diversity_all_sites$exp_het)


# Only urban populations
## distance
div_urb_dist <- diversity_all_sites %>%
   dplyr::filter(u_r_dist == "Urban")

chisq.test(div_urb_dist$obs_hom, div_urb_dist$exp_hom)
chisq.test(div_urb_dist$obs_het, div_urb_dist$exp_het)


## urb score
div_urb_usc <- diversity_all_sites %>%
   dplyr::filter(u_r_usc == "Urban")

chisq.test(div_urb_usc$obs_hom, div_urb_usc$exp_hom)
chisq.test(div_urb_usc$obs_het, div_urb_usc$exp_het)



# Only rural populations
## distance
div_rur_dist <- diversity_all_sites %>%
   dplyr::filter(u_r_dist == "Rural")

chisq.test(div_rur_dist$obs_hom, div_rur_dist$exp_hom)
chisq.test(div_rur_dist$obs_het, div_rur_dist$exp_het)


## urb score
div_rur_usc <- diversity_all_sites %>%
   dplyr::filter(u_r_usc == "Rural")

chisq.test(div_rur_usc$obs_hom, div_rur_usc$exp_hom)
chisq.test(div_rur_usc$obs_het, div_rur_usc$exp_het)


# Looks like the largest differences btwn exp and obs het/hom are in the rural plants when urb = urb score
```
