---
title: How does urbanization predict private alleles, and expected/observed heterozygosity/homozygosity?
output: 
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 3
    number_sections: true
    theme: lumen
---

These analyses use the broadest genetic dataset which aims to include rare variants:
-R = 0.5 & -min_maf = 0.0078

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
```

# Set up notebook
## Load libraries & functions
```{r warning=F}
source(here::here("libraries.R"))
source(here::here("functions.R"))
```

## Import data
### Urbanization data
```{r}
urb <- read.csv(here::here("./clean_data/urb_metrics.csv")) %>%
  dplyr::mutate(site_id = as.factor(site_id),
                patch_id = as.factor(patch_id),
                urb_rur = as.factor(urb_rur),
                quadrant = as.factor(quadrant),
                river_valley = as.factor(river_valley)
                )
```

### Genetic data
```{r}
diversity_variant_sites <-  here::here("./summary_stats/mmaf0.007662_R0.5_populations_output/populations.sumstats_summary.tsv") %>%
  extract_variant_position_summary() %>%
  
# if population doesn't start with "MWI", it should start with "MW"
  # clean names
  janitor::clean_names() %>%
  # rename col 1
  dplyr::rename(pop_id = 1) %>%
  dplyr::mutate(pop_id = as.factor(pop_id)) %>%

# take entries with numeric populations and add "MW" as prefix
  add_MW_IDs()

```

## Join genetic data with urbanization data
```{r}
# left vs. full join because some populations in urb_clean weren't genotyped due to lack of material
diversity_variant_sites %<>%
  left_join(., urb, by = "patch_id") %>%
  dplyr::mutate(patch_id = as.factor(patch_id),
                pop_id = as.factor(pop_id))

```

# Does urbanization predict genetic diversity within pops?
## Explore data
```{r}
str(diversity_variant_sites)

head(diversity_variant_sites)

diversity_variant_sites %>%
  dplyr::select(1, 3, 4, "private", "exp_het", "obs_het", "exp_hom", "obs_hom", "City_dist", "urb_score") %>%
  view

# quick boxplots
boxplot(diversity_variant_sites$obs_het)
boxplot(diversity_variant_sites$exp_het)
boxplot(diversity_variant_sites$obs_hom)
boxplot(diversity_variant_sites$exp_hom)
boxplot(diversity_variant_sites$private)

# quick regressions of genetic vars against distance from cc

qplot(data = diversity_variant_sites,
      City_dist, obs_het)

qplot(data = diversity_variant_sites,
      City_dist, exp_het)

qplot(data = diversity_variant_sites,
      City_dist, obs_hom)

qplot(data = diversity_variant_sites,
      City_dist, exp_hom)

qplot(data = diversity_variant_sites,
      exp_het, obs_het)

qplot(data = diversity_variant_sites,
      City_dist, num_indv)

# regress all selected variables against one another
diversity_variant_sites %>%
  dplyr::select( private, exp_het, obs_het, exp_hom, obs_hom, num_indv, urb_score, City_dist) %>%
  pairs()


# does urb correlated with other population descriptors?

## number of individuals/population
test1 <- glmmTMB(City_dist ~ num_indv, diversity_variant_sites)
performance::check_model(test1)
car::Anova(test1) # yes
r.squaredGLMM(test1) # ~3% variance explained



# private alleles

# how many different values of private alleles are there?
unique(diversity_variant_sites$private)
# 3: 0, 1, and 2.

# how many populations have at least 1 private allele?
diversity_variant_sites %>%
  dplyr::select(private) %>%
  dplyr::filter(private != 0) %>%
  tally()
# there are only 14 populations. That's too small a sample size
# for regressions so I won't analyze this variable.
# In the R = 1, mmaf = 0.05 dataset, there are only zeroes for
# this value anyway.
```

### Does the number of individuals per populations correlate with pop gen stats?
#### obs_het & obs_hom
```{r}
ggplot(diversity_variant_sites, 
       aes(x = num_indv,
           y = obs_het)) + 
  geom_point(aes(color = City_dist)) +
  geom_smooth(stat = "smooth") +
  scale_color_continuous(type = "viridis")


test1 <- glmmTMB(obs_het ~ num_indv,
                 diversity_variant_sites)
performance::check_model(test1)
car::Anova(test1) # yes
r.squaredGLMM(test1) # ~8% variance explained
# SO I will add num_indivs as a random effect to these models

# I'll also add it to the obs_hom models because they are the inverse
```

#### exp_het & exp_hom
```{r}
ggplot(diversity_variant_sites, 
       aes(x = num_indv,
           y = exp_het)) + 
  geom_point(aes(color = City_dist)) +
  geom_smooth(stat = "smooth") +
  scale_color_continuous(type = "viridis")


test1 <- glmmTMB(exp_het ~ num_indv,
                 diversity_variant_sites)
performance::check_model(test1)
car::Anova(test1) # yes
r.squaredGLMM(test1) # ~90% variance explained
# SO I will add num_indivs as a random effect to these models

# I'll also add it to the exp_hom models because they are the inverse
```

## Urbanization is continuous
### Fit linear models/diagnostics
#### Expected heterozygosity (Het_exp)
##### City_dist
```{r}
Het_exp_dist <- glmmTMB(exp_het ~ City_dist
                        + (1|num_indv),
           diversity_variant_sites)

performance::check_model(Het_exp_dist)
```

##### Urb_score
```{r}
Het_exp_usc <- glmmTMB(exp_het ~ urb_score
                       + (1|num_indv),
           diversity_variant_sites)

performance::check_model(Het_exp_usc)
```

#### Observed heterozygosity (Het_obs)
##### City_dist
```{r}
Het_obs_dist <- glmmTMB(obs_het ~ City_dist
                       + (1|num_indv),
           diversity_variant_sites)

performance::check_model(Het_obs_dist)
```

##### Urb_score
```{r}
Het_obs_usc <- glmmTMB(obs_het ~ urb_score
                       + (1|num_indv),
           diversity_variant_sites)

performance::check_model(Het_obs_usc)
```

#### Expected homozygosity (Hom_exp)
##### City_dist
```{r}
Hom_exp_dist <- glmmTMB(exp_hom ~ City_dist
                       + (1|num_indv),
           diversity_variant_sites)

performance::check_model(Hom_exp_dist)
```

##### Urb_score
```{r}
Hom_exp_usc <- glmmTMB(exp_hom ~ urb_score
                       + (1|num_indv),
           diversity_variant_sites)

performance::check_model(Hom_exp_usc)
```


#### Observed homozygosity (Hom_obs)
##### City_dist
```{r}
Hom_obs_dist <- glmmTMB(obs_hom ~ City_dist
                       + (1|num_indv),
           diversity_variant_sites)

performance::check_model(Hom_obs_dist)
```

##### Urb_score
```{r}
Hom_obs_usc <- glmmTMB(obs_hom ~ urb_score
                       + (1|num_indv),
           diversity_variant_sites)

performance::check_model(Hom_obs_usc)
```

### ANOVA & effect sizes
```{r}
# only including heterozygosity models because homozygosity models' statistics are identical
mod_list <- list(Het_exp_dist,
                 Het_exp_usc,
                 Het_obs_dist,
                 Het_obs_usc#,
                 # Hom_exp_dist,
                 # Hom_exp_usc,
                 # Hom_obs_dist,
                 # Hom_obs_usc
                 )

all_anovas <- do.call(rbind, create_anova_df(mod_list)) %>%
  view

# create flextable
all_anovas %>%
  dplyr::select(-Sig) %>%
  flextable::flextable() %>%
  flextable::compose(i = 1, j = 3, part = "header",
                     value = as_paragraph("Î§", as_sup("2"))) %>%
  flextable::compose(i = 1, j = 6, part = "header",
                     value = as_paragraph("R", as_sup("2"))) %>%
  bold(~ p <= 0.05) %>%
  flextable::autofit() %T>%
  flextable::save_as_image(here::here("./Figures_Tables/regressions/het_hom_anova.png"))
```

### Figures
#### Expected heterozygosity
##### City_dist
```{r}
Het_exp_dist_pred <- ggeffects::ggpredict(Het_exp_dist,
                                   terms = c("City_dist"),
                                   type = "fe")

Het_exp_dist_pred_plot <- plot(Het_exp_dist_pred, 
     add.data = F) +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
            #   fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Expected Heterozygosity") +
  ylim(0.07, 0.095) +
  labs(title = "") +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
Het_exp_usc_pred <- ggeffects::ggpredict(Het_exp_usc,
                                   terms = c("urb_score"),
                                   type = "fe")
 
Het_exp_usc_pred_plot <- plot(Het_exp_usc_pred, 
     add.data = F) +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Urbanization Score") +
  ylab("Expected Heterozygosity") +
  labs(title = "") +
    ylim(0.07, 0.095) +
  theme_pubr()+
  labs_pubr() +
  xlim(4, -4)
```

#### Obs heterozygosity
##### City_dist
```{r}
Het_obs_dist_pred <- ggeffects::ggpredict(Het_obs_dist,
                                   terms = c("City_dist"),
                                   type = "fe")

Het_obs_dist_pred_plot <- plot(Het_obs_dist_pred, 
     add.data = F,
     alpha = 0) +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
               fill = "#66a182",
               alpha = 0.5
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Observed Heterozygosity") +
    ylim(0.07, 0.095) +
  labs(title = "") +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
Het_obs_usc_pred <- ggeffects::ggpredict(Het_obs_usc,
                                   terms = c("urb_score"),
                                   type = "fe")
 
Het_obs_usc_pred_plot <- plot(Het_obs_usc_pred, 
     add.data = F,
     alpha = 0) +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
               fill = "#66a182",
               alpha = 0.5
              ) +
  xlab("Urbanization Score") +
  ylab("Observed Heterozygosity") +
  labs(title = "") +
      ylim(0.07, 0.095) +
  theme_pubr()+
  labs_pubr() +
  xlim(4, -4)
```

#### Expected homozygosity
##### City_dist
```{r}
Hom_exp_dist_pred <- ggeffects::ggpredict(Hom_exp_dist,
                                   terms = c("City_dist"),
                                   type = "fe")

Hom_exp_dist_pred_plot <- plot(Hom_exp_dist_pred, 
     add.data = F) +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Expected Homozygosity") +
    ylim(0.9, 0.93) +
  labs(title = "") +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
Hom_exp_usc_pred <- ggeffects::ggpredict(Hom_exp_usc,
                                   terms = c("urb_score"),
                                   type = "fe")
 
Hom_exp_usc_pred_plot <- plot(Hom_exp_usc_pred, 
     add.data = F) +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Urbanization Score") +
  ylab("Expected Homozygosity") +
      ylim(0.9, 0.93) +
  labs(title = "") +
  theme_pubr()+
  labs_pubr() +
  xlim(4, -4)
```

#### Obs homozygosity
##### City_dist
```{r}
Hom_obs_dist_pred <- ggeffects::ggpredict(Hom_obs_dist,
                                   terms = c("City_dist"),
                                   type = "fe")

Hom_obs_dist_pred_plot <- plot(Hom_obs_dist_pred, 
     add.data = F,
     alpha = 0) +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
               fill = "#66a182",
               alpha = 0.5
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Observed Homozygosity") +
    ylim(0.9, 0.93) +
  labs(title = "") +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
Hom_obs_usc_pred <- ggeffects::ggpredict(Hom_obs_usc,
                                   terms = c("urb_score"),
                                   type = "fe")
 
Hom_obs_usc_pred_plot <- plot(Hom_obs_usc_pred, 
     add.data = F,
     alpha = 0) +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
               fill = "#66a182",
               alpha = 0.5
              ) +
  xlab("Urbanization Score") +
  ylab("Observed Homozygosity") +
  ylim(0.9, 0.93) +
  labs(title = "") +
  theme_pubr()+
  labs_pubr() +
  xlim(4, -4)
```

#### Number of individuals per pop
##### City_dist
```{r}
ggplot(diversity_variant_sites,
       aes(x = City_dist,
           y = num_indv)) +
  xlab("Distance to City Center (km)") +
  ylab("Individuals/Population") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
ggplot(diversity_variant_sites,
       aes(x = urb_score,
           y = num_indv)) +
  xlab("Urbanization Score") +
  ylab("Individuals/Population") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### Export all in one pdf
```{r}
# City distance plots
(Het_exp_dist_pred_plot | Het_obs_dist_pred_plot) /
(Hom_exp_dist_pred_plot | Hom_obs_dist_pred_plot )

ggsave(here::here("./Figures_Tables/regressions/het_hom_dist.png"),
       width = 10,
       height = 7,
       units = "in")


# Urbanization score plots
(Het_exp_usc_pred_plot | Het_obs_usc_pred_plot) / 
(Hom_exp_usc_pred_plot | Hom_obs_usc_pred_plot )
  
ggsave(here::here("./Figures_Tables/regressions/het_hom_usc.png"),
       width = 10,
       height = 7,
       units = "in")
```

## Urbanization is categorical
### Fit linear models/diagnostics
#### Expected heterozygosity (Het_exp)
##### u_r_dist
```{r}
Het_exp_dist_cat <- glmmTMB(exp_het ~ u_r_dist +
                              (1|num_indv),
           diversity_variant_sites)

performance::check_model(Het_exp_dist_cat)
```

##### u_r_usc
```{r}
Het_exp_usc_cat <- glmmTMB(exp_het ~ u_r_usc +
                              (1|num_indv),
           diversity_variant_sites)

performance::check_model(Het_exp_usc_cat)
```

#### Observed heterozygosity (Het_obs)
##### u_r_dist
```{r}
Het_obs_dist_cat <- glmmTMB(obs_het ~ u_r_dist +
                              (1|num_indv),
           diversity_variant_sites)

performance::check_model(Het_obs_dist_cat)
```

##### u_r_usc
```{r}
Het_obs_usc_cat <- glmmTMB(obs_het ~ u_r_usc +
                              (1|num_indv),
           diversity_variant_sites)

performance::check_model(Het_obs_usc_cat)
```

#### Expected homozygosity (Hom_exp)
##### u_r_dist
```{r}
Hom_exp_dist_cat <- glmmTMB(exp_hom ~ u_r_dist +
                              (1|num_indv),
           diversity_variant_sites)

performance::check_model(Hom_exp_dist_cat)
```

##### u_r_usc
```{r}
Hom_exp_usc_cat <- glmmTMB(exp_hom ~ u_r_usc +
                              (1|num_indv),
           diversity_variant_sites)

performance::check_model(Hom_exp_usc_cat)
```


#### Observed homozygosity (Hom_obs)
##### u_r_dist
```{r}
Hom_obs_dist_cat <- glmmTMB(obs_hom ~ u_r_dist +
                              (1|num_indv),
           diversity_variant_sites)

performance::check_model(Hom_obs_dist_cat)
```

##### u_r_usc
```{r}
Hom_obs_usc_cat <- glmmTMB(obs_hom ~ u_r_usc +
                              (1|num_indv),
           diversity_variant_sites)

performance::check_model(Hom_obs_usc_cat)
```

### ANOVA & effect sizes
```{r}
# only including heterozygosity models because homozygosity models' statistics are identical
mod_list_cat <- list(
                 Het_exp_dist_cat,
                 Het_exp_usc_cat,
                 Het_obs_dist_cat,
                 Het_obs_usc_cat)

all_anovas_cat <- do.call(rbind, create_anova_df_cat(mod_list_cat)) %>%
  view

# create flextable
all_anovas_cat %>%
  dplyr::select(-Sig) %>%
  flextable::flextable() %>%
  flextable::compose(i = 1, j = 3, part = "header",
                     value = as_paragraph("Î§", as_sup("2"))) %>%
  flextable::compose(i = 1, j = 6, part = "header",
                     value = as_paragraph("R", as_sup("2"))) %>%
  bold(~ p <= 0.05) %>%
  flextable::autofit() %T>%
  flextable::save_as_image(here::here("./Figures_Tables/regressions/het_hom_cat_anova.png"))
```


### Figures
#### Expected heterozygosity
##### Distance
```{r}
Het_exp_dist_pred_cat <- ggeffects::ggpredict(Het_exp_dist_cat,
                                   terms = c("u_r_dist"),
                                   type = "fe")

Het_exp_dist_pred_cat_plot <- ggplot(data = Het_exp_dist_pred_cat,
         aes(x = x, y = predicted)) +
        geom_point(color = "black",
                   size = 2) + 
    geom_errorbar(aes(ymin = conf.low,
                      ymax = conf.high),
                  width = 0.2,
                  color = "black") +
  xlab("Distance to City Center (km)") +
  ylab("Expected Heterozygosity") +
  ylim(0.07, 0.09) +
  labs(title = "") +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
Het_exp_usc_pred_cat <- ggeffects::ggpredict(Het_exp_usc_cat,
                                   terms = c("u_r_usc"),
                                   type = "fe")
 
Het_exp_usc_pred_cat_plot <- ggplot(data = Het_exp_usc_pred_cat,
         aes(x = x, y = predicted)) +
        geom_point(color = "black",
                   size = 2) + 
    geom_errorbar(aes(ymin = conf.low,
                      ymax = conf.high),
                  width = 0.2,
                  color = "black") +
  xlab("Urbanization Score") +
  ylab("Expected Heterozygosity") +
  labs(title = "") +
    ylim(0.07, 0.09) +
  theme_pubr()+
  labs_pubr() 
```

#### Obs heterozygosity
##### Distance
```{r}
Het_obs_dist_pred_cat <- ggeffects::ggpredict(Het_obs_dist_cat,
                                   terms = c("u_r_dist"),
                                   type = "fe") 

Het_obs_dist_pred_cat_plot <- 
  ggplot(data = Het_obs_dist_pred_cat,
         aes(x = x, y = predicted)) +
        geom_point(color = "forestgreen",
                   size = 2) + 
    geom_errorbar(aes(ymin = conf.low,
                      ymax = conf.high),
                  width = 0.2,
                  color = "forestgreen") +
  xlab("Distance to City Center (km)") +
  ylab("Observed Heterozygosity") +
    ylim(0.07, 0.09) +
  labs(title = "") +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
Het_obs_usc_pred_cat <- ggeffects::ggpredict(Het_obs_usc_cat,
                                   terms = c("u_r_usc"),
                                   type = "fe")
 
Het_obs_usc_pred_cat_plot <-
  ggplot(data = Het_obs_usc_pred_cat,
         aes(x = x, y = predicted)) +
        geom_point(color = "forestgreen",
                   size = 2) + 
    geom_errorbar(aes(ymin = conf.low,
                      ymax = conf.high),
                  width = 0.2,
                  color = "forestgreen") +
  xlab("Urbanization Score") +
  ylab("Observed Heterozygosity") +
  labs(title = "") +
      ylim(0.07, 0.09) +
  theme_pubr() +
  labs_pubr() 
```

#### Expected homozygosity
##### Distance
```{r}
Hom_exp_dist_pred_cat <- ggeffects::ggpredict(Hom_exp_dist_cat,
                                   terms = c("u_r_dist"),
                                   type = "fe")

Hom_exp_dist_pred_cat_plot <- ggplot(data = Hom_exp_dist_pred_cat,
         aes(x = x, y = predicted)) +
        geom_point(color = "black",
                   size = 2) + 
    geom_errorbar(aes(ymin = conf.low,
                      ymax = conf.high),
                  width = 0.2,
                  color = "black") +
  xlab("Distance to City Center (km)") +
  ylab("Expected Homozygosity") +
  ylim(0.91, 0.93) +
  labs(title = "") +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
Hom_exp_usc_pred_cat <- ggeffects::ggpredict(Hom_exp_usc_cat,
                                   terms = c("u_r_usc"),
                                   type = "fe")
 
Hom_exp_usc_pred_cat_plot <- ggplot(data = Hom_exp_usc_pred_cat,
         aes(x = x, y = predicted)) +
        geom_point(color = "black",
                   size = 2) + 
    geom_errorbar(aes(ymin = conf.low,
                      ymax = conf.high),
                  width = 0.2,
                  color = "black") +
  xlab("Urbanization Score") +
  ylab("Expected Homozygosity") +
  ylim(0.91, 0.93) +
  labs(title = "") +
  theme_pubr()+
  labs_pubr() 
```

#### Obs homozygosity
##### Distance
```{r}
Hom_obs_dist_pred_cat <- ggeffects::ggpredict(Hom_obs_dist_cat,
                                   terms = c("u_r_dist"),
                                   type = "fe")

Hom_obs_dist_pred_cat_plot <- ggplot(data = Hom_obs_dist_pred_cat,
         aes(x = x, y = predicted)) +
        geom_point(color = "forestgreen",
                   size = 2) + 
    geom_errorbar(aes(ymin = conf.low,
                      ymax = conf.high),
                  width = 0.2,
                  color = "forestgreen") +
  xlab("Distance to City Center (km)") +
  ylab("Observed Homozygosity") +
  ylim(0.91, 0.93) +
  labs(title = "") +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
Hom_obs_usc_pred_cat <- ggeffects::ggpredict(Hom_obs_usc_cat,
                                   terms = c("u_r_usc"),
                                   type = "fe")
 
Hom_obs_usc_pred_cat_plot <- ggplot(data = Hom_obs_usc_pred_cat,
         aes(x = x, y = predicted)) +
        geom_point(color = "forestgreen",
                   size = 2) + 
    geom_errorbar(aes(ymin = conf.low,
                      ymax = conf.high),
                  width = 0.2,
                  color = "forestgreen") +
  xlab("Urbanization Score") +
  ylab("Observed Homozygosity") +
  ylim(0.91, 0.93) +
  labs(title = "") +
  theme_pubr() +
  labs_pubr()
```


#### Export all in one pdf
```{r}
# City distance plots
(Het_exp_dist_pred_cat_plot | Het_obs_dist_pred_cat_plot) /
(Hom_exp_dist_pred_cat_plot | Hom_obs_dist_pred_cat_plot )

ggsave(here::here("./Figures_Tables/regressions/het_hom_dist_categorical.png"),
       width = 7,
       height = 7,
       units = "in")


# Urbanization score plots
(Het_exp_usc_pred_cat_plot | Het_obs_usc_pred_cat_plot) / 
(Hom_exp_usc_pred_cat_plot | Hom_obs_usc_pred_cat_plot )
  
ggsave(here::here("./Figures_Tables/regressions/het_hom_usc_categorical.png"),
       width = 7,
       height = 7,
       units = "in")
```

## Compare expected w/observed heterozygosity, homozygosity
```{r}
# All populations: marginally sig
chisq.test(diversity_variant_sites$obs_hom, diversity_variant_sites$exp_hom)

chisq.test(diversity_variant_sites$obs_het, diversity_variant_sites$exp_het)


# Only urban populations
## distance: not sig
div_urb_dist <- diversity_variant_sites %>%
   dplyr::filter(u_r_dist == "Urban")

chisq.test(div_urb_dist$obs_hom, div_urb_dist$exp_hom)
chisq.test(div_urb_dist$obs_het, div_urb_dist$exp_het)


## urb score: not sig
div_urb_usc <- diversity_variant_sites %>%
   dplyr::filter(u_r_usc == "Urban")

chisq.test(div_urb_usc$obs_hom, div_urb_usc$exp_hom)
chisq.test(div_urb_usc$obs_het, div_urb_usc$exp_het)



# Only rural populations
## distance: not sig
div_rur_dist <- diversity_variant_sites %>%
   dplyr::filter(u_r_dist == "Rural")

chisq.test(div_rur_dist$obs_hom, div_rur_dist$exp_hom)
chisq.test(div_rur_dist$obs_het, div_rur_dist$exp_het)


## urb score: not sig
div_rur_usc <- diversity_variant_sites %>%
   dplyr::filter(u_r_usc == "Rural")

chisq.test(div_rur_usc$obs_hom, div_rur_usc$exp_hom)
chisq.test(div_rur_usc$obs_het, div_rur_usc$exp_het)


# Not much difference btwn exp and obs het/hom, but when there is an effect, it's incorporating all populations
```
