# Set up notebook
## Load libraries & functions
```{r}
source(here::here("libraries.R"))

library(hierfstat)
```

## Import genetic data
### All sites
```{r}
vcf <- read.vcfR(
  here::here("./summary_stats/output_populations_no-IBD/mmaf0.05_R0.75/populations.snps.vcf"),
  verbose = FALSE)

## Convert vcf to genind object
my_genind <- vcfR2genind(vcf)



# read in pop map with populations
pops <- read.csv(here("./genomic_resources/pop_map3.csv"))

# pop map with urban/rural designations
pops_UR <- read.delim(here("./clean_data/pop_map3_UR_dist.txt"),
                      header = F) %>%
  dplyr::rename("sample" = 1,
                "urban_rural" = 2)

# join pop maps
pops <- full_join(pops,
                  pops_UR,
                  by = "sample")


# add pops to genind
strata(my_genind) <- pops
pop(my_genind) <- pops[,"population"]
```

### Urban sites
```{r}
vcf_urb <- read.vcfR(
  here::here("./filtered_vcfs/urban_pops_dist_mmaf0.05_variant_sites.recode.vcf"),
  verbose = FALSE)

## Convert vcf to genind object
my_genind_urb <- vcfR2genind(vcf_urb)


# read in pop map with populations
pops_urb <- read.csv(here("./clean_data/pop_map3_U_dist.csv"))

# add pops to genind
strata(my_genind_urb) <- pops_urb
pop(my_genind_urb) <- pops_urb[,"population"]
```

### Rural sites
```{r}
vcf_rur <- read.vcfR(
  here::here("./filtered_vcfs/rural_pops_dist_mmaf0.05_variant_sites.recode.vcf"),
  verbose = FALSE)

## Convert vcf to genind object
my_genind_rur <- vcfR2genind(vcf_rur)


# read in pop map with populations
pops_rur <- read.csv(here("./clean_data/pop_map3_R_dist.csv"))

# add pops to genind
strata(my_genind_rur) <- pops_rur
pop(my_genind_rur) <- pops_rur[,"population"]
```

# Hierfstat
## Create f-statistic matrix
### All sites
```{r}
# convert it to correct format for package
formatted_genind <- genind2hierfstat(my_genind)

# make sure levels are from outermost to innermost
fsts <- varcomp.glob(levels = data.frame(pops$urban_rural,
                                 pops$population),
             loci = formatted_genind[,-1],
             diploid = TRUE)

# look at variance components summed over all loci
fsts$overall %>%
  as.data.frame()

# look at f-statistic matrix
fsts$F %>%
  as.data.frame()

# Test the significance of the effect of urban/rural groups on genetic differentiation
# p-value = 1
# the g.star matrix is all zeroes. But when I use only a small subset of SNPs, the matrix generates non-zero numbers and p<1, so this is working
test.between(formatted_genind[,-1],
       test.lev = pops$urban_rural,
       rand.unit = pops$population,
       nperm = 1000) 

# test the effect of urban/rural groups on genetic differentiation
test.g(formatted_genind[,-1], 
       level = pops$urban_rural,
       nperm = 1000)
# g.star matrix all zeroes and p = 1
# so effect of urban/rural grouping on genetic differentiation is non-significant

# test the effect of sampling site on genetic differentiation
test.g(formatted_genind[,-1], 
       level = formatted_genind[1], # same as pops$population
       nperm = 1000)
# g.star matrix all zeroes and p = 1
# so effect of sample site is non-significant

# test the significance of F(sampling site/UR group)
test.within(formatted_genind[,-1], 
       within = pops$urban_rural,
       test.lev = pops$population,
       nperm = 100)
# g.star matrix all zeroes and p = 1
# so effect of sample site is non-significant


basic.stats(my_genind,
            diploid = T)

# $overall
#     Ho     Hs     Ht    Dst    Htp   Dstp    Fst   Fstp 
# 0.2132 0.2348 0.2548 0.0200 0.2549 0.0202 0.0784 0.0791 
#    Fis   Dest 
# 0.0921 0.0263 
```

### Urban sites
```{r}
# convert it to correct format for package
formatted_genind_urb <- genind2hierfstat(my_genind_urb)

# make sure levels are from outermost to innermost
fsts_urb <- varcomp.glob(levels = data.frame(pops_urb$population),
             loci = formatted_genind_urb[,-1],
             diploid = TRUE)

# look at variance components summed over all loci
fsts_urb$overall %>%
  as.data.frame()

# look at f-statistic matrix
fsts_urb$F %>%
  as.data.frame()

# slightly negative fst between pops-total means more variation within than between populations. This can stem from uneven sample sizes

basic.stats(my_genind_urb,
            diploid = T)

# $overall
#      Ho      Hs      Ht     Dst     Htp    Dstp     Fst 
#  0.2088  0.2532  0.2520 -0.0011  0.2520 -0.0012 -0.0046 
#    Fstp     Fis    Dest 
# -0.0046  0.1754 -0.0016 
```

### Rural sites
```{r}
# convert it to correct format for package
formatted_genind_rur <- genind2hierfstat(my_genind_rur)

# make sure levels are from outermost to innermost
fsts_rur <- varcomp.glob(levels = data.frame(pops_rur$population),
             loci = formatted_genind_rur[,-1],
             diploid = TRUE)

# look at variance components summed over all loci
fsts_rur$overall %>%
  as.data.frame()

# look at f-statistic matrix
fsts_rur$F %>%
  as.data.frame()



basic.stats(my_genind_rur,
            diploid = T)

# $overall
#     Ho     Hs     Ht    Dst    Htp   Dstp    Fst   Fstp 
# 0.2156 0.2532 0.2559 0.0027 0.2560 0.0027 0.0105 0.0107 
#    Fis   Dest 
# 0.1486 0.0037 
```

## Calculate population-specific fst
### All sites
```{r}
betas(formatted_genind)$betaiovl %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  dplyr::rename("sampling_site" = 1,
                "fst" = 2) %>%
  dplyr::arrange(fst)
```

## Calculate pairwise Fst
### All sites
```{r}
pairwise_fst_matrix <- genet.dist(my_genind,
                                      method = "WC84")

```

### Urban sites
```{r}
# just take the existing matrix and keep the urban populations
urban_pop_list <- pops_urb$population %>%
  as.vector()

pairwise_fst_urb <- read.csv(here::here("./Figures_Tables/fst/pairwise_fst_matrix.csv")) %>%
  dplyr::select(-1) %>%
  # remove added "X" from every colname
  dplyr::rename_with(~gsub("X", "", .), everything()) %>%
  # keep only urban populations in "population" column
  dplyr::filter(population %in% urban_pop_list) %>%
  dplyr::select(1, all_of(urban_pop_list))
```

### Rural sites
```{r}
# just take the existing matrix and keep the rural populations
pairwise_fst_rur <- read.csv(here::here("./Figures_Tables/fst/pairwise_fst_matrix.csv")) %>%
  dplyr::select(-1) %>%
  # remove added "X" from every colname
  dplyr::rename_with(~gsub("X", "", .), everything()) %>%
  # EXCLUDE the urban populations in "population" column
  dplyr::filter(!(population %in% urban_pop_list)) %>%
  dplyr::select(-all_of(urban_pop_list))

```

# Export f-statistics
## Hierarchical
### All sites
```{r}
# first, turn off scientific notation
options(scipen=999) 

# then, export matrix of hierarchical F-statistics type-coefficients
fsts$F %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  dplyr::rename("Level" = 1,
                "Urban/Rural Group" = 2,
                "Sampling Site" = 3,
                "Individual" = 4) %>%
  dplyr::mutate_if(is.numeric, round, 4) %>%
  dplyr::mutate(Level = case_when(Level == "pops.urban_rural" ~ "Urban/Rural Group",
                                  Level == "pops.population" ~ "Sampling Site",
                         TRUE ~ Level)) %>%
  dplyr::mutate_all(~ ifelse(. == 0, "-", .)) %>%
  flextable() %>%
  autofit() %>%
  save_as_docx(path = here::here("./Figures_Tables/fst/hierfstat.docx"))

```

### Urban sites
```{r}
# first, turn off scientific notation
options(scipen=999) 

# then, export matrix of hierarchical F-statistics type-coefficients
fsts_urb$F %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  dplyr::rename("Level" = 1,
                "Sampling Site" = 2,
                "Individual" = 3) %>%
  dplyr::mutate_if(is.numeric, round, 4) %>%
  dplyr::mutate(Level = case_when(Level == "pops_urb.population" ~ "Sampling Site",
                         TRUE ~ Level)) %>%
  dplyr::mutate_all(~ ifelse(. == 0, "-", .)) %>%
  flextable() %>%
  autofit() %>%
  save_as_docx(path = here::here("./Figures_Tables/fst/hierfstat_urban.docx"))

```

### Rural sites
```{r}
# first, turn off scientific notation
options(scipen=999) 

# then, export matrix of hierarchical F-statistics type-coefficients
fsts_rur$F %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  dplyr::rename("Level" = 1,
                "Sampling Site" = 2,
                "Individual" = 3) %>%
  dplyr::mutate_if(is.numeric, round, 4) %>%
  dplyr::mutate(Level = case_when(Level == "pops_rur.population" ~ "Sampling Site",
                         TRUE ~ Level)) %>%
  dplyr::mutate_all(~ ifelse(. == 0, "-", .)) %>%
  flextable() %>%
  autofit() %>%
  save_as_docx(path = here::here("./Figures_Tables/fst/hierfstat_rural.docx"))

```

## Pairwise fst
### All sites
```{r}
# transform dist object into matrix, then export
full_matrix <- pairwise_fst_matrix %>%
  as.matrix() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  dplyr::rename("population" = 1)

write.csv(full_matrix,
          file = here::here("./Figures_Tables/fst/pairwise_fst_matrix.csv"))
```


### Urban sites
```{r}
write.csv(pairwise_fst_urb,
          file = here::here("./Figures_Tables/fst/pairwise_fst_matrix_urban.csv"))
```

### Rural sites
```{r}
write.csv(pairwise_fst_rur,
          file = here::here("./Figures_Tables/fst/pairwise_fst_matrix_rural.csv"))
```