To do:
- export anova
- export figures

chisq.test(diversity$obs_hom, diversity$exp_hom)
chisq.test(diversity$obs_het, diversity$exp_het)

# Set up notebook
## Load libraries
```{r warning=F}
source(here::here("libraries.R"))
```

## Import data
```{r}
# within-population diversity stats
# if population doesn't start with "MWI", it should start with "MW"
div <- read.csv(here::here("./summary_stats/orig_output/populations.sumstats_summary_allpositions.tsv.csv")) %>%
  # clean names
  janitor::clean_names() %>%
  # rename col 1
  dplyr::rename(pop_id = 1) %>%
  dplyr::mutate(pop_id = as.factor(pop_id))

# among-population f statistics (matrix)


# urbanization data
urb <- read.csv(here::here("./clean_data/urb_metrics.csv")) %>%
  dplyr::mutate(site_id = as.factor(site_id),
                patch_id = as.factor(patch_id),
                urb_rur = as.factor(urb_rur),
                quadrant = as.factor(quadrant),
                river_valley = as.factor(river_valley)
                )
```

## Clean data
```{r}
# take entries with numeric populations and add "MW" as prefix

# populations that should start with "MW" but currently start with number
MW_pops <- div %>%
   filter(!str_detect(pop_id, "MWI")) %>%
   filter(!str_detect(pop_id, "UTSC"))

# populations that start with "MWI" or "UTSC"
MWI_UTSC_pops <- anti_join(div,
                           MW_pops)


# split up populations MW001->MW009 vs MW010->MW079
MW_pops_singledigits <- MW_pops[nchar(as.character(MW_pops$pop_id)) == 1, ]
MW_pops_doubledigits <- MW_pops[nchar(as.character(MW_pops$pop_id)) != 1, ]

# add "MW00" to single-digit populations
MW_pops_singledigits %<>%
  dplyr::mutate(patch_id = paste0("MW00", pop_id))

# add "MW0" to double-digit populations
MW_pops_doubledigits %<>%
  dplyr::mutate(patch_id = paste0("MW0", pop_id))

# bring all entries together again
diversity <- full_join(MW_pops_singledigits,
                      MW_pops_doubledigits) %>%
  dplyr::relocate(patch_id, 1:30) %>%
  full_join(.,
            MWI_UTSC_pops) %>%
  # if patch_id column is NA (for MWI and UTSC values), replace it with value from pop_id
  dplyr::mutate(patch_id = coalesce(patch_id, pop_id))
```

## Make plotting theme
```{r}
plot_aesthetics <- c(geom_point(size = 1.5,
                                shape = 1),
                     
                     geom_smooth(method = lm,
                                 color = "black",
                                 size = 0.5,
                                 se = TRUE))

# for ggpredict objects
rep_geoms <- c(geom_smooth(aes(x = x,
                               y = predicted),
                           color = "#66a182",
                           size = 1,
                           se = F),
               geom_ribbon(aes(x = x,
                               ymin = conf.low,
                               ymax = conf.high),
                           fill = "#66a182",
                           alpha = 0.3))

```

# Does urbanization predict genetic diversity within pops?
## Join genetic data with urbanization data
```{r}
# left vs. full join because some populations in urb_clean weren't genotyped due to lack of material
diversity %<>%
  left_join(.,
            urb,
            by = "patch_id") %>%
  dplyr::mutate(patch_id = as.factor(patch_id),
                pop_id = as.factor(pop_id))

```

## Explore data
```{r}
str(diversity)

head(diversity)

diversity %>%
  dplyr::select(1, 3, "exp_het", "obs_het", "exp_hom", "obs_hom", "pi", "City_dist", "urb_score") %>%
  view

# quick boxplots
boxplot(diversity$private)
boxplot(diversity$private / diversity$sites)
boxplot(diversity$fis)
boxplot(diversity$obs_het)
boxplot(diversity$exp_het)
boxplot(diversity$obs_hom)
boxplot(diversity$exp_hom)
boxplot(diversity$pi)

# quick regressions of genetic vars against distance from cc
qplot(data = diversity,
      City_dist, private)

qplot(data = diversity,
      City_dist, private/sites)

qplot(data = diversity,
      City_dist, fis)

qplot(data = diversity,
      City_dist, obs_het)

qplot(data = diversity,
      City_dist, exp_het)

qplot(data = diversity,
      City_dist, obs_hom)

qplot(data = diversity,
      City_dist, exp_hom)

qplot(data = diversity,
      exp_het, obs_het)

qplot(data = diversity,
      City_dist, pi)

# regress all selected variables against one another
diversity %>%
  dplyr::select(private, pi, exp_het, obs_het, exp_hom, obs_hom, fis, urb_score, City_dist) %>%
  pairs()
```

## Fit linear models/diagnostics
### Private alleles
#### Without rarefaction (i.e., dividing by total sites)
##### City_dist
```{r}
priv_dist <- glmmTMB(private ~ City_dist,
           diversity,
           family = "nbinom2")

performance::check_model(priv_dist)
```

##### Urb_score
```{r}
priv_usc <- glmmTMB(private ~ urb_score,
           diversity,
           family = "nbinom2")

performance::check_model(priv_usc)
```

#### WITH rarefaction (i.e., dividing by total sites)
##### City_dist
```{r}
priv_dist_rar <- glmmTMB(private ~ City_dist + (1|sites),
           diversity)

performance::check_model(priv_dist_rar)
```

##### Urb_score
```{r}
priv_usc_rar <- glmmTMB(private ~ urb_score + (1|sites),
           diversity)

performance::check_model(priv_usc_rar)
```

### Observed heterozygosity (Het_obs)
#### City_dist
```{r}
Het_obs_dist <- glmmTMB(obs_het ~ City_dist,
           diversity)

performance::check_model(Het_obs_dist)
```

#### Urb_score
```{r}
Het_obs_usc <- glmmTMB(obs_het ~ urb_score,
           diversity)

performance::check_model(Het_obs_usc)
```

### Expected heterozygosity (Het_exp)
#### City_dist
```{r}
Het_exp_dist <- glmmTMB(exp_het ~ City_dist,
           diversity)

performance::check_model(Het_exp_dist)
```

#### Urb_score
```{r}
Het_exp_usc <- glmmTMB(exp_het ~ urb_score,
           diversity)

performance::check_model(Het_exp_usc)
```

### Observed homozygosity (Hom_obs)
#### City_dist
```{r}
Hom_obs_dist <- glmmTMB(obs_hom ~ City_dist,
           diversity)

performance::check_model(Hom_obs_dist)
```

#### Urb_score
```{r}
Hom_obs_usc <- glmmTMB(obs_hom ~ urb_score,
           diversity)

performance::check_model(Hom_obs_usc)
```

### Expected homozygosity (Hom_exp)
#### City_dist
```{r}
Hom_exp_dist <- glmmTMB(exp_hom ~ City_dist,
           diversity)

performance::check_model(Hom_exp_dist)
```

#### Urb_score
```{r}
Hom_exp_usc <- glmmTMB(exp_hom ~ urb_score,
           diversity)

performance::check_model(Hom_exp_usc)
```

### Inbreeding coefficient (FIS)
This is extremely zero-inflated. I've tried fitting models with zero inflation parameters w/glmmTMB but they don't converge and/or diagnostics don't pass tests. I'll fit manual hurdle models instead.
#### City_dist
```{r}
diversity %<>%
  mutate(fis_bin = case_when(
    fis == 0 ~ 0,
    fis > 0 ~ 1)
  )
  
# binary
fis_dist_bin <- glmmTMB(fis_bin ~ City_dist,
           diversity,
           family = "binomial")

performance::check_model(fis_dist_bin)


# quantitative
fis_dist_quant <- glmmTMB(fis ~ City_dist,
           diversity %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_dist_quant)
```

#### Urb_score
```{r}
# binary
fis_usc_bin <- glmmTMB(fis_bin ~ urb_score,
           diversity,
           family = "binomial")

performance::check_model(fis_usc_bin)


# quantitative
fis_usc_quant <- glmmTMB(fis ~ urb_score,
           diversity %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_usc_quant)
```

### Pi
#### Without rarefaction
##### City_dist
```{r}
pi_dist <- glmmTMB(pi ~ City_dist,
           diversity)

performance::check_model(pi_dist)
```

##### Urb_score
```{r}
pi_usc <- glmmTMB(pi ~ urb_score,
           diversity)

performance::check_model(pi_usc)
```


#### With rarefaction
##### City_dist
```{r}
# have to mult by 10 or else doesn't converge
pi_dist_rar <- glmmTMB(pi*10 ~ City_dist + (1|sites),
           diversity)

performance::check_model(pi_dist_rar)
```

##### Urb_score
```{r}
pi_usc_rar <- glmmTMB(pi*10 ~ urb_score + (1|sites),
           diversity)

performance::check_model(pi_usc_rar)
```


### Sites
#### City_dist
```{r}
sites_dist <- glmmTMB(sites ~ City_dist,
           diversity,
           family = "nbinom2")

performance::check_model(sites_dist)
```

#### Urb_score
```{r}
sites_usc <- glmmTMB(sites ~ urb_score,
           diversity,
           family = "nbinom2")

performance::check_model(sites_usc)
```

### Variant sites
#### City_dist
```{r}
vsites_dist <- glmmTMB(variant_sites ~ City_dist,
           diversity,
           family = "nbinom2")

performance::check_model(vsites_dist)
```

#### Urb_score
```{r}
sites_usc <- glmmTMB(variant_sites ~ urb_score,
           diversity,
           family = "nbinom2")

performance::check_model(vsites_usc)
```

## ANOVA & effect sizes
### Private alleles
#### Without rarefaction
##### City_dist
```{r}
# ANOVA
car::Anova(priv_dist)
summary(priv_dist)

```

##### Urb_score
```{r}
car::Anova(priv_usc)
```

#### With rarefaction
##### City_dist
```{r}
# ANOVA
car::Anova(priv_dist_rar)
summary(priv_dist_rar)

# nearly equivalent to just dividing private alleles by total sites
car::Anova(glmmTMB(private/sites ~ City_dist,
           diversity))



test <- ggeffects::ggpredict(priv_dist_rar)
# Pi at Dist(0) = 315.6427
# Pi at Dist(80) = 127.8714

(127.8714-315.6427)/127.8714
# 147% increase from rural to urban!
```

##### Urb_score
```{r}
car::Anova(priv_usc_rar)
```

### Expected heterozygosity
#### City_dist
```{r}
# ANOVA
car::Anova(Het_exp_dist)
summary(Het_exp_dist)
```

#### Urb_score
```{r}
# ANOVA
car::Anova(Het_exp_usc)
```

### Observed heterozygosity
#### City_dist
```{r}
# ANOVA
car::Anova(Het_obs_dist)

```

#### Urb_score
```{r}
# ANOVA
car::Anova(Het_obs_usc)
```

### Expected homozygosity
#### City_dist
```{r}
# ANOVA
car::Anova(Hom_exp_dist)
summary(Hom_exp_dist)

```

#### Urb_score
```{r}
car::Anova(Hom_exp_usc)
```

### Observed homozygosity
#### City_dist
```{r}
car::Anova(Hom_obs_dist)
```

#### Urb_score
```{r}
car::Anova(Hom_obs_usc)
```

### FIS
#### City_dist
```{r}
# ANOVA
# bin
car::Anova(fis_dist_bin)

# quant
car::Anova(fis_dist_quant)

# Export
```

#### Urb_score
```{r}
# ANOVA
# bin
car::Anova(fis_usc_bin)

# quant
car::Anova(fis_usc_quant)
```

### Pi
#### Without rarefaction
##### City_dist
```{r}
# ANOVA
car::Anova(pi_dist)
summary(pi_dist)
# at 0: 3.526e-03 = 0.003526
# at 80: (-1.254e-05)(80) + 3.526e-03 = 0.0025228

test <- ggeffects::ggpredict(pi_dist)
# Pi at Dist(0) = 0.00353
# Pi at Dist(80) = 0.00252

# estimates are identical whether using summary or ggpredict

(0.0025228-0.003526)/0.0025228
# 40% increase from rural to urban!
```

##### Urb_score
```{r}
# ANOVA
car::Anova(pi_usc)
```

#### With rarefaction
##### City_dist
```{r}
car::Anova(pi_dist_rar)

test <- ggeffects::ggpredict(pi_dist_rar)
# Pi at Dist(0) = 0.03525745
# Pi at Dist(80) = 0.02522897

(0.02522897-0.03525745)/0.02522897
# 40% increase from rural to urban!
```

##### Urb_score
```{r}
car::Anova(pi_usc_rar)
```

### Sites
#### City_dist
```{r}
car::Anova(sites_dist)
summary(sites_dist)

ggeffects::ggpredict(sites_dist)
```

#### Urb_score
```{r}
car::Anova(sites_usc)
```

### Variant sites
#### City_dist
```{r}
car::Anova(vsites_dist)
summary(vsites_dist)

ggeffects::ggpredict(vsites_dist) %>%
  plot
```

#### Urb_score
```{r}
car::Anova(sites_usc)
```

### All
```{r}

mod_list <- list(#priv_dist,
                 priv_dist_rar,
               #  priv_usc,
                 priv_usc_rar,
                 Het_obs_dist,
                 Het_obs_usc,
                 Hom_obs_dist,
                 Hom_obs_usc,
                 fis_dist_bin,
                 fis_dist_quant,
                 fis_usc_bin,
                 fis_usc_quant,
                 # pi_dist,
                 # pi_usc,
                 pi_dist_rar,
                 pi_usc_rar)

anova_l <- lapply(mod_list, function(model) {
  anova_result <- car::Anova(model)
  
  formula <- formula(model)[2] %>%
    unlist()
  
  Rsq <- MuMIn::r.squaredGLMM(model)[1] %>%
  round(digits = 3) %>%
  format(nsmall = 3)
  
  # ADD EFFECT SIZES
  
  anova_tidy <- broom::tidy(anova_result) %>%
    dplyr::mutate("Variable" = as.character(formula),
                  "Rsq" = as.numeric(Rsq)) %>%
    dplyr::rename("Chi_sq" = statistic,
                  "p" = p.value,
                  "Urbanization" = term) %>%
    dplyr::mutate("Sig" = case_when(
      p <= 0.05 & p > 0.01 ~ "*",
      p <= 0.01 & p > 0.001 ~ "**",
      p <= 0.001 ~ "***",
      TRUE ~ "")) %>%
    dplyr::mutate(Chi_sq = format(round(Chi_sq, digits = 3), nsmall = 3),
                  p = format(round(p, digits = 3), nsmall = 3))
  })

all_anovas <- do.call(rbind, anova_l) %>%
  dplyr::arrange(Urbanization, Variable) %>%
  dplyr::select(1,5,2,3,4,7,6) %>%
  dplyr::mutate("Variable" = case_when(
      Variable == "exp_het" ~ "Expected heterozygosity",
      Variable == "obs_het" ~ "Observed heterozygosity",
      Variable == "exp_hom" ~ "Expected homozygosity",
      Variable == "obs_hom" ~ "Observed homozygosity",
      Variable == "fis" ~ "FIS (quantitative)",
      Variable == "fis_bin" ~ "FIS (binary)",
      Variable == "pi * 10" ~ "Pi (standardized)",
      Variable == "private" ~ "Private alleles (standardized)",
     # Variable == "private/sites" ~ "Private alleles/site",
      TRUE ~ "")) %T>%
  view


all_anovas %>%
  dplyr::filter(Sig != "") %>%
  dplyr::select(-1, -Sig) %>%
  dplyr::filter(Variable != "FIS (binary)") %>%
  flextable::flextable() %>%
  flextable::compose(i = 1, j = 2, part = "header",
                     value = as_paragraph("Î§", as_sup("2"))) %>%
  flextable::compose(i = 1, j = 5, part = "header",
                     value = as_paragraph("R", as_sup("2"))) %>%
  flextable::autofit()
```

## Figures
### Private alleles
#### Without rarefaction
##### City_dist
```{r}
ggplot(diversity,
       aes(x = City_dist,
           y = private)) +
  xlab("Distance to City Center (km)") +
  ylab("Private Alleles") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = private)) +
  xlab("Urbanization Score") +
  ylab("Private Alleles") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### With rarefaction
##### City_dist
```{r}
# ggplot(diversity,
#        aes(x = City_dist,
#            y = private/sites)) +
#   xlab("Distance to City Center (km)") +
#   ylab("Private Alleles per Site") +
#   plot_aesthetics +
#   theme_pubr() +
#   labs_pubr() 



priv_dist_rar_pred <- ggeffects::ggpredict(priv_dist_rar,
                                   terms = c("City_dist"),
                                   type = "fe")

plot(priv_dist_rar_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Private alleles") +
  ylim(0, 400) +
  labs(title = "") +
  theme_pubr()

ggsave(here::here("./Figures_Tables/regressions/private_alleles.png"))
```

##### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = private/sites)) +
  xlab("Urbanization Score") +
  ylab("Private Alleles per Site") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

### Expected heterozygosity
#### City_dist
```{r}
ggplot(diversity,
       aes(x = City_dist,
           y = exp_het)) +
  xlab("Distance to City Center (km)") +
  ylab("Expected Heterozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

#### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = exp_het)) +
  xlab("Urbanization Score") +
  ylab("Expected Heterozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)
```

### Obs heterozygosity
#### City_dist
```{r}
ggplot(diversity,
       aes(x = City_dist,
           y = obs_het)) +
  xlab("Distance to City Center (km)") +
  ylab("Observed Heterozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

#### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = obs_het)) +
  xlab("Urbanization Score") +
  ylab("Observed Heterozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

### Expected homozygosity
#### City_dist
```{r}
ggplot(diversity,
       aes(x = City_dist,
           y = exp_hom)) +
  xlab("Distance to City Center (km)") +
  ylab("Expected Homozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

#### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = exp_hom)) +
  xlab("Urbanization Score") +
  ylab("Expected Homozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)
```

### Obs homozygosity
#### City_dist
```{r}
ggplot(diversity,
       aes(x = City_dist,
           y = obs_hom)) +
  xlab("Distance to City Center (km)") +
  ylab("Observed Homozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

#### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = obs_hom)) +
  xlab("Urbanization Score") +
  ylab("Observed Homozygosity") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

### FIS
#### City_dist
```{r}
# binary
ggplot(diversity,
       aes(x = City_dist,
           y = fis_bin)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr() 


# quantitative
ggplot(diversity %>%
         dplyr::filter(fis != 0),
       aes(x = City_dist,
           y = fis)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 
```

#### Urb_score
```{r}
# binary
ggplot(diversity,
       aes(x = urb_score,
           y = fis_bin)) +
  xlab("Urbanization Score") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr()  +
  xlim(4, -4)


# quantitative
ggplot(diversity %>%
         dplyr::filter(fis != 0),
       aes(x = urb_score,
           y = fis)) +
  xlab("Urbanization Score") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

### Pi
#### Without rarefaction
##### City_dist
```{r}
ggplot(diversity,
       aes(x = City_dist,
           y = pi)) +
  xlab("Distance to City Center (km)") +
  ylab("Pi") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = pi)) +
  xlab("Urbanization Score") +
  ylab("Pi") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### With rarefaction
##### City_dist
```{r}
# ggplot(diversity,
#        aes(x = City_dist,
#            y = pi/sites)) +
#   xlab("Distance to City Center (km)") +
#   ylab("Pi/sites") +
#   plot_aesthetics +
#   theme_pubr() +
#   labs_pubr() 

pi_dist_rar_pred <- ggeffects::ggpredict(pi_dist_rar,
                                   terms = c("City_dist"),
                                   type = "fe")

plot(pi_dist_rar_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Pi * 10") +
  ylim(NA, 0.04) +
  labs(title = "") +
  theme_pubr()

ggsave(here::here("./Figures_Tables/regressions/pi.png"))
```

##### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = pi/sites)) +
  xlab("Urbanization Score") +
  ylab("Pi/sites") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

### Sites
#### City_dist
```{r}
ggplot(diversity,
       aes(x = City_dist,
           y = sites)) +
  xlab("Distance to City Center (km)") +
  ylab("Sites") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

#### Urb_score
```{r}
ggplot(diversity,
       aes(x = urb_score,
           y = sites)) +
  xlab("Urbanization Score") +
  ylab("Sites") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

# Does urbanization predict genetic differentiation among populations?