# Set up notebook
## Load libraries & functions
```{r warning=F}
source(here::here("libraries.R"))
source(here::here("functions.R"))
```

## Import files
### Genetic data
#### All individuals
```{r}
# vcf
vcf <- read.vcfR(
  here::here("./summary_stats/output_populations_no-IBD/mmaf0.05_R0.75/populations.snps.vcf"),
  verbose = FALSE)

## Convert vcf to genind object
my_genind <- vcfR2genind(vcf)

```

#### Urban individuals
```{r}
# vcf
vcf_urb <- read.vcfR(
  here::here("./filtered_vcfs/urban_pops_dist_mmaf0.05_variant_sites.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_urb <- vcfR2genind(vcf_urb)

```


#### Rural individuals
```{r}
# vcf
vcf_rur <- read.vcfR(
  here::here("./filtered_vcfs/rural_pops_dist_mmaf0.05_variant_sites.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_rur <- vcfR2genind(vcf_rur)

```



#### 1 individual/population- URBAN ONLY
```{r}
# vcf
vcf_urb_subsamp <- read.vcfR(
  here::here("./filtered_vcfs/1indiv_per_mmaf0.05_variant_sites_URBAN.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_urb_subsamp <- vcfR2genind(vcf_urb_subsamp)
```

#### 1 individual/population- RURAL ONLY
```{r}
# vcf
vcf_rur_subsamp <- read.vcfR(
  here::here("./filtered_vcfs/1indiv_per_mmaf0.05_variant_sites_RURAL.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_rur_subsamp <- vcfR2genind(vcf_rur_subsamp)
```


### Coordinates
#### Create function: Import and create geographic distance matrix
```{r}
create_geog_dist_matrix <- function(vcf_filepath,
                                    pop_map_filepath){
    # lat/longs of sampling sites
  urb <- read.csv(here::here(vcf_filepath))
  
  # sampling site IDs and genetic sample names
  coords <- read.csv(here(pop_map_filepath)) %>%
  
  # if population doesn't start with "MWI", it should start with "MW"
    # rename col 1
    dplyr::rename(pop_id = 2) %>%
    dplyr::mutate(pop_id = as.factor(pop_id)) %>%
  
  # take entries with numeric populations and add "MW" as prefix
    add_MW_IDs() %>%
  
  # left vs. full join because some populations in urb_clean weren't genotyped due to lack of material
    left_join(., urb, by = "patch_id") %>%
    dplyr::mutate(patch_id = as.factor(patch_id),
                  pop_id = as.factor(pop_id)) %>%
    dplyr::select(c(1,2,6,7)) %>%
    dplyr::rename("y" = 3,
                  "x" = 4) %>%
    dplyr::arrange(sample) %>%
    dplyr::select("x", "y")
  
  # convert lat/longs to UTM
  coordinates(coords) <- c("x", "y")
  proj4string(coords) <- CRS("+proj=longlat +datum=WGS84")
  
  res <- spTransform(coords,
                     CRS("+proj=utm +zone=17 ellps=WGS84"))
  
  coords_utm <- res@coords %>%
    as.data.frame()
  
  # create distance matrix (Euclidean)
  geographical_distance_matrix_utm <- dist(coords_utm)
  
  return(geographical_distance_matrix_utm)
}
```

#### All individuals
```{r}
coords_utm <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./genomic_resources/pop_map3.csv")
```

#### Urban
```{r}
coords_utm_urb <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_map3_U_dist.csv")
```

#### Rural
```{r}
coords_utm_rur <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_map3_R_dist.csv")
```

#### 1 individual/population- URBAN ONLY
```{r}
coords_utm_urb_subsamp <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_map3_1indiv_per_pop_URBAN.csv")
```

#### 1 individual/population- RURAL ONLY
```{r}
coords_utm_rur_subsamp <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_map3_1indiv_per_pop_RURAL.csv")
```

# Create genetic distance matrices
```{r}
# all individuals
genetic_distance_matrix <- poppr::prevosti.dist(my_genind)

# urban individuals
genetic_distance_matrix_urb <- poppr::prevosti.dist(my_genind_urb)

# rural individuals
genetic_distance_matrix_rur <- poppr::prevosti.dist(my_genind_rur)

# 1 individual/site: urban
genetic_distance_matrix_urb_subsamp <- poppr::prevosti.dist(my_genind_urb_subsamp)


# 1 individual/site: rural
genetic_distance_matrix_rur_subsamp <- poppr::prevosti.dist(my_genind_rur_subsamp)


# "Prevostiâ€™s distance coefficient: a measurement over all loci of the proportion of unshared alleles. The variance of this sample distance is related to the genetic diversities used in calculating F-statistics... [accurate] at small distances compared with other coefficients...", from this paper: https://journals.ashs.org/hortsci/view/journals/hortsci/41/4/article-p967E.xml?tab_body=pdf
```

# Mantel correlogram (UTM)
## Create breakpoints vector
```{r}
break_points <- c(0, 500, 1000, 1500, 2000, 2500, 3000, 3500, 4000, 4500, 5000, 5500, 6000, 6500, 7000, 7500, 8000, 8500, 9000, 9500, 10000, 10500, 11000, 11500, 12000, 12500, 13000, 13500, 14000, 14500, 15000, 15500, 16000, 16500, 17000, 17500, 18000, 18500, 19000, 19500, 20000, 20500, 21000, 21500, 22000, 22500, 23000, 23500, 24000, 24500, 25000, 25500, 26000, 26500, 27000, 27500, 28000, 28500, 29000, 29500, 30000, 30500, 31000, 31500, 32000, 32500, 33000, 33500, 34000, 34500, 35000, 35500, 36000, 36500, 37000, 37500, 38000, 38500, 39000, 39500, 40000, 40500, 41000, 41500, 42000, 42500, 43000, 43500, 44000, 44500, 45000, 45500, 46000, 46500, 47000, 47500, 48000, 48500, 49000, 49500, 50000, 50500, 51000, 51500, 52000, 52500, 53000, 53500, 54000, 54500, 55000, 55500, 56000, 56500, 57000, 57500, 58000, 58500, 59000, 59500, 60000, 60500, 61000, 61500, 62000, 62500, 63000, 63500, 64000, 64500, 65000, 65500, 66000, 66500, 67000, 67500, 68000, 68500, 69000, 69500, 70000 )
```

## All individuals
```{r}
correlogram_utm <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix,
  D.geo = geographical_distance_matrix_utm,
  nperm = 1000,
  cutoff = F,
 # n.class = 35
   break.pts = break_points
   )

plot(correlogram_utm)
```

## Urban
```{r}
correlogram_utm_urb <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb,
  D.geo = coords_utm_urb,
  nperm = 1000,
  cutoff = F,
 # n.class = 35
   break.pts = break_points
   )

plot(correlogram_utm_urb)
```

## Rural
```{r}
correlogram_utm_rur <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur,
  D.geo = coords_utm_rur,
  nperm = 1000,
  cutoff = F,
 # n.class = 35
   break.pts = break_points
   )

plot(correlogram_utm_rur)
```


## 1 individual/site- Urban
```{r}
correlogram_utm_urb_subsamp <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb_subsamp,
  D.geo = coords_utm_urb_subsamp,
  nperm = 1000,
  cutoff = F,
 # n.class = 35
   break.pts = break_points
   )

plot(correlogram_utm_urb_subsamp)
```

## 1 individual/site- Rural
```{r}
correlogram_utm_rur_subsamp <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur_subsamp,
  D.geo = coords_utm_rur_subsamp,
  nperm = 1000,
  cutoff = F,
 # n.class = 35
   break.pts = break_points
   )

plot(correlogram_utm_rur_subsamp)
```

# Export (UTM)
## All individuals
```{r}
# export basic plot-----
png(here::here("./Figures_Tables/correlogram/mantel_correlogram_UTM.png"), 
    width = 1200, height = 800,
    res = 200) 

plot(correlogram_utm)  

dev.off()


# export table-----
correlogram_utm$mantel.res %>%
  as.data.frame() %>%
  dplyr::select("class.index", "n.dist", "Mantel.cor", "Pr(corrected)") %>%
  drop_na() %>%
  round(3) %>%
  dplyr::mutate(class.index = round(class.index, 0)) %>%
  dplyr::rename("Distance Class (m)" = 1,
                "N" = 2,
                "Mantel r" = 3,
                "p" = 4) %>%
  dplyr::mutate(p = if_else(
                p < 0.001,
                  "<0.001",
                as.character(p)
                )) %>%
  flextable() %>%
  align(align = "center",
        part = "all") %>%
  flextable::bold(i = ~ p <= 0.05, j = 4) %>%
  flextable::italic(i = 1, j = 4, part = "header") %>%
  flextable::autofit() %>%
  save_as_docx(path = here::here("./Figures_Tables/correlogram/correlogram_table_utm.docx"))


# export plot made with ggplot-----
## WITHOUT LABELLED SIGNIFICANT POINTS
corr_data_utm <- correlogram_utm[['mantel.res']] %>%
  as.data.frame() %>%
  dplyr::mutate(Significant = if_else(
    `Pr(corrected)` < 0.05,
    "Significant",
    "Not_significant"
  ))

corr_data_utm_plot1 <- corr_data_utm %>%
  ggplot(aes(x = class.index/1000,
             y = Mantel.cor)) +
  geom_line() +
  geom_point(aes(shape = Significant,
                 fill = Significant),
             size = 2.5) +
  geom_hline(yintercept = 0,
             linetype = "dotted") +
  scale_shape_manual(values = c(21,22)) +
  scale_fill_manual(values = c("white", "black" )) +
  labs(x = "Class Index (km)",
       y = expression(paste("Mantel ", italic(r)))) +
  ylim(-0.05, 0.15) +
  scale_x_continuous(limits = c(0, NA),
       breaks = c(0, 10, 20, 30, 40, 50, 60, 70)) +
  ggpubr::theme_pubr(legend = "none")

ggsave(plot = corr_data_utm_plot1,
       filename = here::here("./Figures_Tables/correlogram/correlogram_UTM_ggplot.png"),
       height = 3,
       width = 8)

## WITH LABELLED SIGNIFICANT POINTS
corr_data_utm_plot1 + 
  ggrepel::geom_text_repel(
    seed = 1,
    data = corr_data_utm %>%
      dplyr::filter(Significant == "Significant"),
    aes(x = class.index/1000,
        y = Mantel.cor,
        label = class.index/1000),
    nudge_x = 2,
    nudge_y = -0.01,
    direction = "both" # Allow labels to be moved in both x and y directions
  )


ggsave(plot = last_plot(),
       filename = here::here("./Figures_Tables/correlogram/correlogram_UTM_ggplot_labelled.png"),
       height = 3,
       width = 8)
```

## Urban
```{r}
# export basic plot-----
png(here::here("./Figures_Tables/correlogram/mantel_correlogram_UTM_urban.png"), 
    width = 1200, height = 800,
    res = 200) 

plot(correlogram_utm_urb)  

dev.off()


# export table-----
correlogram_utm_urb$mantel.res %>%
  as.data.frame() %>%
  dplyr::select("class.index", "n.dist", "Mantel.cor", "Pr(corrected)") %>%
  drop_na() %>%
  round(3) %>%
  dplyr::mutate(class.index = round(class.index, 0)) %>%
  dplyr::rename("Distance Class (m)" = 1,
                "N" = 2,
                "Mantel r" = 3,
                "p" = 4) %>%
  dplyr::mutate(p = if_else(
                p < 0.001,
                  "<0.001",
                as.character(p)
                )) %>%
  flextable() %>%
  align(align = "center",
        part = "all") %>%
  flextable::bold(i = ~ p <= 0.05, j = 4) %>%
  flextable::italic(i = 1, j = 4, part = "header") %>%
  flextable::autofit() %>%
  save_as_docx(path = here::here("./Figures_Tables/correlogram/correlogram_table_utm_urban.docx"))


# export plot made with ggplot-----
## WITHOUT LABELLED SIGNIFICANT POINTS
corr_data_utm_urb <- correlogram_utm_urb[['mantel.res']] %>%
  as.data.frame() %>%
  dplyr::mutate(Significant = if_else(
    `Pr(corrected)` < 0.05,
    "Significant",
    "Not_significant"
  ))

corr_data_utm_plot1_urb <- corr_data_utm_urb %>%
  dplyr::filter(is.na(Mantel.cor) == F) %>%
  ggplot(aes(x = class.index/1000,
             y = Mantel.cor)) +
  geom_line() +
  geom_point(aes(shape = Significant,
                 fill = Significant),
             size = 2.5) +
  geom_hline(yintercept = 0,
             linetype = "dotted") +
  scale_shape_manual(values = c(21,22)) +
  scale_fill_manual(values = c("white", "black" )) +
  labs(x = "Class Index (km)",
       y = expression(paste("Mantel ", italic(r)))) +
  #ylim(-0.05, 0.15) +
  scale_x_continuous(limits = c(0, NA),
       breaks = c(0, 10, 20, 30, 40, 50, 60)) +
  ggpubr::theme_pubr(legend = "none")

ggsave(plot = corr_data_utm_plot1_urb,
       filename = here::here("./Figures_Tables/correlogram/correlogram_UTM_ggplot_urban.png"),
       height = 3,
       width = 8)

## WITH LABELLED SIGNIFICANT POINTS
corr_data_utm_plot1_urb + 
  ggrepel::geom_text_repel(
    seed = 1,
    data = corr_data_utm_urb %>%
      dplyr::filter(Significant == "Significant"),
    aes(x = class.index/1000,
        y = Mantel.cor,
        label = class.index/1000),
    nudge_x = 2,
    nudge_y = -0.01,
    direction = "both" # Allow labels to be moved in both x and y directions
  )


ggsave(plot = last_plot(),
       filename = here::here("./Figures_Tables/correlogram/correlogram_UTM_ggplot_labelled_urban.png"),
       height = 3,
       width = 8)
```

## Rural
```{r}
# export basic plot-----
png(here::here("./Figures_Tables/correlogram/mantel_correlogram_UTM_rural.png"), 
    width = 1200, height = 800,
    res = 200) 

plot(correlogram_utm_rur)  

dev.off()


# export table-----
correlogram_utm_rur$mantel.res %>%
  as.data.frame() %>%
  dplyr::select("class.index", "n.dist", "Mantel.cor", "Pr(corrected)") %>%
  drop_na() %>%
  round(3) %>%
  dplyr::mutate(class.index = round(class.index, 0)) %>%
  dplyr::rename("Distance Class (m)" = 1,
                "N" = 2,
                "Mantel r" = 3,
                "p" = 4) %>%
  dplyr::mutate(p = if_else(
                p < 0.001,
                  "<0.001",
                as.character(p)
                )) %>%
  flextable() %>%
  align(align = "center",
        part = "all") %>%
  flextable::bold(i = ~ p <= 0.05, j = 4) %>%
  flextable::italic(i = 1, j = 4, part = "header") %>%
  flextable::autofit() %>%
  save_as_docx(path = here::here("./Figures_Tables/correlogram/correlogram_table_utm_rural.docx"))


# export plot made with ggplot-----
## WITHOUT LABELLED SIGNIFICANT POINTS
corr_data_utm_rur <- correlogram_utm_rur[['mantel.res']] %>%
  as.data.frame() %>%
  dplyr::mutate(Significant = if_else(
    `Pr(corrected)` < 0.05,
    "Significant",
    "Not_significant"
  ))

corr_data_utm_plot1_rur <- corr_data_utm_rur %>%
  dplyr::filter(is.na(Mantel.cor) == F) %>%
  ggplot(aes(x = class.index/1000,
             y = Mantel.cor)) +
  geom_line() +
  geom_point(aes(shape = Significant,
                 fill = Significant),
             size = 2.5) +
  geom_hline(yintercept = 0,
             linetype = "dotted") +
  scale_shape_manual(values = c(21,22)) +
  scale_fill_manual(values = c("white", "black" )) +
  labs(x = "Class Index (km)",
       y = expression(paste("Mantel ", italic(r)))) +
  #ylim(-0.05, 0.15) +
  scale_x_continuous(limits = c(0, NA),
       breaks = c(0, 10, 20, 30, 40, 50, 60)) +
  ggpubr::theme_pubr(legend = "none")

ggsave(plot = corr_data_utm_plot1_rur,
       filename = here::here("./Figures_Tables/correlogram/correlogram_UTM_ggplot_rural.png"),
       height = 3,
       width = 8)

## WITH LABELLED SIGNIFICANT POINTS
corr_data_utm_plot1_rur + 
  ggrepel::geom_text_repel(
    seed = 1,
    data = corr_data_utm_rur %>%
      dplyr::filter(Significant == "Significant"),
    aes(x = class.index/1000,
        y = Mantel.cor,
        label = class.index/1000),
    nudge_x = 2,
    nudge_y = -0.01,
    direction = "both" # Allow labels to be moved in both x and y directions
  )


ggsave(plot = last_plot(),
       filename = here::here("./Figures_Tables/correlogram/correlogram_UTM_ggplot_labelled_rural.png"),
       height = 3,
       width = 8)
```

## 1 individual/site: Urban
```{r}
# export basic plot-----
png(here::here("./Figures_Tables/correlogram/mantel_correlogram_UTM_urban_subsamp.png"), 
    width = 1200, height = 800,
    res = 200) 

plot(correlogram_utm_urb_subsamp)  

dev.off()


# export table-----
correlogram_utm_urb_subsamp$mantel.res %>%
  as.data.frame() %>%
  dplyr::select("class.index", "n.dist", "Mantel.cor", "Pr(corrected)") %>%
  drop_na() %>%
  round(3) %>%
  dplyr::mutate(class.index = round(class.index, 0)) %>%
  dplyr::rename("Distance Class (m)" = 1,
                "N" = 2,
                "Mantel r" = 3,
                "p" = 4) %>%
  dplyr::mutate(p = if_else(
                p < 0.001,
                  "<0.001",
                as.character(p)
                )) %>%
  flextable() %>%
  align(align = "center",
        part = "all") %>%
  flextable::bold(i = ~ p <= 0.05, j = 4) %>%
  flextable::italic(i = 1, j = 4, part = "header") %>%
  flextable::autofit() %>%
  save_as_docx(path = here::here("./Figures_Tables/correlogram/correlogram_table_utm_urban_subsamp.docx"))


# export plot made with ggplot-----
## WITHOUT LABELLED SIGNIFICANT POINTS
corr_data_utm_urb_subsamp <- correlogram_utm_urb_subsamp[['mantel.res']] %>%
  as.data.frame() %>%
  dplyr::mutate(Significant = if_else(
    `Pr(corrected)` < 0.05,
    "Significant",
    "Not_significant"
  ))

corr_data_utm_plot1_urb_subsamp <- corr_data_utm_urb_subsamp %>%
  dplyr::filter(is.na(Mantel.cor) == F) %>%
  ggplot(aes(x = class.index/1000,
             y = Mantel.cor)) +
  geom_line() +
  geom_point(aes(shape = Significant,
                 fill = Significant),
             size = 2.5) +
  geom_hline(yintercept = 0,
             linetype = "dotted") +
  scale_shape_manual(values = c(21,22)) +
  scale_fill_manual(values = c("white", "black" )) +
  labs(x = "Class Index (km)",
       y = expression(paste("Mantel ", italic(r)))) +
  #ylim(-0.05, 0.15) +
  scale_x_continuous(limits = c(0, NA),
       breaks = c(0, 10, 20, 30, 40, 50, 60)) +
  ggpubr::theme_pubr(legend = "none")

ggsave(plot = corr_data_utm_plot1_urb_subsamp,
       filename = here::here("./Figures_Tables/correlogram/correlogram_UTM_ggplot_urban_subsamp.png"),
       height = 3,
       width = 8)

## WITH LABELLED SIGNIFICANT POINTS
corr_data_utm_plot1_urb_subsamp + 
  ggrepel::geom_text_repel(
    seed = 1,
    data = corr_data_utm_urb_subsamp %>%
      dplyr::filter(Significant == "Significant"),
    aes(x = class.index/1000,
        y = Mantel.cor,
        label = class.index/1000),
    nudge_x = 2,
    nudge_y = -0.01,
    direction = "both" # Allow labels to be moved in both x and y directions
  )


ggsave(plot = last_plot(),
       filename = here::here("./Figures_Tables/correlogram/correlogram_UTM_ggplot_labelled_urban_subsamp.png"),
       height = 3,
       width = 8)
```

## 1 individual/site: Rural
```{r}
# export basic plot-----
png(here::here("./Figures_Tables/correlogram/mantel_correlogram_UTM_rural_subsamp.png"), 
    width = 1200, height = 800,
    res = 200) 

plot(correlogram_utm_rur_subsamp)  

dev.off()


# export table-----
correlogram_utm_rur_subsamp$mantel.res %>%
  as.data.frame() %>%
  dplyr::select("class.index", "n.dist", "Mantel.cor", "Pr(corrected)") %>%
  drop_na() %>%
  round(3) %>%
  dplyr::mutate(class.index = round(class.index, 0)) %>%
  dplyr::rename("Distance Class (m)" = 1,
                "N" = 2,
                "Mantel r" = 3,
                "p" = 4) %>%
  dplyr::mutate(p = if_else(
                p < 0.001,
                  "<0.001",
                as.character(p)
                )) %>%
  flextable() %>%
  align(align = "center",
        part = "all") %>%
  flextable::bold(i = ~ p <= 0.05, j = 4) %>%
  flextable::italic(i = 1, j = 4, part = "header") %>%
  flextable::autofit() %>%
  save_as_docx(path = here::here("./Figures_Tables/correlogram/correlogram_table_utm_rural_subsamp.docx"))


# export plot made with ggplot-----
## WITHOUT LABELLED SIGNIFICANT POINTS
corr_data_utm_rur_subsamp <- correlogram_utm_rur_subsamp[['mantel.res']] %>%
  as.data.frame() %>%
  dplyr::mutate(Significant = if_else(
    `Pr(corrected)` < 0.05,
    "Significant",
    "Not_significant"
  ))

corr_data_utm_plot1_rur_subsamp <- corr_data_utm_rur_subsamp %>%
  dplyr::filter(is.na(Mantel.cor) == F) %>%
  ggplot(aes(x = class.index/1000,
             y = Mantel.cor)) +
  geom_line() +
  geom_point(aes(shape = Significant,
                 fill = Significant),
             size = 2.5) +
  geom_hline(yintercept = 0,
             linetype = "dotted") +
  scale_shape_manual(values = c(21,22)) +
  scale_fill_manual(values = c("white", "black" )) +
  labs(x = "Class Index (km)",
       y = expression(paste("Mantel ", italic(r)))) +
  #ylim(-0.05, 0.15) +
  scale_x_continuous(limits = c(0, NA),
       breaks = c(0, 10, 20, 30, 40, 50, 60)) +
  ggpubr::theme_pubr(legend = "none")

ggsave(plot = corr_data_utm_plot1_rur_subsamp,
       filename = here::here("./Figures_Tables/correlogram/correlogram_UTM_ggplot_rural_subsamp.png"),
       height = 3,
       width = 8)

## WITH LABELLED SIGNIFICANT POINTS
corr_data_utm_plot1_rur_subsamp + 
  ggrepel::geom_text_repel(
    seed = 1,
    data = corr_data_utm_rur_subsamp %>%
      dplyr::filter(Significant == "Significant"),
    aes(x = class.index/1000,
        y = Mantel.cor,
        label = class.index/1000),
    nudge_x = 2,
    nudge_y = -0.01,
    direction = "both" # Allow labels to be moved in both x and y directions
  )


ggsave(plot = last_plot(),
       filename = here::here("./Figures_Tables/correlogram/correlogram_UTM_ggplot_labelled_rural_subsamp.png"),
       height = 3,
       width = 8)
```
