# Set up notebook
## Load libraries & functions
```{r warning=F}
source(here::here("libraries.R"))
source(here::here("functions.R"))
```

## Import files
### Genetic data
#### All individuals
```{r}
# vcf
vcf <- read.vcfR(
  here::here("./summary_stats/output_populations_no-IBD/mmaf0.05_R0.75/populations.snps.vcf"),
  verbose = FALSE)

## Convert vcf to genind object
my_genind <- vcfR2genind(vcf)

```

#### Urban individuals
```{r}
# vcf
vcf_urb <- read.vcfR(
  here::here("./filtered_vcfs/urban_pops_dist_mmaf0.05_variant_sites.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_urb <- vcfR2genind(vcf_urb)

```


#### Rural individuals
```{r}
# vcf
vcf_rur <- read.vcfR(
  here::here("./filtered_vcfs/rural_pops_dist_mmaf0.05_variant_sites.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_rur <- vcfR2genind(vcf_rur)

```



#### 1 individual/population
```{r}
# vcf
vcf_subsamp <- read.vcfR(
  here::here("./filtered_vcfs/1indiv_per_mmaf0.05_variant_sites.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_subsamp <- vcfR2genind(vcf_subsamp)
```


#### 1 individual/population- URBAN ONLY
```{r}
# vcf
vcf_urb_subsamp <- read.vcfR(
  here::here("./filtered_vcfs/1indiv_per_mmaf0.05_variant_sites_URBAN.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_urb_subsamp <- vcfR2genind(vcf_urb_subsamp)
```

#### 1 individual/population- RURAL ONLY
```{r}
# vcf
vcf_rur_subsamp <- read.vcfR(
  here::here("./filtered_vcfs/1indiv_per_mmaf0.05_variant_sites_RURAL.recode.vcf"), verbose = FALSE)

## Convert vcf to genind object
my_genind_rur_subsamp <- vcfR2genind(vcf_rur_subsamp)
```


### Coordinates
#### Create function: Import and create geographic distance matrix
```{r}
create_geog_dist_matrix <- function(vcf_filepath,
                                    pop_map_filepath){
    # lat/longs of sampling sites
  urb <- read.csv(here::here(vcf_filepath))
  
  # sampling site IDs and genetic sample names
  coords <- read.csv(here(pop_map_filepath)) %>%
  
  # if population doesn't start with "MWI", it should start with "MW"
    # rename col 1
    dplyr::rename(pop_id = 2) %>%
    dplyr::mutate(pop_id = as.factor(pop_id)) %>%
  
  # take entries with numeric populations and add "MW" as prefix
    add_MW_IDs() %>%
  
  # left vs. full join because some populations in urb_clean weren't genotyped due to lack of material
    left_join(., urb, by = "patch_id") %>%
    dplyr::mutate(patch_id = as.factor(patch_id),
                  pop_id = as.factor(pop_id)) %>%
    dplyr::select(c(1,2,6,7)) %>%
    dplyr::rename("y" = 3,
                  "x" = 4) %>%
    dplyr::arrange(sample) %>%
    dplyr::select("x", "y")
  
  # convert lat/longs to UTM
  coordinates(coords) <- c("x", "y")
  proj4string(coords) <- CRS("+proj=longlat +datum=WGS84")
  
  res <- spTransform(coords,
                     CRS("+proj=utm +zone=17 ellps=WGS84"))
  
  coords_utm <- res@coords %>%
    as.data.frame()
  
  # create distance matrix (Euclidean)
  geographical_distance_matrix_utm <- dist(coords_utm)
  
  return(geographical_distance_matrix_utm)
}
```

#### All individuals
```{r}
coords_utm <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./genomic_resources/pop_map3.csv")
```

#### Urban
```{r}
coords_utm_urb <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_map3_U_dist.csv")
```

#### Rural
```{r}
coords_utm_rur <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_map3_R_dist.csv")
```
#### 1 individual/population
```{r}
coords_utm_subsamp <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_map3_1indiv_per_pop.csv")
```


#### 1 individual/population- URBAN ONLY
```{r}
coords_utm_urb_subsamp <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_map3_1indiv_per_pop_URBAN.csv")
```

#### 1 individual/population- RURAL ONLY
```{r}
coords_utm_rur_subsamp <- create_geog_dist_matrix(
  "./clean_data/urb_metrics.csv",
  "./clean_data/pop_map3_1indiv_per_pop_RURAL.csv")
```

# Create genetic distance matrices
```{r}
# all individuals
genetic_distance_matrix <- poppr::prevosti.dist(my_genind)

# urban individuals
genetic_distance_matrix_urb <- poppr::prevosti.dist(my_genind_urb)

# rural individuals
genetic_distance_matrix_rur <- poppr::prevosti.dist(my_genind_rur)

# 1 individual/site
genetic_distance_matrix_subsamp <- poppr::prevosti.dist(my_genind_subsamp)

# 1 individual/site: urban
genetic_distance_matrix_urb_subsamp <- poppr::prevosti.dist(my_genind_urb_subsamp)


# 1 individual/site: rural
genetic_distance_matrix_rur_subsamp <- poppr::prevosti.dist(my_genind_rur_subsamp)


# "Prevostiâ€™s distance coefficient: a measurement over all loci of the proportion of unshared alleles. The variance of this sample distance is related to the genetic diversities used in calculating F-statistics... [accurate] at small distances compared with other coefficients...", from this paper: https://journals.ashs.org/hortsci/view/journals/hortsci/41/4/article-p967E.xml?tab_body=pdf
```

# Mantel correlogram (UTM)
## Create breakpoints vectors
### Urban: 2.5km spacing btwn sampling sites
```{r}
break_points_2500m <- c(0, 2500, 5000, 7500, 10000, 12500, 15000, 17500, 20000, 22500, 25000, 27500, 30000, 32500, 35000, 37500, 40000, 42500, 45000, 47500, 50000, 52500, 55000, 57500, 60000, 62500, 65000, 67500, 70000 )
```

### Urban: 3km spacing btwn sampling sites
```{r}
break_points_3000m <- c(0, 3000, 6000, 9000, 12000, 15000, 18000, 21000, 24000, 27000, 30000, 33000, 36000, 39000, 42000, 45000, 48000, 51000, 54000, 57000, 60000, 63000, 66000, 69000)
```

### Rural: 10km spacing btwn sampling sites
```{r}
break_points_10km <- c(0, 10000, 20000, 30000, 40000, 50000, 60000, 70000)
```

### Rural: 12km spacing btwn sampling sites
```{r}
break_points_12km <- c(0, 12000, 24000, 36000, 48000, 60000, 72000)
```

## All individuals
### 2.5km spacing
```{r}
correlogram_utm <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix,
  D.geo = coords_utm,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_2500m
   )

plot(correlogram_utm)
```

### 10km spacing
```{r}
correlogram_utm_10km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix,
  D.geo = coords_utm,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_10km
   )

plot(correlogram_utm_10km)
```

## Urban
### 2.5km spacing
```{r}
correlogram_utm_urb <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb,
  D.geo = coords_utm_urb,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_2500m
   )

plot(correlogram_utm_urb)
```

### 3km spacing
```{r}
correlogram_utm_urb_3km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb,
  D.geo = coords_utm_urb,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_3000m
   )

plot(correlogram_utm_urb_3km)
```

## Rural
### 10km spacing
```{r}
correlogram_utm_rur <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur,
  D.geo = coords_utm_rur,
  nperm = 1000,
  cutoff = T,
   break.pts = break_points_10km
   )

plot(correlogram_utm_rur)
```

### 12km spacing
```{r}
correlogram_utm_rur_12km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur,
  D.geo = coords_utm_rur,
  nperm = 1000,
  cutoff = T,
   break.pts = break_points_12km
   )

plot(correlogram_utm_rur_12km)
```

## 1 individual/site
### 2.5km spacing
```{r}
correlogram_utm_subsamp <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_subsamp,
  D.geo = coords_utm_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_2500m
   )

plot(correlogram_utm_subsamp)
```

### 10km spacing
```{r}
correlogram_utm_subsamp_10km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_subsamp,
  D.geo = coords_utm_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_10km
   )

plot(correlogram_utm_subsamp_10km)
```

## 1 individual/site- Urban
### 2.5km spacing
```{r}
correlogram_utm_urb_subsamp <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb_subsamp,
  D.geo = coords_utm_urb_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_2500m
   )

plot(correlogram_utm_urb_subsamp)
```

### 3km spacing
```{r}
correlogram_utm_urb_subsamp_3km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_urb_subsamp,
  D.geo = coords_utm_urb_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_3000m
   )

plot(correlogram_utm_urb_subsamp_3km)
```

## 1 individual/site- Rural
### 10km spacing
```{r}
correlogram_utm_rur_subsamp <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur_subsamp,
  D.geo = coords_utm_rur_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_10km
   )

plot(correlogram_utm_rur_subsamp)
```

### 12km spacing
```{r}
correlogram_utm_rur_subsamp_12km <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix_rur_subsamp,
  D.geo = coords_utm_rur_subsamp,
  nperm = 1000,
  cutoff = T,
  break.pts = break_points_12km
   )

plot(correlogram_utm_rur_subsamp_12km)
```

# Export tables
## Create function
```{r}
export_corr_table <- function(correlogram_object,
                              output_file_name){
  
  filepath <- paste0(
    "./Figures_Tables/correlogram/tables/",
    deparse(substitute(output_file_name)),
    ".docx")
  
  correlogram_object %>%
    purrr::pluck("mantel.res") %>%
  as.data.frame() %>%
  dplyr::select("class.index", "n.dist", "Mantel.cor", "Pr(corrected)") %>%
  drop_na() %>%
  round(3) %>%
  dplyr::mutate(class.index = round(class.index, 0)) %>%
  dplyr::rename("Distance Class (m)" = 1,
                "N" = 2,
                "Mantel r" = 3,
                "p" = 4) %>%
  dplyr::mutate(p = if_else(
                p < 0.001,
                  "<0.001",
                as.character(p)
                )) %>%
  flextable() %>%
  align(align = "center",
        part = "all") %>%
  flextable::bold(i = ~ p <= 0.05, j = 4) %>%
  flextable::italic(i = 1, j = 4, part = "header") %>%
  flextable::autofit() %>%
  save_as_docx(path = here::here(filepath))
}
```

## All individuals
```{r}
# 2.5 km
export_corr_table(correlogram_utm, all_indiv_2500m)

# 10 km
export_corr_table(correlogram_utm_10km, all_indiv_10km)
```

## Urban
```{r}
# 2.5 km
export_corr_table(correlogram_utm_urb, urban_2500m)

# 3 km
export_corr_table(correlogram_utm_urb_3km, urban_3km)
```

## Rural
```{r}
# 10 km
export_corr_table(correlogram_utm_rur, rural_10km)

# 12 km
export_corr_table(correlogram_utm_rur_12km, rural_12km)
```



## 1 individual/site
```{r}
# 2.5 km
export_corr_table(correlogram_utm_subsamp, one_indiv_2500m)

# 10 km
export_corr_table(correlogram_utm_subsamp_10km, one_indiv_10km)
```

## 1 individual/site: Urban
```{r}
# 2.5 km
export_corr_table(correlogram_utm_urb_subsamp, one_indiv_urban_2500m)

# 3 km
export_corr_table(correlogram_utm_urb_subsamp_3km, one_indiv_urban_3km)
```

## 1 individual/site: Rural
```{r}
# 10 km
export_corr_table(correlogram_utm_rur_subsamp, one_indiv_rural_10km)

# 12 km
export_corr_table(correlogram_utm_rur_subsamp_12km, one_indiv_rural_12km)
```

# Export figures
## Create function
```{r}
export_corr_figure <- function(correlogram_object,
                               output_file_name){
  
    filepath <- paste0(
    "./Figures_Tables/correlogram/figures/",
    deparse(substitute(output_file_name)),
    ".png")
  
    # create df
    corr_data <- correlogram_object %>%
      purrr::pluck("mantel.res") %>%
  as.data.frame() %>%
  dplyr::mutate(Significant = if_else(
    `Pr(corrected)` < 0.05,
    "Significant",
    "Not_significant"
  )) 
    
  # plot
  ggplot(corr_data,
         aes(x = class.index/1000,
             y = Mantel.cor)) +
  geom_line() +
  geom_point(aes(shape = Significant,
                 fill = Significant),
             size = 2.5) +
  geom_hline(yintercept = 0,
             linetype = "dotted") +
  scale_shape_manual(values = c(21,22)) +
  scale_fill_manual(values = c("white", "black" )) +
  labs(x = "Class Index (km)",
       y = expression(paste("Mantel ", italic(r)))) +
  # ylim(-0.05, 0.15) +
  # scale_x_continuous(limits = c(0, NA),
  #      breaks = c(0, 10, 20, 30, 40, 50, 60, 70)) +
  ggpubr::theme_pubr(legend = "none")
  ggrepel::geom_text_repel(
    seed = 1,
    data = corr_data %>%
      dplyr::filter(Significant == "Significant"),
    aes(x = class.index/1000,
        y = Mantel.cor,
        label = class.index/1000),
    nudge_x = 2,
    nudge_y = -0.01,
    direction = "both"
  )
  
  ggsave(plot = last_plot(),
         filename = here::here(filepath),
         height = 3,
         width = 8)
}
```

## All individuals
```{r}
export_corr_figure(correlogram_utm,
                   all_indiv_2500m)

export_corr_figure(correlogram_utm_10km,
                   all_indiv_10km)
```

## Urban
```{r}
# 2.5 km
export_corr_figure(correlogram_utm_urb, urban_2500m)

# 3 km
export_corr_figure(correlogram_utm_urb_3km, urban_3km)
```

## Rural
```{r}
# 10 km
export_corr_figure(correlogram_utm_rur, rural_10km)

# 12 km
export_corr_figure(correlogram_utm_rur_12km, rural_12km)
```

## 1 individual/site
```{r}
# 2.5 km
export_corr_figure(correlogram_utm_subsamp, one_indiv_2500m)

# 10 km
export_corr_figure(correlogram_utm_subsamp_10km, one_indiv_10km)
```

## 1 individual/site: Urban
```{r}
# 2.5 km
export_corr_figure(correlogram_utm_urb_subsamp, one_indiv_urban_2500m)

# 3 km
export_corr_figure(correlogram_utm_urb_subsamp_3km, one_indiv_urban_3km)
```

## 1 individual/site: Rural
```{r}
# 10 km
export_corr_figure(correlogram_utm_rur_subsamp, one_indiv_rural_10km)

# 12 km
export_corr_figure(correlogram_utm_rur_subsamp_12km, one_indiv_rural_12km)
```

