# Set up notebook
## Load libraries & functions
```{r warning=F}
source(here::here("libraries.R"))
source(here::here("functions.R"))
```

## Import files
### Genetic data
```{r}
# vcf
vcf <- read.vcfR(
  here::here("./summary_stats/output_populations_no-IBD/mmaf0.05_R0.75/populations.snps.vcf"),
  verbose = FALSE)

## Convert vcf to genind object
my_genind <- vcfR2genind(vcf)

# tidy up colnames (remove periods)
inputdata1 <- my_genind$tab
colnames(inputdata1) <- gsub("\\.", "_", "") colnames(inputdata1)

```

### Coordinates
```{r}
# lat/longs of sampling sites
urb <- read.csv(here::here("./clean_data/urb_metrics.csv"))

# sampling site IDs and genetic sample names
coords <- read.csv(here("./genomic_resources/pop_map3.csv")) %>%

# if population doesn't start with "MWI", it should start with "MW"
  # rename col 1
  dplyr::rename(pop_id = 2) %>%
  dplyr::mutate(pop_id = as.factor(pop_id)) %>%

# take entries with numeric populations and add "MW" as prefix
  add_MW_IDs() %>%

# left vs. full join because some populations in urb_clean weren't genotyped due to lack of material
  left_join(., urb, by = "patch_id") %>%
  dplyr::mutate(patch_id = as.factor(patch_id),
                pop_id = as.factor(pop_id)) %>%
  dplyr::select(c(1,2,6,7)) %>%
  dplyr::rename("y" = 3,
                "x" = 4) %>%
  dplyr::arrange(sample) %>%
  dplyr::select("x", "y")

# convert lat/longs to UTM
coords2 <- coords
coordinates(coords2) <- c("x", "y")
proj4string(coords2) <- CRS("+proj=longlat +datum=WGS84")

res <- spTransform(coords2,
                   CRS("+proj=utm +zone=17 ellps=WGS84"))
coords_utm <- res@coords %>%
  as.data.frame()
```

# Create distance matrices
## Genetic distances
```{r}
# create genetic distance matrix

# these end up generating this error for the mantel correlog:
# 
# Error in cor(as.vector(xdis), ydis, method = method, use = use) : 
#  missing observations in cov/cor
# genetic_distance_matrix <- poppr::nei.dist(my_genind)
# genetic_distance_matrix <- poppr::edwards.dist(my_genind)
# genetic_distance_matrix <- poppr::rogers.dist(my_genind)
# genetic_distance_matrix <- poppr::reynolds.dist(my_genind)

# these don't return the error:
genetic_distance_matrix <- poppr::prevosti.dist(my_genind)


# "Prevostiâ€™s distance coefficient: a measurement over all loci of the proportion of unshared alleles. The variance of this sample distance is related to the genetic diversities used in calculating F-statistics... [accurate] at small distances compared with other coefficients...", from this paper: https://journals.ashs.org/hortsci/view/journals/hortsci/41/4/article-p967E.xml?tab_body=pdf
```

## Geographic distances
### lat/long
```{r}
# Calculate the geographical distance matrix (Euclidean distance)
geographical_distance_matrix <- dist(coords)
```

### UTM
```{r}
geographical_distance_matrix_utm <- dist(coords_utm)
```

# Mantel correlogram
## Create
### lat/longs
```{r}
# correlogram <- vegan::mantel.correlog(
#   D.eco = genetic_distance_matrix,
#   D.geo = geographical_distance_matrix,
#   nperm = 100, # change back to 10000
#   cutoff = F,
#    n.class = 150,
#    # break.pts = c(0, 500, 1000, 1500, 2000, 2500, 3000, 3500, 4000, 4500, 5000, 5500, 6000, 6500, 7000, 7500, 8000, 8500, 9000, 9500, 10000, 10500, 11000, 11500, 12000, 12500, 13000, 13500, 14000, 14500, 15000, 15500, 16000, 16500, 17000, 17500, 18000, 18500, 19000, 19500, 20000, 20500, 21000, 21500, 22000, 22500, 23000, 23500, 24000, 24500, 25000, 25500, 26000, 26500, 27000, 27500, 28000, 28500, 29000, 29500, 30000, 30500, 31000, 31500, 32000, 32500, 33000, 33500, 34000, 34500, 35000, 35500, 36000, 36500, 37000, 37500, 38000, 38500, 39000, 39500, 40000, 40500, 41000, 41500, 42000, 42500, 43000, 43500, 44000, 44500, 45000, 45500, 46000, 46500, 47000, 47500, 48000, 48500, 49000, 49500, 50000, 50500, 51000, 51500, 52000, 52500, 53000, 53500, 54000, 54500, 55000, 55500, 56000, 56500, 57000, 57500, 58000, 58500, 59000, 59500, 60000, 60500, 61000, 61500, 62000, 62500, 63000, 63500, 64000, 64500, 65000, 65500, 66000, 66500, 67000, 67500, 68000, 68500, 69000, 69500, 70000 )
#   )
# 
# plot(correlogram)
```

### UTM
```{r}
correlogram_utm <- vegan::mantel.correlog(
  D.eco = genetic_distance_matrix,
  D.geo = geographical_distance_matrix_utm,
  nperm = 10000,
  cutoff = F,
 # n.class = 35
   break.pts = c(0, 500, 1000, 1500, 2000, 2500, 3000, 3500, 4000, 4500, 5000, 5500, 6000, 6500, 7000, 7500, 8000, 8500, 9000, 9500, 10000, 10500, 11000, 11500, 12000, 12500, 13000, 13500, 14000, 14500, 15000, 15500, 16000, 16500, 17000, 17500, 18000, 18500, 19000, 19500, 20000, 20500, 21000, 21500, 22000, 22500, 23000, 23500, 24000, 24500, 25000, 25500, 26000, 26500, 27000, 27500, 28000, 28500, 29000, 29500, 30000, 30500, 31000, 31500, 32000, 32500, 33000, 33500, 34000, 34500, 35000, 35500, 36000, 36500, 37000, 37500, 38000, 38500, 39000, 39500, 40000, 40500, 41000, 41500, 42000, 42500, 43000, 43500, 44000, 44500, 45000, 45500, 46000, 46500, 47000, 47500, 48000, 48500, 49000, 49500, 50000, 50500, 51000, 51500, 52000, 52500, 53000, 53500, 54000, 54500, 55000, 55500, 56000, 56500, 57000, 57500, 58000, 58500, 59000, 59500, 60000, 60500, 61000, 61500, 62000, 62500, 63000, 63500, 64000, 64500, 65000, 65500, 66000, 66500, 67000, 67500, 68000, 68500, 69000, 69500, 70000 )
   )

plot(correlogram_utm)
```

## Export
### lat/long
```{r}
# # export plot
# png(here::here("./Figures_Tables/correlogram/mantel_correlogram.png"), 
#     width = 1200, height = 800,
#     res = 200) 
# 
# plot(correlogram)  
# 
# dev.off()
# 
# 
# # export table
# correlogram$mantel.res %>%
#   as.data.frame() %>%
#   dplyr::select("class.index", "n.dist", "Mantel.cor", "Pr(corrected)") %>%
#   drop_na() %>%
#   round(3) %>%
#   dplyr::rename("Distance Class" = 1,
#                 "N" = 2,
#                 "Mantel r" = 3,
#                 "p" = 4) %>%
#   flextable() %>%
#   align(align = "center",
#         part = "all") %>%
#   flextable::bold(i = ~ p <= 0.05, j = 4) %>%
#   flextable::italic(i = 1, j = 4, part = "header") %>%
#   flextable::autofit() %>%
#   save_as_docx(path = here::here("./Figures_Tables/correlogram/correlogram_table.docx"))

```

### UTM
```{r}
# export basic plot-----
png(here::here("./Figures_Tables/correlogram/mantel_correlogram_UTM.png"), 
    width = 1200, height = 800,
    res = 200) 

plot(correlogram_utm)  

dev.off()


# export table-----
correlogram_utm$mantel.res %>%
  as.data.frame() %>%
  dplyr::select("class.index", "n.dist", "Mantel.cor", "Pr(corrected)") %>%
  drop_na() %>%
  round(3) %>%
  dplyr::mutate(class.index = round(class.index, 0)) %>%
  dplyr::rename("Distance Class (m)" = 1,
                "N" = 2,
                "Mantel r" = 3,
                "p" = 4) %>%
  flextable() %>%
  align(align = "center",
        part = "all") %>%
  flextable::bold(i = ~ p <= 0.05, j = 4) %>%
  flextable::italic(i = 1, j = 4, part = "header") %>%
  flextable::autofit() %>%
  save_as_docx(path = here::here("./Figures_Tables/correlogram/correlogram_table_utm.docx"))

# export plot made with ggplot-----
## WITHOUT LABELLED SIGNIFICANT POINTS
corr_data_utm <- correlogram_utm[['mantel.res']] %>%
  as.data.frame() %>%
  dplyr::mutate(Significant = if_else(
    `Pr(corrected)` < 0.05,
    "Significant",
    "Not_significant"
  ))

corr_data_utm_plot1 <- corr_data_utm %>%
  ggplot(aes(x = class.index/1000,
             y = Mantel.cor)) +
  geom_line() +
  geom_point(aes(shape = Significant,
                 fill = Significant),
             size = 2.5) +
  geom_hline(yintercept = 0,
             linetype = "dotted") +
  scale_shape_manual(values = c(21,22)) +
  scale_fill_manual(values = c("white", "black" )) +
  labs(x = "Class Index (km)",
       y = expression(paste("Mantel ", italic(r)))) +
  ylim(-0.05, 0.15) +
  scale_x_continuous(limits = c(0, NA),
       breaks = c(0, 10, 20, 30, 40, 50, 60, 70)) +
  ggpubr::theme_pubr(legend = "none")

ggsave(plot = corr_data_utm_plot1,
       filename = here::here("./Figures_Tables/correlogram/correlogram_UTM_ggplot.png"),
       height = 3,
       width = 8)

## WITH LABELLED SIGNIFICANT POINTS
corr_data_utm_plot1 + 
  ggrepel::geom_text_repel(
    seed = 1,
    data = corr_data_utm %>%
      dplyr::filter(Significant == "Significant"),
    aes(x = class.index/1000,
        y = Mantel.cor,
        label = class.index/1000),
    nudge_x = 2,
    nudge_y = -0.01,
    direction = "both" # Allow labels to be moved in both x and y directions
  )


ggsave(plot = last_plot(),
       filename = here::here("./Figures_Tables/correlogram/correlogram_UTM_ggplot_labelled.png"),
       height = 3,
       width = 8)
```
