# Set up notebook
## Load libraries & functions
```{r warning=F}
source(here::here("libraries.R"))
source(here::here("functions.R"))

# install.packages("memgene")
library(memgene)
library(officer)
```

# MEMgene tutorial
```{r}
## Load the caribou genetic data
caribouData <- read.csv(system.file("extdata/caribou.csv",
package="memgene"))

## Create objects for positional information and genotypes
caribouXY <- caribouData[ ,1:2]
caribouGen <- caribouData[, 3:ncol(caribouData)]

## Produce a proportion of shared alleles genetic distance matrix using the convenience wrapper function provided with the package
caribouDM <- codomToPropShared(caribouGen)

## Run the MEMGENE analysis
## May take several minutes
if (!exists("caribouAnalysis"))
caribouAnalysis <- mgQuick(caribouDM, caribouXY)

# Visualize results
plot(caribouXY, type="n", xlab="", ylab="", axes=FALSE)
mgMap(caribouXY, caribouAnalysis$memgene[, 1], add.plot=TRUE,
legend=TRUE)
box()

# Adj R2
caribouAnalysis$RsqAdj

## Find the proportional variation explained by each MEMGENE variable
caribouMEMGENEProp <- caribouAnalysis$sdev/sum(caribouAnalysis$sdev)
## Neatly print proportions for the first three MEMGENE variables
format(signif(caribouMEMGENEProp, 3)[1:3], scientific=FALSE)
```

# Milkweed data
## Import files
### Genetic data
```{r}
# vcf
vcf <- read.vcfR(
  here::here("./summary_stats/output_populations_no-IBD/mmaf0.05_R0.75/populations.snps.vcf"),
  verbose = FALSE)

## Convert vcf to genind object
my_genind <- vcfR2genind(vcf)

# tidy up colnames (remove periods)
inputdata1 <- my_genind$tab
colnames(inputdata1) <- gsub("\\.", "_", colnames(inputdata1))

## Produce a proportion of shared alleles genetic distance matrix using the convenience wrapper function provided with the package
gen_data <- codomToPropShared(as.data.frame(inputdata1), missingData = NA)
```

### Coordinates
```{r}
# lat/longs of sampling sites
urb <- read.csv(here::here("./clean_data/urb_metrics.csv"))

# sampling site IDs and genetic sample names
coords <- read.csv(here("./genomic_resources/pop_map3.csv")) %>%

# if population doesn't start with "MWI", it should start with "MW"
  # rename col 1
  dplyr::rename(pop_id = 2) %>%
  dplyr::mutate(pop_id = as.factor(pop_id)) %>%

# take entries with numeric populations and add "MW" as prefix
  add_MW_IDs() %>%

# left vs. full join because some populations in urb_clean weren't genotyped due to lack of material
  left_join(., urb, by = "patch_id") %>%
  dplyr::mutate(patch_id = as.factor(patch_id),
                pop_id = as.factor(pop_id)) %>%
  dplyr::select(c(1,2,6,7)) %>%
  dplyr::rename("y" = 3,
                "x" = 4) %>%
  dplyr::arrange(sample) %>%
  dplyr::select("x", "y")
```

## Run analysis
```{r}
## Run the MEMGENE analysis
if (!exists("milkweed_analysis"))
milkweed_analysis <- mgQuick(gen_data, coords)
```

## Visualize results
### Basic plots
```{r}
# with legend
png(#here::here("./Figures_Tables/memgene/memgene_legend.png"),
    paste0("./Figures_Tables/memgene/memgene_legend_", today(), ".png"),
    width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords, type="n", xlab="", ylab="", axes=FALSE)
#base_map
mgMap(coords,
      milkweed_analysis$memgene[, 1], 
      add.plot=TRUE,
      legend=TRUE)
box()

# Close device
dev.off()


# without legend
png(#here::here("./Figures_Tables/memgene/memgene_no_legend.png"),
        paste0("./Figures_Tables/memgene/memgene_nolegend_", today(), ".png"),
        width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords,
      milkweed_analysis$memgene[, 1], 
      add.plot=TRUE,
      legend=FALSE)
box()

# Close device
dev.off()
```

### Plots with terrain basemap
```{r}
# with legend

mem1 <- milkweed_analysis$memgene[, 1] %>%
  as.data.frame() %>%
  dplyr::rename("var" = 1) %>%
  cbind(coords) %>%
  dplyr::mutate("negative" = ifelse(var < 0, "negative", "positive")) %>%
  dplyr::mutate("abs_val" = abs(var))

# black/white discrete
basemap_terrain <- ggmap(get_map(location = c(-80.5, 43.2, -78.5, 44.25), 
              zoom = 11,
                          maptype = 'terrain',
                          source = 'stamen'),
        darken = c(0.3, "white"))

png(paste0("./Figures_Tables/memgene/memgene_terrain_bw_", today(), ".png"),
    width = 15,
    height = 15,
    units = "cm",
    res = 1000)

basemap_terrain +
  geom_point(data = mem1,
             mapping = aes(x = x,
                           y = y, 
                           fill = negative,
                       #    size = abs(var)),
                           size = abs_val),
             shape = 21) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = "MEMgene\nvariable 1",
       size = "MEMgene\nvariable 1") +
  scale_fill_manual(values = c("#333333", "white")) +
  scale_size_continuous(limits = c(0, 0.15))

# Close device
dev.off()




# black/white gradient, instead of discrete
png(paste0("./Figures_Tables/memgene/memgene_terrain_greyscale_", today(), ".png"),
    width = 15,
    height = 15,
    units = "cm",
    res = 1000)

basemap_terrain +
  geom_point(data = mem1,
             mapping = aes(x = x,
                           y = y,
                           fill = var,
                           size = abs_val),
             shape = 21) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = "MEMgene\nvariable 1",
       size = "MEMgene\nvariable 1") +
  scale_fill_gradient(high = "white", low = "black",
                      limits = c(-0.15, 0.1),
                      breaks = c(-0.15, 0, 0.1)) +
  scale_size_continuous(limits = c(0, 0.14)) +
  guides(size = FALSE)

# Close device
dev.off()



# colors: plasma, unidirectional size scale
png(paste0("./Figures_Tables/memgene/memgene_terrain_color_", today(), ".png"),
    width = 15,
    height = 10,
    units = "cm",
    res = 1000)

basemap_terrain +
  geom_point(data = mem1,
             mapping = aes(x = x,
                           y = y,
                           fill = var),
             shape = 21,
             size = 3) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = "MEMGENE\nvariable 1") +
  scale_fill_viridis(option = "plasma")

# Close device
dev.off()

```

### map ALL combined significant memgene variables
#### Without rescaling eigenvalues from 0-1
```{r}
library(sf)
library(tmap)

total_var <- apply(milkweed_analysis$memSelected, 1, sum) %>%
  as.data.frame() %>%
  dplyr::rename("var" = 1)

# sum all SGS explained by all significant memgene variables
# Combine 'var' from 'total_var' with 'coords'
spatial_data <- data.frame(var = total_var$var, coords)

# Convert the combined data frame to an 'sf' object
sf_object <- sf::st_as_sf(spatial_data, coords = c("y", "x"))

base::plot(sf_object, type = "p", cex = 2) 

basemap_terrain +
  geom_point(data = spatial_data,
             mapping = aes(x = x,
                           y = y, 
                           fill = var,
                           size = var),
             shape = 21,
             color = "black",
             ) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = "Variation") +
  scale_fill_viridis(option="plasma") + 
  guides(size = FALSE) 

ggsave(last_plot(),
       file =  paste0("./Figures_Tables/memgene/memgene_allMEMs_", today(), ".png"),
       width = 19,
       height = 19,
       units = "cm")


# colors: black/white, big to little (0) to big circles
# Assuming you have the required libraries loaded:
# library(ggplot2)

basemap_terrain +
  geom_point(data = spatial_data,
             mapping = aes(x = x,
                           y = y, 
                           fill = ifelse(var < 0, "negative", "positive"),
                           size = abs(var)),  # Use abs(var) to get the absolute value for size
             shape = 21#,
            # color = "black"
            ) +
  labs(x = 'Longitude', y = 'Latitude',
       size = "Variation",
       fill = "Variation") +
  scale_fill_manual(values = c("negative" = "darkgrey",
                               "positive" = "white")) +
  scale_size_continuous(
    range = c(1.5, 12 * max(abs(spatial_data$var))), 
    limits = c(0, max(abs(spatial_data$var)))) 



ggsave(last_plot(),
       file =  paste0("./Figures_Tables/memgene/memgene_allMEMs_BW_", today(), ".png"),
       width = 19,
       height = 19,
       units = "cm")
```


## Stats
### Generate
```{r}
# Adj R2
print(paste0("Adjusted R2: ", round(milkweed_analysis$RsqAdj, 3)))

## Find the proportional variation explained by each MEMGENE variable
prop_var_expl <- milkweed_analysis$sdev/sum(milkweed_analysis$sdev)

## Neatly print proportions for the first three MEMGENE variables
print(paste0("Proportion of variation explained by first 3 MEMgene variables: ", toString(format(signif(prop_var_expl, 3)[1:3], scientific=FALSE))))
```

### Export
```{r}
# Create a new Word document
doc <- read_docx() %>%

# Add content to the document
  body_add_par(
    print(
      paste0(
        "Adjusted R2: ", 
        round(milkweed_analysis$RsqAdj, 3)
        )
      )
    ) %>%
  
  body_add_par("") %>%
  
  body_add_par("Proportion of variation explained by ")

for (i in 1:3) {
  doc <- doc %>%
    body_add_par(
      paste0(
        "MEMGENE variable ",
        i,
        " (Moran's eigenvector ",
        c(milkweed_analysis[["whichSelectedPos"]][1],
          milkweed_analysis[["whichSelectedPos"]][2],
          milkweed_analysis[["whichSelectedPos"]][3])[i],
        "): ",
        format(signif(prop_var_expl[i], 3), scientific = FALSE)
        )
    )
  }

# Save the Word document
print(doc,
      target = paste0("./Figures_Tables/memgene/memgene_stats_", today(), ".docx"))

```

