# Set up notebook
## Load libraries & functions
```{r warning=F}
source(here::here("libraries.R"))
source(here::here("functions.R"))
```

# Milkweed data
## Import files
### Genetic data
#### Function
```{r}
process_vcf_data <- function(vcf_filepath) {
  
  # Read vcf file
  vcf <- read.vcfR(vcf_filepath, verbose = FALSE)
  
  # Convert vcf to genind object
  my_genind <- vcfR2genind(vcf)
  
  # Tidy up colnames (remove periods)
  inputdata1 <- my_genind$tab
  colnames(inputdata1) <- gsub("\\.", "_", colnames(inputdata1))
  
  ## Produce a proportion of shared alleles genetic distance matrix using the convenience wrapper function provided with the package
  gen_data <- codomToPropShared(as.data.frame(inputdata1), missingData = NA)
  
  return(gen_data)
}
```

#### All individuals
```{r}
gen_data <- process_vcf_data(here::here("./summary_stats/output_populations_no-IBD/mmaf0.05_R0.75/populations.snps.vcf"))

```

#### JUST URBAN individuals
```{r}
gen_data_urban <- process_vcf_data(
  here::here("./filtered_vcfs/urban_pops_dist_mmaf0.05_variant_sites.recode.vcf")
  )
```

#### 1 individual/population
```{r}
gen_data_subsamp <- process_vcf_data(
  here::here("./filtered_vcfs/1indiv_per_mmaf0.05_variant_sites.recode.vcf")
  )
```

### Coordinates
#### Function
```{r}
process_location_data <- function(urb_filepath, pop_map_filepath) {
  
  # lat/longs of sampling sites
  urb <- read.csv(urb_filepath)
  
  # sampling site IDs and genetic sample names
  coords <- read.csv(pop_map_filepath) %>%
    
    # if population doesn't start with "MWI", it should start with "MW"
  # rename col 1
    dplyr::rename(pop_id = 2) %>%
    dplyr::mutate(pop_id = as.factor(pop_id)) %>%
    
    # take entries with numeric populations and add "MW" as prefix
    add_MW_IDs() %>%

    # left vs. full join because some populations in urb_clean weren't genotyped due to lack of material
    left_join(., urb, by = "patch_id") %>%
    dplyr::mutate(patch_id = as.factor(patch_id),
                  pop_id = as.factor(pop_id)) %>%
    dplyr::select(c(1,2,6,7)) %>%
    dplyr::rename("y" = 3,
                  "x" = 4) %>%
    dplyr::arrange(sample) %>%
    dplyr::select("x", "y")
  
  return(coords)
  }
```

#### All individuals
```{r}
coords <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./genomic_resources/pop_map3.csv"))

```

#### All URBAN individuals
```{r}
coords_urban <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_map3_U_dist.csv"))

# I converted the pop_map3_U_dist.txt into csv, then removed the "MW" prefixes so that the function would work
```

#### 1 individual/population
```{r}
coords_subsamp <- process_location_data(
  here::here("./clean_data/urb_metrics.csv"),
  here::here("./clean_data/pop_map3_1indiv_per_pop.csv"))

```

## Run analysis
### All individuals
```{r}
## Run the MEMGENE analysis
set.seed(1)
if (!exists("milkweed_analysis"))
milkweed_analysis <- mgQuick(gen_data, coords)
```

### Just URBAN individuals
```{r}
set.seed(1)
if (!exists("milkweed_analysis_urban"))
milkweed_analysis_urban <- mgQuick(gen_data_urban, coords_urban)

# No significant MEMs detected
```

### 1 individual per population
```{r}
set.seed(1)
if (!exists("milkweed_analysis_subsamp"))
milkweed_analysis_subsamp <- mgQuick(gen_data_subsamp, coords_subsamp)
# No significant MEMs detected


# I checked multiple ways that this null result is not from a mis-ordering of samples (that the genetic and geographic data align perfectly). One example:
# 
# 
#  my_genind <- vcfR2genind(read.vcfR(  here::here("./filtered_vcfs/1indiv_per_mmaf0.05_variant_sites.recode.vcf"), verbose = FALSE))
# 
# my_genind@tab %>% rownames
#   [1] "1.2_1A10_sorted" "1.2_1A11_sorted" "1.2_1A1_sorted" 
#   [4] "1.2_1A2_sorted"  "1.2_1A3_sorted"  "1.2_1A4_sorted" 
#   [7] "1.2_1A5_sorted"  "1.2_1A7_sorted"  "1.2_1A8_sorted" 
#  [10] "1.2_1A9_sorted"  "1.2_1B10_sorted" "1.2_1B1_sorted" 
#  [13] "1.2_1B2_sorted"  "1.2_1B4_sorted"  "1.2_1B5_sorted" 
#  [16] "1.2_1B6_sorted"  "1.2_1B7_sorted"  "1.2_1B8_sorted" 
#  [19] "1.2_1B9_sorted"  "1.2_1C10_sorted" "1.2_1C2_sorted" 
#  [22] "1.2_1C3_sorted"  "1.2_1C5_sorted"  "1.2_1C6_sorted" 
#  [25] "1.2_1D1_sorted"  "1.2_1D3_sorted"  "1.2_1D4_sorted" 
#  [28] "1.2_1D5_sorted"  "1.2_1D6_sorted"  "1.2_1E11_sorted"
#  [31] "1.2_1E8_sorted"  "1.2_1F1_sorted"  "1.2_1F5_sorted" 
#  [34] "1.2_1F7_sorted"  "1.2_1F8_sorted"  "1.2_1G2_sorted" 
#  [37] "1.2_1G6_sorted"  "1.2_1H10_sorted" "1.2_1H1_sorted" 
#  [40] "1.2_1H2_sorted"  "1.2_1H4_sorted"  "1.2_1H8_sorted" 
#  [43] "1.2_1H9_sorted"  "1.2_2A2_sorted"  "1.2_2A3_sorted" 
#  [46] "1.2_2A5_sorted"  "1.2_2A8_sorted"  "1.2_2B7_sorted" 
#  [49] "1.2_2C8_sorted"  "3A_A11_sorted"   "3A_A12_sorted"  
#  [52] "3A_A4_sorted"    "3A_A5_sorted"    "3A_A6_sorted"   
#  [55] "3A_A7_sorted"    "3A_A8_sorted"    "3A_A9_sorted"   
#  [58] "3A_B10_sorted"   "3A_B11_sorted"   "3A_B12_sorted"  
#  [61] "3A_B1_sorted"    "3A_B4_sorted"    "3A_B5_sorted"   
#  [64] "3A_B6_sorted"    "3A_B7_sorted"    "3A_B8_sorted"   
#  [67] "3A_B9_sorted"    "3A_C10_sorted"   "3A_C11_sorted"  
#  [70] "3A_C12_sorted"   "3A_C4_sorted"    "3A_C5_sorted"   
#  [73] "3A_C6_sorted"    "3A_C7_sorted"    "3A_C8_sorted"   
#  [76] "3A_C9_sorted"    "3A_D10_sorted"   "3A_D11_sorted"  
#  [79] "3A_D12_sorted"   "3A_D4_sorted"    "3A_D5_sorted"   
#  [82] "3A_D6_sorted"    "3A_D7_sorted"    "3A_D8_sorted"   
#  [85] "3A_D9_sorted"    "3A_E10_sorted"   "3A_E11_sorted"  
#  [88] "3A_E12_sorted"   "3A_E3_sorted"    "3A_E4_sorted"   
#  [91] "3A_E5_sorted"    "3A_E6_sorted"    "3A_E8_sorted"   
#  [94] "3A_E9_sorted"    "3A_F10_sorted"   "3A_F11_sorted"  
#  [97] "3A_F12_sorted"   "3A_F3_sorted"    "3A_F4_sorted"   
# [100] "3A_F5_sorted"    "3A_F6_sorted"    "3A_F7_sorted"   
# [103] "3A_F8_sorted"    "3A_F9_sorted"    "3A_G10_sorted"  
# [106] "3A_G11_sorted"   "3A_G12_sorted"   "3A_G3_sorted"   
# [109] "3A_G4_sorted"    "3A_G5_sorted"    "3A_G6_sorted"   
# [112] "3A_G7_sorted"    "3A_G8_sorted"    "3A_G9_sorted"   
# [115] "3A_H10_sorted"   "3A_H11_sorted"   "3A_H3_sorted"   
# [118] "3A_H5_sorted"    "3A_H6_sorted"    "3A_H7_sorted"   
# [121] "3A_H8_sorted"    "3A_H9_sorted"
```

## Visualize results
### Basic plots
#### Function
```{r}
basic_plot_mems <- function(group, # e.g., all_indiv
                            coords_df, 
                            milkweed_analysis_list
                            ){
  
  # MEM1-----
  
  # with legend
png(#here::here("./Figures_Tables/memgene/memgene_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_legend_var1_",
           today(),
           ".png"),
    width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
#base_map
mgMap(coords_df,
      milkweed_analysis_list$memgene[, 1], 
      add.plot=TRUE,
      legend=TRUE)
box()

# Close device
dev.off()


# without legend
png(#here::here("./Figures_Tables/memgene/memgene_no_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_nolegend_var1_",
           today(),
               ".png"),
        width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords_df,
      milkweed_analysis_list$memgene[, 1], 
      add.plot=TRUE,
      legend=FALSE)
box()

# Close device
dev.off()

  # MEM2-----
  
  # with legend
png(#here::here("./Figures_Tables/memgene/memgene_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_legend_var2_",
           today(),
           ".png"),
    width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
#base_map
mgMap(coords_df,
      milkweed_analysis_list$memgene[, 2], 
      add.plot=TRUE,
      legend=TRUE)
box()

# Close device
dev.off()


# without legend
png(#here::here("./Figures_Tables/memgene/memgene_no_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_nolegend_var2_",
           today(),
               ".png"),
        width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords_df,
      milkweed_analysis_list$memgene[, 2], 
      add.plot=TRUE,
      legend=FALSE)
box()

# Close device
dev.off()
  # MEM3-----
  
  # with legend
png(#here::here("./Figures_Tables/memgene/memgene_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_legend_var3_",
           today(),
           ".png"),
    width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
#base_map
mgMap(coords_df,
      milkweed_analysis_list$memgene[, 3], 
      add.plot=TRUE,
      legend=TRUE)
box()

# Close device
dev.off()


# without legend
png(#here::here("./Figures_Tables/memgene/memgene_no_legend.png"),
    paste0("./Figures_Tables/memgene/",
          group,
          "/memgene_nolegend_var3_",
           today(),
               ".png"),
        width = 20,
    height = 20,
    units = "cm",
    res = 1000)

plot(coords_df, type="n", xlab="", ylab="", axes=FALSE)
mgMap(coords_df,
      milkweed_analysis_list$memgene[, 3], 
      add.plot=TRUE,
      legend=FALSE)
box()

# Close device
dev.off()
}
```

#### All individuals
```{r}
# Moran's Eigenvectors 12,20,5
basic_plot_mems("all_indiv",
                coords,
                milkweed_analysis)
```

#### 1 individual/population
```{r}
# FILL IN CORRECTLY
# Moran's Eigenvectors 12,20,5
basic_plot_mems("subsamp",
                coords_subsamp,
                milkweed_analysis_subsamp)
```

### Plots with terrain basemap
#### Code for each var
```{r}
basemap_terrain <- ggmap(get_map(location = c(-80.5, 43.2, -78.5, 44.25), 
              zoom = 11,
                          maptype = 'terrain',
                          source = 'stamen'),
        darken = c(0.3, "white"))


## Find the proportional variation explained by each MEMGENE variable
prop_var_expl <- milkweed_analysis$sdev/sum(milkweed_analysis$sdev)

prop_var_expl_urban <- milkweed_analysis_urban$sdev/sum(milkweed_analysis_urban$sdev)

prop_var_expl_subsamp <- milkweed_analysis_subsamp$sdev/sum(milkweed_analysis_subsamp$sdev)
```

#### Functions
```{r}
# create mem1, mem2, and mem3 objects-----
create_mem <- function(milkweed_analysis_list,
                       coords_df,
                       mem_number #e.g., 1, 2, or 3
                       ){
  
  mem <- milkweed_analysis_list$memgene[, mem_number] %>%
  as.data.frame() %>%
  dplyr::rename("var" = 1) %>%
  cbind(coords_df) %>%
  dplyr::mutate("negative" = ifelse(var < 0, "negative", "positive")) %>%
  dplyr::mutate("abs_val" = abs(var))
  
  return(mem)

}


# create terrain plots-----
terrain_plot_mems <- function(group, # e.g., all_indiv
                              coords_df, 
                              milkweed_analysis_list,
                              mem1_df,
                              mem2_df,
                              mem3_df
){

  for (i in 1:3) {
    
    mem_df <- get(paste0("mem", i, "_df"))  # Access mem1_df, mem2_df, mem3_df dynamically
    
    
    # with legend
    # black/white discrete-----
    basemap_terrain +
      geom_point(data = mem_df,
                 mapping = aes(x = x,
                               y = y, 
                               fill = negative,
                               #    size = abs(var)),
                               size = abs_val),
                 shape = 21) +
      labs(x = 'Longitude', y = 'Latitude',
           fill = paste0("MEMgene\nvariable ", i),
           size = paste0("MEMgene\nvariable ", i)) +
      scale_fill_manual(values = c("#333333", "white")) +
      scale_size_continuous(limits = c(0, 0.15))
    
    ggsave(filename =
             paste0("./Figures_Tables/memgene/",
                    group,
                    "/memgene_terrain_bw_var",
               i,
               "_",
               today(),
               ".png"),
           plot = last_plot(),
           width = 15,
        height = 10,
        units = "cm")

    
    # black/white gradient, instead of discrete-----
    basemap_terrain +
      geom_point(data = mem_df,
                 mapping = aes(x = x,
                               y = y,
                               fill = var,
                               size = abs_val),
                 shape = 21) +
      labs(x = 'Longitude', y = 'Latitude',
           fill = paste0("MEMgene\nvariable ", i),
           size = paste0("MEMgene\nvariable ", i)) +
      scale_fill_gradient(high = "white", low = "black",
                          limits = c(-0.15, 0.1),
                          breaks = c(-0.15, 0, 0.1)) +
      scale_size_continuous(limits = c(0, 0.14)) +
      guides(size = FALSE)

      ggsave(filename =
               paste0("./Figures_Tables/memgene/",
                    group,
               "/memgene_terrain_greyscale_var",
               i,
               "_",
               today(),
               ".png"),
           plot = last_plot(),
           width = 15,
        height = 10,
        units = "cm")

    # colors: plasma, unidirectional size scale-----
    basemap_terrain +
      geom_point(data = mem_df,
                 mapping = aes(x = x,
                               y = y,
                               fill = var),
                 shape = 21,
                 size = 3) +
      labs(x = 'Longitude', y = 'Latitude',
           fill = paste0("MEMgene\nvariable ", i)) +  scale_fill_viridis(option = "plasma")

    ggsave(filename =
             paste0("./Figures_Tables/memgene/",
                    group,
                    "/memgene_terrain_color_var",
               i,
               "_",
               group,
               today(),
               ".png"),
           plot = last_plot(),
           width = 15,
        height = 10,
        units = "cm")

  }
}
```

#### All individuals
```{r}
# create MEM dfs
mem1 <- create_mem(milkweed_analysis, coords, 1)
mem2 <- create_mem(milkweed_analysis, coords, 2)
mem3 <- create_mem(milkweed_analysis, coords, 3)

# create maps
terrain_plot_mems("all_indiv",
                            coords, 
                            milkweed_analysis,
                            mem1,
                            mem2,
                            mem3)
```

#### 1 individual/population
```{r}
# create MEM dfs
mem1_subsamp <- create_mem(milkweed_analysis_subsamp,
                          coords_subsamp, 1)
mem2_subsamp <- create_mem(milkweed_analysis_subsamp,
                          coords_subsamp, 2)
mem3_subsamp <- create_mem(milkweed_analysis_subsamp,
                          coords_subsamp, 3)

# create maps
terrain_plot_mems("subsamp",
                            coords_subsamp, 
                            milkweed_analysis_subsamp,
                            mem1_subsamp,
                            mem2_subsamp,
                            mem3_subsamp)
```


### map ALL combined significant memgene variables
#### Function
```{r}
sig_combined_mem_plot <- function(group,
                                  coords_df,
                                  milkweed_analysis_list
                                  ){
  
  total_var <- apply(milkweed_analysis_list$memSelected, 1, sum) %>%
  as.data.frame() %>%
  dplyr::rename("var" = 1)

# sum all SGS explained by all significant memgene variables
# Combine 'var' from 'total_var' with 'coords'
spatial_data <- data.frame(var = total_var$var,
                           coords_df)

# map with colors-----
basemap_terrain +
  geom_point(data = spatial_data,
             mapping = aes(x = x,
                           y = y, 
                           fill = var,
                           size = var),
             shape = 21,
             color = "black",
             ) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = "Variation") +
  scale_fill_viridis(option="plasma") + 
  guides(size = FALSE) 

ggsave(last_plot(),
       file =  paste0("./Figures_Tables/memgene/",
                      group,
                      "/memgene_allMEMs_", today(), ".png"),
       width = 19,
       height = 19,
       units = "cm")

# colors: black/white, big to little (0) to big circles-----

basemap_terrain +
  geom_point(data = spatial_data,
             mapping = aes(x = x,
                           y = y, 
                           fill = ifelse(var < 0, "negative", "positive"),
                           size = abs(var)),
             shape = 21#,
            # color = "black"
            ) +
  labs(x = 'Longitude', y = 'Latitude',
       size = "Variation",
       fill = "Variation") +
  scale_fill_manual(values = c("negative" = "darkgrey",
                               "positive" = "white")) +
  scale_size_continuous(
    range = c(1.5, 12 * max(abs(spatial_data$var))), 
    limits = c(0, max(abs(spatial_data$var)))) 

ggsave(last_plot(),
       file =  paste0("./Figures_Tables/memgene/",
                      group,
                      "/memgene_allMEMs_BW_", today(), ".png"),
       width = 19,
       height = 19,
       units = "cm")

}
```

#### All individuals
```{r}
sig_combined_mem_plot("all_indiv",
                            coords, 
                            milkweed_analysis)
```

#### 1 individual/population
```{r}
sig_combined_mem_plot("subsamp",
                            coords_subsamp, 
                            milkweed_analysis_subsamp)
```


### Rescaling, then combine MEMs from 0-255
#### Rescale
```{r}
mems <- list(mem1, mem2, mem3,
             mem1_urban, mem2_urban, mem3_urban,
             mem1_subsamp, mem2_subsamp, mem3_subsamp
             )

# rescale mems from 0-255
rescale_var_to_255 <- function(df) {
  df %>%
    mutate(mem_255 = scales::rescale(var, to = c(0, 255)))
}

# map the function to each element of the list
mems <- purrr::map(mems, rescale_var_to_255)

names(mems) <- list("mem1", "mem2", "mem3",
             "mem1_urban", "mem2_urban", "mem3_urban",
             "mem1_subsamp", "mem2_subsamp", "mem3_subsamp")

list2env(mems, globalenv())
```

#### Create individual maps for top 3 MEMs
##### Function
```{r}
rescale_mems <- function(group,
                         milkweed_analysis_list,
                         mem1_df,
                         mem2_df,
                         mem3_df){
  
  # plot with colors: plasma, unidirectional size scale
  
      # MEM1-----
    basemap_terrain +
    geom_point(data = mem1_df,
             mapping = aes(x = x,
                           y = y,
                           fill = mem_255),
             shape = 21,
             size = 3) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = paste0("MEMGENE\nvariable 1"),
       caption = paste0("Moran's Eigenvector: ",
         milkweed_analysis_list$whichSelectedPos[1])) +
  scale_fill_gradient(low = "white",
                      high = "red") 
    

    ggsave(filename = paste0("./Figures_Tables/memgene/",
                             group,
"/memgene_terrain_color_255_var1_",
           today(),
           ".png"),
           plot = last_plot(),
    width = 15,
    height = 10,
    units = "cm")
    
      # MEM2-----
    basemap_terrain +
    geom_point(data = mem2_df,
             mapping = aes(x = x,
                           y = y,
                           fill = mem_255),
             shape = 21,
             size = 3) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = paste0("MEMGENE\nvariable 2"),
       caption = paste0("Moran's Eigenvector: ",
         milkweed_analysis_list$whichSelectedPos[2])) +
  scale_fill_gradient(low = "white",
                      high = "green") 
    

    ggsave(filename = paste0("./Figures_Tables/memgene/",
                             group,
"/memgene_terrain_color_255_var2_",
           today(),
           ".png"),
           plot = last_plot(),
    width = 15,
    height = 10,
    units = "cm")
    
      # MEM3-----
    basemap_terrain +
    geom_point(data = mem3_df,
             mapping = aes(x = x,
                           y = y,
                           fill = mem_255),
             shape = 21,
             size = 3) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = paste0("MEMGENE\nvariable 3"),
       caption = paste0("Moran's Eigenvector: ",
         milkweed_analysis_list$whichSelectedPos[3])) +
  scale_fill_gradient(low = "white",
                      high = "blue") 
    

    ggsave(filename = paste0("./Figures_Tables/memgene/",
                             group,
"/memgene_terrain_color_255_var3_",
           today(),
           ".png"),
           plot = last_plot(),
    width = 15,
    height = 10,
    units = "cm")

}
```

##### All individuals
```{r}
rescale_mems("all_indiv",
             milkweed_analysis,
             mem1,
             mem2,
             mem3)
```

##### 1 individual/population
```{r}
rescale_mems("subsamp",
             milkweed_analysis_subsamp,
             mem1_subsamp,
             mem2_subsamp,
             mem3_subsamp)

```

#### Combine top 3 MEMs into one figure
##### Function
```{r}

```

##### All individuals
```{r}
make_color_df <- function(df, color){
  df %>%
  dplyr::select(x,y,mem_255) %>%
  dplyr::rename_with(~color, mem_255) %>%
  dplyr::mutate(ID = row_number())
}

red_data <- make_color_df(mem1, "red")
green_data <- make_color_df(mem2, "green")
blue_data <- make_color_df(mem3, "blue")

color_data <- inner_join(red_data,
                        green_data,
                        by = c("x","y","ID")) %>%
  inner_join(.,
             blue_data,
             by = c("x","y","ID")) %>%
  dplyr::select(ID, x, y, red, green, blue)

color_data$color <- rgb(color_data$red,
                        color_data$green, 
                        color_data$blue,
                        maxColorValue = 255)



png(paste0("./Figures_Tables/memgene/memgene_terrain_color_255_",
                      "vars1_to_3",
           today(),
           ".png"),
    width = 15,
    height = 10,
    units = "cm",
    res = 1000)

basemap_terrain +
  geom_point(data = color_data,
             mapping = aes(x = x,
                           y = y,
                           fill = color),
             shape = 21,
             size = 3) +
  labs(x = 'Longitude', y = 'Latitude',
       fill = "MEMGENE\nvariables 1-3",
       caption = paste0("Moran's Eigenvectors: ",
         milkweed_analysis$whichSelectedPos[1],
         ", ",
         milkweed_analysis$whichSelectedPos[2],
         ", ",
         milkweed_analysis$whichSelectedPos[3]
         ))  +
    scale_fill_identity()

# Close device
dev.off()


# good visualization of color space in 3D: https://www.whydomath.org/node/wavlets/images/rgbcube.jpg
```


## Stats
### Generate
```{r}
# Adj R2
print(paste0("Adjusted R2: ", round(milkweed_analysis$RsqAdj, 3)))

## Find the proportional variation explained by each MEMGENE variable
prop_var_expl # computed above for figures

## Neatly print proportions for the first three MEMGENE variables
print(paste0("Proportion of variation explained by first 3 MEMgene variables: ", toString(format(signif(prop_var_expl, 3)[1:3], scientific=FALSE))))
```

### Export stats
```{r}
# Extracting relevant data from the milkweed_analysis object
whichSelectedPos <- milkweed_analysis$whichSelectedPos
memgene_values <- milkweed_analysis$mem$valuesMEM
sdev_values <- milkweed_analysis$sdev

# Calculating R-squared values per MEMGENE variable
r_squared_list <- milkweed_analysis$sdev/sum(milkweed_analysis$sdev)

r2_df <- as.data.frame(r_squared_list) %>%
  rownames_to_column("MEMGENE_var") %>%
  slice(1:length(whichSelectedPos)) %>%
  dplyr::rename(R_squared = r_squared_list)


# Creating the dataframe
mg_table <- data.frame(
  MEMGENE_variable = 1:length(whichSelectedPos),
  Moran_Eigenvector = whichSelectedPos,
  R_squared = r2_df$R_squared) %>%
  dplyr::mutate(R_squared = round(R_squared, 3)) %>%
  dplyr::mutate(R_squared = ifelse(
    R_squared < 0.001, 
    "<0.001",
    R_squared)) %>%
    rownames_to_column(var = "Row") %>%
  dplyr::select(-Row) %>%
  dplyr::rename("MEMGENE Variable" = 1,
                "Moran's Eigenvector" = 2,
                "R2" = 3)

mg_table %>%
  flextable() %>%
  flextable::align(., align = "center", part = "all") %>%
  flextable::compose(i = 1,
                     j = 3,
                     part = "header",
                     value = as_paragraph("R", as_sup("2"))) %>%
  flextable::autofit() %>%
  flextable::save_as_docx(.,
                          path = paste0("./Figures_Tables/memgene/memgene_stats_",
                                        today(), 
                                        ".docx"))

```
