
# remove mentions of obs/exp hom/het from this Rmd

# will use narrow genetic dataset
# -R = 1
# -min_maf = 0.05





---
title: Comparing continuous/categorical urbanization data AND all/variant only genetic data
output: 
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 3
    number_sections: true
    theme: lumen
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
```

- looks like there are differences btwn the ch2 and ch3 plants re: pop sizes, which is associated with many effects. What if I removed the ch2 plants? what would I see then?

-Compare data from all_data csv w/ variants csv

# Set up notebook
## Load libraries
```{r warning=F}
source(here::here("libraries.R"))
```

## Import data
### Urbanization data
```{r}
urb <- read.csv(here::here("./clean_data/urb_metrics.csv")) %>%
  dplyr::mutate(site_id = as.factor(site_id),
                patch_id = as.factor(patch_id),
                urb_rur = as.factor(urb_rur),
                quadrant = as.factor(quadrant),
                river_valley = as.factor(river_valley)
                )
```

### Genetic data
#### Functions that read populations.sumstats_summary.tsv files, extracts data re: positions, then puts it into a dataframe
```{r}

## Extract variant positions
extract_variant_position_summary <- function(file) {
  
    text <- readLines(file)
    begin_line <- grep("# Variant positions", text)
    end_line <- grep("# All positions", text)
    data_text <- text[(begin_line + 1):(end_line - 1)]
    data_df <- read.table(text = data_text, header = FALSE, sep = "\t")
    header <- text[begin_line + 1]
    colnames(data_df) <- unlist(strsplit(header, "\t"))

  return(data_df)
}


# Extract ALL positions 
extract_all_position_summary <- function(file) {
  
    text <- readLines(file)
    begin_line <- grep("# All positions", text)
    end_line <- length(text)
    data_text <- text[(begin_line + 1):(end_line - 1)]
    data_df <- read.table(text = data_text, header = FALSE, sep = "\t")
    header <- text[begin_line + 1]
    colnames(data_df) <- unlist(strsplit(header, "\t"))

  return(data_df)
}

```

#### Read in genetic data
```{r}

input_file <- here::here("./summary_stats/mmaf0.05_R1_populations_output/populations.sumstats_summary.tsv")


div <- extract_variant_position_summary(input_file) %>%
# if population doesn't start with "MWI", it should start with "MW"
  # clean names
  janitor::clean_names() %>%
  # rename col 1
  dplyr::rename(pop_id = 1) %>%
  dplyr::mutate(pop_id = as.factor(pop_id))


div_all <- extract_all_position_summary(input_file) %>%
# if population doesn't start with "MWI", it should start with "MW"
  # clean names
  janitor::clean_names() %>%
  # rename col 1
  dplyr::rename(pop_id = 1) %>%
  dplyr::mutate(pop_id = as.factor(pop_id))

```

#### Add "MW" to numeric populations
```{r}
# take entries with numeric populations and add "MW" as prefix
add_MW_IDs <- function(div_df) {
  # Filter out populations that start with "MWI" or "UTSC"
  MW_pops <- div_df %>%
    filter(!str_detect(pop_id, "MWI")) %>%
    filter(!str_detect(pop_id, "UTSC"))

  # Get populations that start with "MWI" or "UTSC"
  MWI_UTSC_pops <- anti_join(div_df, MW_pops)

  # Split up populations MW001->MW009 vs MW010->MW079
  MW_pops_singledigits <- MW_pops[nchar(as.character(MW_pops$pop_id)) == 1, ]
  MW_pops_doubledigits <- MW_pops[nchar(as.character(MW_pops$pop_id)) != 1, ]

  # Add "MW00" to single-digit populations
  MW_pops_singledigits %<>%
    dplyr::mutate(patch_id = paste0("MW00", pop_id))

  # Add "MW0" to double-digit populations
  MW_pops_doubledigits %<>%
    dplyr::mutate(patch_id = paste0("MW0", pop_id))

  # Bring all entries together again
  all_pops <- full_join(MW_pops_singledigits,
                         MW_pops_doubledigits) %>%
    dplyr::relocate(patch_id, .before = 1) %>%
    full_join(.,
              MWI_UTSC_pops) %>%
    # If patch_id column is NA (for MWI and UTSC values), replace it with value from pop_id
    dplyr::mutate(patch_id = coalesce(patch_id, pop_id))

  return(all_pops)
}

diversity_all_sites <- add_MW_IDs(div_all)
diversity_variant_sites <- add_MW_IDs(div)

```

## Join genetic data with urbanization data
```{r}
# left vs. full join because some populations in urb_clean weren't genotyped due to lack of material
diversity_all_sites %<>%
  left_join(., urb, by = "patch_id") %>%
  dplyr::mutate(patch_id = as.factor(patch_id),
                pop_id = as.factor(pop_id))

diversity_variant_sites %<>%
  left_join(., urb, by = "patch_id") %>%
  dplyr::mutate(patch_id = as.factor(patch_id),
                pop_id = as.factor(pop_id))


# make new columns: u_r_dist & u_r_usc (urban or rural)
# this column is different from urb_rur, which I made when scouting sites for the 100 populations. I don't have those types of designations for the sites I imported from my previous observational study (note the NAs). So I'll calculate two new columns that assign a population an "urban" or "rural" status based on its distance from city center (u_r_dist) or urbanization score (u_r_usc)
diversity_variant_sites %<>%
  dplyr::mutate(u_r_dist = case_when(
    City_dist <= 40 ~ "Urban",
    TRUE ~ "Rural")) %>%
  dplyr::mutate(u_r_usc = case_when(
    urb_score > 0 ~ "Urban",
    TRUE ~ "Rural"))


diversity_all_sites %<>%
  dplyr::mutate(u_r_dist = case_when(
    City_dist <= 40 ~ "Urban",
    TRUE ~ "Rural")) %>%
  dplyr::mutate(u_r_usc = case_when(
    urb_score > 0 ~ "Urban",
    TRUE ~ "Rural"))

```

## Make plotting theme
```{r}
plot_aesthetics <- c(geom_point(size = 1.5,
                                shape = 1),
                     
                     geom_smooth(method = lm,
                                 color = "black",
                                 size = 0.5,
                                 se = TRUE))

# for ggpredict objects
rep_geoms <- c(geom_smooth(aes(x = x,
                               y = predicted),
                           color = "#66a182",
                           size = 1,
                           se = F),
               geom_ribbon(aes(x = x,
                               ymin = conf.low,
                               ymax = conf.high),
                           fill = "#66a182",
                           alpha = 0.3))

```

# Does urbanization predict genetic diversity within pops?
## Explore data
### All sites
```{r}
str(diversity_all_sites)

head(diversity_all_sites)

diversity_all_sites %>%
  dplyr::select(1, 3, "pi", "num_indv", "City_dist", "urb_score") %>%
  view

# quick boxplots
boxplot(diversity_all_sites$private)
boxplot(diversity_all_sites$fis)
boxplot(diversity_all_sites$pi)

# quick regressions of genetic vars against distance from cc
qplot(data = diversity_all_sites,
      City_dist, private)

qplot(data = diversity_all_sites,
      City_dist, fis)

qplot(data = diversity_all_sites,
      City_dist, pi)

qplot(data = diversity_all_sites,
      City_dist, num_indv)

# regress all selected variables against one another
diversity_all_sites %>%
  dplyr::select(private, pi, num_indv, fis, urb_score, City_dist) %>%
  pairs()

# look at distributions of urb vs rural pops
qplot(data = diversity_all_sites,
      u_r_dist, pi)

qplot(data = diversity_all_sites,
      u_r_dist, private)

qplot(data = diversity_all_sites,
      u_r_usc, pi)

qplot(data = diversity_all_sites,
      u_r_usc, private)


# does urb correlated with other population descriptors?

## number of individuals/population
test1 <- glmmTMB(City_dist ~ num_indv,
                 diversity_all_sites)
performance::check_model(test1)
car::Anova(test1) # yes (marginally significant)
r.squaredGLMM(test1) # ~3% variance explained


## number of polymorphic sites/population
test1 <- glmmTMB(City_dist ~ polymorphic_sites,
                 diversity_all_sites)
performance::check_model(test1)
car::Anova(test1) # yes
r.squaredGLMM(test1) # ~4% variance explained



# num indivs vs fis
qplot(diversity_all_sites$num_indv,
      diversity_all_sites$fis)
```

### Variant sites
```{r}
str(diversity_variant_sites)

head(diversity_variant_sites)

diversity_variant_sites %>%
  dplyr::select(1, 3, 4, "pi", "City_dist", "urb_score") %>%
  view

# quick boxplots
boxplot(diversity_variant_sites$fis)
boxplot(diversity_variant_sites$pi)

# quick regressions of genetic vars against distance from cc
qplot(data = diversity_variant_sites,
      City_dist, private)

qplot(data = diversity_variant_sites,
      City_dist, private/num_indv)

qplot(data = diversity_variant_sites,
      City_dist, fis)

qplot(data = diversity_variant_sites,
      City_dist, pi)

qplot(data = diversity_variant_sites,
      City_dist, num_indv)

# regress all selected variables against one another
diversity_variant_sites %>%
  dplyr::select(private, pi, num_indv, fis, urb_score, City_dist) %>%
  pairs()

# look at distributions of urb vs rural pops
qplot(data = diversity_variant_sites,
      u_r_dist, pi)

qplot(data = diversity_variant_sites,
      u_r_dist, private)

qplot(data = diversity_variant_sites,
      u_r_usc, pi)

qplot(data = diversity_variant_sites,
      u_r_usc, private)


# does urb correlated with other population descriptors?

## number of individuals/population
test1 <- glmmTMB(City_dist ~ num_indv, diversity_variant_sites)
performance::check_model(test1)
car::Anova(test1) # yes
r.squaredGLMM(test1) # ~3% variance explained
```

## Urbanization is continuous / data = all sites
### Fit linear models/diagnostics
#### Private alleles
This is 0 for each individual.
#### Inbreeding coefficient (FIS)
This is extremely zero-inflated. I've tried fitting models with zero inflation parameters w/glmmTMB but they don't converge and/or diagnostics don't pass tests. I'll fit manual hurdle models instead.
##### City_dist
```{r}
diversity_all_sites %<>%
  mutate(fis_bin = case_when(
    fis == 0 ~ 0,
    fis > 0 ~ 1)
  )
  
# binary
fis_dist_bin <- glmmTMB(fis_bin ~ City_dist,
           diversity_all_sites,
           family = "binomial")

performance::check_model(fis_dist_bin)


# quantitative
fis_dist_quant <- glmmTMB(fis ~  City_dist ,
           diversity_all_sites %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_dist_quant)
```

##### Urb_score
```{r}
# binary
fis_usc_bin <- glmmTMB(fis_bin ~ urb_score,
           diversity_all_sites,
           family = "binomial")

performance::check_model(fis_usc_bin)


# quantitative
fis_usc_quant <- glmmTMB(fis ~ urb_score ,
           diversity_all_sites %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_usc_quant)
```

#### Pi
##### City_dist
```{r}
pi_dist <- glmmTMB(pi ~ City_dist +
                         (1|num_indv),
           diversity_all_sites)

performance::check_model(pi_dist)
```

##### Urb_score
```{r}
pi_usc <- glmmTMB(pi ~ urb_score +
                        (1|num_indv),
           diversity_all_sites)

performance::check_model(pi_usc)
```

### ANOVA & effect sizes
```{r}

mod_list <- list(priv_dist,
                 priv_usc,
                 fis_dist_bin,
                 fis_dist_quant,
                 fis_usc_bin,
                 fis_usc_quant,
                 pi_dist,
                 pi_usc)

create_anova_df <- lapply(mod_list, function(model) {
  anova_result <- car::Anova(model)
  
  formula <- formula(model)[2] %>%
    unlist()
  
  Rsq <- MuMIn::r.squaredGLMM(model)[1] %>%
  round(digits = 3) %>%
  format(nsmall = 3)
  
  # ADD EFFECT SIZES
  
  anova_tidy <- broom::tidy(anova_result) %>%
    dplyr::mutate("Variable" = as.character(formula),
                  "Rsq" = as.numeric(Rsq)) %>%
    dplyr::rename("Chi_sq" = statistic,
                  "p" = p.value,
                  "Urbanization" = term) %>%
    dplyr::mutate("Sig" = case_when(
      p <= 0.05 & p > 0.01 ~ "*",
      p <= 0.01 & p > 0.001 ~ "**",
      p <= 0.001 ~ "***",
      TRUE ~ "")) %>%
    dplyr::mutate(Chi_sq = format(round(Chi_sq, digits = 3), nsmall = 3),
                  p = format(round(p, digits = 3), nsmall = 3)) %>%
  dplyr::arrange(Urbanization, Variable) %>%
  dplyr::mutate("Variable" = case_when(
      Variable == "fis * 10" ~ "FIS (quantitative)",
      Variable == "fis" ~ "FIS (quantitative)",
      Variable == "fis_bin" ~ "FIS (binary)",
      Variable == "pi" ~ "Pi",
      Variable == "private" ~ "Private alleles",
      TRUE ~ ""))
  })

all_anovas <- do.call(rbind, create_anova_df) %>%
  dplyr::select(1,5,2,3,4,7,6) %T>%
  view

# create flextable
all_anovas %>%
  dplyr::select(-Sig) %>%
  flextable::flextable() %>%
  flextable::compose(i = 1, j = 2, part = "header",
                     value = as_paragraph("Χ", as_sup("2"))) %>%
  flextable::compose(i = 1, j = 5, part = "header",
                     value = as_paragraph("R", as_sup("2"))) %>%
  bold(~ p <= 0.05) %>%
  flextable::autofit()
```

### Figures
#### Private alleles
##### City_dist
```{r}
priv_dist_pred <- ggeffects::ggpredict(priv_dist,
                                   terms = c("City_dist"),
                                   type = "fe")

plot(priv_dist_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Private Alleles") +
 # ylim(0, 500) +
  labs(title = "") +
  theme_pubr() +
  labs_pubr() 

#ggsave(here::here("./Figures_Tables/regressions/private_alleles.png"))
```

##### Urb_score
```{r}
priv_usc_pred <- ggeffects::ggpredict(priv_usc,
                                   terms = c("urb_score"),
                                   type = "fe")

plot(priv_usc_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Urbanization Score") +
  ylab("Private Alleles") +
 # ylim(0, 500) +
  labs(title = "")+
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)
```

#### FIS
##### City_dist
```{r}
# binary
ggplot(diversity_all_sites,
       aes(x = City_dist,
           y = fis_bin)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr() 


# quantitative
ggplot(diversity_all_sites %>%
         dplyr::filter(fis != 0),
       aes(x = City_dist,
           y = fis)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 
```

##### Urb_score
```{r}
# binary
ggplot(diversity_all_sites,
       aes(x = urb_score,
           y = fis_bin)) +
  xlab("Urbanization Score") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr()  +
  xlim(4, -4)


# quantitative
ggplot(diversity_all_sites %>%
         dplyr::filter(fis != 0),
       aes(x = urb_score,
           y = fis)) +
  xlab("Urbanization Score") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### Pi
##### City_dist
```{r}

pi_dist_pred <- ggeffects::ggpredict(pi_dist,
                                   terms = c("City_dist"),
                                   type = "fe")

plot(pi_dist_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Pi") +
  #ylim(NA, 0.04) +
  labs(title = "") +
  theme_pubr()

ggsave(here::here("./Figures_Tables/regressions/pi.png"))
```

###### Urb_score
```{r}
pi_usc_pred <- ggeffects::ggpredict(pi_usc,
                                   terms = c("urb_score"),
                                   type = "fe")

plot(pi_usc_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Urbanization Score") +
  ylab("Pi") +
  #ylim(NA, 0.04) +
  labs(title = "") +
  theme_pubr()+
  xlim(4, -4)


```

#### Number of individuals per pop
##### City_dist
```{r}
ggplot(diversity_all_sites,
       aes(x = City_dist,
           y = num_indv)) +
  xlab("Distance to City Center (km)") +
  ylab("Individuals/Population") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
ggplot(diversity_all_sites,
       aes(x = urb_score,
           y = num_indv)) +
  xlab("Urbanization Score") +
  ylab("Individuals/Population") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

## Urbanization is categorical / data = all sites
### Fit linear models/diagnostics
#### Private alleles
##### u_r_dist
```{r}
priv_dist_cat <- glmmTMB(private ~ u_r_dist +
                           (1|num_indv) 
                    #   +    (1|sites)
                         ,
           diversity_all_sites,
           family = "nbinom2")

performance::check_model(priv_dist_cat)
car::Anova(priv_dist_cat) # becomes insignificant after accounting for sites and indivs/pop
```

##### u_r_usc
```{r}
priv_usc_cat <- glmmTMB(private ~ u_r_usc  +
                           (1|num_indv) 
                     #    +  (1|sites)
                        ,
           diversity_all_sites,
           family = "nbinom2")

performance::check_model(priv_usc_cat)
```

#### Inbreeding coefficient (FIS)
This is extremely zero-inflated. I've tried fitting models with zero inflation parameters w/glmmTMB but they don't converge and/or diagnostics don't pass tests. I'll fit manual hurdle models instead.
##### u_r_dist
```{r}
diversity_all_sites %<>%
  mutate(fis_bin = case_when(
    fis == 0 ~ 0,
    fis > 0 ~ 1)
  )
  
# binary
fis_dist_bin_cat <- glmmTMB(fis_bin ~ u_r_dist 
                        + (1|num_indv)
                        ,
           diversity_all_sites,
           family = "binomial")

performance::check_model(fis_dist_bin_cat)


# quantitative
fis_dist_quant_cat <- glmmTMB(fis ~  u_r_dist 
                        + (1|num_indv)
                        ,
           diversity_all_sites %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_dist_quant_cat)
```

##### u_r_usc
```{r}
# binary
fis_usc_bin_cat <- glmmTMB(fis_bin ~ u_r_usc 
                        + (1|num_indv)
                        ,
           diversity_all_sites,
           family = "binomial")

performance::check_model(fis_usc_bin_cat)


# quantitative
fis_usc_quant_cat <- glmmTMB(fis * 100 ~ u_r_usc 
                        + (1|num_indv)
                        ,
           diversity_all_sites %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_usc_quant_cat)
```

#### Pi
##### u_r_dist
```{r}
pi_dist_cat <- glmmTMB(pi ~ u_r_dist +
                         (1|num_indv),
           diversity_all_sites)

performance::check_model(pi_dist_cat)
```

##### u_r_usc
```{r}
pi_usc_cat <- glmmTMB(pi ~ u_r_usc +
                        (1|num_indv),
           diversity_all_sites)

performance::check_model(pi_usc_cat)
```

### ANOVA & effect sizes
```{r}

mod_list_cat <- list(priv_dist_cat,
                 priv_usc_cat,
                 fis_dist_bin_cat,
                 fis_dist_quant_cat,
                 fis_usc_bin_cat,
                 fis_usc_quant_cat,
                 pi_dist_cat,
                 pi_usc_cat)

create_anova_df_cat <- lapply(mod_list_cat, function(model) {
  anova_result <- car::Anova(model)
  
  formula <- formula(model)[2] %>%
    unlist()
  
  Rsq <- MuMIn::r.squaredGLMM(model)[1] %>%
  round(digits = 3) %>%
  format(nsmall = 3)
  
  # ADD EFFECT SIZES
  
  anova_tidy <- broom::tidy(anova_result) %>%
    dplyr::mutate("Variable" = as.character(formula),
                  "Rsq" = as.numeric(Rsq)) %>%
    dplyr::rename("Chi_sq" = statistic,
                  "p" = p.value,
                  "Urbanization" = term) %>%
    dplyr::mutate("Sig" = case_when(
      p <= 0.05 & p > 0.01 ~ "*",
      p <= 0.01 & p > 0.001 ~ "**",
      p <= 0.001 ~ "***",
      TRUE ~ "")) %>%
    dplyr::mutate(Chi_sq = format(round(Chi_sq, digits = 3), nsmall = 3),
                  p = format(round(p, digits = 3), nsmall = 3)) %>%
  dplyr::arrange(Urbanization, Variable) %>%
  dplyr::mutate("Variable" = case_when(
      Variable == "fis" ~ "FIS (quantitative)",
      Variable == "fis * 100" ~ "FIS (quantitative)",
      Variable == "fis_bin" ~ "FIS (binary)",
      Variable == "pi" ~ "Pi",
      Variable == "private" ~ "Private alleles",
      TRUE ~ "")) %>%
  dplyr::mutate("Urbanization" = case_when(
      Urbanization == "u_r_dist" ~ "Distance (categorical)",
      Urbanization == "u_r_usc" ~ "Urbanization Score (categorical)",
      TRUE ~ ""))
  })

all_anovas_cat <- do.call(rbind, create_anova_df_cat) %>%
  dplyr::select(1,5,2,3,4,7,6) %T>%
  view

# create flextable
all_anovas_cat %>%
  dplyr::select( -Sig) %>%
  flextable::flextable() %>%
  flextable::compose(i = 1, j = 2, part = "header",
                     value = as_paragraph("Χ", as_sup("2"))) %>%
  flextable::compose(i = 1, j = 5, part = "header",
                     value = as_paragraph("R", as_sup("2"))) %>%
  bold(~ p <= 0.05) %>%
  flextable::autofit()
```


### Figures
#### Private alleles
##### City_dist
```{r}
priv_dist_pred <- ggeffects::ggpredict(priv_dist,
                                   terms = c("City_dist"),
                                   type = "fe")

plot(priv_dist_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Private Alleles") +
 # ylim(0, 500) +
  labs(title = "") +
  theme_pubr() +
  labs_pubr() 

#ggsave(here::here("./Figures_Tables/regressions/private_alleles.png"))
```

##### Urb_score
```{r}
priv_usc_pred <- ggeffects::ggpredict(priv_usc,
                                   terms = c("urb_score"),
                                   type = "fe")

plot(priv_usc_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Urbanization Score") +
  ylab("Private Alleles") +
 # ylim(0, 500) +
  labs(title = "")+
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)
```

#### FIS
##### City_dist
```{r}
# binary
ggplot(diversity_all_sites,
       aes(x = City_dist,
           y = fis_bin)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr() 


# quantitative
ggplot(diversity_all_sites %>%
         dplyr::filter(fis != 0),
       aes(x = City_dist,
           y = fis)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 
```

##### Urb_score
```{r}
# binary
ggplot(diversity_all_sites,
       aes(x = urb_score,
           y = fis_bin)) +
  xlab("Urbanization Score") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr()  +
  xlim(4, -4)


# quantitative
ggplot(diversity_all_sites %>%
         dplyr::filter(fis != 0),
       aes(x = urb_score,
           y = fis)) +
  xlab("Urbanization Score") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### Pi
##### City_dist
```{r}

pi_dist_pred <- ggeffects::ggpredict(pi_dist,
                                   terms = c("City_dist"),
                                   type = "fe")

plot(pi_dist_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Pi") +
  #ylim(NA, 0.04) +
  labs(title = "") +
  theme_pubr()

ggsave(here::here("./Figures_Tables/regressions/pi.png"))
```

###### Urb_score
```{r}
pi_usc_pred <- ggeffects::ggpredict(pi_usc,
                                   terms = c("urb_score"),
                                   type = "fe")

plot(pi_usc_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Urbanization Score") +
  ylab("Pi") +
  #ylim(NA, 0.04) +
  labs(title = "") +
  theme_pubr()+
  xlim(4, -4)


```

For the next section (only variant sites considered), I just copy/pasted the above code and switched the data. The object names are the SAME. Beware.

## Urbanization is continuous / data = ONLY VARIANT sites
### Fit linear models/diagnostics
#### Private alleles
##### City_dist
```{r}
priv_dist <- glmmTMB(private ~ City_dist +
                           (1|num_indv) 
                    #   +    (1|sites)
                         ,
           diversity_variant_sites,
           family = "nbinom2")

performance::check_model(priv_dist)
car::Anova(priv_dist)
```

##### Urb_score
```{r}
priv_usc <- glmmTMB(private ~ urb_score  +
                           (1|num_indv) 
                     #    +  (1|sites)
                        ,
           diversity_variant_sites,
           family = "nbinom2")

performance::check_model(priv_usc)
```

#### Inbreeding coefficient (FIS)
This is extremely zero-inflated. I've tried fitting models with zero inflation parameters w/glmmTMB but they don't converge and/or diagnostics don't pass tests. I'll fit manual hurdle models instead.
##### City_dist
```{r}
diversity_variant_sites %<>%
  mutate(fis_bin = case_when(
    fis == 0 ~ 0,
    fis > 0 ~ 1)
  )
  
# binary
fis_dist_bin <- glmmTMB(fis_bin ~ City_dist 
                        + (1|num_indv)
                        ,
           diversity_variant_sites,
           family = "binomial")

performance::check_model(fis_dist_bin)


# quantitative
# wont' converge unless I multiply it
fis_dist_quant <- glmmTMB(fis * 10 ~  City_dist 
                        + (1|num_indv)
                        ,
           diversity_variant_sites %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_dist_quant)
```

##### Urb_score
```{r}
# binary
fis_usc_bin <- glmmTMB(fis_bin ~ urb_score 
                        + (1|num_indv)
                        ,
           diversity_variant_sites,
           family = "binomial")

performance::check_model(fis_usc_bin)


# quantitative
fis_usc_quant <- glmmTMB(fis ~ urb_score 
                        + (1|num_indv)
                        ,
           diversity_variant_sites %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_usc_quant)
```

#### Pi
##### City_dist
```{r}
pi_dist <- glmmTMB(pi ~ City_dist +
                         (1|num_indv),
           diversity_variant_sites)

performance::check_model(pi_dist)
```

##### Urb_score
```{r}
pi_usc <- glmmTMB(pi ~ urb_score +
                        (1|num_indv),
           diversity_variant_sites)

performance::check_model(pi_usc)
```

### ANOVA & effect sizes
```{r}

mod_list <- list(priv_dist,
                 priv_usc,
                 fis_dist_bin,
                 fis_dist_quant,
                 fis_usc_bin,
                 fis_usc_quant,
                 pi_dist,
                 pi_usc)

create_anova_df <- lapply(mod_list, function(model) {
  anova_result <- car::Anova(model)
  
  formula <- formula(model)[2] %>%
    unlist()
  
  Rsq <- MuMIn::r.squaredGLMM(model)[1] %>%
  round(digits = 3) %>%
  format(nsmall = 3)
  
  # ADD EFFECT SIZES
  
  anova_tidy <- broom::tidy(anova_result) %>%
    dplyr::mutate("Variable" = as.character(formula),
                  "Rsq" = as.numeric(Rsq)) %>%
    dplyr::rename("Chi_sq" = statistic,
                  "p" = p.value,
                  "Urbanization" = term) %>%
    dplyr::mutate("Sig" = case_when(
      p <= 0.05 & p > 0.01 ~ "*",
      p <= 0.01 & p > 0.001 ~ "**",
      p <= 0.001 ~ "***",
      TRUE ~ "")) %>%
    dplyr::mutate(Chi_sq = format(round(Chi_sq, digits = 3), nsmall = 3),
                  p = format(round(p, digits = 3), nsmall = 3)) %>%
  dplyr::arrange(Urbanization, Variable) %>%
  dplyr::mutate("Variable" = case_when(
      Variable == "exp_het" ~ "Expected heterozygosity",
      Variable == "obs_het" ~ "Observed heterozygosity",
      Variable == "exp_hom" ~ "Expected homozygosity",
      Variable == "obs_hom" ~ "Observed homozygosity",
      Variable == "fis" ~ "FIS (quantitative)",
      Variable == "fis * 10" ~ "FIS (quantitative)",
      Variable == "fis_bin" ~ "FIS (binary)",
      Variable == "pi" ~ "Pi",
      Variable == "private" ~ "Private alleles",
      TRUE ~ ""))
  })

all_anovas <- do.call(rbind, create_anova_df) %>%
  dplyr::select(1,5,2,3,4,7,6) %T>%
  view

# create flextable
all_anovas %>%
  dplyr::select(-Sig) %>%
  flextable::flextable() %>%
  flextable::compose(i = 1, j = 2, part = "header",
                     value = as_paragraph("Χ", as_sup("2"))) %>%
  flextable::compose(i = 1, j = 5, part = "header",
                     value = as_paragraph("R", as_sup("2"))) %>%
  bold(~ p <= 0.05) %>%
  flextable::autofit()
```

### Figures
#### Private alleles
##### City_dist
```{r}
priv_dist_pred <- ggeffects::ggpredict(priv_dist,
                                   terms = c("City_dist"),
                                   type = "fe")

plot(priv_dist_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Private Alleles") +
 # ylim(0, 500) +
  labs(title = "") +
  theme_pubr() +
  labs_pubr() 

#ggsave(here::here("./Figures_Tables/regressions/private_alleles.png"))
```

##### Urb_score
```{r}
priv_usc_pred <- ggeffects::ggpredict(priv_usc,
                                   terms = c("urb_score"),
                                   type = "fe")

plot(priv_usc_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Urbanization Score") +
  ylab("Private Alleles") +
 # ylim(0, 500) +
  labs(title = "")+
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)
```

#### FIS
##### City_dist
```{r}
# binary
ggplot(diversity_variant_sites,
       aes(x = City_dist,
           y = fis_bin)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr() 


# quantitative
ggplot(diversity_variant_sites %>%
         dplyr::filter(fis != 0),
       aes(x = City_dist,
           y = fis)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 
```

##### Urb_score
```{r}
# binary
ggplot(diversity_variant_sites,
       aes(x = urb_score,
           y = fis_bin)) +
  xlab("Urbanization Score") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr()  +
  xlim(4, -4)


# quantitative
ggplot(diversity_variant_sites %>%
         dplyr::filter(fis != 0),
       aes(x = urb_score,
           y = fis)) +
  xlab("Urbanization Score") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### Pi
##### City_dist
```{r}

pi_dist_pred <- ggeffects::ggpredict(pi_dist,
                                   terms = c("City_dist"),
                                   type = "fe")

plot(pi_dist_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Pi") +
  #ylim(NA, 0.04) +
  labs(title = "") +
  theme_pubr()

ggsave(here::here("./Figures_Tables/regressions/pi.png"))
```

###### Urb_score
```{r}
pi_usc_pred <- ggeffects::ggpredict(pi_usc,
                                   terms = c("urb_score"),
                                   type = "fe")

plot(pi_usc_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Urbanization Score") +
  ylab("Pi") +
  #ylim(NA, 0.04) +
  labs(title = "") +
  theme_pubr()+
  xlim(4, -4)


```

#### Number of individuals per pop
##### City_dist
```{r}
ggplot(diversity_variant_sites,
       aes(x = City_dist,
           y = num_indv)) +
  xlab("Distance to City Center (km)") +
  ylab("Individuals/Population") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 

```

##### Urb_score
```{r}
ggplot(diversity_variant_sites,
       aes(x = urb_score,
           y = num_indv)) +
  xlab("Urbanization Score") +
  ylab("Individuals/Population") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

## Urbanization is categorical / data = ONLY VARIANT sites
### Fit linear models/diagnostics
#### Private alleles
##### u_r_dist
```{r}
priv_dist_cat <- glmmTMB(private ~ u_r_dist +
                           (1|num_indv) 
                    #   +    (1|sites)
                         ,
           diversity_variant_sites,
           family = "nbinom2")

performance::check_model(priv_dist_cat)
car::Anova(priv_dist_cat) # becomes insignificant after accounting for sites and indivs/pop
```

##### u_r_usc
```{r}
priv_usc_cat <- glmmTMB(private ~ u_r_usc  +
                           (1|num_indv) 
                     #    +  (1|sites)
                        ,
           diversity_variant_sites,
           family = "nbinom2")

performance::check_model(priv_usc_cat)
```

#### Inbreeding coefficient (FIS)
This is extremely zero-inflated. I've tried fitting models with zero inflation parameters w/glmmTMB but they don't converge and/or diagnostics don't pass tests. I'll fit manual hurdle models instead.
##### u_r_dist
```{r}
diversity_variant_sites %<>%
  mutate(fis_bin = case_when(
    fis == 0 ~ 0,
    fis > 0 ~ 1)
  )
  
# binary
fis_dist_bin_cat <- glmmTMB(fis_bin ~ u_r_dist 
                        + (1|num_indv)
                        ,
           diversity_variant_sites,
           family = "binomial")

performance::check_model(fis_dist_bin_cat)


# quantitative
fis_dist_quant_cat <- glmmTMB(fis ~  u_r_dist 
                        + (1|num_indv)
                        ,
           diversity_variant_sites %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_dist_quant_cat)
```

##### u_r_usc
```{r}
# binary
fis_usc_bin_cat <- glmmTMB(fis_bin ~ u_r_usc 
                        + (1|num_indv)
                        ,
           diversity_variant_sites,
           family = "binomial")

performance::check_model(fis_usc_bin_cat)


# quantitative
fis_usc_quant_cat <- glmmTMB(fis * 100 ~ u_r_usc 
                        + (1|num_indv)
                        ,
           diversity_variant_sites %>%
             dplyr::filter(fis != 0))

performance::check_model(fis_usc_quant_cat)
```

#### Pi
##### u_r_dist
```{r}
pi_dist_cat <- glmmTMB(pi ~ u_r_dist +
                         (1|num_indv),
           diversity_variant_sites)

performance::check_model(pi_dist_cat)
```

##### u_r_usc
```{r}
pi_usc_cat <- glmmTMB(pi ~ u_r_usc +
                        (1|num_indv),
           diversity_variant_sites)

performance::check_model(pi_usc_cat)
```

### ANOVA & effect sizes
```{r}

mod_list_cat <- list(priv_dist_cat,
                 priv_usc_cat,
                 fis_dist_bin_cat,
                 fis_dist_quant_cat,
                 fis_usc_bin_cat,
                 fis_usc_quant_cat,
                 pi_dist_cat,
                 pi_usc_cat)

create_anova_df_cat <- lapply(mod_list_cat, function(model) {
  anova_result <- car::Anova(model)
  
  formula <- formula(model)[2] %>%
    unlist()
  
  Rsq <- MuMIn::r.squaredGLMM(model)[1] %>%
  round(digits = 3) %>%
  format(nsmall = 3)
  
  # ADD EFFECT SIZES
  
  anova_tidy <- broom::tidy(anova_result) %>%
    dplyr::mutate("Variable" = as.character(formula),
                  "Rsq" = as.numeric(Rsq)) %>%
    dplyr::rename("Chi_sq" = statistic,
                  "p" = p.value,
                  "Urbanization" = term) %>%
    dplyr::mutate("Sig" = case_when(
      p <= 0.05 & p > 0.01 ~ "*",
      p <= 0.01 & p > 0.001 ~ "**",
      p <= 0.001 ~ "***",
      TRUE ~ "")) %>%
    dplyr::mutate(Chi_sq = format(round(Chi_sq, digits = 3), nsmall = 3),
                  p = format(round(p, digits = 3), nsmall = 3)) %>%
  dplyr::arrange(Urbanization, Variable) %>%
  dplyr::mutate("Variable" = case_when(
      Variable == "fis" ~ "FIS (quantitative)",
      Variable == "fis * 100" ~ "FIS (quantitative)",
      Variable == "fis_bin" ~ "FIS (binary)",
      Variable == "pi" ~ "Pi",
      Variable == "private" ~ "Private alleles",
      TRUE ~ "")) %>%
  dplyr::mutate("Urbanization" = case_when(
      Urbanization == "u_r_dist" ~ "Distance (categorical)",
      Urbanization == "u_r_usc" ~ "Urbanization Score (categorical)",
      TRUE ~ ""))
  })

all_anovas_cat <- do.call(rbind, create_anova_df_cat) %>%
  dplyr::select(1,5,2,3,4,7,6) %T>%
  view

# create flextable
all_anovas_cat %>%
  dplyr::select(-Sig) %>%
  flextable::flextable() %>%
  flextable::compose(i = 1, j = 2, part = "header",
                     value = as_paragraph("Χ", as_sup("2"))) %>%
  flextable::compose(i = 1, j = 5, part = "header",
                     value = as_paragraph("R", as_sup("2"))) %>%
  bold(~ p <= 0.05) %>%
  flextable::autofit()
```


### Figures
#### Private alleles
##### City_dist
```{r}
priv_dist_pred <- ggeffects::ggpredict(priv_dist,
                                   terms = c("City_dist"),
                                   type = "fe")

plot(priv_dist_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Private Alleles") +
 # ylim(0, 500) +
  labs(title = "") +
  theme_pubr() +
  labs_pubr() 

#ggsave(here::here("./Figures_Tables/regressions/private_alleles.png"))
```

##### Urb_score
```{r}
priv_usc_pred <- ggeffects::ggpredict(priv_usc,
                                   terms = c("urb_score"),
                                   type = "fe")

plot(priv_usc_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Urbanization Score") +
  ylab("Private Alleles") +
 # ylim(0, 500) +
  labs(title = "")+
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)
```

#### FIS
##### City_dist
```{r}
# binary
ggplot(diversity_variant_sites,
       aes(x = City_dist,
           y = fis_bin)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr() 


# quantitative
ggplot(diversity_variant_sites %>%
         dplyr::filter(fis != 0),
       aes(x = City_dist,
           y = fis)) +
  xlab("Distance to City Center (km)") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() 
```

##### Urb_score
```{r}
# binary
ggplot(diversity_variant_sites,
       aes(x = urb_score,
           y = fis_bin)) +
  xlab("Urbanization Score") +
  ylab("FIS (binary)") + 
  geom_point(size = 1.5,
             shape = 1) +
  stat_smooth(method = "glm",
              color = "black",
              size = 0.5,
              method.args = list(family="binomial"),
              se = T) +
  theme_pubr() +
  labs_pubr()  +
  xlim(4, -4)


# quantitative
ggplot(diversity_variant_sites %>%
         dplyr::filter(fis != 0),
       aes(x = urb_score,
           y = fis)) +
  xlab("Urbanization Score") +
  ylab("FIS (Quantitative)") +
  plot_aesthetics +
  theme_pubr() +
  labs_pubr() +
  xlim(4, -4)

```

#### Pi
##### City_dist
```{r}

pi_dist_pred <- ggeffects::ggpredict(pi_dist,
                                   terms = c("City_dist"),
                                   type = "fe")

plot(pi_dist_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Distance to City Center (km)") +
  ylab("Pi") +
  #ylim(NA, 0.04) +
  labs(title = "") +
  theme_pubr()

ggsave(here::here("./Figures_Tables/regressions/pi.png"))
```

###### Urb_score
```{r}
pi_usc_pred <- ggeffects::ggpredict(pi_usc,
                                   terms = c("urb_score"),
                                   type = "fe")

plot(pi_usc_pred, 
     add.data = F,
     colors = "bw") +
  geom_ribbon(aes(x = x,
                  ymin = conf.low,
                  ymax = conf.high),
              # ,
              # fill = "#66a182",
               alpha = 0.1
              ) +
  xlab("Urbanization Score") +
  ylab("Pi") +
  #ylim(NA, 0.04) +
  labs(title = "") +
  theme_pubr()+
  xlim(4, -4)


```
