# Set up notebook
## Load libraries & functions
```{r}
source("libraries.R")
source("functions.R")

# if(!requireNamespace("corMLPE", quietly = TRUE)) remotes::install_github("nspope/corMLPE")
# 
# if(!requireNamespace("radish", quietly = TRUE)) remotes::install_github("nspope/radish")

library(radish)
library(raster)
library(igraph)

# # install package with tutorial data
# if(!requireNamespace("GeNetIt", quietly = TRUE)) remotes::install_github("jeffreyevans/GeNetIt")
# 
# # tutorial data
# data(ralu.site, package="GeNetIt")
```

## Load data
### Raster data (LULC)
```{r}
solris_raster <- raster::raster(here::here("./raw_data/SOLRIS_Version_3_0/SOLRIS_Version_3_0_LAMBERT.tif"))


# inspect
# print(solris_raster)
# plot(solris_raster)

# crop raster
gta_solris <- raster::crop(solris_raster,
                    extent(1285000, # xmin
                           1450000, # xmax
                           11840000, # ymin
                           11970000)) # ymax


# check classes not warped
unique(gta_solris)

rm(solris_raster)


# Add LULC classes to raster- PAUSING THIS FOR NOW. MAY NOT BE NECESSARY
# gta_classes <- unique(gta_solris)
# 
# print(gta_classes)
# 
# # took these labels from 'Southern Ontario Land Resource Information System Version 3.0 Data Specifications.pdf'
# factor_labels <- c("Open Beach",
#                    "Treed Sand Dune",
#                    "Open Cliff and Talus",
#                    "Treed Cliff and Talus",
#                    "Shrub Alvar",
#                    "Open Tallgrass Prairie",
#                    "Tallgrass Savannah",
#                    "Tallgrass Woodland",
#                    "Forest",
#                    "Coniferous Forest",
#                    "Mixed Forest",
#                    "Deciduous Forest",
#                    "Treed Swamp",
#                    "Thicket Swamp",
#                    "Fen",
#                    "Bog",
#                    "Marsh",
#                    "Open Water",
#                    "Plantation",
#                    "Hedge Rows",
#                    "Tilled",
#                    "Transportation",
#                    "Built Up Area- Pervious",
#                    "Built Up Area- Impervious",
#                    "Extraction- Aggregate",
#                    "Extraction- Peat/Topsoil",
#                    "Undifferentiated")
# 
# reclass_matrix <- cbind(gta_classes,
#                         factor_labels) 
# 
# 
# # Use 'reclassify' to convert numeric values into factor classes
# gta_solris_factor <- raster::reclassify(gta_solris,
#                                         reclass_matrix)

```


### Genetic data
```{r}
# format required: genetic distances

# vcf
vcf <- read.vcfR(
  here::here("./summary_stats/output_populations_no-IBD/mmaf0.05_R0.75/populations.snps.vcf"),
  verbose = FALSE)

## Convert vcf to genind object
my_genind <- vcfR2genind(vcf)

# tidy up colnames (remove periods)
inputdata1 <- my_genind$tab

genetic_distance_matrix <- poppr::prevosti.dist(my_genind)

```


### Geographic sampling coordinates
#### import coords
```{r}
# lat/longs of sampling sites
urb <- read.csv(here::here("./clean_data/urb_metrics.csv"))

# sampling site IDs and genetic sample names
coords <- read.csv(here("./genomic_resources/pop_map3.csv")) %>%

# if population doesn't start with "MWI", it should start with "MW"
  # rename col 1
  dplyr::rename(pop_id = 2) %>%
  dplyr::mutate(pop_id = as.factor(pop_id)) %>%

# take entries with numeric populations and add "MW" as prefix
  add_MW_IDs() %>%

# left vs. full join because some populations in urb_clean weren't genotyped due to lack of material
  left_join(., urb, by = "patch_id") %>%
  dplyr::mutate(patch_id = as.factor(patch_id),
                pop_id = as.factor(pop_id)) %>%
  dplyr::select(c(1,2,6,7)) %>%
  dplyr::rename("y" = 3,
                "x" = 4) %>%
  dplyr::arrange(sample) %>%
  dplyr::select("x", "y")

```

#### convert to correct proj/cs
```{r}

# Convert the dataframe to a SpatialPoints object
sp_points <- SpatialPoints(coords, 
                           proj4string = CRS("+proj=longlat +datum=WGS84 +no_defs"))

# Project the SpatialPoints object to match the raster's CRS
sp_points_proj <- spTransform(sp_points, crs(gta_solris))

# convert to spatialpoints obj
sp_points_proj <- SpatialPoints(sp_points_proj, 
                           proj4string = CRS("+proj=longlat +datum=WGS84 +no_defs"))

# Plot the raster
plot(gta_solris)

# Overlay the projected SpatialPoints on the raster
points(sp_points_proj, col = "red", pch = 20)
```

# Model fitting
## Scale covariates
```{r}

# convert into raster stack and scale
scale_covs <- stack(scale(gta_solris)) ## Must be a raster stack

# started 10:30am 8/3/2023
# error at 12pm: Error: cannot allocate vector of size 10.0 Gb

surface <- conductance_surface(covariates = scale_covs,
                               coords = sp_points_proj,
                               directions = 8)
```

## Fit radish models
```{r}
fit_mlpe <- radish(genetic_distance_matrix ~ solris_utm,
                   data = surface, 
                   conductance_model = radish::loglinear_conductance, 
                   measurement_model = radish::mlpe)

# examine model
summary(fit_mlpe)

## we are modeling conductances --> positive coefficient estimates indicate increasing conductance (e.g. rates of movement/gene flow) at higher values of the covariate, while negative coefficients indicate lower conductance at higher values of the covariate. 
```

