# Set up notebook
## Load libraries & functions
```{r}
source("libraries.R")
source("functions.R")

# if(!requireNamespace("corMLPE", quietly = TRUE)) remotes::install_github("nspope/corMLPE")
# 
# if(!requireNamespace("radish", quietly = TRUE)) remotes::install_github("nspope/radish")

library(radish)
library(raster)
library(igraph)

# install package with tutorial data
if(!requireNamespace("GeNetIt", quietly = TRUE)) remotes::install_github("jeffreyevans/GeNetIt")

# tutorial data
data(ralu.site, package="GeNetIt")
```

## Load data
### Raster data (LULC)
#### Ontario Land Cover compilation
```{r}
ontario_raster <- raster::raster(here::here("./raw_data/OntarioLandCoverComp-v2/OntarioLandCoverComp-v2/OLCC_V2_TIFF/OLCC_V2_TIFF.tif"))


# inspect
print(ontario_raster)
plot(ontario_raster)

# crop raster
gta <- raster::crop(ontario_raster,
                    extent(1285000, # xmin
                           1450000, # xmax
                           11840000, # ymin
                           11970000)) # ymax

# remove large ontario raster
rm(ontario_raster)


# Define the target CRS (UTM Zone 17N)
target_crs <- CRS("+proj=utm +zone=17 +ellps=WGS84 +datum=WGS84 +units=m +no_defs")

# Reproject the raster to UTM (meters)
gta_utm <- projectRaster(gta, crs = target_crs)

# convert into raster stack
gta_stack <- raster::stack(gta_utm)

```

#### SOLRIS
```{r}
solris_raster <- raster::raster(here::here("./raw_data/SOLRIS_Version_3_0/SOLRIS_Version_3_0_LAMBERT.tif"))


# inspect
print(solris_raster)
plot(solris_raster)

# crop raster
gta_solris <- raster::crop(solris_raster,
                    extent(1285000, # xmin
                           1450000, # xmax
                           11840000, # ymin
                           11970000)) # ymax

# remove large ontario raster
rm(solris_raster)


# Define the target CRS (UTM Zone 17N)
target_crs <- CRS("+proj=utm +zone=17 +ellps=WGS84 +datum=WGS84 +units=m +no_defs")

# Reproject the raster to UTM (meters)
solris_utm <- projectRaster(gta_solris, crs = target_crs)

# convert into raster stack
solris_gta_stack <- raster::stack(solris_utm)

```


### Genetic data
```{r}
# format required: genetic distances

# vcf
vcf <- read.vcfR(
  here::here("./summary_stats/output_populations_no-IBD/mmaf0.05_R0.75/populations.snps.vcf"),
  verbose = FALSE)

## Convert vcf to genind object
my_genind <- vcfR2genind(vcf)

# tidy up colnames (remove periods)
inputdata1 <- my_genind$tab

genetic_distance_matrix <- poppr::prevosti.dist(my_genind)

```


### Geographic sampling coordinates
```{r}
# lat/longs of sampling sites
urb <- read.csv(here::here("./clean_data/urb_metrics.csv"))

# sampling site IDs and genetic sample names
coords <- read.csv(here("./genomic_resources/pop_map3.csv")) %>%

# if population doesn't start with "MWI", it should start with "MW"
  # rename col 1
  dplyr::rename(pop_id = 2) %>%
  dplyr::mutate(pop_id = as.factor(pop_id)) %>%

# take entries with numeric populations and add "MW" as prefix
  add_MW_IDs() %>%

# left vs. full join because some populations in urb_clean weren't genotyped due to lack of material
  left_join(., urb, by = "patch_id") %>%
  dplyr::mutate(patch_id = as.factor(patch_id),
                pop_id = as.factor(pop_id)) %>%
  dplyr::select(c(1,2,6,7)) %>%
  dplyr::rename("y" = 3,
                "x" = 4) %>%
  dplyr::arrange(sample) %>%
  dplyr::select("x", "y")


# import UTM coordinates
coords2 <- coords
coordinates(coords2) <- c("x", "y")
proj4string(coords2) <- CRS("+proj=longlat +datum=WGS84 +units=m")
res <- spTransform(coords2,
                   CRS("+proj=utm +zone=17 ellps=WGS84 +units=m"))
coords_utm <- res@coords %>%
  as.data.frame()

# convert to sp SpatialPoints object
coords_utm_sp <- sp::SpatialPoints(coords = coords_utm,
                               proj4string = CRS("+proj=utm +zone=17 ellps=WGS84 +units=m"))

# verify points align with raster by overlaying the SpatialPoints on the raster
plot(solris_utm)

points(coords_utm_sp, col = "red", pch = 16)
```

# Model fitting
## Scale covariates
```{r}
scale_covs <- stack(scale(solris_gta_stack)) ## Must be a raster stack

surface <- conductance_surface(covariates = scale_covs,
                               coords = coords_utm_sp,
                               directions = 8)


```

## Fit radish models
```{r}
fit_mlpe <- radish(genetic_distance_matrix ~ solris_utm,
                   data = surface, 
                   conductance_model = radish::loglinear_conductance, 
                   measurement_model = radish::mlpe)

# examine model
summary(fit_mlpe)

## we are modeling conductances --> positive coefficient estimates indicate increasing conductance (e.g. rates of movement/gene flow) at higher values of the covariate, while negative coefficients indicate lower conductance at higher values of the covariate. 
```

