# Set up notebook
## load libraries
```{r}
source("libraries.R")
```

# Create sfs
## Load vcf2sfs code
```{r}
# can't install this package so copied the R code into my own file, load it

source(here::here("./R_scripts/exploration/vcf2sfs_code.R"))
```

## get path to vcf
```{r}
vcf_address <- here::here("./pi_fst_TajD_estimates/pi/global_pi/global_pi_vcf-all/populations.all.vcf")

```

## Notes for understanding vcfR::maf()'s output

*Each row is a locus.
*nAllele = the number of different alleles sequenced at this
 locus. At most, it's 2n = 261*2=522 because it's a diploid-
 so there's at most 2 alleles per locus for each individual.
 It can drop to numbers like 50 if only, for example,
 50/2n=50/2*261=50/522=~10% of the loci were sequenced.
*Count = the number of times a specific allele was seen at
 this locus.
*Frequency = Count/nAlleles to give the total frequency of
 that specific allele.

There can be bars in the 1 bins on the x axis because of
 missing values. Some loci were only sequenced in a few of
 the individuals (depending on R), so if an allele was seen
 only once but its locus was sequenced in only ~25% of
 individuals, it could still have a frequency >0.05 and
 wind up in the "1 individual" bin.

This holds up because the mmaf = 0.05, R = 1 plot should
 have no missing data and a min minor allele frequency of
 5%, so the first individual should show up at
 0.05*2n=0.05*2*261=26 individuals. It does.
 
## Read the VCF file and the popmap file and create a gt object
### SFS for all individuals in 1 population
```{r}
mygt <- vcf2gt(vcf_address,
               here::here("./genomic_resources/pop_map3_all1pop_for_sfs.txt"))

# look at all unique populations
unique(mygt$popmap)
```

### SFS for URBAN or RURAL individuals in 1 population
#### Urbanization = dist to CC
```{r}
mygt_urb <- vcf2gt(vcf_address,
               here::here("./clean_data/pop_map3_UR_dist.txt"))

# look at all unique populations
unique(mygt_urb$popmap)
```

#### Urbanization = urbanization score
```{r}
mygt_urb2 <- vcf2gt(vcf_address,
               here::here("./clean_data/pop_map3_UR_usc.txt"))

# look at all unique populations
unique(mygt_urb2$popmap)
```

## Create sfs
### SFS for all individuals in 1 population
```{r}
sfs <- gt2sfs.raw(mygt,
           "1") %>%
        as.data.frame() %>%
  dplyr::filter(X1 != 0) %T>%
  write_delim(here::here("./StairwayPlot/sfs.txt"))

```

### SFS for URBAN individuals in 1 population
#### Urbanization = dist to CC
```{r}
sfs_urb <- gt2sfs.raw(mygt_urb,
           "Urban") %>%
        as.data.frame() %>%
  dplyr::filter(Urban != 0) %T>%
  write_delim(here::here("./StairwayPlot/sfs_urban.txt"))

```

#### Urbanization = urbanization score
```{r}
sfs_urb2 <- gt2sfs.raw(mygt_urb2,
           "Urban") %>%
        as.data.frame() %>%
  dplyr::filter(Urban != 0) %T>%
  write_delim(here::here("./StairwayPlot/sfs_urban_urbscore.txt"))

```

### SFS for RURAL individuals in 1 population
#### Urbanization = dist to CC
```{r}
sfs_rur <- gt2sfs.raw(mygt_urb,
           "Rural") %>%
        as.data.frame() %>%
  dplyr::filter(Rural != 0) %T>%
  write_delim(here::here("./StairwayPlot/sfs_rural.txt"))

```

#### Urbanization = urbanization score
```{r}
sfs_rur2 <- gt2sfs.raw(mygt_urb2,
           "Rural") %>%
        as.data.frame() %>%
  dplyr::filter(Rural != 0) %T>%
  write_delim(here::here("./StairwayPlot/sfs_rural_urbscore.txt"))

```

## Prepare to export sfs
### SFS for all individuals in 1 population
```{r}
# get freqs for SFS
freqs_for_stairwayplot <- sfs %>%
                   dplyr::select(Freq) %>%
  pull()


freqs_string <- paste(freqs_for_stairwayplot,
                      collapse = " ")

# Write the string to a txt file
writeLines(freqs_string,
           here::here("./StairwayPlot/sfs_freqs.txt"))

# length of SFS
length(freqs_for_stairwayplot)
```

### SFS for URBAN individuals in 1 population
#### Urbanization = dist to CC
```{r}
# get freqs for SFS
sfs_urb %>%
  dplyr::select(Freq) %>%
  pull() %>%
  paste(collapse = " ") %T>%
# Write the string to a txt file
  writeLines(here::here("./StairwayPlot/sfs_freqs_urb.txt"))

# length of SFS
length(sfs_urb %>%
  dplyr::select(Freq) %>%
  pull())
```

#### Urbanization = urbanization score
```{r}
# get freqs for SFS
sfs_urb2 %>%
  dplyr::select(Freq) %>%
  pull() %>%
  paste(collapse = " ") %T>%
# Write the string to a txt file
  writeLines(here::here("./StairwayPlot/sfs_freqs_urb_urbscore.txt"))

# length of SFS
length(sfs_urb2 %>%
  dplyr::select(Freq) %>%
  pull())
```

### SFS for RURAL individuals in 1 population
#### Urbanization = dist to CC
```{r}
# get freqs for SFS
sfs_rur %>%
  dplyr::select(Freq) %>%
  pull() %>%
  paste(collapse = " ") %T>%
# Write the string to a txt file
  writeLines(here::here("./StairwayPlot/sfs_freqs_rur.txt"))

# length of SFS
length(sfs_rur %>%
  dplyr::select(Freq) %>%
  pull())
```

#### Urbanization = urbanization score
```{r}
# get freqs for SFS
sfs_rur2 %>%
  dplyr::select(Freq) %>%
  pull() %>%
  paste(collapse = " ") %T>%
# Write the string to a txt file
  writeLines(here::here("./StairwayPlot/sfs_freqs_rur_urbscore.txt"))

# length of SFS
length(sfs_rur2 %>%
  dplyr::select(Freq) %>%
  pull())
```

# Stairway Plot figures
## Import stairway plot data
### mu = 1.0x10^-8 & generation time = 1 year
#### All sites in 1 population
```{r}
full_fig_1e8 <- read.delim(here::here("./StairwayPlot/mu_1e-8/gen_time_1yr/full_mmaf0.007813_mut_rate_1e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

### mu = 1.0x10^-8 & generation time = 1.5 years
#### All sites in 1 population
```{r}
full_fig_1e8_1.5yrs <- read.delim(here::here("./StairwayPlot/mu_1e-8/gen_time_1.5yr/full_mmaf0.007813_mut_rate_1e-8_gt1.5.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

### mu = 1.2x10^-8 & generation time = 1 year
#### All sites in 1 population
```{r}
full_fig <- read.delim(here::here("./StairwayPlot/mu_1.2e-8/gen_time_1yr/full_mmaf0.007813.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

#### Urban
##### Dist to CC
```{r}
urb_fig <- read.delim(here::here("./StairwayPlot/mu_1.2e-8/gen_time_1yr/urban_dist_mmaf0.007813.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

##### Urb Score
```{r}
urb_fig2 <- read.delim(here::here("./StairwayPlot/mu_1.2e-8/gen_time_1yr/urban_usc_mmaf0.007813.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

#### Rural
##### Dist to CC
```{r}
rur_fig <- read.delim(here::here("./StairwayPlot/mu_1.2e-8/gen_time_1yr/rural_dist_mmaf0.007813.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

##### Urb score
```{r}
rur_fig2 <- read.delim(here::here("./StairwayPlot/mu_1.2e-8/gen_time_1yr/rural_usc_mmaf0.007813.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

#### Merge dfs
##### Urb = dist to CC
```{r}
all_dist <- dplyr::full_join(urb_fig %>%
                    dplyr::mutate(Group = "Urban"),
                  rur_fig %>%
                    dplyr::mutate(Group = "Rural")) %>%
  dplyr::full_join(.,
                   full_fig %>%
                    dplyr::mutate(Group = "Global"))

# remove duplicate rows
all_dist %<>%
  distinct(.keep_all = TRUE)
```


##### Urb = urb score
```{r}
all_usc <- dplyr::full_join(urb_fig2 %>%
                    dplyr::mutate(Group = "Urban"),
                  rur_fig2 %>%
                    dplyr::mutate(Group = "Rural")) %>%
  dplyr::full_join(.,
                   full_fig %>%
                    dplyr::mutate(Group = "Global"))

# remove duplicate rows
all_usc %<>%
  distinct(.keep_all = TRUE)
```

### mu = 1.2x10^-8 & generation time = 1.5 years
#### All sites in 1 population
```{r}
full_fig_1.5yrs <- read.delim(here::here("./StairwayPlot/mu_1.2e-8/gen_time_1.5yr/full_mmaf0.007813_mut_rate_1.2e-8_gt1.5.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

### mu = 1.8x10^-8 & generation time = 1 year
#### All sites in 1 population
```{r}
full_fig_1.8e8 <- read.delim(here::here("./StairwayPlot/mu_1.8e-8/gen_time_1yr/full_mmaf0.007813_mut_rate_1.8e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

#### Urban
##### Dist to CC
```{r}
urb_fig_1.8e8 <- read.delim(here::here("./StairwayPlot/mu_1.8e-8/gen_time_1yr/urban_dist_mmaf0.007813_mut_rate_1.8e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

##### Urb Score
```{r}
urb_fig2_1.8e8 <- read.delim(here::here("./StairwayPlot/mu_1.8e-8/gen_time_1yr/urban_usc_mmaf0.007813_mut_rate_1.8e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

#### Rural
##### Dist to CC
```{r}
rur_fig_1.8e8 <- read.delim(here::here("./StairwayPlot/mu_1.8e-8/gen_time_1yr/rural_dist_mmaf0.007813_mut_rate_1.8e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

##### Urb score
```{r}
rur_fig2_1.8e8 <- read.delim(here::here("./StairwayPlot/mu_1.8e-8/gen_time_1yr/rural_usc_mmaf0.007813_mut_rate_1.8e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

#### Merge dfs
##### Urb = dist to CC
```{r}
all_dist_1.8e8 <- dplyr::full_join(urb_fig_1.8e8 %>%
                    dplyr::mutate(Group = "Urban"),
                  rur_fig_1.8e8 %>%
                    dplyr::mutate(Group = "Rural")) %>%
  dplyr::full_join(.,
                   full_fig_1.8e8 %>%
                    dplyr::mutate(Group = "Global"))

# remove duplicate rows
all_dist_1.8e8 %<>%
  distinct(.keep_all = TRUE)
```


##### Urb = urb score
```{r}
all_usc_1.8e8 <- dplyr::full_join(urb_fig2_1.8e8 %>%
                    dplyr::mutate(Group = "Urban"),
                  rur_fig2_1.8e8 %>%
                    dplyr::mutate(Group = "Rural")) %>%
  dplyr::full_join(.,
                   full_fig_1.8e8 %>%
                    dplyr::mutate(Group = "Global"))

# remove duplicate rows
all_usc_1.8e8 %<>%
  distinct(.keep_all = TRUE)
```

### mu = 1.8x10^-8 & generation time = 1.5 years
#### All sites in 1 population
```{r}
full_fig_1.8e8_1.5yr <- read.delim(here::here("./StairwayPlot/mu_1.8e-8/gen_time_1.5yr/full_mmaf0.007813_mut_rate_1.8e-8_gt1.5.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

### mu = 2.6x10^-8 & generation time = 1 year
#### All sites in 1 population
```{r}
full_fig_2.6e8 <- read.delim(here::here("./StairwayPlot/mu_2.6e-8/gen_time_1yr/full_mmaf0.007813_mut_rate_2.6e-8.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

### mu = 2.6x10^-8 & generation time = 1.5 years
#### All sites in 1 population
```{r}
full_fig_2.6e8_1.5yr <- read.delim(here::here("./StairwayPlot/mu_2.6e-8/gen_time_1.5yr/full_mmaf0.007813_mut_rate_2.6e-8_gt1.5.final.summary")) %>%
  dplyr::select(-c(theta_per_site_median,
                   theta_per_site_2.5.,
                   theta_per_site_97.5.,
                   n_estimation,
                   mutation_per_site)) %>%
  # remove duplicate rows
  distinct(.keep_all = TRUE)

```

## Create functions
### Export plots
#### All, urban, and rural separately
```{r}
# regular y axis
make_ne_plot_reg_y <- function(df, exported_filepath){
  
  max_y <- df %>%
    dplyr::select(Ne_median) %>%
    dplyr::filter(Ne_median == max(Ne_median)) %>%
    dplyr::slice(1) %>%
    dplyr::mutate(max = Ne_median*1.1) %>%
    dplyr::select(max) %>%
    as.numeric()
  
  # full plot with red horizontal line at 500 years ago
  full_plot <- ggplot(
    df,
    aes(
      x = year,
      y = Ne_median))  +
    
    geom_segment( # add blue vertical line at 12,000 yrs (glaciers retreat)
      aes(
        x = 12000,
        xend = 12000,
        y = 0,
        yend = max_y),
      color = "blue",
      linetype = 5 ) +
    
    geom_segment( # add red vertical line at 500 yrs
      aes(
        x = 500,
        xend = 500,
        y = 0,
        yend = max_y),
      color = "red",
      linetype = 2 ) +
    geom_point(size = 0.75) +
    scale_x_continuous(labels = scales::comma,
                       limits = c(NA, 40000)) +
    scale_y_continuous(breaks=c(0, 25000, 50000, 75000, 100000, 125000, 150000),
                       labels = scales::comma,
                       limits = c(0, 
                                  max_y
                                  )
                       ) +
    xlab("Years before present") +
    ylab("Median Nₑ (1k individuals)") +
    theme_bw() +
    labs_pubr() +
    theme(plot.margin = unit(c(0.5,1,0.5,0.5), "cm"))
  
  ### Plot inset (past 2300 years)
  full_inset <- full_plot +
    scale_x_continuous(limits = c(NA, 2300),
                       labels = scales::comma) +
    theme(axis.title = element_blank(),
          plot.background = element_rect(
            colour = "black", fill = "white", size = 1))
  
  ### Overlay inset in full figure
  cowplot::ggdraw() +
    draw_plot(full_plot) +
    draw_plot(full_inset,
              x = 0.45,
              y = 0.4, 
              width = 0.45,
              height = 0.5)
  
  
  #### Export
  ggsave(filename = here::here(exported_filepath),
         height = 5,
         width = 7,
         dpi = 200)
}
```

#### All plots together
```{r}
export_all_plots <- function(df_distCC, df_urbscore, mu, gen_time){
  
  # All plots will have regular y axis
  
  # first, find maximum y axis values
    max_y_distCC <- df_distCC %>%
    dplyr::select(Ne_median) %>%
    dplyr::filter(Ne_median == max(Ne_median)) %>%
    dplyr::slice(1) %>%
    dplyr::mutate(max = Ne_median*1.1) %>%
    dplyr::select(max) %>%
    as.numeric()
    
    max_y_urbscore <- df_urbscore %>%
    dplyr::select(Ne_median) %>%
    dplyr::filter(Ne_median == max(Ne_median)) %>%
    dplyr::slice(1) %>%
    dplyr::mutate(max = Ne_median*1.1) %>%
    dplyr::select(max) %>%
    as.numeric()
  
  # DISTANCE TO CC-----
  exported_filename_distCC <- paste0("./Figures_Tables/Stairway_plot/mu_",
                              mu,
                              "/gen_time_",
                              gen_time,
                              "/distance",
                              "/all_with_inset.png")
  
  all_plot <- ggplot(df_distCC)  +
    
    geom_segment( # add red vertical line at 12,000 yrs (glaciers retreat)
      aes(
        x = 12000,
        xend = 12000,
        y = 0,
        yend = max_y_distCC),
      color = "blue",
      linetype = 5 ) +
    
    geom_segment( # add red vertical line at 500 yrs
      aes(
        x = 500,
        xend = 500,
        y = 0,
        yend = max_y_distCC),
      color = "red",
      linetype = 2 ) +

    geom_point(aes(x = year,
                   y = Ne_median,
                   color = Group),
               size = 0.25) +
  
    scale_x_continuous(labels = scales::comma,
                       limits = c(NA, 40000)) +
    scale_y_continuous(breaks=c(0, 25000, 50000, 75000, 100000, 125000, 150000),
                       labels = scales::comma,
                       limits = c(0, 
                                  max_y_distCC
                                  )) +
    xlab("Years before present") +
    ylab("Median Nₑ (1k individuals)") +
    theme_bw() +
    labs_pubr() +
  scale_color_viridis(discrete=TRUE) +
    theme(plot.margin = unit(c(0.5,1,0.5,0.5), "cm")) +
  # make legend point size larger
  guides(colour = guide_legend(
    override.aes = list(size=3)))
  

  ### Plot inset (past 2300 years)
all_plot_inset <- all_plot +
    scale_x_continuous(limits = c(NA, 2300),
                       labels = scales::comma) +
    theme(axis.title = element_blank(),
          plot.background = element_rect(
            colour = "black",
            fill = "white",
            size = 1),
          legend.position = "none")

  
  
#### Export
ggsave(plot =
### Overlay inset in full figure
cowplot::ggdraw() +
    draw_plot(all_plot) +
    draw_plot(all_plot_inset,
              x = 0.4,
              y = 0.4, 
              width = 0.4,
              height = 0.5),     
           filename = here::here(exported_filename_distCC),
         height = 5,
         width = 9,
         dpi = 200)
  


  # URBANIZATION SCORE-----

 exported_filename_urbscore <- paste0("./Figures_Tables/Stairway_plot/mu_",
                              mu,
                              "/gen_time_",
                              gen_time,
                              "/urb_score",
                              "/all_with_inset.png")

all_plot_usc <- ggplot(df_urbscore)  +
    
    geom_segment( # add red vertical line at 12,000 yrs (glaciers retreat)
      aes(
        x = 12000,
        xend = 12000,
        y = 0,
        yend = max_y_urbscore),
      color = "blue",
      linetype = 5 ) +
    
    geom_segment( # add red vertical line at 500 yrs
      aes(
        x = 500,
        xend = 500,
        y = 0,
        yend = max_y_urbscore),
      color = "red",
      linetype = 2 ) +

    geom_point(aes(x = year,
                   y = Ne_median,
                   color = Group),
               size = 0.25) +
  
    scale_x_continuous(labels = scales::comma,
                       limits = c(NA, 40000)) +
    scale_y_continuous(breaks=c(0, 25000, 50000, 75000, 100000, 125000, 150000),
                       labels = scales::comma,
                       limits = c(0, 
                                  max_y_urbscore)) +
    xlab("Years before present") +
    ylab("Median Nₑ (1k individuals)") +
  theme_bw() +
    labs_pubr() +
  scale_color_viridis(discrete=TRUE) +
    theme(plot.margin = unit(c(0.5,1,0.5,0.5), "cm")) +
  # make legend point size larger
  guides(colour = guide_legend(
    override.aes = list(size=3)))
  

  ### Plot inset (past 2300 years)
all_plot_inset_usc <- all_plot_usc +
    scale_x_continuous(limits = c(NA, 2300),
                       labels = scales::comma) +
    theme(axis.title = element_blank(),
          plot.background = element_rect(
            colour = "black",
            fill = "white",
            size = 1),
          legend.position = "none")

  
  
#### Export
ggsave(plot =
### Overlay inset in full figure
cowplot::ggdraw() +
    draw_plot(all_plot_usc) +
    draw_plot(all_plot_inset_usc,
              x = 0.4,
              y = 0.4, 
              width = 0.4,
              height = 0.5),     
  filename = here::here(exported_filename_urbscore),
         height = 5,
         width = 9,
         dpi = 200)
}
```

### Find Ne highs, lows
```{r}
Ne_stats <- function(Ne_df){

max_ne <- max(Ne_df$Ne_median)  

# find year of max Ne
max_ne_years <- Ne_df %>%
  dplyr::filter(Ne_median == max_ne) %>%
   dplyr::select(year) %>%
   dplyr::mutate(year = round(year, 0))

print(paste0("Years of max Ne: ", max_ne_years))


# find median Ne around 750 years ago
median_ne_750YA <- Ne_df %>%
  dplyr::filter(year > 750 & year < 751) %>%
  dplyr::select(Ne_median) %>%
  dplyr::mutate(Ne_median = round(Ne_median, 0)) %>%
  slice(1) %>%
  as.numeric()

print(paste0("Median Ne ~750 years ago: ", median_ne_750YA*1000))

min_year <- min(Ne_df$year)  

# find most recent median Ne
most_recent_ne <- Ne_df %>%
  dplyr::filter(year == min_year) %>%
  dplyr::select(Ne_median) %>%
  dplyr::mutate(Ne_median = round(Ne_median, 0)) %>%
  slice(1) %>%
  as.numeric()

print(paste0("Most recent median Ne: ", most_recent_ne*1000))


# find % decrease from max to min
perc_decrease <- round(((most_recent_ne-median_ne_750YA)/median_ne_750YA ) * 100, 3)

print(paste0("% decrease in Ne: ", perc_decrease))
}

```

## Export plots
### Mu = 1.0e-8 & gen time = 1 yr
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_1e8,
                   "./Figures_Tables/Stairway_plot/mu_1e-8/gen_time_1yr/full_with_inset.png")
```

### Mu = 1.0e-8 & gen time = 1.5 yrs
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_1e8_1.5yrs,
                   "./Figures_Tables/Stairway_plot/mu_1e-8/gen_time_1.5yr/full_with_inset.png")
```

### Mu = 1.2e-8 & gen time = 1 yr
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig,
                   "./Figures_Tables/Stairway_plot/mu_1.2e-8/gen_time_1yr/full_with_inset.png")
```

#### Export all plots
```{r}
export_all_plots(df_distCC = all_dist,
                 df_urbscore = all_usc,
                 mu = "1.2e-8",
                 gen_time = "1yr")
```

### Mu = 1.2e-8 & gen time = 1.5 yrs
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_1.5yrs,
                   "./Figures_Tables/Stairway_plot/mu_1.2e-8/gen_time_1.5yr/full_with_inset.png")
```

### Mu = 1.8e-8 & gen time = 1 yr
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_1.8e8,
                   "./Figures_Tables/Stairway_plot/mu_1.8e-8/gen_time_1yr/full_with_inset.png")
```

#### Export all plots
```{r}
export_all_plots(df_distCC = all_dist_1.8e8,
                 df_urbscore = all_usc_1.8e8,
                 mu = "1.8e-8",
                 gen_time = "1yr")
```


### Mu = 1.8e-8 & gen time = 1.5 yrs
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_1.8e8_1.5yr,
                   "./Figures_Tables/Stairway_plot/mu_1.8e-8/gen_time_1.5yr/full_with_inset.png")
```

### Mu = 2.6e-8 & gen time = 1 yr
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_2.6e8,
                   "./Figures_Tables/Stairway_plot/mu_2.6e-8/gen_time_1yr/full_with_inset.png")
```

### Mu = 2.6e-8 & gen time = 1.5 yrs
#### Export main plot
```{r}
make_ne_plot_reg_y(full_fig_2.6e8_1.5yr,
                   "./Figures_Tables/Stairway_plot/mu_2.6e-8/gen_time_1.5yr/full_with_inset.png")
```

### Just the inset (gen time = 1 yr, mu = 1.2e-8)
#### Past 200 years
```{r}
ggplot(
    full_fig,
    aes(
      x = year,
      y = Ne_median))  +
    geom_point(size = 0.75) +
    scale_x_continuous(limits = c(NA, 200),
                       labels = scales::comma) +
    scale_y_continuous(breaks=c(0, 10000, 20000, 30000, 40000, 50000),
                  labels = scales::comma,
                  limits = c(1, 50000)) +
    xlab("Years before present") +
    ylab("Median Nₑ (1k individuals)") +
  theme_bw() +
    labs_pubr() +
    theme(plot.margin = unit(c(0.5,1,0.5,0.5), "cm"),
          plot.background = element_rect(
            colour = "black", fill = "white", size = 1))

  
# Export
ggsave(filename = here::here("./Figures_Tables/Stairway_plot/Insets/full_inset_200yrs.png"),
         height = 5,
         width = 7,
         dpi = 200)
```

#### Past 800 years
```{r}
ggplot(
    full_fig,
    aes(
      x = year,
      y = Ne_median))  +
    geom_point(size = 0.75) +
    scale_x_continuous(limits = c(NA, 800),
                       labels = scales::comma) +
    scale_y_continuous(breaks=c(0, 20000, 40000, 60000, 80000, 100000),
                  labels = scales::comma,
                  limits = c(1, 105000)) +
    xlab("Years before present") +
    ylab("Median Nₑ (1k individuals)") +
  theme_bw() +
    labs_pubr() +
    theme(plot.margin = unit(c(0.5,1,0.5,0.5), "cm"),
          plot.background = element_rect(
            colour = "black", fill = "white", size = 1))

  
# Export
ggsave(filename = here::here("./Figures_Tables/Stairway_plot/Insets/full_inset_800yrs.png"),
         height = 5,
         width = 7,
         dpi = 200)
```

#### Past 2500 years
```{r}
ggplot(
    full_fig,
    aes(
      x = year,
      y = Ne_median))  +
    geom_point(size = 0.75) +
    scale_x_continuous(limits = c(NA, 2500),
                       labels = scales::comma) +
    scale_y_continuous(breaks=c(0, 30000, 60000, 90000, 120000, 150000),
                  labels = scales::comma,
                  limits = c(1, 150000)) +
    xlab("Years before present") +
    ylab("Median Nₑ (1k individuals)") +
  theme_bw() +
    labs_pubr() +
    theme(plot.margin = unit(c(0.5,1,0.5,0.5), "cm"),
          plot.background = element_rect(
            colour = "black", fill = "white", size = 1))

  
# Export
ggsave(filename = here::here("./Figures_Tables/Stairway_plot/Insets/full_inset_2500yrs.png"),
         height = 5,
         width = 7,
         dpi = 200)
```


## Calculate Ne stats
### mu = 1.8 x 10^-8, gen time = 1 yr
```{r}
Ne_stats(full_fig_1.8e8)
Ne_stats(urb_fig_1.8e8)
#Ne_stats(urb_fig2_1.8e8)
Ne_stats(rur_fig_1.8e8)
#Ne_stats(rur_fig2_1.8e8)
```

## Import human population data
### Import Canada human population data
#### Past-2007
```{r}
# this data doesn't have human population info- more about dwellings-----
# library(sf)
# human_pop <- st_read(here::here("./raw_data/Toronto_human_population_data_HWagner_20230918/GTADwellings.shp"))


# # scrape historic Canadian population data
# # 1867-1970 data Source: CANSIM, table 075-0001 (persons) ii
# # 1971 to present (2007) data Source: CANSIM, table 051-0001 (persons) iii
# 
# library(rvest)
# 
# # Read the HTML code from the website
# webpage <- read_html("https://www150.statcan.gc.ca/n1/pub/98-187-x/4151287-eng.htm")
# 
# # Using CSS selectors to scrape the relevant section
# human_pop <- html_nodes(webpage,
#                             'td div') %>%
#   html_text() %>%
#   as.data.frame() %>%
#   dplyr::rename("Year_and_Population" = 1) %T>%
#   # export
#   write.csv(.,
#           file =
#             here::here("./raw_data/human_population/human_population_Canada_raw.csv"))
# 
# human_pop %<>%
#   dplyr::rename("Year" = 1) %>%
#   separate(.,
#            "Year",
#            into = c("Year", "Population"),
#            sep = " +") %>%
#   dplyr::filter(is.na(Population) == F) %>%
# 
# # manual data cleaning
# # specific rows without spaces splitting year and population
#   dplyr::mutate(Year = ifelse(row_number() == 123,
#                                     1868, Year),
#                 Population = ifelse(row_number() == 123,
#                                     "3,511,000", Population),
#                 
#                 Year = ifelse(row_number() == 146,
#                                     1891, Year),
#                 Population = ifelse(row_number() == 146,
#                                     "4,833,000", Population),
#                 
# 
#                 Year = ifelse(row_number() == 152,
#                                     1897, Year),
#                 Population = ifelse(row_number() == 152,
#                                     "5,122,000", Population)
#                 ,
#                 
# 
#                 Year = ifelse(row_number() == 168,
#                                     1913, Year),
#                 Population = ifelse(row_number() == 168,
#                                     "7,632,000", Population)
#                 ) %>%
#   # remove years before 1867- census isn't uniform and numbers vary wildly
#   dplyr::slice(-(1:121)) %>%
# 
# # Split the Population column by comma
#   separate(.,
#            "Population",
#            into = c("Population1", "Population2", "Population3"),
#            sep = ",+") %>%
#   # only keep last 3 digits before last comma- there were some superscripts in original table (e.g., 50, 51...) that were added on as extra values to these population counts and have to be removed
#   dplyr::mutate(Population3 = str_sub(
#     Population3, start = 1, end = 3)) %>%
#   dplyr::mutate(Population = paste0(Population1, Population2, Population3)) %>%
#   dplyr::select(-c(Population1, Population2, Population3)) %>%
#     # convert to years before present, just like x axis in Ne dfs
#   dplyr::mutate(Population = as.numeric(Population),
#                 Year = as.numeric(Year),
#                 years_before_present = 2023 - Year) %T>%
#   # export
#   write.csv(.,
#           file =
#             here::here("./clean_data/human_population_Canada_clean.csv"))


human_pop <- read.csv(here::here("./clean_data/human_population_Canada_clean.csv"))
```

#### 2008-2022
```{r}
all_pop <- read.csv(here::here("./raw_data/human_population/human_population_Canada_raw_after2007.csv")) %>%
  # keep only Q1 values
  dplyr::filter(grepl('Q1', Geography)) %>%
  # remove Q1
  dplyr::mutate(Geography = str_replace(Geography,
                                        "Q1 ", "")) %>%
  dplyr::rename(Year = 1,
                Population = 2) %>%
    # convert to years before present, just like x axis in Ne dfs
  dplyr::mutate(Population = gsub(",", "", Population),
                Population = as.numeric(Population),
                Year = as.numeric(Year),
                years_before_present = 2023 - Year) %>%
  # join with older data
  dplyr::full_join(human_pop,
                   .,
                   by = c("Year", "Population", "years_before_present")) 

# %T>%
#   # export
#   write.csv(.,
#           file =
#             here::here("./clean_data/human_population_Canada_clean.csv"))

```

### Import Ontario human population data
```{r}
ontario_pop <- read.csv(here::here("./raw_data/human_population/Ontario_human_population/ontario_pop.csv")) %>%
  dplyr::mutate(Population = gsub(",", "", Population)) %>%
  dplyr::mutate(Year = as.numeric(Year),
                Population = as.numeric(Population),
                years_before_present = 2023 - Year) %T>%
  # export
  write.csv(.,
          file =
            here::here("./clean_data/human_population_Ontario_clean.csv"))
    
```

### Import Toronto human population data
```{r}
toronto_pop <- read.csv(here::here("./raw_data/human_population/Toronto_human_population_Wikipedia/Toronto_city_population.csv")) %>%
  dplyr::select(Year, City) %>%
  dplyr::rename(Population = City) %>%
  dplyr::mutate( years_before_present = 2023 - Year) %T>%
  # export
  write.csv(.,
          file =
            here::here("./clean_data/human_population_Toronto_clean.csv"))
    
```



## Export human population graphs
### Canada
```{r}
ggplot(
    all_pop,
    aes(
      x = years_before_present,
      y = Population/1000000))  +
    # geom_segment( # add vertical line at 1867
    #   aes(
    #     x = 156,
    #     xend = 156,
    #     y = 0,
    #     yend = 40),
    #   color = "blue",
    #   linetype = 4 ) +
    geom_point(size = 0.75) +
      geom_line() +
    scale_y_continuous(labels = scales::comma,
                       limits = c(0, NA)) +
    xlab("Years before present") +
    ylab("Human population in\n Canada (millions)") +
    theme_bw() +
    labs_pubr() +
    scale_x_continuous(limits = c(0, 156),
                       sec.axis = sec_axis(~ 2023 - .,
                                 name = "Year",
                                 breaks = c(2023, 2000, 1950, 1900, 1867))) +
    theme(plot.background = element_rect(
            colour = "black", fill = "white", size = 1))

ggsave(last_plot(),
       file = here::here("./Figures_Tables/Stairway_plot/Human_population/Canada_population.png"),
       height = 5,
         width = 7,
         dpi = 200)
  
```

### Ontario
```{r}
ggplot(
    ontario_pop,
    aes(
      x = years_before_present,
      y = Population/1000000))  +
    # geom_segment( # add vertical line at 1867
    #   aes(
    #     x = 156,
    #     xend = 156,
    #     y = 0,
    #     yend = 40),
    #   color = "blue",
    #   linetype = 4 ) +
    geom_point(size = 0.75) +
    geom_line() +
    scale_y_continuous(limits = c(0, 15)) +
    xlab("Years before present") +
    ylab("Human population in\n Ontario (millions)") +
    theme_bw() +
    labs_pubr() +
    scale_x_continuous(limits = c(0, 200),
                       sec.axis = sec_axis(~ 2023 - .,
                                 name = "Year",
                                 breaks = c(2023, 2000, 1950, 1900, 1850, 1824))) +
    theme(plot.background = element_rect(
            colour = "black", fill = "white", size = 1))

ggsave(last_plot(),
       file = here::here("./Figures_Tables/Stairway_plot/Human_population/Ontario_population.png"),
       height = 5,
         width = 7,
         dpi = 200)
  
```


### Toronto
```{r}
ggplot(
    toronto_pop,
    aes(
      x = years_before_present,
      y = Population/1000000))  +
    geom_point(size = 0.75) +
    geom_line() +
    scale_y_continuous(limits = c(0, 3)) +
    xlab("Years before present") +
    ylab("Human population in\n Toronto (millions)") +
    theme_bw() +
    labs_pubr() +
    scale_x_continuous(limits = c(0, 200),
                       sec.axis = sec_axis(~ 2023 - .,
                                 name = "Year",
                                 breaks = c(2023, 2000, 1950, 1900, 1850, 1824))) +
    theme(plot.background = element_rect(
            colour = "black", fill = "white", size = 1))

ggsave(last_plot(),
       file = here::here("./Figures_Tables/Stairway_plot/Human_population/Toronto_population.png"),
       height = 5,
         width = 7,
         dpi = 200)
  
```

