# Set up notebook
## Load libraries & functions
```{r}
source(here::here("libraries.R"))
```

# Visualize Tajima's D across the genome
## Import files
### Global (all sites in one large population)
#### Window size: 1 million bp
```{r}
global_taj <- read.delim(
  here::here("./pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/global_TajD_window_1mil/global_TajD_window_1mil.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(global_taj$TajimaD)
```

#### Window size: 100,000 bp
```{r}
global_taj_100k <- read.delim(
  here::here("./pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/global_TajD_window_100k/global_TajD_window_100k.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(global_taj_100k$TajimaD)
```

### Urban (based on distance to CC)
#### Window size: 1 million bp
```{r}
urb_taj <- read.delim(
  here::here("./pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Urban_dist_TajD_window_1mil/Urban_dist_TajD_window_1mil.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(urb_taj$TajimaD)
```

#### Window size: 100,000 bp
```{r}
urb_taj_100k <- read.delim(
  here::here("./pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Urban_dist_TajD_window_100k/Urban_dist_TajD_window_100k.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(urb_taj_100k$TajimaD)
```

### Rural (based on distance to CC)
#### Window size: 1 million bp
```{r}
rur_taj <- read.delim(
  here::here("./pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Rural_dist_TajD_window_1mil/Rural_dist_TajD_window_1mil.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(rur_taj$TajimaD)
```

#### Window size: 100,000 bp
```{r}
rur_taj_100k <- read.delim(
  here::here("./pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Rural_dist_TajD_window_100k/Rural_dist_TajD_window_100k.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(rur_taj_100k$TajimaD)
```

### Urban (based on urb score)
#### Window size: 1 million bp
```{r}
urb_taj_usc <- read.delim(
  here::here("./pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Urban_usc_TajD_window_1mil/Urban_usc_TajD_window_1mil.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(urb_taj_usc$TajimaD)
```

#### Window size: 100,000 bp
```{r}
urb_taj_100k_usc <- read.delim(
  here::here("./pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Urban_usc_TajD_window_100k/Urban_usc_TajD_window_100k.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(urb_taj_100k_usc$TajimaD)
```

### Rural (based on urb score)
#### Window size: 1 million bp
```{r}
rur_taj_usc <- read.delim(
  here::here("./pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Rural_usc_TajD_window_1mil/Rural_usc_TajD_window_1mil.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(rur_taj_usc$TajimaD)
```

#### Window size: 100,000 bp
```{r}
rur_taj_100k_usc <- read.delim(
  here::here("./pi_fst_TajD_estimates/TajD_mmaf0.007813_R0.5_no-IBD_vcf-all/Rural_usc_TajD_window_100k/Rural_usc_TajD_window_100k.Tajima.D")) %>%
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(rur_taj_100k_usc$TajimaD)
```

## Visualize
### Global (all sites in one large population)
#### Window size: 1 million bp
```{r}
global_taj_plot <- ggplot(global_taj,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(global_taj$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=50, 
           y=3,
           size = 4.5,
           label = paste("Mean Tajima's D:",
                         round(mean(global_taj$TajimaD),3)))

global_taj_plot

# add "0" x axis mark
global_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(global_taj_plot, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
global_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(global_taj_plot, 
                   x = 0.95,  y = 3.2, as.numeric(sum(global_taj$BIN_START)),
                   size = 12))


global_taj_plot

ggsave(here::here("./Figures_Tables/tajimas_d/global_taj_d.png"), 
       plot = global_taj_plot,
       width = 20,
       height = 10,
       units = "cm")

mean(global_taj$TajimaD)
sum(global_taj$BIN_START) 
```


#### Window size: 100,000 bp
```{r}
global_taj_plot2 <- ggplot(global_taj_100k,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(global_taj_100k$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=120, 
           y=3.5,
           size = 4.5,
           label = paste("Mean Tajima's D:",
                         round(mean(global_taj_100k$TajimaD),3)))

global_taj_plot2

# add "0" x axis mark
global_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(global_taj_plot2, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
global_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(global_taj_plot2, 
                   x = 0.93,  y = 3.2, as.numeric(sum(global_taj_100k$BIN_START)),
                   size = 12))


global_taj_plot2

ggsave(here::here("./Figures_Tables/tajimas_d/global_taj_d_window100k.png"), 
       plot = global_taj_plot2,
       width = 20,
       height = 10,
       units = "cm")

mean(global_taj_100k$TajimaD)
sum(global_taj_100k$BIN_START) 
```

### Urban (based on distance to CC)
#### Window size: 1 million bp
```{r}
urb_taj_plot <- ggplot(urb_taj,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(urb_taj$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=50, 
           y=3,
           size = 4.5,
           label = paste("Mean Tajima's D:",
                         round(mean(urb_taj$TajimaD),3)))

urb_taj_plot

# add "0" x axis mark
urb_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
urb_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot, 
                   x = 0.95,  y = 3.2, as.numeric(sum(urb_taj$BIN_START)),
                   size = 12))


urb_taj_plot

ggsave(here::here("./Figures_Tables/tajimas_d/urb_taj_d.png"), 
       plot = urb_taj_plot,
       width = 20,
       height = 10,
       units = "cm")

mean(urb_taj$TajimaD)
sum(urb_taj$BIN_START) 
```


#### Window size: 100,000 bp
```{r}
urb_taj_plot2 <- ggplot(urb_taj_100k,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(urb_taj_100k$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=120, 
           y=3.5,
           size = 4.5,
           label = paste("Mean Tajima's D:",
                         round(mean(urb_taj_100k$TajimaD),3)))

urb_taj_plot2

# add "0" x axis mark
urb_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot2, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
urb_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot2, 
                   x = 0.93,  y = 3.2, format(as.numeric(sum(urb_taj_100k$BIN_START)), scientific = TRUE),
                   size = 12))


urb_taj_plot2

ggsave(here::here("./Figures_Tables/tajimas_d/urb_taj_d_window100k.png"), 
       plot = urb_taj_plot2,
       width = 20,
       height = 10,
       units = "cm")

mean(urb_taj_100k$TajimaD)
sum(urb_taj_100k$BIN_START) 
```


### Rural (based on distance to CC)
#### Window size: 1 million bp
```{r}
rur_taj_plot <- ggplot(rur_taj,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(rur_taj$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=50, 
           y=3,
           size = 4.5,
           label = paste("Mean Tajima's D:",
                         round(mean(rur_taj$TajimaD),3)))

rur_taj_plot

# add "0" x axis mark
rur_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
rur_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot, 
                   x = 0.95,  y = 3.2, as.numeric(sum(rur_taj$BIN_START)),
                   size = 12))


rur_taj_plot

ggsave(here::here("./Figures_Tables/tajimas_d/rur_taj_d.png"), 
       plot = rur_taj_plot,
       width = 20,
       height = 10,
       units = "cm")

mean(rur_taj$TajimaD)
sum(rur_taj$BIN_START) 
```


#### Window size: 100,000 bp
```{r}
rur_taj_plot2 <- ggplot(rur_taj_100k,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(rur_taj_100k$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=120, 
           y=3.5,
           size = 4.5,
           label = paste("Mean Tajima's D:",
                         round(mean(rur_taj_100k$TajimaD),3)))

rur_taj_plot2

# add "0" x axis mark
rur_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot2, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
rur_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot2, 
                   x = 0.93,  y = 3.2, format(as.numeric(sum(rur_taj_100k$BIN_START)), scientific = TRUE),
                   size = 12))


rur_taj_plot2

ggsave(here::here("./Figures_Tables/tajimas_d/rur_taj_d_window100k.png"), 
       plot = rur_taj_plot2,
       width = 20,
       height = 10,
       units = "cm")

mean(rur_taj_100k$TajimaD)
sum(rur_taj_100k$BIN_START) 
```

### Urban (based on urb score)
#### Window size: 1 million bp
```{r}
urb_taj_plot_usc <- ggplot(urb_taj_usc,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(urb_taj_usc$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=50, 
           y=3,
           size = 4.5,
           label = paste("Mean Tajima's D:",
                         round(mean(urb_taj_usc$TajimaD),3)))

urb_taj_plot_usc

# add "0" x axis mark
urb_taj_plot_usc <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot_usc, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
urb_taj_plot_usc <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot_usc, 
                   x = 0.95,  y = 3.2, as.numeric(sum(urb_taj_usc$BIN_START)),
                   size = 12))


urb_taj_plot_usc

ggsave(here::here("./Figures_Tables/tajimas_d/urb_taj_d_usc.png"), 
       plot = urb_taj_plot_usc,
       width = 20,
       height = 10,
       units = "cm")

mean(urb_taj_usc$TajimaD)
sum(urb_taj_usc$BIN_START) 
```


#### Window size: 100,000 bp
```{r}
urb_taj_plot2_usc <- ggplot(urb_taj_100k_usc,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(urb_taj_100k_usc$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=120, 
           y=3.5,
           size = 4.5,
           label = paste("Mean Tajima's D:",
                         round(mean(urb_taj_100k_usc$TajimaD),3)))

urb_taj_plot2_usc

# add "0" x axis mark
urb_taj_plot2_usc <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot2_usc, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
urb_taj_plot2_usc <- cowplot::ggdraw(
  cowplot::add_sub(urb_taj_plot2_usc, 
                   x = 0.93,  y = 3.2, as.numeric(sum(urb_taj_100k_usc$BIN_START)),
                   size = 12))


urb_taj_plot2_usc

ggsave(here::here("./Figures_Tables/tajimas_d/urb_taj_d_window100k_usc.png"), 
       plot = urb_taj_plot2_usc,
       width = 20,
       height = 10,
       units = "cm")

mean(urb_taj_100k_usc$TajimaD)
sum(urb_taj_100k_usc$BIN_START) 
```


### Rural (based on urb score)
#### Window size: 1 million bp
```{r}
rur_taj_plot_usc <- ggplot(rur_taj_usc,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(rur_taj_usc$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=50, 
           y=3,
           size = 4.5,
           label = paste("Mean Tajima's D:",
                         round(mean(rur_taj_usc$TajimaD),3)))

rur_taj_plot_usc

# add "0" x axis mark
rur_taj_plot_usc <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot_usc, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
rur_taj_plot_usc <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot_usc, 
                   x = 0.95,  y = 3.2, as.numeric(sum(rur_taj_usc$BIN_START)),
                   size = 12))


rur_taj_plot_usc

ggsave(here::here("./Figures_Tables/tajimas_d/rur_taj_d_usc.png"), 
       plot = rur_taj_plot_usc,
       width = 20,
       height = 10,
       units = "cm")

mean(rur_taj_usc$TajimaD)
sum(rur_taj_usc$BIN_START) 
```


#### Window size: 100,000 bp
```{r}
rur_taj_plot2_usc <- ggplot(rur_taj_100k_usc,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(rur_taj_100k_usc$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=120, 
           y=3.5,
           size = 4.5,
           label = paste("Mean Tajima's D:",
                         round(mean(rur_taj_100k_usc$TajimaD),3)))

rur_taj_plot2_usc

# add "0" x axis mark
rur_taj_plot2_usc <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot2_usc, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
rur_taj_plot2_usc <- cowplot::ggdraw(
  cowplot::add_sub(rur_taj_plot2_usc, 
                   x = 0.93,  y = 3.2, format(as.numeric(sum(rur_taj_100k_usc$BIN_START)), scientific = TRUE),
                   size = 12))


rur_taj_plot2_usc

ggsave(here::here("./Figures_Tables/tajimas_d/rur_taj_d_window100k_usc.png"), 
       plot = rur_taj_plot2_usc,
       width = 20,
       height = 10,
       units = "cm")

mean(rur_taj_100k_usc$TajimaD)
sum(rur_taj_100k_usc$BIN_START) 
```

# Export means
```{r}
tajd_table <- data.frame(
  Group = c("Global",
            "Urban",
            "Urban",
            "Rural",
            "Rural"),
  Urbanization = c("-",
                   "Distance to City Center",
                   "Urbanization Score",
                   "Distance to City Center",
                   "Urbanization Score"),
  TajD_window_1mil = c(
    round(mean(global_taj$TajimaD),3),
    round(mean(urb_taj$TajimaD),3),
    round(mean(rur_taj$TajimaD),3),
    round(mean(urb_taj_usc$TajimaD),3),
    round(mean(rur_taj_usc$TajimaD),3)
    ),
  TajD_window_100k = c(
    round(mean(global_taj_100k$TajimaD),3),
    round(mean(urb_taj_100k$TajimaD),3),
    round(mean(rur_taj_100k$TajimaD),3),
    round(mean(urb_taj_100k_usc$TajimaD),3),
    round(mean(rur_taj_100k_usc$TajimaD),3)
  )
)

tajd_table %>%
  dplyr::rename("Window = 1,000,000 bp" = 3,
                "Window = 100,000 bp" = 4) %>%
  flextable::flextable() %>%
  flextable::align(align = "center", i = c(1:5), part = "body") %>%
  flextable::align(align = "left", j = 1:2, part = "all") %>%
  flextable::align(align = "center", i = 1, j = c(3,4), part = "header") %>%
  flextable::autofit() %>%
  flextable::save_as_docx(path = here::here(
    "./Figures_Tables/tajimas_d/taj_d_means.docx"))
```

