# Set up notebook
## Load libraries & functions
```{r warning=F}
source(here::here("libraries.R"))
```

# Visualize Tajima's D across the genome
## Import files
### Global (all sites in one large population)
#### Window size: 1 million bp
```{r}
global_taj <- read.csv(
  here::here("./global_TajD/global_TajD.csv")) %>%
  dplyr::rename("CHROM" = 1) %>%
  
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

range(global_taj$TajimaD)
```

#### Window size: 100,000 bp
```{r}
global_taj_100k <- read.csv(
  here::here("./global_TajD/global_TajD_window100k.csv")) %>%
  dplyr::rename("CHROM" = 1) %>%
  
  # remove windows with 0 SNPs
  dplyr::filter(N_SNPS != 0) 

```

## Visualize
### Global (all sites in one large population)
#### Window size: 1 million bp
```{r}
global_taj_plot <- ggplot(global_taj,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(global_taj$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=50, 
           y=3,
           size = 4.5,
           label = paste("Mean Tajima's D:",
                         round(mean(global_taj$TajimaD),3)))

global_taj_plot

# add "0" x axis mark
global_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(global_taj_plot, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
global_taj_plot <- cowplot::ggdraw(
  cowplot::add_sub(global_taj_plot, 
                   x = 0.95,  y = 3.2, as.numeric(sum(global_taj$BIN_START)),
                   size = 12))


global_taj_plot

ggsave(here::here("./Figures_Tables/tajimas_d/global_taj_d.png"), 
       plot = global_taj_plot,
       width = 20,
       height = 10,
       units = "cm")

mean(global_taj$TajimaD)
sum(global_taj$BIN_START) 
```


#### Window size: 100,000 bp
```{r}
global_taj_plot2 <- ggplot(global_taj_100k,
       aes(x = paste(CHROM,BIN_START),
           y = TajimaD,
       group = 1)) +
  geom_point(shape = 21) +
  xlab("Nucleotide Position") +
  ylab("Tajima's D") +
  geom_hline(yintercept = mean(global_taj_100k$TajimaD),
              linetype = "longdash",
              color = "black",
             linewidth = 1.1) +
  scale_x_discrete(breaks = c(paste("AS1.0ch01", 0.0e+00), paste("Scaff685",0.0e+00))) +
  theme_pubr() +
  theme(axis.text.x = element_text(color = "white")) +
  annotate("text",
           x=120, 
           y=3.5,
           size = 4.5,
           label = paste("Mean Tajima's D:",
                         round(mean(global_taj_100k$TajimaD),3)))

global_taj_plot2

# add "0" x axis mark
global_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(global_taj_plot2, 
                   x = 0,  y = 1.9, "0",
                   size = 13))

# add ending x axis mark
global_taj_plot2 <- cowplot::ggdraw(
  cowplot::add_sub(global_taj_plot2, 
                   x = 0.93,  y = 3.2, as.numeric(sum(global_taj_100k$BIN_START)),
                   size = 12))


global_taj_plot2

ggsave(here::here("./Figures_Tables/tajimas_d/global_taj_d_window100k.png"), 
       plot = global_taj_plot2,
       width = 20,
       height = 10,
       units = "cm")

mean(global_taj_100k$TajimaD)
sum(global_taj_100k$BIN_START) 
```



## Test significance
### Global
```{r}
# import vcf
library(vcfR)
vcf <- read.vcfR(
  here::here("./global_pi_vcf-all/populations.all.vcf"),
  verbose = FALSE)

# this doesn't work... says 0 segregating sites
# convert to dnabin format
dnabin <- vcfR2DNAbin(vcf)

rm(vcf)
pegas::tajima.test(dnabin)
```

